{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "80a7f636",
   "metadata": {},
   "source": [
    "<H1>MAPK8IP3 genotype phenotype correlations</H1>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c3b3b7e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/robin/PycharmProjects/genophenocorr/gpc_venv/lib/python3.9/site-packages/urllib3/__init__.py:35: NotOpenSSLWarning: urllib3 v2 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded HPO v2023-10-09\n",
      "hpotk version 0.5.0\n",
      "Using genophenocorr version 0.1.1dev\n"
     ]
    }
   ],
   "source": [
    "import genophenocorr\n",
    "import hpotk\n",
    "from genophenocorr.preprocessing import configure_caching_cohort_creator, load_phenopacket_folder\n",
    "from genophenocorr.view import CohortViewable, StatsViewable\n",
    "from genophenocorr.model import VariantEffect\n",
    "from genophenocorr.analysis import configure_cohort_analysis, CohortAnalysisConfiguration\n",
    "from genophenocorr.analysis.predicate import PatientCategories\n",
    "\n",
    "from IPython.display import display, HTML\n",
    "\n",
    "store = hpotk.configure_ontology_store()\n",
    "hpo = store.load_minimal_hpo(release='v2023-10-09')\n",
    "print(f'Loaded HPO v{hpo.version}')\n",
    "print(f'hpotk version {hpotk.__version__}')\n",
    "print(f\"Using genophenocorr version {genophenocorr.__version__}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5694f0d4",
   "metadata": {},
   "source": [
    "### Setup\n",
    "\n",
    "Here we specificy the path to the folder with the phenopackets to be analyzed, and the transcript to be used for the analysis (in general, the MANE transcript should be used). Here, we use the transcript `NM_001318852.2` which is the MANE transcript of the *MAPK8IP3* gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c9a4489e",
   "metadata": {},
   "outputs": [],
   "source": [
    "fpath_phenopackets = 'phenopackets'\n",
    "MAPK8IP3_id = 'NM_001318852.2'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "021d1952",
   "metadata": {},
   "source": [
    "### Create the cohort\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d68d06a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Patients Created: 100%|██████████| 20/20 [00:00<00:00, 443.17it/s]\n",
      "Validated under none policy\n",
      "No errors or warnings were found\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<!DOCTYPE html>\n",
       "<html lang=\"en\">\n",
       "<head>\n",
       "  <meta charset=\"utf-8\">\n",
       "  <title>Cohort</title>\n",
       "  <style>table {\n",
       "    border-collapse: collapse;\n",
       "    margin: 25px 0;\n",
       "      font-size: 0.9em;\n",
       "      font-family: sans-serif;\n",
       "      min-width: 400px;\n",
       "      box-shadow: 0 0 20px rgba(0, 0, 0, 0.15);\n",
       "  }\n",
       "\n",
       "  .table .column-1 {\n",
       "      text-align: left;\n",
       "  }\n",
       "  th {\n",
       "    background-color: LightSkyBlue;\n",
       "    border: 1px solid #dddddd;\n",
       "    text-align: left;\n",
       "    padding: 2px;\n",
       "  font-weight: bold;  \n",
       "  font-size: 120%;\n",
       "}\n",
       "\n",
       "\n",
       "  tr {\n",
       "    border: 1px solid #dddddd;\n",
       "  }\n",
       "\n",
       "  td {\n",
       "    padding: 2px;\n",
       "    font-weight: bold;\n",
       "  }\n",
       "\n",
       "  tr:nth-child(even) {\n",
       "    background-color: #f2f2f2;\n",
       "  }\n",
       "\n",
       "  .table td,tr {\n",
       "    text-align: right;\n",
       "}\n",
       "\n",
       ".lft {\n",
       "    text-align: left;\n",
       "}​\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "caption {\n",
       "  caption-side: bottom;\n",
       "  text-align: left;\n",
       "  padding-bottom: 10px;\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       "  </style>\n",
       "</head>\n",
       "\n",
       "<body>\n",
       "  <h1>genophenocorr cohort analysis</h1>\n",
       "  <p>Successfully ingested 20 phenopackets.</p>\n",
       "  \n",
       "  <p>No errors encountered.</p>\n",
       "  \n",
       " \n",
       "    <h3>Top 10 HPO Terms</h3>\n",
       "    <table>\n",
       "      <caption>A total of 128 HPO terms were used to annotated the cohort. </caption>\n",
       "        <tbody>\n",
       "            <tr class=\"strng\">\n",
       "                <th class=\"lft\">HPO Term</th>\n",
       "                <th>ID</th>\n",
       "                <th>Annotation Count</th>\n",
       "              </tr>\n",
       "              \n",
       "              <tr>\n",
       "                <td class=\"lft\">Global developmental delay</td>\n",
       "                <td>HP:0001263</td>\n",
       "                <td>19</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Hypotonia</td>\n",
       "                <td>HP:0001252</td>\n",
       "                <td>11</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Thin corpus callosum</td>\n",
       "                <td>HP:0033725</td>\n",
       "                <td>9</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Intellectual disability, moderate</td>\n",
       "                <td>HP:0002342</td>\n",
       "                <td>7</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Delayed ability to walk</td>\n",
       "                <td>HP:0031936</td>\n",
       "                <td>6</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Spastic diplegia</td>\n",
       "                <td>HP:0001264</td>\n",
       "                <td>6</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Intellectual disability</td>\n",
       "                <td>HP:0001249</td>\n",
       "                <td>6</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Seizure</td>\n",
       "                <td>HP:0001250</td>\n",
       "                <td>6</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Cerebral atrophy</td>\n",
       "                <td>HP:0002059</td>\n",
       "                <td>5</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Thin upper lip vermilion</td>\n",
       "                <td>HP:0000219</td>\n",
       "                <td>5</td>\n",
       "                </tr>\n",
       "            \n",
       "                </tbody>\n",
       "    </table>\n",
       "\n",
       " <h3>Top 10 Variants</h3>\n",
       " <table>\n",
       "  <caption>\n",
       "    Variants are shown according to NM_001318852.2. A total of 20 unique variants were identified in the cohort.\n",
       "  </caption>\n",
       "        <tbody>\n",
       "            <tr class=\"strng\">\n",
       "                <th class=\"lft\">Variant</th>\n",
       "                <th>Variant name</th>\n",
       "                <th>Variant Count</th>\n",
       "              </tr>\n",
       "              \n",
       "              <tr>\n",
       "                <td class=\"lft\">16_1762843_1762843_C_T</td>\n",
       "                <td>c.1735C>T</td>\n",
       "                <td>6</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">16_1767834_1767834_C_T</td>\n",
       "                <td>c.3439C>T</td>\n",
       "                <td>5</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">16_1760409_1760409_T_C</td>\n",
       "                <td>c.1334T>C</td>\n",
       "                <td>2</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">16_1706384_1706384_C_G</td>\n",
       "                <td>c.45C>G</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">16_1762388_1762388_G_A</td>\n",
       "                <td>c.1577G>A</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">16_1706418_1706418_G_T</td>\n",
       "                <td>c.79G>T</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">16_1706402_1706403_CG_C</td>\n",
       "                <td>c.65del</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">16_1748705_1748705_G_A</td>\n",
       "                <td>c.1201G>A</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">16_1766768_1766768_C_G</td>\n",
       "                <td>c.2985C>G</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">16_1706450_1706450_C_G</td>\n",
       "                <td>c.111C>G</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "         </tbody>\n",
       "    </table>\n",
       "    <h3>Diseases</h3>\n",
       "    <table>\n",
       "        <tbody>\n",
       "            <tr class=\"strng\">\n",
       "                <th class=\"lft\">Disease ID</th>\n",
       "                <th>Disease Name</th>\n",
       "                <th>Annotation Count</th>\n",
       "              </tr>\n",
       "              \n",
       "              <tr>\n",
       "                <td class=\"lft\">OMIM:618443</td>\n",
       "                <td>Neurodevelopmental disorder with or without variable brain abnormalities</td>\n",
       "                <td>20</td>\n",
       "                </tr>\n",
       "            \n",
       "           </tbody>\n",
       "    </table>\n",
       "    \n",
       "  <p>Call this function with transcript to see table with variant effect counts.</p>\n",
       "  \n",
       "\n",
       "\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cohort_creator = configure_caching_cohort_creator(hpo)\n",
    "cohort = load_phenopacket_folder(fpath_phenopackets, cohort_creator)\n",
    "cv = CohortViewable(hpo=hpo)\n",
    "html = cv.process(cohort=cohort, transcript_id=MAPK8IP3_id)\n",
    "\n",
    "display(HTML(html))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d732e6f",
   "metadata": {},
   "source": [
    "## Configure the analysis\n",
    "For the first run, we will not remove unneeded terms prior to statistical analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c4567056",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_config = CohortAnalysisConfiguration()\n",
    "analysis_config.missing_implies_excluded = True\n",
    "analysis_config.pval_correction = 'fdr_bh'\n",
    "analysis_config.min_perc_patients_w_hpo = 0.1\n",
    "analysis_config.heuristic_strategy()\n",
    "analysis = configure_cohort_analysis(cohort, hpo, config=analysis_config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0938ee95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>FRAMESHIFT_VARIANT on NM_001318852.2</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Yes</th>\n",
       "      <th colspan=\"2\" halign=\"left\">No</th>\n",
       "      <th colspan=\"2\" halign=\"left\"></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>p value</th>\n",
       "      <th>Corrected p value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Abnormality of movement [HP:0100022]</th>\n",
       "      <td>0/1</td>\n",
       "      <td>0%</td>\n",
       "      <td>8/13</td>\n",
       "      <td>62%</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cerebral palsy [HP:0100021]</th>\n",
       "      <td>0/1</td>\n",
       "      <td>0%</td>\n",
       "      <td>3/13</td>\n",
       "      <td>23%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cerebral visual impairment [HP:0100704]</th>\n",
       "      <td>0/1</td>\n",
       "      <td>0%</td>\n",
       "      <td>2/6</td>\n",
       "      <td>33%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hypotonia [HP:0001252]</th>\n",
       "      <td>1/1</td>\n",
       "      <td>100%</td>\n",
       "      <td>10/17</td>\n",
       "      <td>59%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Autistic behavior [HP:0000729]</th>\n",
       "      <td>1/1</td>\n",
       "      <td>100%</td>\n",
       "      <td>3/6</td>\n",
       "      <td>50%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gait disturbance [HP:0001288]</th>\n",
       "      <td>0/1</td>\n",
       "      <td>0%</td>\n",
       "      <td>5/10</td>\n",
       "      <td>50%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Seizure [HP:0001250]</th>\n",
       "      <td>0/1</td>\n",
       "      <td>0%</td>\n",
       "      <td>6/17</td>\n",
       "      <td>35%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inability to walk [HP:0002540]</th>\n",
       "      <td>0/1</td>\n",
       "      <td>0%</td>\n",
       "      <td>4/10</td>\n",
       "      <td>40%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "FRAMESHIFT_VARIANT on NM_001318852.2      Yes             No          \\\n",
       "                                        Count Percent  Count Percent   \n",
       "Abnormality of movement [HP:0100022]      0/1      0%   8/13     62%   \n",
       "Cerebral palsy [HP:0100021]               0/1      0%   3/13     23%   \n",
       "Cerebral visual impairment [HP:0100704]   0/1      0%    2/6     33%   \n",
       "Hypotonia [HP:0001252]                    1/1    100%  10/17     59%   \n",
       "Autistic behavior [HP:0000729]            1/1    100%    3/6     50%   \n",
       "Gait disturbance [HP:0001288]             0/1      0%   5/10     50%   \n",
       "Seizure [HP:0001250]                      0/1      0%   6/17     35%   \n",
       "Inability to walk [HP:0002540]            0/1      0%   4/10     40%   \n",
       "\n",
       "FRAMESHIFT_VARIANT on NM_001318852.2                                 \n",
       "                                          p value Corrected p value  \n",
       "Abnormality of movement [HP:0100022]     0.428571               1.0  \n",
       "Cerebral palsy [HP:0100021]              1.000000               1.0  \n",
       "Cerebral visual impairment [HP:0100704]  1.000000               1.0  \n",
       "Hypotonia [HP:0001252]                   1.000000               1.0  \n",
       "Autistic behavior [HP:0000729]           1.000000               1.0  \n",
       "Gait disturbance [HP:0001288]            1.000000               1.0  \n",
       "Seizure [HP:0001250]                     1.000000               1.0  \n",
       "Inability to walk [HP:0002540]           1.000000               1.0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frameshift = analysis.compare_by_variant_effect(VariantEffect.FRAMESHIFT_VARIANT, tx_id=MAPK8IP3_id)\n",
    "frameshift.summarize(hpo, PatientCategories.YES)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "cdf2c9e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<!DOCTYPE html>\n",
       "<html lang=\"en\">\n",
       "<head>\n",
       "  <meta charset=\"utf-8\">\n",
       "  <title>Cohort</title>\n",
       "  <style>table {\n",
       "    border-collapse: collapse;\n",
       "    margin: 25px 0;\n",
       "      font-size: 0.9em;\n",
       "      font-family: sans-serif;\n",
       "      min-width: 400px;\n",
       "      box-shadow: 0 0 20px rgba(0, 0, 0, 0.15);\n",
       "  }\n",
       "\n",
       "\n",
       "  th {\n",
       "    background-color: #f2f2f2;\n",
       "    border: 1px solid #dddddd;\n",
       "    text-align: left;\n",
       "    padding: 2px;\n",
       "  }\n",
       "\n",
       "  tr {\n",
       "    border: 1px solid #dddddd;\n",
       "  }\n",
       "\n",
       "  td {\n",
       "    padding: 2px;\n",
       "    font-weight: bold;\n",
       "  }\n",
       "\n",
       "  tr:nth-child(even) {\n",
       "    background-color: #f2f2f2;\n",
       "  }\n",
       "\n",
       "  .table td,tr {\n",
       "    text-align: left;\n",
       "}\n",
       "\n",
       ".table td:first-child, tr:first-child {\n",
       "    text-align: left;\n",
       "}​\n",
       "\n",
       "  </style>\n",
       "</head>\n",
       "\n",
       "<body>\n",
       "  <h1>Statistical analysis: fdr_bh (heuristic sampler)</h1>\n",
       "  <p>Performed statistical tests for 8 HPO terms.</p>\n",
       "    <table>\n",
       "      <caption>Using  heuristic sampler, the following numbers of terms were omitted from statistical analysis. </caption>\n",
       "        <tbody>\n",
       "            <tr>\n",
       "                <th>Skipped</th>\n",
       "                <th>Count</th>\n",
       "              </tr>\n",
       "              \n",
       "              <tr>\n",
       "                <td>Skipping term with only 1 observations (not powered for 2x2)</td>\n",
       "                <td>150</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term with only 2 observations (not powered for 2x2)</td>\n",
       "                <td>33</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term with only 3 observations (not powered for 2x2)</td>\n",
       "                <td>27</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term because one genotype had zero observations</td>\n",
       "                <td>22</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term with only 4 observations (not powered for 2x2)</td>\n",
       "                <td>17</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term with only 6 observations (not powered for 2x2)</td>\n",
       "                <td>15</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term because all genotypes have same HPO observed proportions</td>\n",
       "                <td>14</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping top level term</td>\n",
       "                <td>12</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term with only 5 observations (not powered for 2x2)</td>\n",
       "                <td>12</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term because no genotype has more than one observed HPO count</td>\n",
       "                <td>3</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping non phenotype term </td>\n",
       "                <td>2</td>\n",
       "                </tr>\n",
       "            \n",
       "                </tbody>\n",
       "    </table>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "report = frameshift.mtc_filter_report\n",
    "sv = StatsViewable(hpo_mtc_report=report)\n",
    "display(HTML(sv.process(cohort=cohort)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8537a91a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>FRAMESHIFT_VARIANT on NM_001318852.2</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Yes</th>\n",
       "      <th colspan=\"2\" halign=\"left\">No</th>\n",
       "      <th colspan=\"2\" halign=\"left\"></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>p value</th>\n",
       "      <th>Corrected p value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Abnormality of movement [HP:0100022]</th>\n",
       "      <td>0/1</td>\n",
       "      <td>0%</td>\n",
       "      <td>8/13</td>\n",
       "      <td>62%</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cerebral palsy [HP:0100021]</th>\n",
       "      <td>0/1</td>\n",
       "      <td>0%</td>\n",
       "      <td>3/13</td>\n",
       "      <td>23%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cerebral visual impairment [HP:0100704]</th>\n",
       "      <td>0/1</td>\n",
       "      <td>0%</td>\n",
       "      <td>2/6</td>\n",
       "      <td>33%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hypotonia [HP:0001252]</th>\n",
       "      <td>1/1</td>\n",
       "      <td>100%</td>\n",
       "      <td>10/17</td>\n",
       "      <td>59%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Autistic behavior [HP:0000729]</th>\n",
       "      <td>1/1</td>\n",
       "      <td>100%</td>\n",
       "      <td>3/6</td>\n",
       "      <td>50%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gait disturbance [HP:0001288]</th>\n",
       "      <td>0/1</td>\n",
       "      <td>0%</td>\n",
       "      <td>5/10</td>\n",
       "      <td>50%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Seizure [HP:0001250]</th>\n",
       "      <td>0/1</td>\n",
       "      <td>0%</td>\n",
       "      <td>6/17</td>\n",
       "      <td>35%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inability to walk [HP:0002540]</th>\n",
       "      <td>0/1</td>\n",
       "      <td>0%</td>\n",
       "      <td>4/10</td>\n",
       "      <td>40%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "FRAMESHIFT_VARIANT on NM_001318852.2      Yes             No          \\\n",
       "                                        Count Percent  Count Percent   \n",
       "Abnormality of movement [HP:0100022]      0/1      0%   8/13     62%   \n",
       "Cerebral palsy [HP:0100021]               0/1      0%   3/13     23%   \n",
       "Cerebral visual impairment [HP:0100704]   0/1      0%    2/6     33%   \n",
       "Hypotonia [HP:0001252]                    1/1    100%  10/17     59%   \n",
       "Autistic behavior [HP:0000729]            1/1    100%    3/6     50%   \n",
       "Gait disturbance [HP:0001288]             0/1      0%   5/10     50%   \n",
       "Seizure [HP:0001250]                      0/1      0%   6/17     35%   \n",
       "Inability to walk [HP:0002540]            0/1      0%   4/10     40%   \n",
       "\n",
       "FRAMESHIFT_VARIANT on NM_001318852.2                                 \n",
       "                                          p value Corrected p value  \n",
       "Abnormality of movement [HP:0100022]     0.428571               1.0  \n",
       "Cerebral palsy [HP:0100021]              1.000000               1.0  \n",
       "Cerebral visual impairment [HP:0100704]  1.000000               1.0  \n",
       "Hypotonia [HP:0001252]                   1.000000               1.0  \n",
       "Autistic behavior [HP:0000729]           1.000000               1.0  \n",
       "Gait disturbance [HP:0001288]            1.000000               1.0  \n",
       "Seizure [HP:0001250]                     1.000000               1.0  \n",
       "Inability to walk [HP:0002540]           1.000000               1.0  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frameshift = analysis.compare_by_variant_effect(VariantEffect.FRAMESHIFT_VARIANT, tx_id=MAPK8IP3_id)\n",
    "frameshift.summarize(hpo, PatientCategories.YES)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "86707437",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<!DOCTYPE html>\n",
       "<html lang=\"en\">\n",
       "<head>\n",
       "  <meta charset=\"utf-8\">\n",
       "  <title>Cohort</title>\n",
       "  <style>table {\n",
       "    border-collapse: collapse;\n",
       "    margin: 25px 0;\n",
       "      font-size: 0.9em;\n",
       "      font-family: sans-serif;\n",
       "      min-width: 400px;\n",
       "      box-shadow: 0 0 20px rgba(0, 0, 0, 0.15);\n",
       "  }\n",
       "\n",
       "\n",
       "  th {\n",
       "    background-color: #f2f2f2;\n",
       "    border: 1px solid #dddddd;\n",
       "    text-align: left;\n",
       "    padding: 2px;\n",
       "  }\n",
       "\n",
       "  tr {\n",
       "    border: 1px solid #dddddd;\n",
       "  }\n",
       "\n",
       "  td {\n",
       "    padding: 2px;\n",
       "    font-weight: bold;\n",
       "  }\n",
       "\n",
       "  tr:nth-child(even) {\n",
       "    background-color: #f2f2f2;\n",
       "  }\n",
       "\n",
       "  .table td,tr {\n",
       "    text-align: left;\n",
       "}\n",
       "\n",
       ".table td:first-child, tr:first-child {\n",
       "    text-align: left;\n",
       "}​\n",
       "\n",
       "  </style>\n",
       "</head>\n",
       "\n",
       "<body>\n",
       "  <h1>Statistical analysis: fdr_bh (heuristic sampler)</h1>\n",
       "  <p>Performed statistical tests for 8 HPO terms.</p>\n",
       "    <table>\n",
       "      <caption>Using  heuristic sampler, the following numbers of terms were omitted from statistical analysis. </caption>\n",
       "        <tbody>\n",
       "            <tr>\n",
       "                <th>Skipped</th>\n",
       "                <th>Count</th>\n",
       "              </tr>\n",
       "              \n",
       "              <tr>\n",
       "                <td>Skipping term with only 1 observations (not powered for 2x2)</td>\n",
       "                <td>150</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term with only 2 observations (not powered for 2x2)</td>\n",
       "                <td>33</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term with only 3 observations (not powered for 2x2)</td>\n",
       "                <td>27</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term because one genotype had zero observations</td>\n",
       "                <td>22</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term with only 4 observations (not powered for 2x2)</td>\n",
       "                <td>17</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term with only 6 observations (not powered for 2x2)</td>\n",
       "                <td>15</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term because all genotypes have same HPO observed proportions</td>\n",
       "                <td>14</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping top level term</td>\n",
       "                <td>12</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term with only 5 observations (not powered for 2x2)</td>\n",
       "                <td>12</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term because no genotype has more than one observed HPO count</td>\n",
       "                <td>3</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping non phenotype term </td>\n",
       "                <td>2</td>\n",
       "                </tr>\n",
       "            \n",
       "                </tbody>\n",
       "    </table>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cv = StatsViewable(hpo_mtc_report=frameshift.mtc_filter_report)\n",
    "html = cv.process(cohort=cohort)\n",
    "\n",
    "display(HTML(html))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ac309bf",
   "metadata": {},
   "source": [
    "## Correlation analysis for c.1735C>T\n",
    "\n",
    "`NM_001318852.2:c.1735C>T` is the most commonly encountered variant in our cohort. In the following code, we investigate whether this variant displays significant genotype-phenotype correlations.\n",
    "\n",
    "For the purpose of the analysis, the variant is denoted by its key: `16_1762843_1762843_C_T`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8e46c391",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>&gt;=1 allele of the variant 16_1762843_1762843_C_T</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Yes</th>\n",
       "      <th colspan=\"2\" halign=\"left\">No</th>\n",
       "      <th colspan=\"2\" halign=\"left\"></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>p value</th>\n",
       "      <th>Corrected p value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Seizure [HP:0001250]</th>\n",
       "      <td>4/5</td>\n",
       "      <td>80%</td>\n",
       "      <td>2/13</td>\n",
       "      <td>15%</td>\n",
       "      <td>0.021709</td>\n",
       "      <td>0.123810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inability to walk [HP:0002540]</th>\n",
       "      <td>3/3</td>\n",
       "      <td>100%</td>\n",
       "      <td>1/8</td>\n",
       "      <td>12%</td>\n",
       "      <td>0.024242</td>\n",
       "      <td>0.123810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Autistic behavior [HP:0000729]</th>\n",
       "      <td>0/3</td>\n",
       "      <td>0%</td>\n",
       "      <td>4/4</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.123810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gait disturbance [HP:0001288]</th>\n",
       "      <td>3/3</td>\n",
       "      <td>100%</td>\n",
       "      <td>2/8</td>\n",
       "      <td>25%</td>\n",
       "      <td>0.060606</td>\n",
       "      <td>0.196970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Abnormality of movement [HP:0100022]</th>\n",
       "      <td>4/4</td>\n",
       "      <td>100%</td>\n",
       "      <td>4/10</td>\n",
       "      <td>40%</td>\n",
       "      <td>0.084915</td>\n",
       "      <td>0.220779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hypotonia [HP:0001252]</th>\n",
       "      <td>2/5</td>\n",
       "      <td>40%</td>\n",
       "      <td>9/13</td>\n",
       "      <td>69%</td>\n",
       "      <td>0.325980</td>\n",
       "      <td>0.557143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Thin corpus callosum [HP:0033725]</th>\n",
       "      <td>6/6</td>\n",
       "      <td>100%</td>\n",
       "      <td>3/4</td>\n",
       "      <td>75%</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.557143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Spastic diplegia [HP:0001264]</th>\n",
       "      <td>4/4</td>\n",
       "      <td>100%</td>\n",
       "      <td>2/3</td>\n",
       "      <td>67%</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.557143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Short philtrum [HP:0000322]</th>\n",
       "      <td>0/3</td>\n",
       "      <td>0%</td>\n",
       "      <td>2/4</td>\n",
       "      <td>50%</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.557143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Anteverted nares [HP:0000463]</th>\n",
       "      <td>0/3</td>\n",
       "      <td>0%</td>\n",
       "      <td>2/4</td>\n",
       "      <td>50%</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.557143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prominent nasal bridge [HP:0000426]</th>\n",
       "      <td>1/3</td>\n",
       "      <td>33%</td>\n",
       "      <td>3/4</td>\n",
       "      <td>75%</td>\n",
       "      <td>0.485714</td>\n",
       "      <td>0.567460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Autism [HP:0000717]</th>\n",
       "      <td>0/5</td>\n",
       "      <td>0%</td>\n",
       "      <td>2/10</td>\n",
       "      <td>20%</td>\n",
       "      <td>0.523810</td>\n",
       "      <td>0.567460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cerebral palsy [HP:0100021]</th>\n",
       "      <td>1/4</td>\n",
       "      <td>25%</td>\n",
       "      <td>2/10</td>\n",
       "      <td>20%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       ">=1 allele of the variant 16_1762843_1762843_C_T   Yes            No          \\\n",
       "                                                 Count Percent Count Percent   \n",
       "Seizure [HP:0001250]                               4/5     80%  2/13     15%   \n",
       "Inability to walk [HP:0002540]                     3/3    100%   1/8     12%   \n",
       "Autistic behavior [HP:0000729]                     0/3      0%   4/4    100%   \n",
       "Gait disturbance [HP:0001288]                      3/3    100%   2/8     25%   \n",
       "Abnormality of movement [HP:0100022]               4/4    100%  4/10     40%   \n",
       "Hypotonia [HP:0001252]                             2/5     40%  9/13     69%   \n",
       "Thin corpus callosum [HP:0033725]                  6/6    100%   3/4     75%   \n",
       "Spastic diplegia [HP:0001264]                      4/4    100%   2/3     67%   \n",
       "Short philtrum [HP:0000322]                        0/3      0%   2/4     50%   \n",
       "Anteverted nares [HP:0000463]                      0/3      0%   2/4     50%   \n",
       "Prominent nasal bridge [HP:0000426]                1/3     33%   3/4     75%   \n",
       "Autism [HP:0000717]                                0/5      0%  2/10     20%   \n",
       "Cerebral palsy [HP:0100021]                        1/4     25%  2/10     20%   \n",
       "\n",
       ">=1 allele of the variant 16_1762843_1762843_C_T                              \n",
       "                                                   p value Corrected p value  \n",
       "Seizure [HP:0001250]                              0.021709          0.123810  \n",
       "Inability to walk [HP:0002540]                    0.024242          0.123810  \n",
       "Autistic behavior [HP:0000729]                    0.028571          0.123810  \n",
       "Gait disturbance [HP:0001288]                     0.060606          0.196970  \n",
       "Abnormality of movement [HP:0100022]              0.084915          0.220779  \n",
       "Hypotonia [HP:0001252]                            0.325980          0.557143  \n",
       "Thin corpus callosum [HP:0033725]                 0.400000          0.557143  \n",
       "Spastic diplegia [HP:0001264]                     0.428571          0.557143  \n",
       "Short philtrum [HP:0000322]                       0.428571          0.557143  \n",
       "Anteverted nares [HP:0000463]                     0.428571          0.557143  \n",
       "Prominent nasal bridge [HP:0000426]               0.485714          0.567460  \n",
       "Autism [HP:0000717]                               0.523810          0.567460  \n",
       "Cerebral palsy [HP:0100021]                       1.000000          1.000000  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_key = '16_1762843_1762843_C_T'\n",
    "by_variant = analysis.compare_by_variant_key(variant_key=variant_key)\n",
    "by_variant.summarize(hpo, PatientCategories.YES)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4045f858",
   "metadata": {},
   "source": [
    "TODO - finalize!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "exploratory",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
