{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9443554a",
   "metadata": {},
   "source": [
    "# KBG Syndrome\n",
    "\n",
    "[KBG syndrome (KBGS)](https://omim.org/entry/148050) is caused by heterozygous mutation in the ANKRD11 gene. In this notebook, we have used\n",
    "[pyphetools](https://github.com/monarch-initiative/pyphetools) to parse the clinical data included in the supplemental files of\n",
    "[Martinez-Cayuelas E, et al. Clinical description, molecular delineation and genotype-phenotype correlation in 340 patients with KBG syndrome: addition of 67 new patients](https://pubmed.ncbi.nlm.nih.gov/36446582).\n",
    "\n",
    "The authors identified a significantly higher frequency of patients with a triangular face in carriers of sequence variants compared to CNVs. Other associations found were short stature and variants in exon 9, a lower incidence of ID/ADHD/ASD in carriers of the c.1903_1907del variant and the size of the deletion, in CNV carriers, with the presence of macrodontia and hand anomalies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cd2a5b1b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/robin/PycharmProjects/genophenocorr/gpc_venv/lib/python3.9/site-packages/urllib3/__init__.py:35: NotOpenSSLWarning: urllib3 v2 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded HPO v2023-10-09\n",
      "Using genophenocorr version 0.1.1dev\n"
     ]
    }
   ],
   "source": [
    "import genophenocorr\n",
    "import hpotk\n",
    "\n",
    "store = hpotk.configure_ontology_store()\n",
    "hpo = store.load_minimal_hpo(release='v2023-10-09')\n",
    "print(f'Loaded HPO v{hpo.version}')\n",
    "print(f\"Using genophenocorr version {genophenocorr.__version__}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "566b13c6",
   "metadata": {},
   "source": [
    "## Settings\n",
    "Specify the transcript to be used to encode the variants (the phenopackets contain VCF representations of small variants)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97cd3439",
   "metadata": {},
   "source": [
    "### Pick transcript\n",
    "\n",
    "We choose the [MANE Select](https://www.ncbi.nlm.nih.gov/nuccore/NM_013275.6) transcript for *ANKRD11*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "86b84c66",
   "metadata": {},
   "outputs": [],
   "source": [
    "tx_id = 'NM_013275.6'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2bf0ff1",
   "metadata": {},
   "source": [
    "## Load Phenopackets\n",
    "\n",
    "We will load phenopacket JSON files located in `phenopackets` folder that is next to the notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d7623bcb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Patients Created: 100%|██████████| 337/337 [00:00<00:00, 454.45it/s]\n",
      "Validated under none policy\n",
      "Showing errors and warnings\n",
      "337 phenopacket(s) found at `phenopackets`\n",
      "  patient #0\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P2[PMID_36446582_Novara_2017_P2]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P2 has no variants to work with\n",
      "  patient #1\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P13[PMID_36446582_Goldenberg2016_P13]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P13 has no variants to work with\n",
      "  patient #3\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Ockeloen2015_P20[PMID_36446582_Ockeloen2015_P20]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Ockeloen2015_P20 has no variants to work with\n",
      "  patient #7\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P22[PMID_36446582_Kutkowska_Kazmierczak2021_P22]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska_Kazmierczak2021_P22 has no variants to work with\n",
      "  patient #8\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P10[PMID_36446582_Novara_2017_P10]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P10 has no variants to work with\n",
      "  patient #12\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Gnazzo, 2020_P31[PMID_36446582_Gnazzo_2020_P31]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Gnazzo_2020_P31 has no variants to work with\n",
      "  patient #16\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P29[PMID_36446582_Goldenberg2016_P29]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P29 has no variants to work with\n",
      "  patient #17\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P18[PMID_36446582_Kutkowska_Kazmierczak2021_P18]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska_Kazmierczak2021_P18 has no variants to work with\n",
      "  patient #22\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG36[PMID_36446582_KBG36]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG36 has no variants to work with\n",
      "  patient #27\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P1[PMID_36446582_Goldenberg2016_P1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P1 has no variants to work with\n",
      "  patient #32\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P33[PMID_36446582_Goldenberg2016_P33]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P33 has no variants to work with\n",
      "  patient #34\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P14[PMID_36446582_Kutkowska_Kazmierczak2021_P14]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska_Kazmierczak2021_P14 has no variants to work with\n",
      "  patient #39\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P15[PMID_36446582_Kutkowska_Kazmierczak2021_P15]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska_Kazmierczak2021_P15 has no variants to work with\n",
      "  patient #42\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P32[PMID_36446582_Goldenberg2016_P32]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P32 has no variants to work with\n",
      "  patient #43\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Bucerzan2020[PMID_36446582_Bucerzan2020]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Bucerzan2020 has no variants to work with\n",
      "  patient #46\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P24[PMID_36446582_Goldenberg2016_P24]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P24 has no variants to work with\n",
      "  patient #50\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Miyatake, 2013[PMID_36446582_Miyatake_2013]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Miyatake_2013 has no variants to work with\n",
      "  patient #56\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG1[PMID_36446582_KBG1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG1 has no variants to work with\n",
      "  patient #60\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Sacharow, 2012_P2[PMID_36446582_Sacharow_2012_P2]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Sacharow_2012_P2 has no variants to work with\n",
      "  patient #62\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P19[PMID_36446582_Kutkowska_Kazmierczak2021_P19]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska_Kazmierczak2021_P19 has no variants to work with\n",
      "  patient #64\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P28[PMID_36446582_Goldenberg2016_P28]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P28 has no variants to work with\n",
      "  patient #68\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Gnazzo, 2020_P30[PMID_36446582_Gnazzo_2020_P30]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Gnazzo_2020_P30 has no variants to work with\n",
      "  patient #71\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P23[PMID_36446582_Kutkowska_Kazmierczak2021_P23]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska_Kazmierczak2021_P23 has no variants to work with\n",
      "  patient #72\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P11[PMID_36446582_Novara_2017_P11]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P11 has no variants to work with\n",
      "  patient #76\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P12[PMID_36446582_Goldenberg2016_P12]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P12 has no variants to work with\n",
      "  patient #77\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P3[PMID_36446582_Novara_2017_P3]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P3 has no variants to work with\n",
      "  patient #78\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Khalifa, 2013_P1B[PMID_36446582_Khalifa_2013_P1B]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Khalifa_2013_P1B has no variants to work with\n",
      "  patient #79\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Isrie, 2012_P2[PMID_36446582_Isrie_2012_P2]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Isrie_2012_P2 has no variants to work with\n",
      "  patient #80\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P19[PMID_36446582_Goldenberg2016_P19]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P19 has no variants to work with\n",
      "  patient #83\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P8[PMID_36446582_Novara_2017_P8]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P8 has no variants to work with\n",
      "  patient #85\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG30[PMID_36446582_KBG30]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG30 has no variants to work with\n",
      "  patient #91\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Scarano, 2013_P11[PMID_36446582_Scarano_2013_P11]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Scarano_2013_P11 has no variants to work with\n",
      "  patient #93\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Crippa2015_P1[PMID_36446582_Crippa2015_P1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Crippa2015_P1 has no variants to work with\n",
      "  patient #100\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P4[PMID_36446582_Novara_2017_P4]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P4 has no variants to work with\n",
      "  patient #103\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Parenti2021_P23[PMID_36446582_Parenti2021_P23]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Parenti2021_P23 has no variants to work with\n",
      "  patient #135\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient VanDongen2019_P5[PMID_36446582_VanDongen2019_P5]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_VanDongen2019_P5 has no variants to work with\n",
      "  patient #140\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P5[PMID_36446582_Novara_2017_P5]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P5 has no variants to work with\n",
      "  patient #142\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Willemsen2010_P1[PMID_36446582_Willemsen2010_P1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Willemsen2010_P1 has no variants to work with\n",
      "  patient #147\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P22[PMID_36446582_Goldenberg2016_P22]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P22 has no variants to work with\n",
      "  patient #148\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Scarano, 2013_P10[PMID_36446582_Scarano_2013_P10]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Scarano_2013_P10 has no variants to work with\n",
      "  patient #156\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P9[PMID_36446582_Novara_2017_P9]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P9 has no variants to work with\n",
      "  patient #159\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P18[PMID_36446582_Goldenberg2016_P18]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P18 has no variants to work with\n",
      "  patient #165\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Willemsen2010_P2[PMID_36446582_Willemsen2010_P2]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Willemsen2010_P2 has no variants to work with\n",
      "  patient #168\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Spengler, 2013[PMID_36446582_Spengler_2013]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Spengler_2013 has no variants to work with\n",
      "  patient #173\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Youngs2011[PMID_36446582_Youngs2011]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Youngs2011 has no variants to work with\n",
      "  patient #174\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P21[PMID_36446582_Goldenberg2016_P21]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P21 has no variants to work with\n",
      "  patient #175\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Crippa2015_P3[PMID_36446582_Crippa2015_P3]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Crippa2015_P3 has no variants to work with\n",
      "  patient #186\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG32[PMID_36446582_KBG32]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG32 has no variants to work with\n",
      "  patient #200\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Palumbo 2016[PMID_36446582_Palumbo_2016]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Palumbo_2016 has no variants to work with\n",
      "  patient #210\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P7[PMID_36446582_Novara_2017_P7]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P7 has no variants to work with\n",
      "  patient #214\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient VanDongen2019_P7[PMID_36446582_VanDongen2019_P7]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_VanDongen2019_P7 has no variants to work with\n",
      "  patient #223\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG52[PMID_36446582_KBG52]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG52 has no variants to work with\n",
      "  patient #233\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG33[PMID_36446582_KBG33]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG33 has no variants to work with\n",
      "  patient #238\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P4[PMID_36446582_Goldenberg2016_P4]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P4 has no variants to work with\n",
      "  patient #239\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Scarano, 2013_P12[PMID_36446582_Scarano_2013_P12]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Scarano_2013_P12 has no variants to work with\n",
      "  patient #241\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P20[PMID_36446582_Goldenberg2016_P20]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P20 has no variants to work with\n",
      "  patient #242\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Crippa2015_P2[PMID_36446582_Crippa2015_P2]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Crippa2015_P2 has no variants to work with\n",
      "  patient #244\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P36[PMID_36446582_Goldenberg2016_P36]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P36 has no variants to work with\n",
      "  patient #248\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Willemsen2010_P3[PMID_36446582_Willemsen2010_P3]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Willemsen2010_P3 has no variants to work with\n",
      "  patient #253\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Sacharow, 2012_P1[PMID_36446582_Sacharow_2012_P1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Sacharow_2012_P1 has no variants to work with\n",
      "  patient #256\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Srivastava, 2017_P1[PMID_36446582_Srivastava_2017_P1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Srivastava_2017_P1 has no variants to work with\n",
      "  patient #266\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P12[PMID_36446582_Novara_2017_P12]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P12 has no variants to work with\n",
      "  patient #268\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P20[PMID_36446582_Kutkowska_Kazmierczak2021_P20]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska_Kazmierczak2021_P20 has no variants to work with\n",
      "  patient #274\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Isrie, 2012_P1[PMID_36446582_Isrie_2012_P1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Isrie_2012_P1 has no variants to work with\n",
      "  patient #275\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Khalifa, 2013_P1A[PMID_36446582_Khalifa_2013_P1A]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Khalifa_2013_P1A has no variants to work with\n",
      "  patient #278\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Willemsen2010_P4[PMID_36446582_Willemsen2010_P4]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Willemsen2010_P4 has no variants to work with\n",
      "  patient #279\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Gnazzo, 2020_P29[PMID_36446582_Gnazzo_2020_P29]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Gnazzo_2020_P29 has no variants to work with\n",
      "  patient #282\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P16[PMID_36446582_Kutkowska_Kazmierczak2021_P16]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska_Kazmierczak2021_P16 has no variants to work with\n",
      "  patient #288\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P3[PMID_36446582_Goldenberg2016_P3]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P3 has no variants to work with\n",
      "  patient #299\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Lim2014[PMID_36446582_Lim2014]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Lim2014 has no variants to work with\n",
      "  patient #300\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG35[PMID_36446582_KBG35]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG35 has no variants to work with\n",
      "  patient #304\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P2[PMID_36446582_Goldenberg2016_P2]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P2 has no variants to work with\n",
      "  patient #306\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P26[PMID_36446582_Goldenberg2016_P26]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P26 has no variants to work with\n",
      "  patient #309\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P17[PMID_36446582_Kutkowska_Kazmierczak2021_P17]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska_Kazmierczak2021_P17 has no variants to work with\n",
      "  patient #316\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P1[PMID_36446582_Novara_2017_P1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P1 has no variants to work with\n",
      "  patient #317\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P10[PMID_36446582_Goldenberg2016_P10]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P10 has no variants to work with\n",
      "  patient #320\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG39[PMID_36446582_KBG39]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG39 has no variants to work with\n",
      "  patient #323\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P21[PMID_36446582_Kutkowska_Kazmierczak2021_P21]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska_Kazmierczak2021_P21 has no variants to work with\n",
      "  patient #326\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Behnert, 2018[PMID_36446582_Behnert_2018]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Behnert_2018 has no variants to work with\n",
      "  patient #335\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG15[PMID_36446582_KBG15]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG15 has no variants to work with\n"
     ]
    }
   ],
   "source": [
    "from genophenocorr.preprocessing import configure_caching_cohort_creator, load_phenopacket_folder\n",
    "\n",
    "fpath_phenopackets = 'phenopackets'\n",
    "cohort_creator = configure_caching_cohort_creator(hpo, timeout=20)\n",
    "cohort = load_phenopacket_folder(fpath_phenopackets, cohort_creator)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f442ac4",
   "metadata": {},
   "source": [
    "## Summarize the cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "070a06ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<!DOCTYPE html>\n",
       "<html lang=\"en\">\n",
       "<head>\n",
       "  <meta charset=\"utf-8\">\n",
       "  <title>Cohort</title>\n",
       "  <style>table {\n",
       "    border-collapse: collapse;\n",
       "    margin: 25px 0;\n",
       "      font-size: 0.9em;\n",
       "      font-family: sans-serif;\n",
       "      min-width: 400px;\n",
       "      box-shadow: 0 0 20px rgba(0, 0, 0, 0.15);\n",
       "  }\n",
       "\n",
       "\n",
       "  th {\n",
       "    background-color: #f2f2f2;\n",
       "    border: 1px solid #dddddd;\n",
       "    text-align: left;\n",
       "    padding: 2px;\n",
       "  }\n",
       "\n",
       "  tr {\n",
       "    border: 1px solid #dddddd;\n",
       "  }\n",
       "\n",
       "  td {\n",
       "    padding: 2px;\n",
       "    font-weight: bold;\n",
       "  }\n",
       "\n",
       "  tr:nth-child(even) {\n",
       "    background-color: #f2f2f2;\n",
       "  }\n",
       "\n",
       "  .table td,tr {\n",
       "    text-align: right;\n",
       "}\n",
       "\n",
       ".table td:first-child, tr:first-child {\n",
       "    text-align: left;\n",
       "}​\n",
       "\n",
       "  </style>\n",
       "</head>\n",
       "\n",
       "<body>\n",
       "  <h1>genophenocorr cohort analysis</h1>\n",
       "  <p>Successfully ingested 337 phenopackets.</p>\n",
       "  \n",
       "  <p>Not able to ingest 80 individuals.</p>\n",
       "  \n",
       "  <p>A total of 93 HPO terms were used to annotated the cohort. </p>\n",
       "    <h3>Top 10 HPO Terms</h3>\n",
       "    <table>\n",
       "        <tbody>\n",
       "            <tr>\n",
       "                <th>HPO Term</th>\n",
       "                <th>ID</th>\n",
       "                <th>Annotation Count</th>\n",
       "              </tr>\n",
       "              \n",
       "              <tr>\n",
       "                <td>Macrodontia</td>\n",
       "                <td>HP:0001572</td>\n",
       "                <td>211</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Intellectual disability</td>\n",
       "                <td>HP:0001249</td>\n",
       "                <td>194</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Abnormality of the hand</td>\n",
       "                <td>HP:0001155</td>\n",
       "                <td>189</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Global developmental delay</td>\n",
       "                <td>HP:0001263</td>\n",
       "                <td>176</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Delayed speech and language development</td>\n",
       "                <td>HP:0000750</td>\n",
       "                <td>160</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Short stature</td>\n",
       "                <td>HP:0004322</td>\n",
       "                <td>150</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Thick eyebrow</td>\n",
       "                <td>HP:0000574</td>\n",
       "                <td>126</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Long philtrum</td>\n",
       "                <td>HP:0000343</td>\n",
       "                <td>121</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Bulbous nose</td>\n",
       "                <td>HP:0000414</td>\n",
       "                <td>89</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Triangular face</td>\n",
       "                <td>HP:0000325</td>\n",
       "                <td>83</td>\n",
       "                </tr>\n",
       "            \n",
       "                </tbody>\n",
       "    </table>\n",
       " <h3>Variants</h3>\n",
       " <p>A total of 257 unique variants were identified in the cohort.</p>\n",
       " <h3>Top 10 Variants</h3>\n",
       " <table>\n",
       "        <tbody>\n",
       "            <tr>\n",
       "                <th>Variant</th>\n",
       "                <th>Variant name</th>\n",
       "                <th>Variant Count</th>\n",
       "              </tr>\n",
       "              \n",
       "              <tr>\n",
       "                <td>16_89284634_89284639_GTGTTT_G</td>\n",
       "                <td>todo</td>\n",
       "                <td>34</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>16_89284129_89284134_CTTTTT_C</td>\n",
       "                <td>todo</td>\n",
       "                <td>10</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>16_89284140_89284144_TTTTC_T</td>\n",
       "                <td>todo</td>\n",
       "                <td>9</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>16_89285157_89285161_GTTTC_G</td>\n",
       "                <td>todo</td>\n",
       "                <td>8</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>16_89275180_89275180_A_AG</td>\n",
       "                <td>todo</td>\n",
       "                <td>5</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>16_89279749_89279749_C_CG</td>\n",
       "                <td>todo</td>\n",
       "                <td>5</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>16_89284345_89284345_G_A</td>\n",
       "                <td>todo</td>\n",
       "                <td>3</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>16_89283314_89283318_CCTTT_C</td>\n",
       "                <td>todo</td>\n",
       "                <td>3</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>16_89284565_89284565_G_C</td>\n",
       "                <td>todo</td>\n",
       "                <td>3</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>16_89284358_89284360_GAT_G</td>\n",
       "                <td>todo</td>\n",
       "                <td>3</td>\n",
       "                </tr>\n",
       "            \n",
       "                </tbody>\n",
       "    </table>\n",
       "    <h3>Diseases</h3>\n",
       "    <table>\n",
       "        <tbody>\n",
       "            <tr>\n",
       "                <th>Disease</th>\n",
       "                <th>Annotation Count</th>\n",
       "              </tr>\n",
       "              \n",
       "              <tr>\n",
       "                <td>OMIM:148050</td>\n",
       "                <td>337</td>\n",
       "                </tr>\n",
       "            \n",
       "           </tbody>\n",
       "    </table>\n",
       "    \n",
       "    <h3>Variant categories</h3>\n",
       "    <table>\n",
       "        <tbody>\n",
       "            <tr>\n",
       "                <th>Variant effect</th>\n",
       "                <th>Annotation Count</th>\n",
       "              </tr>\n",
       "              \n",
       "              <tr>\n",
       "                <td>FRAMESHIFT_VARIANT</td>\n",
       "                <td>175</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>STOP_GAINED</td>\n",
       "                <td>67</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>MISSENSE_VARIANT</td>\n",
       "                <td>7</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>SPLICE_ACCEPTOR_VARIANT</td>\n",
       "                <td>4</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>SPLICE_REGION_VARIANT</td>\n",
       "                <td>2</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>SPLICE_DONOR_VARIANT</td>\n",
       "                <td>3</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>INFRAME_DELETION</td>\n",
       "                <td>2</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>CODING_SEQUENCE_VARIANT</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "           </tbody>\n",
       "    </table>\n",
       "  \n",
       "\n",
       "    has_transcript\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.display import display, HTML\n",
    "from genophenocorr.view import CohortViewable\n",
    "\n",
    "cv = CohortViewable(hpo=hpo, transcript_id=tx_id)\n",
    "html = cv.process(cohort=cohort)\n",
    "\n",
    "display(HTML(html))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbbc2cc1",
   "metadata": {},
   "source": [
    "## Configure the analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ec6c204f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from genophenocorr.analysis import configure_cohort_analysis, CohortAnalysisConfiguration\n",
    "from genophenocorr.analysis.predicate import PatientCategories\n",
    "\n",
    "analysis_config = CohortAnalysisConfiguration()\n",
    "analysis_config.missing_implies_excluded = True\n",
    "analysis_config.pval_correction = 'fdr_bh'\n",
    "analysis_config.min_perc_patients_w_hpo = 0.1\n",
    "analysis = configure_cohort_analysis(cohort, hpo, config=analysis_config)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ff62a78",
   "metadata": {},
   "source": [
    "Test for presence of genotype-phenotype correlations between frameshift variants vs. others."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "91809a38",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>FRAMESHIFT_VARIANT on NM_013275.6</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Yes</th>\n",
       "      <th colspan=\"2\" halign=\"left\">No</th>\n",
       "      <th colspan=\"2\" halign=\"left\"></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>p value</th>\n",
       "      <th>Corrected p value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Abnormality of the hand [HP:0001155]</th>\n",
       "      <td>95/144</td>\n",
       "      <td>66%</td>\n",
       "      <td>60/71</td>\n",
       "      <td>85%</td>\n",
       "      <td>0.005661</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EEG abnormality [HP:0002353]</th>\n",
       "      <td>7/33</td>\n",
       "      <td>21%</td>\n",
       "      <td>9/16</td>\n",
       "      <td>56%</td>\n",
       "      <td>0.022884</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Feeding difficulties [HP:0011968]</th>\n",
       "      <td>33/89</td>\n",
       "      <td>37%</td>\n",
       "      <td>26/45</td>\n",
       "      <td>58%</td>\n",
       "      <td>0.027584</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Low anterior hairline [HP:0000294]</th>\n",
       "      <td>40/58</td>\n",
       "      <td>69%</td>\n",
       "      <td>15/30</td>\n",
       "      <td>50%</td>\n",
       "      <td>0.105274</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Intellectual disability [HP:0001249]</th>\n",
       "      <td>99/119</td>\n",
       "      <td>83%</td>\n",
       "      <td>59/64</td>\n",
       "      <td>92%</td>\n",
       "      <td>0.115195</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Long philtrum [HP:0000343]</th>\n",
       "      <td>66/82</td>\n",
       "      <td>80%</td>\n",
       "      <td>38/48</td>\n",
       "      <td>79%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Abnormal eyebrow morphology [HP:0000534]</th>\n",
       "      <td>77/77</td>\n",
       "      <td>100%</td>\n",
       "      <td>39/39</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Abnormality of the digestive system [HP:0025031]</th>\n",
       "      <td>35/35</td>\n",
       "      <td>100%</td>\n",
       "      <td>29/29</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hyperactivity [HP:0000752]</th>\n",
       "      <td>34/34</td>\n",
       "      <td>100%</td>\n",
       "      <td>19/19</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Generalized non-motor (absence) seizure [HP:0002121]</th>\n",
       "      <td>1/4</td>\n",
       "      <td>25%</td>\n",
       "      <td>2/4</td>\n",
       "      <td>50%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>190 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "FRAMESHIFT_VARIANT on NM_013275.6                      Yes             No  \\\n",
       "                                                     Count Percent  Count   \n",
       "Abnormality of the hand [HP:0001155]                95/144     66%  60/71   \n",
       "EEG abnormality [HP:0002353]                          7/33     21%   9/16   \n",
       "Feeding difficulties [HP:0011968]                    33/89     37%  26/45   \n",
       "Low anterior hairline [HP:0000294]                   40/58     69%  15/30   \n",
       "Intellectual disability [HP:0001249]                99/119     83%  59/64   \n",
       "...                                                    ...     ...    ...   \n",
       "Long philtrum [HP:0000343]                           66/82     80%  38/48   \n",
       "Abnormal eyebrow morphology [HP:0000534]             77/77    100%  39/39   \n",
       "Abnormality of the digestive system [HP:0025031]     35/35    100%  29/29   \n",
       "Hyperactivity [HP:0000752]                           34/34    100%  19/19   \n",
       "Generalized non-motor (absence) seizure [HP:000...     1/4     25%    2/4   \n",
       "\n",
       "FRAMESHIFT_VARIANT on NM_013275.6                                     \\\n",
       "                                                   Percent   p value   \n",
       "Abnormality of the hand [HP:0001155]                   85%  0.005661   \n",
       "EEG abnormality [HP:0002353]                           56%  0.022884   \n",
       "Feeding difficulties [HP:0011968]                      58%  0.027584   \n",
       "Low anterior hairline [HP:0000294]                     50%  0.105274   \n",
       "Intellectual disability [HP:0001249]                   92%  0.115195   \n",
       "...                                                    ...       ...   \n",
       "Long philtrum [HP:0000343]                             79%  1.000000   \n",
       "Abnormal eyebrow morphology [HP:0000534]              100%  1.000000   \n",
       "Abnormality of the digestive system [HP:0025031]      100%  1.000000   \n",
       "Hyperactivity [HP:0000752]                            100%  1.000000   \n",
       "Generalized non-motor (absence) seizure [HP:000...     50%  1.000000   \n",
       "\n",
       "FRAMESHIFT_VARIANT on NM_013275.6                                     \n",
       "                                                   Corrected p value  \n",
       "Abnormality of the hand [HP:0001155]                             1.0  \n",
       "EEG abnormality [HP:0002353]                                     1.0  \n",
       "Feeding difficulties [HP:0011968]                                1.0  \n",
       "Low anterior hairline [HP:0000294]                               1.0  \n",
       "Intellectual disability [HP:0001249]                             1.0  \n",
       "...                                                              ...  \n",
       "Long philtrum [HP:0000343]                                       1.0  \n",
       "Abnormal eyebrow morphology [HP:0000534]                         1.0  \n",
       "Abnormality of the digestive system [HP:0025031]                 1.0  \n",
       "Hyperactivity [HP:0000752]                                       1.0  \n",
       "Generalized non-motor (absence) seizure [HP:000...               1.0  \n",
       "\n",
       "[190 rows x 6 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from genophenocorr.model import VariantEffect\n",
    "\n",
    "frameshift = analysis.compare_by_variant_effect(VariantEffect.FRAMESHIFT_VARIANT, tx_id=tx_id)\n",
    "frameshift.summarize(hpo, PatientCategories.YES)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "67266357",
   "metadata": {},
   "outputs": [],
   "source": [
    "from genophenocorr.analysis import configure_cohort_analysis, CohortAnalysisConfiguration\n",
    "from genophenocorr.analysis.predicate import PatientCategories\n",
    "\n",
    "analysis_config = CohortAnalysisConfiguration()\n",
    "analysis_config.missing_implies_excluded = True\n",
    "analysis_config.pval_correction = 'fdr_bh'\n",
    "analysis_config.min_perc_patients_w_hpo = 0.1\n",
    "analysis_config.heuristic_strategy()\n",
    "analysis = configure_cohort_analysis(cohort, hpo, config=analysis_config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3d3bdfea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>FRAMESHIFT_VARIANT on NM_013275.6</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Yes</th>\n",
       "      <th colspan=\"2\" halign=\"left\">No</th>\n",
       "      <th colspan=\"2\" halign=\"left\"></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>p value</th>\n",
       "      <th>Corrected p value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Low anterior hairline [HP:0000294]</th>\n",
       "      <td>40/58</td>\n",
       "      <td>69%</td>\n",
       "      <td>15/30</td>\n",
       "      <td>50%</td>\n",
       "      <td>0.105274</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Microretrognathia [HP:0000308]</th>\n",
       "      <td>13/36</td>\n",
       "      <td>36%</td>\n",
       "      <td>3/16</td>\n",
       "      <td>19%</td>\n",
       "      <td>0.330525</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sleep abnormality [HP:0002360]</th>\n",
       "      <td>15/40</td>\n",
       "      <td>38%</td>\n",
       "      <td>4/17</td>\n",
       "      <td>24%</td>\n",
       "      <td>0.370187</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Exaggerated cupid's bow [HP:0002263]</th>\n",
       "      <td>11/35</td>\n",
       "      <td>31%</td>\n",
       "      <td>2/12</td>\n",
       "      <td>17%</td>\n",
       "      <td>0.464459</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Thick eyebrow [HP:0000574]</th>\n",
       "      <td>68/82</td>\n",
       "      <td>83%</td>\n",
       "      <td>38/49</td>\n",
       "      <td>78%</td>\n",
       "      <td>0.494742</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Abnormality of the forehead [HP:0000290]</th>\n",
       "      <td>40/40</td>\n",
       "      <td>100%</td>\n",
       "      <td>15/15</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Conductive hearing impairment [HP:0000405]</th>\n",
       "      <td>18/18</td>\n",
       "      <td>100%</td>\n",
       "      <td>8/8</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Functional abnormality of the middle ear [HP:0011452]</th>\n",
       "      <td>18/18</td>\n",
       "      <td>100%</td>\n",
       "      <td>8/8</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Abnormality of the chin [HP:0000306]</th>\n",
       "      <td>8/8</td>\n",
       "      <td>100%</td>\n",
       "      <td>5/5</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prominent nose [HP:0000448]</th>\n",
       "      <td>12/28</td>\n",
       "      <td>43%</td>\n",
       "      <td>5/13</td>\n",
       "      <td>38%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>127 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "FRAMESHIFT_VARIANT on NM_013275.6                     Yes             No  \\\n",
       "                                                    Count Percent  Count   \n",
       "Low anterior hairline [HP:0000294]                  40/58     69%  15/30   \n",
       "Microretrognathia [HP:0000308]                      13/36     36%   3/16   \n",
       "Sleep abnormality [HP:0002360]                      15/40     38%   4/17   \n",
       "Exaggerated cupid's bow [HP:0002263]                11/35     31%   2/12   \n",
       "Thick eyebrow [HP:0000574]                          68/82     83%  38/49   \n",
       "...                                                   ...     ...    ...   \n",
       "Abnormality of the forehead [HP:0000290]            40/40    100%  15/15   \n",
       "Conductive hearing impairment [HP:0000405]          18/18    100%    8/8   \n",
       "Functional abnormality of the middle ear [HP:00...  18/18    100%    8/8   \n",
       "Abnormality of the chin [HP:0000306]                  8/8    100%    5/5   \n",
       "Prominent nose [HP:0000448]                         12/28     43%   5/13   \n",
       "\n",
       "FRAMESHIFT_VARIANT on NM_013275.6                                     \\\n",
       "                                                   Percent   p value   \n",
       "Low anterior hairline [HP:0000294]                     50%  0.105274   \n",
       "Microretrognathia [HP:0000308]                         19%  0.330525   \n",
       "Sleep abnormality [HP:0002360]                         24%  0.370187   \n",
       "Exaggerated cupid's bow [HP:0002263]                   17%  0.464459   \n",
       "Thick eyebrow [HP:0000574]                             78%  0.494742   \n",
       "...                                                    ...       ...   \n",
       "Abnormality of the forehead [HP:0000290]              100%  1.000000   \n",
       "Conductive hearing impairment [HP:0000405]            100%  1.000000   \n",
       "Functional abnormality of the middle ear [HP:00...    100%  1.000000   \n",
       "Abnormality of the chin [HP:0000306]                  100%  1.000000   \n",
       "Prominent nose [HP:0000448]                            38%  1.000000   \n",
       "\n",
       "FRAMESHIFT_VARIANT on NM_013275.6                                     \n",
       "                                                   Corrected p value  \n",
       "Low anterior hairline [HP:0000294]                               1.0  \n",
       "Microretrognathia [HP:0000308]                                   1.0  \n",
       "Sleep abnormality [HP:0002360]                                   1.0  \n",
       "Exaggerated cupid's bow [HP:0002263]                             1.0  \n",
       "Thick eyebrow [HP:0000574]                                       1.0  \n",
       "...                                                              ...  \n",
       "Abnormality of the forehead [HP:0000290]                         1.0  \n",
       "Conductive hearing impairment [HP:0000405]                       1.0  \n",
       "Functional abnormality of the middle ear [HP:00...               1.0  \n",
       "Abnormality of the chin [HP:0000306]                             1.0  \n",
       "Prominent nose [HP:0000448]                                      1.0  \n",
       "\n",
       "[127 rows x 6 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frameshift = analysis.compare_by_variant_effect(VariantEffect.FRAMESHIFT_VARIANT, tx_id=tx_id)\n",
    "frameshift.summarize(hpo, PatientCategories.YES)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9841f6f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from genophenocorr.view import StatsViewable\n",
    "sv = StatsViewable(filter_method_name=analysis_config.mtc_strategy, mtc_name=\"nana\", filter_results_map=frameshift.mtc_filter_report,term_count=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "eb0dc510",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'HpoMtcReport' object has no attribute 'items'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[13], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m display(HTML(\u001b[43msv\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mprocess\u001b[49m\u001b[43m(\u001b[49m\u001b[43mframeshift\u001b[49m\u001b[43m)\u001b[49m))\n",
      "File \u001b[0;32m~/PycharmProjects/genophenocorr/src/genophenocorr/view/_stats.py:49\u001b[0m, in \u001b[0;36mStatsViewable.process\u001b[0;34m(self, cohort)\u001b[0m\n\u001b[1;32m     36\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mprocess\u001b[39m(\n\u001b[1;32m     37\u001b[0m         \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m     38\u001b[0m         cohort: Cohort,\n\u001b[1;32m     39\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28mstr\u001b[39m:\n\u001b[1;32m     40\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m     41\u001b[0m \u001b[38;5;124;03m    Create an HTML that should be shown with display(HTML(..)) of the ipython package.\u001b[39;00m\n\u001b[1;32m     42\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     47\u001b[0m \u001b[38;5;124;03m        str: an HTML string with parameterized template for rendering\u001b[39;00m\n\u001b[1;32m     48\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 49\u001b[0m     context \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_prepare_context\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcohort\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     50\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_cohort_template\u001b[38;5;241m.\u001b[39mrender(context)\n",
      "File \u001b[0;32m~/PycharmProjects/genophenocorr/src/genophenocorr/view/_stats.py:58\u001b[0m, in \u001b[0;36mStatsViewable._prepare_context\u001b[0;34m(self, cohort)\u001b[0m\n\u001b[1;32m     52\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_prepare_context\u001b[39m(\n\u001b[1;32m     53\u001b[0m         \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m     54\u001b[0m         cohort: Cohort,\n\u001b[1;32m     55\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m typing\u001b[38;5;241m.\u001b[39mMapping[\u001b[38;5;28mstr\u001b[39m, typing\u001b[38;5;241m.\u001b[39mAny]:\n\u001b[1;32m     57\u001b[0m     skip_counts \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m()\n\u001b[0;32m---> 58\u001b[0m     sort_by_value \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mdict\u001b[39m(\u001b[38;5;28msorted\u001b[39m(\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_results_map\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mitems\u001b[49m(), key\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mlambda\u001b[39;00m item: item[\u001b[38;5;241m1\u001b[39m], reverse\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m))\n\u001b[1;32m     59\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m skipped, c \u001b[38;5;129;01min\u001b[39;00m sort_by_value\u001b[38;5;241m.\u001b[39mitems():\n\u001b[1;32m     60\u001b[0m         skip_counts\u001b[38;5;241m.\u001b[39mappend({\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSkipped\u001b[39m\u001b[38;5;124m\"\u001b[39m: skipped, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcount\u001b[39m\u001b[38;5;124m\"\u001b[39m: c})\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'HpoMtcReport' object has no attribute 'items'"
     ]
    }
   ],
   "source": [
    "display(HTML(sv.process(frameshift)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a83a5ca",
   "metadata": {},
   "source": [
    "Test for presence of genotype-phenotype correlations between subjects with >=1 allele of a variant vs. the other subjects:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "57dd0e23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>&gt;=1 allele of the variant 16_89284634_89284639_GTGTTT_G</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Yes</th>\n",
       "      <th colspan=\"2\" halign=\"left\">No</th>\n",
       "      <th colspan=\"2\" halign=\"left\"></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>p value</th>\n",
       "      <th>Corrected p value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Sleep abnormality [HP:0002360]</th>\n",
       "      <td>7/9</td>\n",
       "      <td>78%</td>\n",
       "      <td>12/48</td>\n",
       "      <td>25%</td>\n",
       "      <td>0.004267</td>\n",
       "      <td>0.810716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Autistic behavior [HP:0000729]</th>\n",
       "      <td>1/6</td>\n",
       "      <td>17%</td>\n",
       "      <td>39/66</td>\n",
       "      <td>59%</td>\n",
       "      <td>0.081923</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Intellectual disability [HP:0001249]</th>\n",
       "      <td>14/19</td>\n",
       "      <td>74%</td>\n",
       "      <td>144/164</td>\n",
       "      <td>88%</td>\n",
       "      <td>0.147465</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Synophrys [HP:0000664]</th>\n",
       "      <td>5/14</td>\n",
       "      <td>36%</td>\n",
       "      <td>48/82</td>\n",
       "      <td>59%</td>\n",
       "      <td>0.148677</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EEG abnormality [HP:0002353]</th>\n",
       "      <td>0/6</td>\n",
       "      <td>0%</td>\n",
       "      <td>16/43</td>\n",
       "      <td>37%</td>\n",
       "      <td>0.158804</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Growth abnormality [HP:0001507]</th>\n",
       "      <td>12/12</td>\n",
       "      <td>100%</td>\n",
       "      <td>108/108</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Abnormal eyebrow morphology [HP:0000534]</th>\n",
       "      <td>13/13</td>\n",
       "      <td>100%</td>\n",
       "      <td>103/103</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Abnormality of the digestive system [HP:0025031]</th>\n",
       "      <td>6/6</td>\n",
       "      <td>100%</td>\n",
       "      <td>58/58</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hyperactivity [HP:0000752]</th>\n",
       "      <td>5/5</td>\n",
       "      <td>100%</td>\n",
       "      <td>48/48</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Generalized non-motor (absence) seizure [HP:0002121]</th>\n",
       "      <td>0/0</td>\n",
       "      <td>0%</td>\n",
       "      <td>3/8</td>\n",
       "      <td>38%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>190 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       ">=1 allele of the variant 16_89284634_89284639_GTGTTT_G    Yes          \\\n",
       "                                                         Count Percent   \n",
       "Sleep abnormality [HP:0002360]                             7/9     78%   \n",
       "Autistic behavior [HP:0000729]                             1/6     17%   \n",
       "Intellectual disability [HP:0001249]                     14/19     74%   \n",
       "Synophrys [HP:0000664]                                    5/14     36%   \n",
       "EEG abnormality [HP:0002353]                               0/6      0%   \n",
       "...                                                        ...     ...   \n",
       "Growth abnormality [HP:0001507]                          12/12    100%   \n",
       "Abnormal eyebrow morphology [HP:0000534]                 13/13    100%   \n",
       "Abnormality of the digestive system [HP:0025031]           6/6    100%   \n",
       "Hyperactivity [HP:0000752]                                 5/5    100%   \n",
       "Generalized non-motor (absence) seizure [HP:000...         0/0      0%   \n",
       "\n",
       ">=1 allele of the variant 16_89284634_89284639_GTGTTT_G       No          \\\n",
       "                                                           Count Percent   \n",
       "Sleep abnormality [HP:0002360]                             12/48     25%   \n",
       "Autistic behavior [HP:0000729]                             39/66     59%   \n",
       "Intellectual disability [HP:0001249]                     144/164     88%   \n",
       "Synophrys [HP:0000664]                                     48/82     59%   \n",
       "EEG abnormality [HP:0002353]                               16/43     37%   \n",
       "...                                                          ...     ...   \n",
       "Growth abnormality [HP:0001507]                          108/108    100%   \n",
       "Abnormal eyebrow morphology [HP:0000534]                 103/103    100%   \n",
       "Abnormality of the digestive system [HP:0025031]           58/58    100%   \n",
       "Hyperactivity [HP:0000752]                                 48/48    100%   \n",
       "Generalized non-motor (absence) seizure [HP:000...           3/8     38%   \n",
       "\n",
       ">=1 allele of the variant 16_89284634_89284639_GTGTTT_G            \\\n",
       "                                                          p value   \n",
       "Sleep abnormality [HP:0002360]                           0.004267   \n",
       "Autistic behavior [HP:0000729]                           0.081923   \n",
       "Intellectual disability [HP:0001249]                     0.147465   \n",
       "Synophrys [HP:0000664]                                   0.148677   \n",
       "EEG abnormality [HP:0002353]                             0.158804   \n",
       "...                                                           ...   \n",
       "Growth abnormality [HP:0001507]                          1.000000   \n",
       "Abnormal eyebrow morphology [HP:0000534]                 1.000000   \n",
       "Abnormality of the digestive system [HP:0025031]         1.000000   \n",
       "Hyperactivity [HP:0000752]                               1.000000   \n",
       "Generalized non-motor (absence) seizure [HP:000...       1.000000   \n",
       "\n",
       ">=1 allele of the variant 16_89284634_89284639_GTGTTT_G                    \n",
       "                                                        Corrected p value  \n",
       "Sleep abnormality [HP:0002360]                                   0.810716  \n",
       "Autistic behavior [HP:0000729]                                   1.000000  \n",
       "Intellectual disability [HP:0001249]                             1.000000  \n",
       "Synophrys [HP:0000664]                                           1.000000  \n",
       "EEG abnormality [HP:0002353]                                     1.000000  \n",
       "...                                                                   ...  \n",
       "Growth abnormality [HP:0001507]                                  1.000000  \n",
       "Abnormal eyebrow morphology [HP:0000534]                         1.000000  \n",
       "Abnormality of the digestive system [HP:0025031]                 1.000000  \n",
       "Hyperactivity [HP:0000752]                                       1.000000  \n",
       "Generalized non-motor (absence) seizure [HP:000...               1.000000  \n",
       "\n",
       "[190 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var_single = analysis.compare_by_variant_key('16_89284634_89284639_GTGTTT_G')\n",
    "var_single.summarize(hpo, PatientCategories.YES)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74eaa536",
   "metadata": {},
   "source": [
    "Or between subjects with one variant vs. the other variant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0dfa0a50",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>&gt;=1 allele of either variant 16_89284129_89284134_CTTTTT_C or variant 16_89284634_89284639_GTGTTT_G</th>\n",
       "      <th colspan=\"2\" halign=\"left\">First</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Second</th>\n",
       "      <th colspan=\"2\" halign=\"left\"></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>p value</th>\n",
       "      <th>Corrected p value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Autistic behavior [HP:0000729]</th>\n",
       "      <td>2/2</td>\n",
       "      <td>100%</td>\n",
       "      <td>1/6</td>\n",
       "      <td>17%</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Macrodontia [HP:0001572]</th>\n",
       "      <td>5/9</td>\n",
       "      <td>56%</td>\n",
       "      <td>22/27</td>\n",
       "      <td>81%</td>\n",
       "      <td>0.184067</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hypertelorism [HP:0000316]</th>\n",
       "      <td>3/3</td>\n",
       "      <td>100%</td>\n",
       "      <td>5/10</td>\n",
       "      <td>50%</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cryptorchidism [HP:0000028]</th>\n",
       "      <td>0/4</td>\n",
       "      <td>0%</td>\n",
       "      <td>5/12</td>\n",
       "      <td>42%</td>\n",
       "      <td>0.244505</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Low anterior hairline [HP:0000294]</th>\n",
       "      <td>3/3</td>\n",
       "      <td>100%</td>\n",
       "      <td>7/13</td>\n",
       "      <td>54%</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Long philtrum [HP:0000343]</th>\n",
       "      <td>4/4</td>\n",
       "      <td>100%</td>\n",
       "      <td>12/13</td>\n",
       "      <td>92%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Abnormal eyebrow morphology [HP:0000534]</th>\n",
       "      <td>5/5</td>\n",
       "      <td>100%</td>\n",
       "      <td>13/13</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Abnormality of the digestive system [HP:0025031]</th>\n",
       "      <td>1/1</td>\n",
       "      <td>100%</td>\n",
       "      <td>6/6</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hyperactivity [HP:0000752]</th>\n",
       "      <td>0/0</td>\n",
       "      <td>0%</td>\n",
       "      <td>5/5</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Generalized non-motor (absence) seizure [HP:0002121]</th>\n",
       "      <td>1/1</td>\n",
       "      <td>100%</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>190 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       ">=1 allele of either variant 16_89284129_89284134_CTTTTT_C or variant 16_89284634_89284639_GTGTTT_G First  \\\n",
       "                                                                                                    Count   \n",
       "Autistic behavior [HP:0000729]                                                                        2/2   \n",
       "Macrodontia [HP:0001572]                                                                              5/9   \n",
       "Hypertelorism [HP:0000316]                                                                            3/3   \n",
       "Cryptorchidism [HP:0000028]                                                                           0/4   \n",
       "Low anterior hairline [HP:0000294]                                                                    3/3   \n",
       "...                                                                                                   ...   \n",
       "Long philtrum [HP:0000343]                                                                            4/4   \n",
       "Abnormal eyebrow morphology [HP:0000534]                                                              5/5   \n",
       "Abnormality of the digestive system [HP:0025031]                                                      1/1   \n",
       "Hyperactivity [HP:0000752]                                                                            0/0   \n",
       "Generalized non-motor (absence) seizure [HP:000...                                                    1/1   \n",
       "\n",
       ">=1 allele of either variant 16_89284129_89284134_CTTTTT_C or variant 16_89284634_89284639_GTGTTT_G          \\\n",
       "                                                                                                    Percent   \n",
       "Autistic behavior [HP:0000729]                                                                         100%   \n",
       "Macrodontia [HP:0001572]                                                                                56%   \n",
       "Hypertelorism [HP:0000316]                                                                             100%   \n",
       "Cryptorchidism [HP:0000028]                                                                              0%   \n",
       "Low anterior hairline [HP:0000294]                                                                     100%   \n",
       "...                                                                                                     ...   \n",
       "Long philtrum [HP:0000343]                                                                             100%   \n",
       "Abnormal eyebrow morphology [HP:0000534]                                                               100%   \n",
       "Abnormality of the digestive system [HP:0025031]                                                       100%   \n",
       "Hyperactivity [HP:0000752]                                                                               0%   \n",
       "Generalized non-motor (absence) seizure [HP:000...                                                     100%   \n",
       "\n",
       ">=1 allele of either variant 16_89284129_89284134_CTTTTT_C or variant 16_89284634_89284639_GTGTTT_G Second  \\\n",
       "                                                                                                     Count   \n",
       "Autistic behavior [HP:0000729]                                                                         1/6   \n",
       "Macrodontia [HP:0001572]                                                                             22/27   \n",
       "Hypertelorism [HP:0000316]                                                                            5/10   \n",
       "Cryptorchidism [HP:0000028]                                                                           5/12   \n",
       "Low anterior hairline [HP:0000294]                                                                    7/13   \n",
       "...                                                                                                    ...   \n",
       "Long philtrum [HP:0000343]                                                                           12/13   \n",
       "Abnormal eyebrow morphology [HP:0000534]                                                             13/13   \n",
       "Abnormality of the digestive system [HP:0025031]                                                       6/6   \n",
       "Hyperactivity [HP:0000752]                                                                             5/5   \n",
       "Generalized non-motor (absence) seizure [HP:000...                                                     0/0   \n",
       "\n",
       ">=1 allele of either variant 16_89284129_89284134_CTTTTT_C or variant 16_89284634_89284639_GTGTTT_G          \\\n",
       "                                                                                                    Percent   \n",
       "Autistic behavior [HP:0000729]                                                                          17%   \n",
       "Macrodontia [HP:0001572]                                                                                81%   \n",
       "Hypertelorism [HP:0000316]                                                                              50%   \n",
       "Cryptorchidism [HP:0000028]                                                                             42%   \n",
       "Low anterior hairline [HP:0000294]                                                                      54%   \n",
       "...                                                                                                     ...   \n",
       "Long philtrum [HP:0000343]                                                                              92%   \n",
       "Abnormal eyebrow morphology [HP:0000534]                                                               100%   \n",
       "Abnormality of the digestive system [HP:0025031]                                                       100%   \n",
       "Hyperactivity [HP:0000752]                                                                             100%   \n",
       "Generalized non-motor (absence) seizure [HP:000...                                                       0%   \n",
       "\n",
       ">=1 allele of either variant 16_89284129_89284134_CTTTTT_C or variant 16_89284634_89284639_GTGTTT_G            \\\n",
       "                                                                                                      p value   \n",
       "Autistic behavior [HP:0000729]                                                                       0.107143   \n",
       "Macrodontia [HP:0001572]                                                                             0.184067   \n",
       "Hypertelorism [HP:0000316]                                                                           0.230769   \n",
       "Cryptorchidism [HP:0000028]                                                                          0.244505   \n",
       "Low anterior hairline [HP:0000294]                                                                   0.250000   \n",
       "...                                                                                                       ...   \n",
       "Long philtrum [HP:0000343]                                                                           1.000000   \n",
       "Abnormal eyebrow morphology [HP:0000534]                                                             1.000000   \n",
       "Abnormality of the digestive system [HP:0025031]                                                     1.000000   \n",
       "Hyperactivity [HP:0000752]                                                                           1.000000   \n",
       "Generalized non-motor (absence) seizure [HP:000...                                                   1.000000   \n",
       "\n",
       ">=1 allele of either variant 16_89284129_89284134_CTTTTT_C or variant 16_89284634_89284639_GTGTTT_G                    \n",
       "                                                                                                    Corrected p value  \n",
       "Autistic behavior [HP:0000729]                                                                                    1.0  \n",
       "Macrodontia [HP:0001572]                                                                                          1.0  \n",
       "Hypertelorism [HP:0000316]                                                                                        1.0  \n",
       "Cryptorchidism [HP:0000028]                                                                                       1.0  \n",
       "Low anterior hairline [HP:0000294]                                                                                1.0  \n",
       "...                                                                                                               ...  \n",
       "Long philtrum [HP:0000343]                                                                                        1.0  \n",
       "Abnormal eyebrow morphology [HP:0000534]                                                                          1.0  \n",
       "Abnormality of the digestive system [HP:0025031]                                                                  1.0  \n",
       "Hyperactivity [HP:0000752]                                                                                        1.0  \n",
       "Generalized non-motor (absence) seizure [HP:000...                                                                1.0  \n",
       "\n",
       "[190 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var_double = analysis.compare_by_variant_keys('16_89284129_89284134_CTTTTT_C', '16_89284634_89284639_GTGTTT_G')\n",
    "var_double.summarize(hpo, PatientCategories.YES)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a12b8117",
   "metadata": {},
   "source": [
    "TODO - finalize!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
