{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9443554a",
   "metadata": {},
   "source": [
    "# KBG Syndrome\n",
    "\n",
    "[KBG syndrome (KBGS)](https://omim.org/entry/148050) is caused by heterozygous mutation in the ANKRD11 gene. In this notebook, we have used\n",
    "[pyphetools](https://github.com/monarch-initiative/pyphetools) to parse the clinical data included in the supplemental files of\n",
    "[Martinez-Cayuelas E, et al. Clinical description, molecular delineation and genotype-phenotype correlation in 340 patients with KBG syndrome: addition of 67 new patients](https://pubmed.ncbi.nlm.nih.gov/36446582).\n",
    "\n",
    "The authors identified a significantly higher frequency of patients with a triangular face in carriers of sequence variants compared to CNVs. Other associations found were short stature and variants in exon 9, a lower incidence of ID/ADHD/ASD in carriers of the c.1903_1907del variant and the size of the deletion, in CNV carriers, with the presence of macrodontia and hand anomalies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cd2a5b1b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded HPO v2023-10-09\n",
      "Using genophenocorr version 0.1.1dev\n"
     ]
    }
   ],
   "source": [
    "import genophenocorr\n",
    "import hpotk\n",
    "\n",
    "store = hpotk.configure_ontology_store()\n",
    "hpo = store.load_minimal_hpo(release='v2023-10-09')\n",
    "print(f'Loaded HPO v{hpo.version}')\n",
    "print(f\"Using genophenocorr version {genophenocorr.__version__}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "566b13c6",
   "metadata": {},
   "source": [
    "## Settings\n",
    "Specify the transcript to be used to encode the variants (the phenopackets contain VCF representations of small variants)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97cd3439",
   "metadata": {},
   "source": [
    "### Pick transcript\n",
    "\n",
    "We choose the [MANE Select](https://www.ncbi.nlm.nih.gov/nuccore/NM_013275.6) transcript for *ANKRD11*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "86b84c66",
   "metadata": {},
   "outputs": [],
   "source": [
    "tx_id = 'NM_013275.6'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2bf0ff1",
   "metadata": {},
   "source": [
    "## Load Phenopackets\n",
    "\n",
    "We will load phenopacket JSON files located in `phenopackets` folder that is next to the notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d7623bcb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Patients Created: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████| 337/337 [00:01<00:00, 216.15it/s]\n",
      "Validated under none policy\n",
      "Showing errors and warnings\n",
      "337 phenopacket(s) found at `phenopackets`\n",
      "  patient #0\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P2[PMID_36446582_Novara_2017_P2]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P2 has no variants to work with\n",
      "  patient #1\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P13[PMID_36446582_Goldenberg2016_P13]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P13 has no variants to work with\n",
      "  patient #3\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Ockeloen2015_P20[PMID_36446582_Ockeloen2015_P20]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Ockeloen2015_P20 has no variants to work with\n",
      "  patient #7\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P22[PMID_36446582_Kutkowska_Kazmierczak2021_P22]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska_Kazmierczak2021_P22 has no variants to work with\n",
      "  patient #8\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P10[PMID_36446582_Novara_2017_P10]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P10 has no variants to work with\n",
      "  patient #12\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Gnazzo, 2020_P31[PMID_36446582_Gnazzo_2020_P31]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Gnazzo_2020_P31 has no variants to work with\n",
      "  patient #16\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P29[PMID_36446582_Goldenberg2016_P29]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P29 has no variants to work with\n",
      "  patient #17\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P18[PMID_36446582_Kutkowska_Kazmierczak2021_P18]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska_Kazmierczak2021_P18 has no variants to work with\n",
      "  patient #22\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG36[PMID_36446582_KBG36]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG36 has no variants to work with\n",
      "  patient #27\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P1[PMID_36446582_Goldenberg2016_P1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P1 has no variants to work with\n",
      "  patient #32\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P33[PMID_36446582_Goldenberg2016_P33]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P33 has no variants to work with\n",
      "  patient #34\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P14[PMID_36446582_Kutkowska_Kazmierczak2021_P14]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska_Kazmierczak2021_P14 has no variants to work with\n",
      "  patient #39\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P15[PMID_36446582_Kutkowska_Kazmierczak2021_P15]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska_Kazmierczak2021_P15 has no variants to work with\n",
      "  patient #42\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P32[PMID_36446582_Goldenberg2016_P32]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P32 has no variants to work with\n",
      "  patient #43\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Bucerzan2020[PMID_36446582_Bucerzan2020]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Bucerzan2020 has no variants to work with\n",
      "  patient #46\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P24[PMID_36446582_Goldenberg2016_P24]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P24 has no variants to work with\n",
      "  patient #50\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Miyatake, 2013[PMID_36446582_Miyatake_2013]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Miyatake_2013 has no variants to work with\n",
      "  patient #56\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG1[PMID_36446582_KBG1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG1 has no variants to work with\n",
      "  patient #60\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Sacharow, 2012_P2[PMID_36446582_Sacharow_2012_P2]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Sacharow_2012_P2 has no variants to work with\n",
      "  patient #62\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P19[PMID_36446582_Kutkowska_Kazmierczak2021_P19]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska_Kazmierczak2021_P19 has no variants to work with\n",
      "  patient #64\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P28[PMID_36446582_Goldenberg2016_P28]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P28 has no variants to work with\n",
      "  patient #68\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Gnazzo, 2020_P30[PMID_36446582_Gnazzo_2020_P30]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Gnazzo_2020_P30 has no variants to work with\n",
      "  patient #71\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P23[PMID_36446582_Kutkowska_Kazmierczak2021_P23]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska_Kazmierczak2021_P23 has no variants to work with\n",
      "  patient #72\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P11[PMID_36446582_Novara_2017_P11]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P11 has no variants to work with\n",
      "  patient #76\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P12[PMID_36446582_Goldenberg2016_P12]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P12 has no variants to work with\n",
      "  patient #77\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P3[PMID_36446582_Novara_2017_P3]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P3 has no variants to work with\n",
      "  patient #78\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Khalifa, 2013_P1B[PMID_36446582_Khalifa_2013_P1B]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Khalifa_2013_P1B has no variants to work with\n",
      "  patient #79\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Isrie, 2012_P2[PMID_36446582_Isrie_2012_P2]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Isrie_2012_P2 has no variants to work with\n",
      "  patient #80\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P19[PMID_36446582_Goldenberg2016_P19]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P19 has no variants to work with\n",
      "  patient #83\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P8[PMID_36446582_Novara_2017_P8]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P8 has no variants to work with\n",
      "  patient #85\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG30[PMID_36446582_KBG30]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG30 has no variants to work with\n",
      "  patient #91\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Scarano, 2013_P11[PMID_36446582_Scarano_2013_P11]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Scarano_2013_P11 has no variants to work with\n",
      "  patient #93\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Crippa2015_P1[PMID_36446582_Crippa2015_P1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Crippa2015_P1 has no variants to work with\n",
      "  patient #100\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P4[PMID_36446582_Novara_2017_P4]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P4 has no variants to work with\n",
      "  patient #103\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Parenti2021_P23[PMID_36446582_Parenti2021_P23]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Parenti2021_P23 has no variants to work with\n",
      "  patient #135\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient VanDongen2019_P5[PMID_36446582_VanDongen2019_P5]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_VanDongen2019_P5 has no variants to work with\n",
      "  patient #140\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P5[PMID_36446582_Novara_2017_P5]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P5 has no variants to work with\n",
      "  patient #142\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Willemsen2010_P1[PMID_36446582_Willemsen2010_P1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Willemsen2010_P1 has no variants to work with\n",
      "  patient #147\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P22[PMID_36446582_Goldenberg2016_P22]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P22 has no variants to work with\n",
      "  patient #148\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Scarano, 2013_P10[PMID_36446582_Scarano_2013_P10]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Scarano_2013_P10 has no variants to work with\n",
      "  patient #156\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P9[PMID_36446582_Novara_2017_P9]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P9 has no variants to work with\n",
      "  patient #159\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P18[PMID_36446582_Goldenberg2016_P18]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P18 has no variants to work with\n",
      "  patient #165\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Willemsen2010_P2[PMID_36446582_Willemsen2010_P2]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Willemsen2010_P2 has no variants to work with\n",
      "  patient #168\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Spengler, 2013[PMID_36446582_Spengler_2013]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Spengler_2013 has no variants to work with\n",
      "  patient #173\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Youngs2011[PMID_36446582_Youngs2011]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Youngs2011 has no variants to work with\n",
      "  patient #174\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P21[PMID_36446582_Goldenberg2016_P21]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P21 has no variants to work with\n",
      "  patient #175\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Crippa2015_P3[PMID_36446582_Crippa2015_P3]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Crippa2015_P3 has no variants to work with\n",
      "  patient #186\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG32[PMID_36446582_KBG32]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG32 has no variants to work with\n",
      "  patient #200\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Palumbo 2016[PMID_36446582_Palumbo_2016]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Palumbo_2016 has no variants to work with\n",
      "  patient #210\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P7[PMID_36446582_Novara_2017_P7]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P7 has no variants to work with\n",
      "  patient #214\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient VanDongen2019_P7[PMID_36446582_VanDongen2019_P7]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_VanDongen2019_P7 has no variants to work with\n",
      "  patient #223\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG52[PMID_36446582_KBG52]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG52 has no variants to work with\n",
      "  patient #233\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG33[PMID_36446582_KBG33]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG33 has no variants to work with\n",
      "  patient #238\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P4[PMID_36446582_Goldenberg2016_P4]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P4 has no variants to work with\n",
      "  patient #239\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Scarano, 2013_P12[PMID_36446582_Scarano_2013_P12]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Scarano_2013_P12 has no variants to work with\n",
      "  patient #241\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P20[PMID_36446582_Goldenberg2016_P20]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P20 has no variants to work with\n",
      "  patient #242\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Crippa2015_P2[PMID_36446582_Crippa2015_P2]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Crippa2015_P2 has no variants to work with\n",
      "  patient #244\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P36[PMID_36446582_Goldenberg2016_P36]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P36 has no variants to work with\n",
      "  patient #248\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Willemsen2010_P3[PMID_36446582_Willemsen2010_P3]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Willemsen2010_P3 has no variants to work with\n",
      "  patient #253\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Sacharow, 2012_P1[PMID_36446582_Sacharow_2012_P1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Sacharow_2012_P1 has no variants to work with\n",
      "  patient #256\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Srivastava, 2017_P1[PMID_36446582_Srivastava_2017_P1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Srivastava_2017_P1 has no variants to work with\n",
      "  patient #266\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P12[PMID_36446582_Novara_2017_P12]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P12 has no variants to work with\n",
      "  patient #268\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P20[PMID_36446582_Kutkowska_Kazmierczak2021_P20]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska_Kazmierczak2021_P20 has no variants to work with\n",
      "  patient #274\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Isrie, 2012_P1[PMID_36446582_Isrie_2012_P1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Isrie_2012_P1 has no variants to work with\n",
      "  patient #275\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Khalifa, 2013_P1A[PMID_36446582_Khalifa_2013_P1A]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Khalifa_2013_P1A has no variants to work with\n",
      "  patient #278\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Willemsen2010_P4[PMID_36446582_Willemsen2010_P4]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Willemsen2010_P4 has no variants to work with\n",
      "  patient #279\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Gnazzo, 2020_P29[PMID_36446582_Gnazzo_2020_P29]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Gnazzo_2020_P29 has no variants to work with\n",
      "  patient #282\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P16[PMID_36446582_Kutkowska_Kazmierczak2021_P16]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska_Kazmierczak2021_P16 has no variants to work with\n",
      "  patient #288\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P3[PMID_36446582_Goldenberg2016_P3]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P3 has no variants to work with\n",
      "  patient #299\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Lim2014[PMID_36446582_Lim2014]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Lim2014 has no variants to work with\n",
      "  patient #300\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG35[PMID_36446582_KBG35]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG35 has no variants to work with\n",
      "  patient #304\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P2[PMID_36446582_Goldenberg2016_P2]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P2 has no variants to work with\n",
      "  patient #306\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P26[PMID_36446582_Goldenberg2016_P26]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P26 has no variants to work with\n",
      "  patient #309\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P17[PMID_36446582_Kutkowska_Kazmierczak2021_P17]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska_Kazmierczak2021_P17 has no variants to work with\n",
      "  patient #316\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P1[PMID_36446582_Novara_2017_P1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara_2017_P1 has no variants to work with\n",
      "  patient #317\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P10[PMID_36446582_Goldenberg2016_P10]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P10 has no variants to work with\n",
      "  patient #320\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG39[PMID_36446582_KBG39]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG39 has no variants to work with\n",
      "  patient #323\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P21[PMID_36446582_Kutkowska_Kazmierczak2021_P21]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska_Kazmierczak2021_P21 has no variants to work with\n",
      "  patient #326\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Behnert, 2018[PMID_36446582_Behnert_2018]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Behnert_2018 has no variants to work with\n",
      "  patient #335\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG15[PMID_36446582_KBG15]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG15 has no variants to work with\n"
     ]
    }
   ],
   "source": [
    "from genophenocorr.preprocessing import configure_caching_cohort_creator, load_phenopacket_folder\n",
    "\n",
    "fpath_phenopackets = 'phenopackets'\n",
    "cohort_creator = configure_caching_cohort_creator(hpo, timeout=20)\n",
    "cohort = load_phenopacket_folder(fpath_phenopackets, cohort_creator)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f442ac4",
   "metadata": {},
   "source": [
    "## Summarize the cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "070a06ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<!DOCTYPE html>\n",
       "<html lang=\"en\">\n",
       "<head>\n",
       "  <meta charset=\"utf-8\">\n",
       "  <title>Cohort</title>\n",
       "  <style>table {\n",
       "    border-collapse: collapse;\n",
       "    margin: 25px 0;\n",
       "      font-size: 0.9em;\n",
       "      font-family: sans-serif;\n",
       "      min-width: 400px;\n",
       "      box-shadow: 0 0 20px rgba(0, 0, 0, 0.15);\n",
       "  }\n",
       "\n",
       "  .table .column-1 {\n",
       "      text-align: left;\n",
       "  }\n",
       "  th {\n",
       "    background-color: LightSkyBlue;\n",
       "    border: 1px solid #dddddd;\n",
       "    text-align: left;\n",
       "    padding: 2px;\n",
       "  font-weight: bold;  \n",
       "  font-size: 120%;\n",
       "}\n",
       "\n",
       "\n",
       "  tr {\n",
       "    border: 1px solid #dddddd;\n",
       "  }\n",
       "\n",
       "  td {\n",
       "    padding: 2px;\n",
       "    font-weight: bold;\n",
       "  }\n",
       "\n",
       "  tr:nth-child(even) {\n",
       "    background-color: #f2f2f2;\n",
       "  }\n",
       "\n",
       "  .table td,tr {\n",
       "    text-align: right;\n",
       "}\n",
       "\n",
       ".lft {\n",
       "    text-align: left;\n",
       "}​\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "caption {\n",
       "  caption-side: bottom;\n",
       "  text-align: left;\n",
       "  padding-bottom: 10px;\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       "  </style>\n",
       "</head>\n",
       "\n",
       "<body>\n",
       "  <h1>genophenocorr cohort analysis</h1>\n",
       "  <p>Successfully ingested 337 phenopackets.</p>\n",
       "  \n",
       "  <p>Not able to ingest 80 individuals.</p>\n",
       "  \n",
       " \n",
       "    <h3>Top 10 HPO Terms</h3>\n",
       "    <table>\n",
       "      <caption>A total of 93 HPO terms were used to annotated the cohort. </caption>\n",
       "        <tbody>\n",
       "            <tr class=\"strng\">\n",
       "                <th class=\"lft\">HPO Term</th>\n",
       "                <th>ID</th>\n",
       "                <th>Annotation Count</th>\n",
       "              </tr>\n",
       "              \n",
       "              <tr>\n",
       "                <td class=\"lft\">Macrodontia</td>\n",
       "                <td>HP:0001572</td>\n",
       "                <td>211</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Intellectual disability</td>\n",
       "                <td>HP:0001249</td>\n",
       "                <td>194</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Abnormality of the hand</td>\n",
       "                <td>HP:0001155</td>\n",
       "                <td>189</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Global developmental delay</td>\n",
       "                <td>HP:0001263</td>\n",
       "                <td>176</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Delayed speech and language development</td>\n",
       "                <td>HP:0000750</td>\n",
       "                <td>160</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Short stature</td>\n",
       "                <td>HP:0004322</td>\n",
       "                <td>150</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Thick eyebrow</td>\n",
       "                <td>HP:0000574</td>\n",
       "                <td>126</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Long philtrum</td>\n",
       "                <td>HP:0000343</td>\n",
       "                <td>121</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Bulbous nose</td>\n",
       "                <td>HP:0000414</td>\n",
       "                <td>89</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Triangular face</td>\n",
       "                <td>HP:0000325</td>\n",
       "                <td>83</td>\n",
       "                </tr>\n",
       "            \n",
       "                </tbody>\n",
       "    </table>\n",
       "\n",
       " <h3>Top 10 Variants</h3>\n",
       " <table>\n",
       "  <caption>\n",
       "    Variants are shown according to NM_013275.6. A total of 257 unique variants were identified in the cohort.\n",
       "  </caption>\n",
       "        <tbody>\n",
       "            <tr class=\"strng\">\n",
       "                <th class=\"lft\">Variant</th>\n",
       "                <th>Variant name</th>\n",
       "                <th>Variant Count</th>\n",
       "              </tr>\n",
       "              \n",
       "              <tr>\n",
       "                <td class=\"lft\">16_89284634_89284639_GTGTTT_G</td>\n",
       "                <td>c.1903_1907del</td>\n",
       "                <td>34</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">16_89284129_89284134_CTTTTT_C</td>\n",
       "                <td>c.2408_2412del</td>\n",
       "                <td>10</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">16_89284140_89284144_TTTTC_T</td>\n",
       "                <td>c.2398_2401del</td>\n",
       "                <td>9</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">16_89285157_89285161_GTTTC_G</td>\n",
       "                <td>c.1381_1384del</td>\n",
       "                <td>8</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">16_89279749_89279749_C_CG</td>\n",
       "                <td>c.6792dup</td>\n",
       "                <td>5</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">16_89275180_89275180_A_AG</td>\n",
       "                <td>c.7481dup</td>\n",
       "                <td>5</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">16_89283314_89283318_CCTTT_C</td>\n",
       "                <td>c.3224_3227del</td>\n",
       "                <td>3</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">16_89284358_89284360_GAT_G</td>\n",
       "                <td>c.2182_2183del</td>\n",
       "                <td>3</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">16_89282710_89282710_T_A</td>\n",
       "                <td>c.3832A>T</td>\n",
       "                <td>3</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">16_89282136_89282136_C_T</td>\n",
       "                <td>c.4406G>A</td>\n",
       "                <td>3</td>\n",
       "                </tr>\n",
       "            \n",
       "         </tbody>\n",
       "    </table>\n",
       "    <h3>Diseases</h3>\n",
       "    <table>\n",
       "        <tbody>\n",
       "            <tr class=\"strng\">\n",
       "                <th class=\"lft\">Disease</th>\n",
       "                <th>Annotation Count</th>\n",
       "              </tr>\n",
       "              \n",
       "              <tr>\n",
       "                <td class=\"lft\">OMIM:148050</td>\n",
       "                <td>337</td>\n",
       "                </tr>\n",
       "            \n",
       "           </tbody>\n",
       "    </table>\n",
       "    \n",
       "    <h3>Variant categories</h3>\n",
       "    <table>\n",
       "        <tbody>\n",
       "            <tr class=\"strng\">\n",
       "                <th class=\"lft\">Variant effect</th>\n",
       "                <th>Annotation Count</th>\n",
       "              </tr>\n",
       "              \n",
       "              <tr>\n",
       "                <td class=\"lft\">FRAMESHIFT_VARIANT</td>\n",
       "                <td>175</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">STOP_GAINED</td>\n",
       "                <td>67</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">MISSENSE_VARIANT</td>\n",
       "                <td>7</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">SPLICE_ACCEPTOR_VARIANT</td>\n",
       "                <td>4</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">SPLICE_DONOR_VARIANT</td>\n",
       "                <td>3</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">CODING_SEQUENCE_VARIANT</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">INFRAME_DELETION</td>\n",
       "                <td>2</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">SPLICE_REGION_VARIANT</td>\n",
       "                <td>2</td>\n",
       "                </tr>\n",
       "            \n",
       "           </tbody>\n",
       "    </table>\n",
       "  \n",
       "\n",
       "\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.display import display, HTML\n",
    "from genophenocorr.view import CohortViewable\n",
    "\n",
    "cv = CohortViewable(hpo=hpo, transcript_id=tx_id)\n",
    "html = cv.process(cohort=cohort, transcript_id=tx_id)\n",
    "\n",
    "display(HTML(html))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbbc2cc1",
   "metadata": {},
   "source": [
    "## Configure the analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ec6c204f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from genophenocorr.analysis import configure_cohort_analysis, CohortAnalysisConfiguration\n",
    "from genophenocorr.analysis.predicate import PatientCategories\n",
    "\n",
    "analysis_config = CohortAnalysisConfiguration()\n",
    "analysis_config.missing_implies_excluded = True\n",
    "analysis_config.pval_correction = 'fdr_bh'\n",
    "analysis_config.min_perc_patients_w_hpo = 0.1\n",
    "analysis = configure_cohort_analysis(cohort, hpo, config=analysis_config)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ff62a78",
   "metadata": {},
   "source": [
    "Test for presence of genotype-phenotype correlations between frameshift variants vs. others."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "91809a38",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>FRAMESHIFT_VARIANT on NM_013275.6</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Yes</th>\n",
       "      <th colspan=\"2\" halign=\"left\">No</th>\n",
       "      <th colspan=\"2\" halign=\"left\"></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>p value</th>\n",
       "      <th>Corrected p value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Abnormality of the hand [HP:0001155]</th>\n",
       "      <td>95/144</td>\n",
       "      <td>66%</td>\n",
       "      <td>60/71</td>\n",
       "      <td>85%</td>\n",
       "      <td>0.005661</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EEG abnormality [HP:0002353]</th>\n",
       "      <td>7/33</td>\n",
       "      <td>21%</td>\n",
       "      <td>9/16</td>\n",
       "      <td>56%</td>\n",
       "      <td>0.022884</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Feeding difficulties [HP:0011968]</th>\n",
       "      <td>33/89</td>\n",
       "      <td>37%</td>\n",
       "      <td>26/45</td>\n",
       "      <td>58%</td>\n",
       "      <td>0.027584</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Low anterior hairline [HP:0000294]</th>\n",
       "      <td>40/58</td>\n",
       "      <td>69%</td>\n",
       "      <td>15/30</td>\n",
       "      <td>50%</td>\n",
       "      <td>0.105274</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Intellectual disability [HP:0001249]</th>\n",
       "      <td>99/119</td>\n",
       "      <td>83%</td>\n",
       "      <td>59/64</td>\n",
       "      <td>92%</td>\n",
       "      <td>0.115195</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Abnormality of the head [HP:0000234]</th>\n",
       "      <td>153/153</td>\n",
       "      <td>100%</td>\n",
       "      <td>74/74</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aplasia/hypoplasia affecting bones of the axial skeleton [HP:0009122]</th>\n",
       "      <td>13/13</td>\n",
       "      <td>100%</td>\n",
       "      <td>3/3</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Abnormal cardiovascular system physiology [HP:0011025]</th>\n",
       "      <td>1/1</td>\n",
       "      <td>100%</td>\n",
       "      <td>2/2</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Non-motor seizure [HP:0033259]</th>\n",
       "      <td>1/1</td>\n",
       "      <td>100%</td>\n",
       "      <td>2/2</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prominent nose [HP:0000448]</th>\n",
       "      <td>12/28</td>\n",
       "      <td>43%</td>\n",
       "      <td>5/13</td>\n",
       "      <td>38%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>190 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "FRAMESHIFT_VARIANT on NM_013275.6                       Yes             No  \\\n",
       "                                                      Count Percent  Count   \n",
       "Abnormality of the hand [HP:0001155]                 95/144     66%  60/71   \n",
       "EEG abnormality [HP:0002353]                           7/33     21%   9/16   \n",
       "Feeding difficulties [HP:0011968]                     33/89     37%  26/45   \n",
       "Low anterior hairline [HP:0000294]                    40/58     69%  15/30   \n",
       "Intellectual disability [HP:0001249]                 99/119     83%  59/64   \n",
       "...                                                     ...     ...    ...   \n",
       "Abnormality of the head [HP:0000234]                153/153    100%  74/74   \n",
       "Aplasia/hypoplasia affecting bones of the axial...    13/13    100%    3/3   \n",
       "Abnormal cardiovascular system physiology [HP:0...      1/1    100%    2/2   \n",
       "Non-motor seizure [HP:0033259]                          1/1    100%    2/2   \n",
       "Prominent nose [HP:0000448]                           12/28     43%   5/13   \n",
       "\n",
       "FRAMESHIFT_VARIANT on NM_013275.6                                     \\\n",
       "                                                   Percent   p value   \n",
       "Abnormality of the hand [HP:0001155]                   85%  0.005661   \n",
       "EEG abnormality [HP:0002353]                           56%  0.022884   \n",
       "Feeding difficulties [HP:0011968]                      58%  0.027584   \n",
       "Low anterior hairline [HP:0000294]                     50%  0.105274   \n",
       "Intellectual disability [HP:0001249]                   92%  0.115195   \n",
       "...                                                    ...       ...   \n",
       "Abnormality of the head [HP:0000234]                  100%  1.000000   \n",
       "Aplasia/hypoplasia affecting bones of the axial...    100%  1.000000   \n",
       "Abnormal cardiovascular system physiology [HP:0...    100%  1.000000   \n",
       "Non-motor seizure [HP:0033259]                        100%  1.000000   \n",
       "Prominent nose [HP:0000448]                            38%  1.000000   \n",
       "\n",
       "FRAMESHIFT_VARIANT on NM_013275.6                                     \n",
       "                                                   Corrected p value  \n",
       "Abnormality of the hand [HP:0001155]                             1.0  \n",
       "EEG abnormality [HP:0002353]                                     1.0  \n",
       "Feeding difficulties [HP:0011968]                                1.0  \n",
       "Low anterior hairline [HP:0000294]                               1.0  \n",
       "Intellectual disability [HP:0001249]                             1.0  \n",
       "...                                                              ...  \n",
       "Abnormality of the head [HP:0000234]                             1.0  \n",
       "Aplasia/hypoplasia affecting bones of the axial...               1.0  \n",
       "Abnormal cardiovascular system physiology [HP:0...               1.0  \n",
       "Non-motor seizure [HP:0033259]                                   1.0  \n",
       "Prominent nose [HP:0000448]                                      1.0  \n",
       "\n",
       "[190 rows x 6 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from genophenocorr.model import VariantEffect\n",
    "\n",
    "frameshift = analysis.compare_by_variant_effect(VariantEffect.FRAMESHIFT_VARIANT, tx_id=tx_id)\n",
    "frameshift.summarize(hpo, PatientCategories.YES)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "67266357",
   "metadata": {},
   "outputs": [],
   "source": [
    "from genophenocorr.analysis import configure_cohort_analysis, CohortAnalysisConfiguration\n",
    "from genophenocorr.analysis.predicate import PatientCategories\n",
    "\n",
    "analysis_config = CohortAnalysisConfiguration()\n",
    "analysis_config.missing_implies_excluded = True\n",
    "analysis_config.pval_correction = 'fdr_bh'\n",
    "analysis_config.min_perc_patients_w_hpo = 0.1\n",
    "analysis_config.heuristic_strategy()\n",
    "analysis = configure_cohort_analysis(cohort, hpo, config=analysis_config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3d3bdfea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>FRAMESHIFT_VARIANT on NM_013275.6</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Yes</th>\n",
       "      <th colspan=\"2\" halign=\"left\">No</th>\n",
       "      <th colspan=\"2\" halign=\"left\"></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>p value</th>\n",
       "      <th>Corrected p value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Abnormality of the hand [HP:0001155]</th>\n",
       "      <td>95/144</td>\n",
       "      <td>66%</td>\n",
       "      <td>60/71</td>\n",
       "      <td>85%</td>\n",
       "      <td>0.005661</td>\n",
       "      <td>0.209453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EEG abnormality [HP:0002353]</th>\n",
       "      <td>7/33</td>\n",
       "      <td>21%</td>\n",
       "      <td>9/16</td>\n",
       "      <td>56%</td>\n",
       "      <td>0.022884</td>\n",
       "      <td>0.340205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Feeding difficulties [HP:0011968]</th>\n",
       "      <td>33/89</td>\n",
       "      <td>37%</td>\n",
       "      <td>26/45</td>\n",
       "      <td>58%</td>\n",
       "      <td>0.027584</td>\n",
       "      <td>0.340205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Low anterior hairline [HP:0000294]</th>\n",
       "      <td>40/58</td>\n",
       "      <td>69%</td>\n",
       "      <td>15/30</td>\n",
       "      <td>50%</td>\n",
       "      <td>0.105274</td>\n",
       "      <td>0.794281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Intellectual disability [HP:0001249]</th>\n",
       "      <td>99/119</td>\n",
       "      <td>83%</td>\n",
       "      <td>59/64</td>\n",
       "      <td>92%</td>\n",
       "      <td>0.115195</td>\n",
       "      <td>0.794281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Delayed skeletal maturation [HP:0002750]</th>\n",
       "      <td>37/86</td>\n",
       "      <td>43%</td>\n",
       "      <td>17/28</td>\n",
       "      <td>61%</td>\n",
       "      <td>0.128802</td>\n",
       "      <td>0.794281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inguinal hernia [HP:0000023]</th>\n",
       "      <td>2/31</td>\n",
       "      <td>6%</td>\n",
       "      <td>4/17</td>\n",
       "      <td>24%</td>\n",
       "      <td>0.166824</td>\n",
       "      <td>0.881785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Intrauterine growth retardation [HP:0001511]</th>\n",
       "      <td>6/39</td>\n",
       "      <td>15%</td>\n",
       "      <td>7/23</td>\n",
       "      <td>30%</td>\n",
       "      <td>0.202568</td>\n",
       "      <td>0.935054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bulbous nose [HP:0000414]</th>\n",
       "      <td>45/71</td>\n",
       "      <td>63%</td>\n",
       "      <td>29/39</td>\n",
       "      <td>74%</td>\n",
       "      <td>0.291265</td>\n",
       "      <td>0.935054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hypertelorism [HP:0000316]</th>\n",
       "      <td>34/63</td>\n",
       "      <td>54%</td>\n",
       "      <td>25/38</td>\n",
       "      <td>66%</td>\n",
       "      <td>0.299330</td>\n",
       "      <td>0.935054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Microretrognathia [HP:0000308]</th>\n",
       "      <td>13/36</td>\n",
       "      <td>36%</td>\n",
       "      <td>3/16</td>\n",
       "      <td>19%</td>\n",
       "      <td>0.330525</td>\n",
       "      <td>0.935054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sleep abnormality [HP:0002360]</th>\n",
       "      <td>15/40</td>\n",
       "      <td>38%</td>\n",
       "      <td>4/17</td>\n",
       "      <td>24%</td>\n",
       "      <td>0.370187</td>\n",
       "      <td>0.935054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Short stature [HP:0004322]</th>\n",
       "      <td>73/126</td>\n",
       "      <td>58%</td>\n",
       "      <td>42/65</td>\n",
       "      <td>65%</td>\n",
       "      <td>0.436104</td>\n",
       "      <td>0.935054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Delayed speech and language development [HP:0000750]</th>\n",
       "      <td>77/92</td>\n",
       "      <td>84%</td>\n",
       "      <td>44/49</td>\n",
       "      <td>90%</td>\n",
       "      <td>0.448471</td>\n",
       "      <td>0.935054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Exaggerated cupid's bow [HP:0002263]</th>\n",
       "      <td>11/35</td>\n",
       "      <td>31%</td>\n",
       "      <td>2/12</td>\n",
       "      <td>17%</td>\n",
       "      <td>0.464459</td>\n",
       "      <td>0.935054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mandibular prognathia [HP:0000303]</th>\n",
       "      <td>8/33</td>\n",
       "      <td>24%</td>\n",
       "      <td>5/14</td>\n",
       "      <td>36%</td>\n",
       "      <td>0.486071</td>\n",
       "      <td>0.935054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cryptorchidism [HP:0000028]</th>\n",
       "      <td>20/51</td>\n",
       "      <td>39%</td>\n",
       "      <td>15/31</td>\n",
       "      <td>48%</td>\n",
       "      <td>0.492270</td>\n",
       "      <td>0.935054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Thick eyebrow [HP:0000574]</th>\n",
       "      <td>68/82</td>\n",
       "      <td>83%</td>\n",
       "      <td>38/49</td>\n",
       "      <td>78%</td>\n",
       "      <td>0.494742</td>\n",
       "      <td>0.935054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Triangular face [HP:0000325]</th>\n",
       "      <td>45/63</td>\n",
       "      <td>71%</td>\n",
       "      <td>23/36</td>\n",
       "      <td>64%</td>\n",
       "      <td>0.501819</td>\n",
       "      <td>0.935054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Microcephaly [HP:0000252]</th>\n",
       "      <td>18/74</td>\n",
       "      <td>24%</td>\n",
       "      <td>12/39</td>\n",
       "      <td>31%</td>\n",
       "      <td>0.505435</td>\n",
       "      <td>0.935054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Motor delay [HP:0001270]</th>\n",
       "      <td>36/52</td>\n",
       "      <td>69%</td>\n",
       "      <td>19/25</td>\n",
       "      <td>76%</td>\n",
       "      <td>0.600188</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Focal-onset seizure [HP:0007359]</th>\n",
       "      <td>4/8</td>\n",
       "      <td>50%</td>\n",
       "      <td>4/6</td>\n",
       "      <td>67%</td>\n",
       "      <td>0.627040</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Low-set ears [HP:0000369]</th>\n",
       "      <td>16/30</td>\n",
       "      <td>53%</td>\n",
       "      <td>5/11</td>\n",
       "      <td>45%</td>\n",
       "      <td>0.733718</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Recurrent otitis media [HP:0000403]</th>\n",
       "      <td>17/36</td>\n",
       "      <td>47%</td>\n",
       "      <td>7/17</td>\n",
       "      <td>41%</td>\n",
       "      <td>0.772020</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Protruding ear [HP:0000411]</th>\n",
       "      <td>32/44</td>\n",
       "      <td>73%</td>\n",
       "      <td>16/23</td>\n",
       "      <td>70%</td>\n",
       "      <td>0.783007</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Autistic behavior [HP:0000729]</th>\n",
       "      <td>26/48</td>\n",
       "      <td>54%</td>\n",
       "      <td>14/24</td>\n",
       "      <td>58%</td>\n",
       "      <td>0.804713</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Macrodontia [HP:0001572]</th>\n",
       "      <td>115/145</td>\n",
       "      <td>79%</td>\n",
       "      <td>55/68</td>\n",
       "      <td>81%</td>\n",
       "      <td>0.856244</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Global developmental delay [HP:0001263]</th>\n",
       "      <td>86/91</td>\n",
       "      <td>95%</td>\n",
       "      <td>46/48</td>\n",
       "      <td>96%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Long philtrum [HP:0000343]</th>\n",
       "      <td>66/82</td>\n",
       "      <td>80%</td>\n",
       "      <td>38/48</td>\n",
       "      <td>79%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Anteverted nares [HP:0000463]</th>\n",
       "      <td>41/58</td>\n",
       "      <td>71%</td>\n",
       "      <td>22/30</td>\n",
       "      <td>73%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Attention deficit hyperactivity disorder [HP:0007018]</th>\n",
       "      <td>34/44</td>\n",
       "      <td>77%</td>\n",
       "      <td>19/24</td>\n",
       "      <td>79%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Strabismus [HP:0000486]</th>\n",
       "      <td>12/51</td>\n",
       "      <td>24%</td>\n",
       "      <td>4/19</td>\n",
       "      <td>21%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Febrile seizure (within the age range of 3 months to 6 years) [HP:0002373]</th>\n",
       "      <td>3/5</td>\n",
       "      <td>60%</td>\n",
       "      <td>2/3</td>\n",
       "      <td>67%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hypotonia [HP:0001252]</th>\n",
       "      <td>14/35</td>\n",
       "      <td>40%</td>\n",
       "      <td>7/18</td>\n",
       "      <td>39%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prominent nose [HP:0000448]</th>\n",
       "      <td>12/28</td>\n",
       "      <td>43%</td>\n",
       "      <td>5/13</td>\n",
       "      <td>38%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Generalized-onset seizure [HP:0002197]</th>\n",
       "      <td>8/9</td>\n",
       "      <td>89%</td>\n",
       "      <td>4/5</td>\n",
       "      <td>80%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Generalized non-motor (absence) seizure [HP:0002121]</th>\n",
       "      <td>1/4</td>\n",
       "      <td>25%</td>\n",
       "      <td>2/4</td>\n",
       "      <td>50%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "FRAMESHIFT_VARIANT on NM_013275.6                       Yes             No  \\\n",
       "                                                      Count Percent  Count   \n",
       "Abnormality of the hand [HP:0001155]                 95/144     66%  60/71   \n",
       "EEG abnormality [HP:0002353]                           7/33     21%   9/16   \n",
       "Feeding difficulties [HP:0011968]                     33/89     37%  26/45   \n",
       "Low anterior hairline [HP:0000294]                    40/58     69%  15/30   \n",
       "Intellectual disability [HP:0001249]                 99/119     83%  59/64   \n",
       "Delayed skeletal maturation [HP:0002750]              37/86     43%  17/28   \n",
       "Inguinal hernia [HP:0000023]                           2/31      6%   4/17   \n",
       "Intrauterine growth retardation [HP:0001511]           6/39     15%   7/23   \n",
       "Bulbous nose [HP:0000414]                             45/71     63%  29/39   \n",
       "Hypertelorism [HP:0000316]                            34/63     54%  25/38   \n",
       "Microretrognathia [HP:0000308]                        13/36     36%   3/16   \n",
       "Sleep abnormality [HP:0002360]                        15/40     38%   4/17   \n",
       "Short stature [HP:0004322]                           73/126     58%  42/65   \n",
       "Delayed speech and language development [HP:000...    77/92     84%  44/49   \n",
       "Exaggerated cupid's bow [HP:0002263]                  11/35     31%   2/12   \n",
       "Mandibular prognathia [HP:0000303]                     8/33     24%   5/14   \n",
       "Cryptorchidism [HP:0000028]                           20/51     39%  15/31   \n",
       "Thick eyebrow [HP:0000574]                            68/82     83%  38/49   \n",
       "Triangular face [HP:0000325]                          45/63     71%  23/36   \n",
       "Microcephaly [HP:0000252]                             18/74     24%  12/39   \n",
       "Motor delay [HP:0001270]                              36/52     69%  19/25   \n",
       "Focal-onset seizure [HP:0007359]                        4/8     50%    4/6   \n",
       "Low-set ears [HP:0000369]                             16/30     53%   5/11   \n",
       "Recurrent otitis media [HP:0000403]                   17/36     47%   7/17   \n",
       "Protruding ear [HP:0000411]                           32/44     73%  16/23   \n",
       "Autistic behavior [HP:0000729]                        26/48     54%  14/24   \n",
       "Macrodontia [HP:0001572]                            115/145     79%  55/68   \n",
       "Global developmental delay [HP:0001263]               86/91     95%  46/48   \n",
       "Long philtrum [HP:0000343]                            66/82     80%  38/48   \n",
       "Anteverted nares [HP:0000463]                         41/58     71%  22/30   \n",
       "Attention deficit hyperactivity disorder [HP:00...    34/44     77%  19/24   \n",
       "Strabismus [HP:0000486]                               12/51     24%   4/19   \n",
       "Febrile seizure (within the age range of 3 mont...      3/5     60%    2/3   \n",
       "Hypotonia [HP:0001252]                                14/35     40%   7/18   \n",
       "Prominent nose [HP:0000448]                           12/28     43%   5/13   \n",
       "Generalized-onset seizure [HP:0002197]                  8/9     89%    4/5   \n",
       "Generalized non-motor (absence) seizure [HP:000...      1/4     25%    2/4   \n",
       "\n",
       "FRAMESHIFT_VARIANT on NM_013275.6                                     \\\n",
       "                                                   Percent   p value   \n",
       "Abnormality of the hand [HP:0001155]                   85%  0.005661   \n",
       "EEG abnormality [HP:0002353]                           56%  0.022884   \n",
       "Feeding difficulties [HP:0011968]                      58%  0.027584   \n",
       "Low anterior hairline [HP:0000294]                     50%  0.105274   \n",
       "Intellectual disability [HP:0001249]                   92%  0.115195   \n",
       "Delayed skeletal maturation [HP:0002750]               61%  0.128802   \n",
       "Inguinal hernia [HP:0000023]                           24%  0.166824   \n",
       "Intrauterine growth retardation [HP:0001511]           30%  0.202568   \n",
       "Bulbous nose [HP:0000414]                              74%  0.291265   \n",
       "Hypertelorism [HP:0000316]                             66%  0.299330   \n",
       "Microretrognathia [HP:0000308]                         19%  0.330525   \n",
       "Sleep abnormality [HP:0002360]                         24%  0.370187   \n",
       "Short stature [HP:0004322]                             65%  0.436104   \n",
       "Delayed speech and language development [HP:000...     90%  0.448471   \n",
       "Exaggerated cupid's bow [HP:0002263]                   17%  0.464459   \n",
       "Mandibular prognathia [HP:0000303]                     36%  0.486071   \n",
       "Cryptorchidism [HP:0000028]                            48%  0.492270   \n",
       "Thick eyebrow [HP:0000574]                             78%  0.494742   \n",
       "Triangular face [HP:0000325]                           64%  0.501819   \n",
       "Microcephaly [HP:0000252]                              31%  0.505435   \n",
       "Motor delay [HP:0001270]                               76%  0.600188   \n",
       "Focal-onset seizure [HP:0007359]                       67%  0.627040   \n",
       "Low-set ears [HP:0000369]                              45%  0.733718   \n",
       "Recurrent otitis media [HP:0000403]                    41%  0.772020   \n",
       "Protruding ear [HP:0000411]                            70%  0.783007   \n",
       "Autistic behavior [HP:0000729]                         58%  0.804713   \n",
       "Macrodontia [HP:0001572]                               81%  0.856244   \n",
       "Global developmental delay [HP:0001263]                96%  1.000000   \n",
       "Long philtrum [HP:0000343]                             79%  1.000000   \n",
       "Anteverted nares [HP:0000463]                          73%  1.000000   \n",
       "Attention deficit hyperactivity disorder [HP:00...     79%  1.000000   \n",
       "Strabismus [HP:0000486]                                21%  1.000000   \n",
       "Febrile seizure (within the age range of 3 mont...     67%  1.000000   \n",
       "Hypotonia [HP:0001252]                                 39%  1.000000   \n",
       "Prominent nose [HP:0000448]                            38%  1.000000   \n",
       "Generalized-onset seizure [HP:0002197]                 80%  1.000000   \n",
       "Generalized non-motor (absence) seizure [HP:000...     50%  1.000000   \n",
       "\n",
       "FRAMESHIFT_VARIANT on NM_013275.6                                     \n",
       "                                                   Corrected p value  \n",
       "Abnormality of the hand [HP:0001155]                        0.209453  \n",
       "EEG abnormality [HP:0002353]                                0.340205  \n",
       "Feeding difficulties [HP:0011968]                           0.340205  \n",
       "Low anterior hairline [HP:0000294]                          0.794281  \n",
       "Intellectual disability [HP:0001249]                        0.794281  \n",
       "Delayed skeletal maturation [HP:0002750]                    0.794281  \n",
       "Inguinal hernia [HP:0000023]                                0.881785  \n",
       "Intrauterine growth retardation [HP:0001511]                0.935054  \n",
       "Bulbous nose [HP:0000414]                                   0.935054  \n",
       "Hypertelorism [HP:0000316]                                  0.935054  \n",
       "Microretrognathia [HP:0000308]                              0.935054  \n",
       "Sleep abnormality [HP:0002360]                              0.935054  \n",
       "Short stature [HP:0004322]                                  0.935054  \n",
       "Delayed speech and language development [HP:000...          0.935054  \n",
       "Exaggerated cupid's bow [HP:0002263]                        0.935054  \n",
       "Mandibular prognathia [HP:0000303]                          0.935054  \n",
       "Cryptorchidism [HP:0000028]                                 0.935054  \n",
       "Thick eyebrow [HP:0000574]                                  0.935054  \n",
       "Triangular face [HP:0000325]                                0.935054  \n",
       "Microcephaly [HP:0000252]                                   0.935054  \n",
       "Motor delay [HP:0001270]                                    1.000000  \n",
       "Focal-onset seizure [HP:0007359]                            1.000000  \n",
       "Low-set ears [HP:0000369]                                   1.000000  \n",
       "Recurrent otitis media [HP:0000403]                         1.000000  \n",
       "Protruding ear [HP:0000411]                                 1.000000  \n",
       "Autistic behavior [HP:0000729]                              1.000000  \n",
       "Macrodontia [HP:0001572]                                    1.000000  \n",
       "Global developmental delay [HP:0001263]                     1.000000  \n",
       "Long philtrum [HP:0000343]                                  1.000000  \n",
       "Anteverted nares [HP:0000463]                               1.000000  \n",
       "Attention deficit hyperactivity disorder [HP:00...          1.000000  \n",
       "Strabismus [HP:0000486]                                     1.000000  \n",
       "Febrile seizure (within the age range of 3 mont...          1.000000  \n",
       "Hypotonia [HP:0001252]                                      1.000000  \n",
       "Prominent nose [HP:0000448]                                 1.000000  \n",
       "Generalized-onset seizure [HP:0002197]                      1.000000  \n",
       "Generalized non-motor (absence) seizure [HP:000...          1.000000  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frameshift = analysis.compare_by_variant_effect(VariantEffect.FRAMESHIFT_VARIANT, tx_id=tx_id)\n",
    "frameshift.summarize(hpo, PatientCategories.YES)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9841f6f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from genophenocorr.view import StatsViewable\n",
    "sv = StatsViewable(frameshift.mtc_filter_report)\n",
    "\n",
    "#(filter_method_name=analysis_config.mtc_strategy, mtc_name=\"nana\", filter_results_map=frameshift.mtc_filter_report,term_count=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "eb0dc510",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<!DOCTYPE html>\n",
       "<html lang=\"en\">\n",
       "<head>\n",
       "  <meta charset=\"utf-8\">\n",
       "  <title>Cohort</title>\n",
       "  <style>table {\n",
       "    border-collapse: collapse;\n",
       "    margin: 25px 0;\n",
       "      font-size: 0.9em;\n",
       "      font-family: sans-serif;\n",
       "      min-width: 400px;\n",
       "      box-shadow: 0 0 20px rgba(0, 0, 0, 0.15);\n",
       "  }\n",
       "\n",
       "\n",
       "  th {\n",
       "    background-color: #f2f2f2;\n",
       "    border: 1px solid #dddddd;\n",
       "    text-align: left;\n",
       "    padding: 2px;\n",
       "  }\n",
       "\n",
       "  tr {\n",
       "    border: 1px solid #dddddd;\n",
       "  }\n",
       "\n",
       "  td {\n",
       "    padding: 2px;\n",
       "    font-weight: bold;\n",
       "  }\n",
       "\n",
       "  tr:nth-child(even) {\n",
       "    background-color: #f2f2f2;\n",
       "  }\n",
       "\n",
       "  .table td,tr {\n",
       "    text-align: right;\n",
       "}\n",
       "\n",
       ".table td:first-child, tr:first-child {\n",
       "    text-align: left;\n",
       "}​\n",
       "\n",
       "  </style>\n",
       "</head>\n",
       "\n",
       "<body>\n",
       "  <h1>Statistical analysis: fdr_bh (heuristic sampler)</h1>\n",
       "  <p>Performed statistical tests for 37 HPO terms.</p>\n",
       "  <p>Using  heuristic sampler, the following numbers of terms were omitted from statistical analysis. </p>\n",
       "    <h3>Omitted from statisical analysis</h3>\n",
       "    <table>\n",
       "        <tbody>\n",
       "            <tr>\n",
       "                <th>Skipped</th>\n",
       "                <th>Count</th>\n",
       "              </tr>\n",
       "              \n",
       "              <tr>\n",
       "                <td>Skipping top level term</td>\n",
       "                <td>12</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term with only 3 observations (not powered for 2x2)</td>\n",
       "                <td>8</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term with only 6 observations (not powered for 2x2)</td>\n",
       "                <td>5</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping non phenotype term </td>\n",
       "                <td>2</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term with only 5 observations (not powered for 2x2)</td>\n",
       "                <td>2</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0012759 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0012638 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0011446 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0012758 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0001999 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000271 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000234 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000534 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0030669 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0032039 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000315 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0001595 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0011138 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000288 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000177 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000159 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000163 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0031816 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000153 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000405 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0011452 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000370 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0031703 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000365 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000364 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0031704 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0002719 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0032101 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0010978 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000388 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0012649 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0012647 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000347 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0009118 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0009116 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0009122 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0009121 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0011842 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000924 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0009115 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000929 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000277 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0030791 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0011821 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000278 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0011339 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000664 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0002219 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000998 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0011362 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0005288 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0005105 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000366 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000429 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0010938 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000752 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000734 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:5200241 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000708 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000736 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:5200044 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0100543 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0030178 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0001311 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0002463 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0034434 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0002167 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0001250 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0100886 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0012372 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0011458 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0025032 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000436 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0006482 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000164 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0002817 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0001510 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000002 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000927 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0011843 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000035 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0011961 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000027 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0008669 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000025 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0012874 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000080 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000078 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0040195 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000240 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0007364 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0002977 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0002011 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0012639 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0002060 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0100547 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0012443 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0001671 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0001627 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0030680 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000549 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000496 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0012373 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000357 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000377 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000356 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0010438 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0001713 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000599 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0009553 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0100037 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0001965 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0011361 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0010720 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000290 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000306 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0003808 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0011804 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0003011 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0000407 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0011389 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0001631 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0011994 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>Skipping term HP:0005120 because all genotypes have same HPO observed proportions</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "                </tbody>\n",
       "    </table>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(HTML(sv.process(frameshift)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a83a5ca",
   "metadata": {},
   "source": [
    "Test for presence of genotype-phenotype correlations between subjects with >=1 allele of a variant vs. the other subjects:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "57dd0e23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>&gt;=1 allele of the variant 16_89284634_89284639_GTGTTT_G</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Yes</th>\n",
       "      <th colspan=\"2\" halign=\"left\">No</th>\n",
       "      <th colspan=\"2\" halign=\"left\"></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>p value</th>\n",
       "      <th>Corrected p value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Sleep abnormality [HP:0002360]</th>\n",
       "      <td>7/9</td>\n",
       "      <td>78%</td>\n",
       "      <td>12/48</td>\n",
       "      <td>25%</td>\n",
       "      <td>0.004267</td>\n",
       "      <td>0.153609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Autistic behavior [HP:0000729]</th>\n",
       "      <td>1/6</td>\n",
       "      <td>17%</td>\n",
       "      <td>39/66</td>\n",
       "      <td>59%</td>\n",
       "      <td>0.081923</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Intellectual disability [HP:0001249]</th>\n",
       "      <td>14/19</td>\n",
       "      <td>74%</td>\n",
       "      <td>144/164</td>\n",
       "      <td>88%</td>\n",
       "      <td>0.147465</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Synophrys [HP:0000664]</th>\n",
       "      <td>5/14</td>\n",
       "      <td>36%</td>\n",
       "      <td>48/82</td>\n",
       "      <td>59%</td>\n",
       "      <td>0.148677</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EEG abnormality [HP:0002353]</th>\n",
       "      <td>0/6</td>\n",
       "      <td>0%</td>\n",
       "      <td>16/43</td>\n",
       "      <td>37%</td>\n",
       "      <td>0.158804</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Low-set ears [HP:0000369]</th>\n",
       "      <td>5/6</td>\n",
       "      <td>83%</td>\n",
       "      <td>16/35</td>\n",
       "      <td>46%</td>\n",
       "      <td>0.183611</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Motor delay [HP:0001270]</th>\n",
       "      <td>6/11</td>\n",
       "      <td>55%</td>\n",
       "      <td>49/66</td>\n",
       "      <td>74%</td>\n",
       "      <td>0.277389</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Abnormality of the hand [HP:0001155]</th>\n",
       "      <td>22/27</td>\n",
       "      <td>81%</td>\n",
       "      <td>133/188</td>\n",
       "      <td>71%</td>\n",
       "      <td>0.358444</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Recurrent otitis media [HP:0000403]</th>\n",
       "      <td>2/7</td>\n",
       "      <td>29%</td>\n",
       "      <td>22/46</td>\n",
       "      <td>48%</td>\n",
       "      <td>0.436242</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Long philtrum [HP:0000343]</th>\n",
       "      <td>12/13</td>\n",
       "      <td>92%</td>\n",
       "      <td>92/117</td>\n",
       "      <td>79%</td>\n",
       "      <td>0.463475</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Short stature [HP:0004322]</th>\n",
       "      <td>12/23</td>\n",
       "      <td>52%</td>\n",
       "      <td>103/168</td>\n",
       "      <td>61%</td>\n",
       "      <td>0.496540</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Anteverted nares [HP:0000463]</th>\n",
       "      <td>8/13</td>\n",
       "      <td>62%</td>\n",
       "      <td>55/75</td>\n",
       "      <td>73%</td>\n",
       "      <td>0.505804</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bulbous nose [HP:0000414]</th>\n",
       "      <td>10/13</td>\n",
       "      <td>77%</td>\n",
       "      <td>64/97</td>\n",
       "      <td>66%</td>\n",
       "      <td>0.540179</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Low anterior hairline [HP:0000294]</th>\n",
       "      <td>7/13</td>\n",
       "      <td>54%</td>\n",
       "      <td>48/75</td>\n",
       "      <td>64%</td>\n",
       "      <td>0.542610</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Global developmental delay [HP:0001263]</th>\n",
       "      <td>15/16</td>\n",
       "      <td>94%</td>\n",
       "      <td>117/123</td>\n",
       "      <td>95%</td>\n",
       "      <td>0.583703</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mandibular prognathia [HP:0000303]</th>\n",
       "      <td>2/5</td>\n",
       "      <td>40%</td>\n",
       "      <td>11/42</td>\n",
       "      <td>26%</td>\n",
       "      <td>0.606967</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Microretrognathia [HP:0000308]</th>\n",
       "      <td>2/5</td>\n",
       "      <td>40%</td>\n",
       "      <td>14/47</td>\n",
       "      <td>30%</td>\n",
       "      <td>0.637363</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Attention deficit hyperactivity disorder [HP:0007018]</th>\n",
       "      <td>5/7</td>\n",
       "      <td>71%</td>\n",
       "      <td>48/61</td>\n",
       "      <td>79%</td>\n",
       "      <td>0.644784</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Exaggerated cupid's bow [HP:0002263]</th>\n",
       "      <td>1/7</td>\n",
       "      <td>14%</td>\n",
       "      <td>12/40</td>\n",
       "      <td>30%</td>\n",
       "      <td>0.654898</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Protruding ear [HP:0000411]</th>\n",
       "      <td>5/6</td>\n",
       "      <td>83%</td>\n",
       "      <td>43/61</td>\n",
       "      <td>70%</td>\n",
       "      <td>0.666587</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hypotonia [HP:0001252]</th>\n",
       "      <td>2/7</td>\n",
       "      <td>29%</td>\n",
       "      <td>19/46</td>\n",
       "      <td>41%</td>\n",
       "      <td>0.689725</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Delayed speech and language development [HP:0000750]</th>\n",
       "      <td>13/16</td>\n",
       "      <td>81%</td>\n",
       "      <td>108/125</td>\n",
       "      <td>86%</td>\n",
       "      <td>0.701519</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Thick eyebrow [HP:0000574]</th>\n",
       "      <td>10/13</td>\n",
       "      <td>77%</td>\n",
       "      <td>96/118</td>\n",
       "      <td>81%</td>\n",
       "      <td>0.712830</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hypertelorism [HP:0000316]</th>\n",
       "      <td>5/10</td>\n",
       "      <td>50%</td>\n",
       "      <td>54/91</td>\n",
       "      <td>59%</td>\n",
       "      <td>0.737500</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Feeding difficulties [HP:0011968]</th>\n",
       "      <td>6/15</td>\n",
       "      <td>40%</td>\n",
       "      <td>53/119</td>\n",
       "      <td>45%</td>\n",
       "      <td>0.789475</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Delayed skeletal maturation [HP:0002750]</th>\n",
       "      <td>10/20</td>\n",
       "      <td>50%</td>\n",
       "      <td>44/94</td>\n",
       "      <td>47%</td>\n",
       "      <td>0.810453</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Triangular face [HP:0000325]</th>\n",
       "      <td>6/8</td>\n",
       "      <td>75%</td>\n",
       "      <td>62/91</td>\n",
       "      <td>68%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Focal-onset seizure [HP:0007359]</th>\n",
       "      <td>1/1</td>\n",
       "      <td>100%</td>\n",
       "      <td>7/13</td>\n",
       "      <td>54%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Macrodontia [HP:0001572]</th>\n",
       "      <td>22/27</td>\n",
       "      <td>81%</td>\n",
       "      <td>148/186</td>\n",
       "      <td>80%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cryptorchidism [HP:0000028]</th>\n",
       "      <td>5/12</td>\n",
       "      <td>42%</td>\n",
       "      <td>30/70</td>\n",
       "      <td>43%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Microcephaly [HP:0000252]</th>\n",
       "      <td>4/16</td>\n",
       "      <td>25%</td>\n",
       "      <td>26/97</td>\n",
       "      <td>27%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Strabismus [HP:0000486]</th>\n",
       "      <td>2/10</td>\n",
       "      <td>20%</td>\n",
       "      <td>14/60</td>\n",
       "      <td>23%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Febrile seizure (within the age range of 3 months to 6 years) [HP:0002373]</th>\n",
       "      <td>1/1</td>\n",
       "      <td>100%</td>\n",
       "      <td>4/7</td>\n",
       "      <td>57%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Intrauterine growth retardation [HP:0001511]</th>\n",
       "      <td>1/7</td>\n",
       "      <td>14%</td>\n",
       "      <td>12/55</td>\n",
       "      <td>22%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inguinal hernia [HP:0000023]</th>\n",
       "      <td>0/4</td>\n",
       "      <td>0%</td>\n",
       "      <td>6/44</td>\n",
       "      <td>14%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prominent nose [HP:0000448]</th>\n",
       "      <td>2/4</td>\n",
       "      <td>50%</td>\n",
       "      <td>15/37</td>\n",
       "      <td>41%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       ">=1 allele of the variant 16_89284634_89284639_GTGTTT_G    Yes          \\\n",
       "                                                         Count Percent   \n",
       "Sleep abnormality [HP:0002360]                             7/9     78%   \n",
       "Autistic behavior [HP:0000729]                             1/6     17%   \n",
       "Intellectual disability [HP:0001249]                     14/19     74%   \n",
       "Synophrys [HP:0000664]                                    5/14     36%   \n",
       "EEG abnormality [HP:0002353]                               0/6      0%   \n",
       "Low-set ears [HP:0000369]                                  5/6     83%   \n",
       "Motor delay [HP:0001270]                                  6/11     55%   \n",
       "Abnormality of the hand [HP:0001155]                     22/27     81%   \n",
       "Recurrent otitis media [HP:0000403]                        2/7     29%   \n",
       "Long philtrum [HP:0000343]                               12/13     92%   \n",
       "Short stature [HP:0004322]                               12/23     52%   \n",
       "Anteverted nares [HP:0000463]                             8/13     62%   \n",
       "Bulbous nose [HP:0000414]                                10/13     77%   \n",
       "Low anterior hairline [HP:0000294]                        7/13     54%   \n",
       "Global developmental delay [HP:0001263]                  15/16     94%   \n",
       "Mandibular prognathia [HP:0000303]                         2/5     40%   \n",
       "Microretrognathia [HP:0000308]                             2/5     40%   \n",
       "Attention deficit hyperactivity disorder [HP:00...         5/7     71%   \n",
       "Exaggerated cupid's bow [HP:0002263]                       1/7     14%   \n",
       "Protruding ear [HP:0000411]                                5/6     83%   \n",
       "Hypotonia [HP:0001252]                                     2/7     29%   \n",
       "Delayed speech and language development [HP:000...       13/16     81%   \n",
       "Thick eyebrow [HP:0000574]                               10/13     77%   \n",
       "Hypertelorism [HP:0000316]                                5/10     50%   \n",
       "Feeding difficulties [HP:0011968]                         6/15     40%   \n",
       "Delayed skeletal maturation [HP:0002750]                 10/20     50%   \n",
       "Triangular face [HP:0000325]                               6/8     75%   \n",
       "Focal-onset seizure [HP:0007359]                           1/1    100%   \n",
       "Macrodontia [HP:0001572]                                 22/27     81%   \n",
       "Cryptorchidism [HP:0000028]                               5/12     42%   \n",
       "Microcephaly [HP:0000252]                                 4/16     25%   \n",
       "Strabismus [HP:0000486]                                   2/10     20%   \n",
       "Febrile seizure (within the age range of 3 mont...         1/1    100%   \n",
       "Intrauterine growth retardation [HP:0001511]               1/7     14%   \n",
       "Inguinal hernia [HP:0000023]                               0/4      0%   \n",
       "Prominent nose [HP:0000448]                                2/4     50%   \n",
       "\n",
       ">=1 allele of the variant 16_89284634_89284639_GTGTTT_G       No          \\\n",
       "                                                           Count Percent   \n",
       "Sleep abnormality [HP:0002360]                             12/48     25%   \n",
       "Autistic behavior [HP:0000729]                             39/66     59%   \n",
       "Intellectual disability [HP:0001249]                     144/164     88%   \n",
       "Synophrys [HP:0000664]                                     48/82     59%   \n",
       "EEG abnormality [HP:0002353]                               16/43     37%   \n",
       "Low-set ears [HP:0000369]                                  16/35     46%   \n",
       "Motor delay [HP:0001270]                                   49/66     74%   \n",
       "Abnormality of the hand [HP:0001155]                     133/188     71%   \n",
       "Recurrent otitis media [HP:0000403]                        22/46     48%   \n",
       "Long philtrum [HP:0000343]                                92/117     79%   \n",
       "Short stature [HP:0004322]                               103/168     61%   \n",
       "Anteverted nares [HP:0000463]                              55/75     73%   \n",
       "Bulbous nose [HP:0000414]                                  64/97     66%   \n",
       "Low anterior hairline [HP:0000294]                         48/75     64%   \n",
       "Global developmental delay [HP:0001263]                  117/123     95%   \n",
       "Mandibular prognathia [HP:0000303]                         11/42     26%   \n",
       "Microretrognathia [HP:0000308]                             14/47     30%   \n",
       "Attention deficit hyperactivity disorder [HP:00...         48/61     79%   \n",
       "Exaggerated cupid's bow [HP:0002263]                       12/40     30%   \n",
       "Protruding ear [HP:0000411]                                43/61     70%   \n",
       "Hypotonia [HP:0001252]                                     19/46     41%   \n",
       "Delayed speech and language development [HP:000...       108/125     86%   \n",
       "Thick eyebrow [HP:0000574]                                96/118     81%   \n",
       "Hypertelorism [HP:0000316]                                 54/91     59%   \n",
       "Feeding difficulties [HP:0011968]                         53/119     45%   \n",
       "Delayed skeletal maturation [HP:0002750]                   44/94     47%   \n",
       "Triangular face [HP:0000325]                               62/91     68%   \n",
       "Focal-onset seizure [HP:0007359]                            7/13     54%   \n",
       "Macrodontia [HP:0001572]                                 148/186     80%   \n",
       "Cryptorchidism [HP:0000028]                                30/70     43%   \n",
       "Microcephaly [HP:0000252]                                  26/97     27%   \n",
       "Strabismus [HP:0000486]                                    14/60     23%   \n",
       "Febrile seizure (within the age range of 3 mont...           4/7     57%   \n",
       "Intrauterine growth retardation [HP:0001511]               12/55     22%   \n",
       "Inguinal hernia [HP:0000023]                                6/44     14%   \n",
       "Prominent nose [HP:0000448]                                15/37     41%   \n",
       "\n",
       ">=1 allele of the variant 16_89284634_89284639_GTGTTT_G            \\\n",
       "                                                          p value   \n",
       "Sleep abnormality [HP:0002360]                           0.004267   \n",
       "Autistic behavior [HP:0000729]                           0.081923   \n",
       "Intellectual disability [HP:0001249]                     0.147465   \n",
       "Synophrys [HP:0000664]                                   0.148677   \n",
       "EEG abnormality [HP:0002353]                             0.158804   \n",
       "Low-set ears [HP:0000369]                                0.183611   \n",
       "Motor delay [HP:0001270]                                 0.277389   \n",
       "Abnormality of the hand [HP:0001155]                     0.358444   \n",
       "Recurrent otitis media [HP:0000403]                      0.436242   \n",
       "Long philtrum [HP:0000343]                               0.463475   \n",
       "Short stature [HP:0004322]                               0.496540   \n",
       "Anteverted nares [HP:0000463]                            0.505804   \n",
       "Bulbous nose [HP:0000414]                                0.540179   \n",
       "Low anterior hairline [HP:0000294]                       0.542610   \n",
       "Global developmental delay [HP:0001263]                  0.583703   \n",
       "Mandibular prognathia [HP:0000303]                       0.606967   \n",
       "Microretrognathia [HP:0000308]                           0.637363   \n",
       "Attention deficit hyperactivity disorder [HP:00...       0.644784   \n",
       "Exaggerated cupid's bow [HP:0002263]                     0.654898   \n",
       "Protruding ear [HP:0000411]                              0.666587   \n",
       "Hypotonia [HP:0001252]                                   0.689725   \n",
       "Delayed speech and language development [HP:000...       0.701519   \n",
       "Thick eyebrow [HP:0000574]                               0.712830   \n",
       "Hypertelorism [HP:0000316]                               0.737500   \n",
       "Feeding difficulties [HP:0011968]                        0.789475   \n",
       "Delayed skeletal maturation [HP:0002750]                 0.810453   \n",
       "Triangular face [HP:0000325]                             1.000000   \n",
       "Focal-onset seizure [HP:0007359]                         1.000000   \n",
       "Macrodontia [HP:0001572]                                 1.000000   \n",
       "Cryptorchidism [HP:0000028]                              1.000000   \n",
       "Microcephaly [HP:0000252]                                1.000000   \n",
       "Strabismus [HP:0000486]                                  1.000000   \n",
       "Febrile seizure (within the age range of 3 mont...       1.000000   \n",
       "Intrauterine growth retardation [HP:0001511]             1.000000   \n",
       "Inguinal hernia [HP:0000023]                             1.000000   \n",
       "Prominent nose [HP:0000448]                              1.000000   \n",
       "\n",
       ">=1 allele of the variant 16_89284634_89284639_GTGTTT_G                    \n",
       "                                                        Corrected p value  \n",
       "Sleep abnormality [HP:0002360]                                   0.153609  \n",
       "Autistic behavior [HP:0000729]                                   1.000000  \n",
       "Intellectual disability [HP:0001249]                             1.000000  \n",
       "Synophrys [HP:0000664]                                           1.000000  \n",
       "EEG abnormality [HP:0002353]                                     1.000000  \n",
       "Low-set ears [HP:0000369]                                        1.000000  \n",
       "Motor delay [HP:0001270]                                         1.000000  \n",
       "Abnormality of the hand [HP:0001155]                             1.000000  \n",
       "Recurrent otitis media [HP:0000403]                              1.000000  \n",
       "Long philtrum [HP:0000343]                                       1.000000  \n",
       "Short stature [HP:0004322]                                       1.000000  \n",
       "Anteverted nares [HP:0000463]                                    1.000000  \n",
       "Bulbous nose [HP:0000414]                                        1.000000  \n",
       "Low anterior hairline [HP:0000294]                               1.000000  \n",
       "Global developmental delay [HP:0001263]                          1.000000  \n",
       "Mandibular prognathia [HP:0000303]                               1.000000  \n",
       "Microretrognathia [HP:0000308]                                   1.000000  \n",
       "Attention deficit hyperactivity disorder [HP:00...               1.000000  \n",
       "Exaggerated cupid's bow [HP:0002263]                             1.000000  \n",
       "Protruding ear [HP:0000411]                                      1.000000  \n",
       "Hypotonia [HP:0001252]                                           1.000000  \n",
       "Delayed speech and language development [HP:000...               1.000000  \n",
       "Thick eyebrow [HP:0000574]                                       1.000000  \n",
       "Hypertelorism [HP:0000316]                                       1.000000  \n",
       "Feeding difficulties [HP:0011968]                                1.000000  \n",
       "Delayed skeletal maturation [HP:0002750]                         1.000000  \n",
       "Triangular face [HP:0000325]                                     1.000000  \n",
       "Focal-onset seizure [HP:0007359]                                 1.000000  \n",
       "Macrodontia [HP:0001572]                                         1.000000  \n",
       "Cryptorchidism [HP:0000028]                                      1.000000  \n",
       "Microcephaly [HP:0000252]                                        1.000000  \n",
       "Strabismus [HP:0000486]                                          1.000000  \n",
       "Febrile seizure (within the age range of 3 mont...               1.000000  \n",
       "Intrauterine growth retardation [HP:0001511]                     1.000000  \n",
       "Inguinal hernia [HP:0000023]                                     1.000000  \n",
       "Prominent nose [HP:0000448]                                      1.000000  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var_single = analysis.compare_by_variant_key('16_89284634_89284639_GTGTTT_G')\n",
    "var_single.summarize(hpo, PatientCategories.YES)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74eaa536",
   "metadata": {},
   "source": [
    "Or between subjects with one variant vs. the other variant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0dfa0a50",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "PatientCategory(cat_id=0, name=No, description=The patient does not belong to the group.)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/Documents/Programs/genophenocorr/enviro/lib/python3.10/site-packages/pandas/core/indexes/base.py:3790\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3789\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3790\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3791\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32mindex.pyx:152\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mindex.pyx:181\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7080\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7088\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: PatientCategory(cat_id=0, name=No, description=The patient does not belong to the group.)",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[12], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m var_double \u001b[38;5;241m=\u001b[39m \u001b[43manalysis\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcompare_by_variant_keys\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m16_89284129_89284134_CTTTTT_C\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m16_89284634_89284639_GTGTTT_G\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      2\u001b[0m var_double\u001b[38;5;241m.\u001b[39msummarize(hpo, PatientCategories\u001b[38;5;241m.\u001b[39mYES)\n",
      "File \u001b[0;32m~/Documents/Programs/genophenocorr/src/genophenocorr/analysis/_gp_impl.py:79\u001b[0m, in \u001b[0;36mGpCohortAnalysis.compare_by_variant_keys\u001b[0;34m(self, variant_key1, variant_key2)\u001b[0m\n\u001b[1;32m     77\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcompare_by_variant_keys\u001b[39m(\u001b[38;5;28mself\u001b[39m, variant_key1: \u001b[38;5;28mstr\u001b[39m, variant_key2: \u001b[38;5;28mstr\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m GenotypePhenotypeAnalysisResult:\n\u001b[1;32m     78\u001b[0m     predicate \u001b[38;5;241m=\u001b[39m VariantsPredicate(variant_key1, variant_key2)\n\u001b[0;32m---> 79\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_apply_poly_predicate_on_hpo_terms\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpredicate\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/Programs/genophenocorr/src/genophenocorr/analysis/_gp_impl.py:160\u001b[0m, in \u001b[0;36mGpCohortAnalysis._apply_poly_predicate_on_hpo_terms\u001b[0;34m(self, gt_predicate)\u001b[0m\n\u001b[1;32m    156\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(gt_predicate, GenotypePolyPredicate), \\\n\u001b[1;32m    157\u001b[0m     \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(gt_predicate)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m is not an instance of `GenotypePolyPredicate`\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m    159\u001b[0m pheno_predicates \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_prepare_phenotype_predicates()\n\u001b[0;32m--> 160\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_apply_poly_predicate\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpheno_predicates\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgt_predicate\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/Programs/genophenocorr/src/genophenocorr/analysis/_gp_impl.py:168\u001b[0m, in \u001b[0;36mGpCohortAnalysis._apply_poly_predicate\u001b[0;34m(self, pheno_predicates, gt_predicate)\u001b[0m\n\u001b[1;32m    163\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_apply_poly_predicate\u001b[39m(\n\u001b[1;32m    164\u001b[0m         \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    165\u001b[0m         pheno_predicates: typing\u001b[38;5;241m.\u001b[39mIterable[PhenotypePolyPredicate[P]],\n\u001b[1;32m    166\u001b[0m         gt_predicate: GenotypePolyPredicate,\n\u001b[1;32m    167\u001b[0m ):\n\u001b[0;32m--> 168\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_gp_analyzer\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43manalyze\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    169\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpatients\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_patient_list\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    170\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpheno_predicates\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpheno_predicates\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    171\u001b[0m \u001b[43m        \u001b[49m\u001b[43mgt_predicate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgt_predicate\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    172\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/Programs/genophenocorr/src/genophenocorr/analysis/_gp_analysis.py:372\u001b[0m, in \u001b[0;36mFisherExactAnalyzer.analyze\u001b[0;34m(self, patients, pheno_predicates, gt_predicate)\u001b[0m\n\u001b[1;32m    368\u001b[0m categories, n_usable, all_counts \u001b[38;5;241m=\u001b[39m GPAnalyzer\u001b[38;5;241m.\u001b[39m_count_patients(\n\u001b[1;32m    369\u001b[0m     patients, pheno_predicates, gt_predicate,\n\u001b[1;32m    370\u001b[0m )\n\u001b[1;32m    371\u001b[0m \u001b[38;5;66;03m# 1.5) Filter terms for MTC\u001b[39;00m\n\u001b[0;32m--> 372\u001b[0m n_usable, all_counts, filtered_terms_d \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_hpo_mtc_filter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfilter_terms_to_test\u001b[49m\u001b[43m(\u001b[49m\u001b[43mn_usable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mall_counts\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    374\u001b[0m \u001b[38;5;66;03m# 2) Statistical tests\u001b[39;00m\n\u001b[1;32m    375\u001b[0m pheno_idx \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mIndex(n_usable\u001b[38;5;241m.\u001b[39mindex, name\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mp_val\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m~/Documents/Programs/genophenocorr/src/genophenocorr/analysis/_gp_analysis.py:229\u001b[0m, in \u001b[0;36mHeuristicSamplerMtcFilter.filter_terms_to_test\u001b[0;34m(self, n_usable, all_counts)\u001b[0m\n\u001b[1;32m    227\u001b[0m     filtered_terms_d[reason] \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m    228\u001b[0m     \u001b[38;5;28;01mcontinue\u001b[39;00m\n\u001b[0;32m--> 229\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[43mHeuristicSamplerMtcFilter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgenotypes_have_same_HPO_proportions\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcounts_frame\u001b[49m\u001b[43m)\u001b[49m:\n\u001b[1;32m    230\u001b[0m     reason \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSkipping term \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mhpo_tk_term_id\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m because all genotypes have same HPO observed proportions\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    231\u001b[0m     filtered_terms_d[reason] \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n",
      "File \u001b[0;32m~/Documents/Programs/genophenocorr/src/genophenocorr/analysis/_gp_analysis.py:138\u001b[0m, in \u001b[0;36mHeuristicSamplerMtcFilter.genotypes_have_same_HPO_proportions\u001b[0;34m(counts)\u001b[0m\n\u001b[1;32m    136\u001b[0m DELTA \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0.01\u001b[39m\n\u001b[1;32m    137\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m counts\u001b[38;5;241m.\u001b[39mshape \u001b[38;5;241m==\u001b[39m (\u001b[38;5;241m2\u001b[39m, \u001b[38;5;241m2\u001b[39m):\n\u001b[0;32m--> 138\u001b[0m     num1 \u001b[38;5;241m=\u001b[39m  \u001b[43mcounts\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloc\u001b[49m\u001b[43m[\u001b[49m\u001b[43mPatientCategories\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mYES\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mPatientCategories\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mNO\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m    139\u001b[0m     denom1 \u001b[38;5;241m=\u001b[39m  counts\u001b[38;5;241m.\u001b[39mloc[PatientCategories\u001b[38;5;241m.\u001b[39mYES, PatientCategories\u001b[38;5;241m.\u001b[39mNO] \u001b[38;5;241m+\u001b[39m  counts\u001b[38;5;241m.\u001b[39mloc[PatientCategories\u001b[38;5;241m.\u001b[39mNO, PatientCategories\u001b[38;5;241m.\u001b[39mNO]\n\u001b[1;32m    140\u001b[0m     num2 \u001b[38;5;241m=\u001b[39m counts\u001b[38;5;241m.\u001b[39mloc[PatientCategories\u001b[38;5;241m.\u001b[39mYES, PatientCategories\u001b[38;5;241m.\u001b[39mYES]\n",
      "File \u001b[0;32m~/Documents/Programs/genophenocorr/enviro/lib/python3.10/site-packages/pandas/core/indexing.py:1147\u001b[0m, in \u001b[0;36m_LocationIndexer.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1145\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_is_scalar_access(key):\n\u001b[1;32m   1146\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj\u001b[38;5;241m.\u001b[39m_get_value(\u001b[38;5;241m*\u001b[39mkey, takeable\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_takeable)\n\u001b[0;32m-> 1147\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_tuple\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1148\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1149\u001b[0m     \u001b[38;5;66;03m# we by definition only have the 0th axis\u001b[39;00m\n\u001b[1;32m   1150\u001b[0m     axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxis \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m0\u001b[39m\n",
      "File \u001b[0;32m~/Documents/Programs/genophenocorr/enviro/lib/python3.10/site-packages/pandas/core/indexing.py:1330\u001b[0m, in \u001b[0;36m_LocIndexer._getitem_tuple\u001b[0;34m(self, tup)\u001b[0m\n\u001b[1;32m   1328\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m suppress(IndexingError):\n\u001b[1;32m   1329\u001b[0m     tup \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_expand_ellipsis(tup)\n\u001b[0;32m-> 1330\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_lowerdim\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtup\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1332\u001b[0m \u001b[38;5;66;03m# no multi-index, so validate all of the indexers\u001b[39;00m\n\u001b[1;32m   1333\u001b[0m tup \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_tuple_indexer(tup)\n",
      "File \u001b[0;32m~/Documents/Programs/genophenocorr/enviro/lib/python3.10/site-packages/pandas/core/indexing.py:1063\u001b[0m, in \u001b[0;36m_LocationIndexer._getitem_lowerdim\u001b[0;34m(self, tup)\u001b[0m\n\u001b[1;32m   1061\u001b[0m             \u001b[38;5;28;01mreturn\u001b[39;00m section\n\u001b[1;32m   1062\u001b[0m         \u001b[38;5;66;03m# This is an elided recursive call to iloc/loc\u001b[39;00m\n\u001b[0;32m-> 1063\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mgetattr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43msection\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mname\u001b[49m\u001b[43m)\u001b[49m\u001b[43m[\u001b[49m\u001b[43mnew_key\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m   1065\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m IndexingError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnot applicable\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/Documents/Programs/genophenocorr/enviro/lib/python3.10/site-packages/pandas/core/indexing.py:1153\u001b[0m, in \u001b[0;36m_LocationIndexer.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1150\u001b[0m axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxis \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m0\u001b[39m\n\u001b[1;32m   1152\u001b[0m maybe_callable \u001b[38;5;241m=\u001b[39m com\u001b[38;5;241m.\u001b[39mapply_if_callable(key, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj)\n\u001b[0;32m-> 1153\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_axis\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmaybe_callable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/Programs/genophenocorr/enviro/lib/python3.10/site-packages/pandas/core/indexing.py:1393\u001b[0m, in \u001b[0;36m_LocIndexer._getitem_axis\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1391\u001b[0m \u001b[38;5;66;03m# fall thru to straight lookup\u001b[39;00m\n\u001b[1;32m   1392\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_key(key, axis)\n\u001b[0;32m-> 1393\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_label\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/Programs/genophenocorr/enviro/lib/python3.10/site-packages/pandas/core/indexing.py:1343\u001b[0m, in \u001b[0;36m_LocIndexer._get_label\u001b[0;34m(self, label, axis)\u001b[0m\n\u001b[1;32m   1341\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_get_label\u001b[39m(\u001b[38;5;28mself\u001b[39m, label, axis: AxisInt):\n\u001b[1;32m   1342\u001b[0m     \u001b[38;5;66;03m# GH#5567 this will fail if the label is not present in the axis.\u001b[39;00m\n\u001b[0;32m-> 1343\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mxs\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlabel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/Programs/genophenocorr/enviro/lib/python3.10/site-packages/pandas/core/generic.py:4236\u001b[0m, in \u001b[0;36mNDFrame.xs\u001b[0;34m(self, key, axis, level, drop_level)\u001b[0m\n\u001b[1;32m   4234\u001b[0m             new_index \u001b[38;5;241m=\u001b[39m index[loc]\n\u001b[1;32m   4235\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 4236\u001b[0m     loc \u001b[38;5;241m=\u001b[39m \u001b[43mindex\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4238\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(loc, np\u001b[38;5;241m.\u001b[39mndarray):\n\u001b[1;32m   4239\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m loc\u001b[38;5;241m.\u001b[39mdtype \u001b[38;5;241m==\u001b[39m np\u001b[38;5;241m.\u001b[39mbool_:\n",
      "File \u001b[0;32m~/Documents/Programs/genophenocorr/enviro/lib/python3.10/site-packages/pandas/core/indexes/base.py:3797\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3792\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[1;32m   3793\u001b[0m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[1;32m   3794\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[1;32m   3795\u001b[0m     ):\n\u001b[1;32m   3796\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[0;32m-> 3797\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3798\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3799\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3800\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3801\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3802\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: PatientCategory(cat_id=0, name=No, description=The patient does not belong to the group.)"
     ]
    }
   ],
   "source": [
    "var_double = analysis.compare_by_variant_keys('16_89284129_89284134_CTTTTT_C', '16_89284634_89284639_GTGTTT_G')\n",
    "var_double.summarize(hpo, PatientCategories.YES)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a12b8117",
   "metadata": {},
   "source": [
    "TODO - finalize!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "enviro",
   "language": "python",
   "name": "enviro"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
