{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9443554a",
   "metadata": {},
   "source": [
    "# KBG Syndrome\n",
    "\n",
    "Data from [Martinez-Cayuelas E, et al. Clinical description, molecular delineation and genotype-phenotype correlation in 340 patients with KBG syndrome: addition of 67 new patients](https://pubmed.ncbi.nlm.nih.gov/36446582).\n",
    "\n",
    "We investigate subjects with mutations in *ANKRD11*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cd2a5b1b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using genophenocorr version 0.1.1dev\n"
     ]
    }
   ],
   "source": [
    "import genophenocorr\n",
    "\n",
    "print(f\"Using genophenocorr version {genophenocorr.__version__}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8f5f366",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80a3e9ab",
   "metadata": {},
   "source": [
    "### Load HPO\n",
    "\n",
    "We use HPO `v2023-10-09` release for this analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d7623bcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import hpotk\n",
    "\n",
    "fpath_hpo = 'https://github.com/obophenotype/human-phenotype-ontology/releases/download/v2023-10-09/hp.json'\n",
    "hpo = hpotk.load_minimal_ontology(fpath_hpo)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f6cd5b3",
   "metadata": {},
   "source": [
    "### Load Phenopackets\n",
    "\n",
    "We will load phenopacket JSON files located in `phenopackets` folder that is next to the notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "070a06ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Patients Created: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████| 328/328 [00:01<00:00, 196.20it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "328\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Validated under none policy\n",
      "Showing errors and warnings\n",
      "328 phenopacket(s) found at `phenopackets`\n",
      "  patient #0\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P2[PMID_36446582_Novara,_2017_P2]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara,_2017_P2 has no variants to work with\n",
      "  patient #1\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P13[PMID_36446582_Goldenberg2016_P13]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P13 has no variants to work with\n",
      "  patient #3\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Ockeloen2015_P20[PMID_36446582_Ockeloen2015_P20]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Ockeloen2015_P20 has no variants to work with\n",
      "  patient #7\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P22[PMID_36446582_Kutkowska-Kazmierczak2021_P22]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska-Kazmierczak2021_P22 has no variants to work with\n",
      "  patient #10\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Gnazzo, 2020_P31[PMID_36446582_Gnazzo,_2020_P31]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Gnazzo,_2020_P31 has no variants to work with\n",
      "  patient #13\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P29[PMID_36446582_Goldenberg2016_P29]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P29 has no variants to work with\n",
      "  patient #14\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P18[PMID_36446582_Kutkowska-Kazmierczak2021_P18]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska-Kazmierczak2021_P18 has no variants to work with\n",
      "  patient #24\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P1[PMID_36446582_Goldenberg2016_P1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P1 has no variants to work with\n",
      "  patient #29\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P33[PMID_36446582_Goldenberg2016_P33]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P33 has no variants to work with\n",
      "  patient #31\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P14[PMID_36446582_Kutkowska-Kazmierczak2021_P14]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska-Kazmierczak2021_P14 has no variants to work with\n",
      "  patient #36\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P15[PMID_36446582_Kutkowska-Kazmierczak2021_P15]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska-Kazmierczak2021_P15 has no variants to work with\n",
      "  patient #39\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P32[PMID_36446582_Goldenberg2016_P32]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P32 has no variants to work with\n",
      "  patient #40\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Bucerzan2020[PMID_36446582_Bucerzan2020]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Bucerzan2020 has no variants to work with\n",
      "  patient #43\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P24[PMID_36446582_Goldenberg2016_P24]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P24 has no variants to work with\n",
      "  patient #47\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Miyatake, 2013[PMID_36446582_Miyatake,_2013]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Miyatake,_2013 has no variants to work with\n",
      "  patient #53\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG1[PMID_36446582_KBG1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG1 has no variants to work with\n",
      "  patient #57\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Sacharow, 2012_P2[PMID_36446582_Sacharow,_2012_P2]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Sacharow,_2012_P2 has no variants to work with\n",
      "  patient #59\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P19[PMID_36446582_Kutkowska-Kazmierczak2021_P19]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska-Kazmierczak2021_P19 has no variants to work with\n",
      "  patient #61\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P28[PMID_36446582_Goldenberg2016_P28]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P28 has no variants to work with\n",
      "  patient #65\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Gnazzo, 2020_P30[PMID_36446582_Gnazzo,_2020_P30]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Gnazzo,_2020_P30 has no variants to work with\n",
      "  patient #67\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P23[PMID_36446582_Kutkowska-Kazmierczak2021_P23]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska-Kazmierczak2021_P23 has no variants to work with\n",
      "  patient #68\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P11[PMID_36446582_Novara,_2017_P11]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara,_2017_P11 has no variants to work with\n",
      "  patient #72\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P12[PMID_36446582_Goldenberg2016_P12]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P12 has no variants to work with\n",
      "  patient #73\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P3[PMID_36446582_Novara,_2017_P3]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara,_2017_P3 has no variants to work with\n",
      "  patient #74\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Khalifa, 2013_P1B[PMID_36446582_Khalifa,_2013_P1B]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Khalifa,_2013_P1B has no variants to work with\n",
      "  patient #75\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Isrie, 2012_P2[PMID_36446582_Isrie,_2012_P2]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Isrie,_2012_P2 has no variants to work with\n",
      "  patient #76\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P19[PMID_36446582_Goldenberg2016_P19]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P19 has no variants to work with\n",
      "  patient #79\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P8[PMID_36446582_Novara,_2017_P8]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara,_2017_P8 has no variants to work with\n",
      "  patient #82\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG26[PMID_36446582_KBG26]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG26 has no variants to work with\n",
      "  patient #86\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Scarano, 2013_P11[PMID_36446582_Scarano,_2013_P11]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Scarano,_2013_P11 has no variants to work with\n",
      "  patient #88\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Crippa2015_P1[PMID_36446582_Crippa2015_P1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Crippa2015_P1 has no variants to work with\n",
      "  patient #95\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P4[PMID_36446582_Novara,_2017_P4]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara,_2017_P4 has no variants to work with\n",
      "  patient #98\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Parenti2021_P23[PMID_36446582_Parenti2021_P23]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Parenti2021_P23 has no variants to work with\n",
      "  patient #133\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P5[PMID_36446582_Novara,_2017_P5]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara,_2017_P5 has no variants to work with\n",
      "  patient #135\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Willemsen2010_P1[PMID_36446582_Willemsen2010_P1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Willemsen2010_P1 has no variants to work with\n",
      "  patient #139\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P22[PMID_36446582_Goldenberg2016_P22]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P22 has no variants to work with\n",
      "  patient #140\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Scarano, 2013_P10[PMID_36446582_Scarano,_2013_P10]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Scarano,_2013_P10 has no variants to work with\n",
      "  patient #147\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P9[PMID_36446582_Novara,_2017_P9]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara,_2017_P9 has no variants to work with\n",
      "  patient #150\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P18[PMID_36446582_Goldenberg2016_P18]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P18 has no variants to work with\n",
      "  patient #156\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Willemsen2010_P2[PMID_36446582_Willemsen2010_P2]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Willemsen2010_P2 has no variants to work with\n",
      "  patient #159\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Spengler, 2013[PMID_36446582_Spengler,_2013]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Spengler,_2013 has no variants to work with\n",
      "  patient #164\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Youngs2011[PMID_36446582_Youngs2011]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Youngs2011 has no variants to work with\n",
      "  patient #165\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P21[PMID_36446582_Goldenberg2016_P21]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P21 has no variants to work with\n",
      "  patient #166\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Crippa2015_P3[PMID_36446582_Crippa2015_P3]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Crippa2015_P3 has no variants to work with\n",
      "  patient #191\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Palumbo 2016[PMID_36446582_Palumbo_2016]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Palumbo_2016 has no variants to work with\n",
      "  patient #202\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P7[PMID_36446582_Novara,_2017_P7]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara,_2017_P7 has no variants to work with\n",
      "  patient #203\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG9[PMID_36446582_KBG9]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG9 has no variants to work with\n",
      "  patient #226\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG25[PMID_36446582_KBG25]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG25 has no variants to work with\n",
      "  patient #229\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P4[PMID_36446582_Goldenberg2016_P4]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P4 has no variants to work with\n",
      "  patient #230\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Scarano, 2013_P12[PMID_36446582_Scarano,_2013_P12]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Scarano,_2013_P12 has no variants to work with\n",
      "  patient #232\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P20[PMID_36446582_Goldenberg2016_P20]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P20 has no variants to work with\n",
      "  patient #233\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Crippa2015_P2[PMID_36446582_Crippa2015_P2]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Crippa2015_P2 has no variants to work with\n",
      "  patient #235\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P36[PMID_36446582_Goldenberg2016_P36]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P36 has no variants to work with\n",
      "  patient #239\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Willemsen2010_P3[PMID_36446582_Willemsen2010_P3]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Willemsen2010_P3 has no variants to work with\n",
      "  patient #244\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Sacharow, 2012_P1[PMID_36446582_Sacharow,_2012_P1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Sacharow,_2012_P1 has no variants to work with\n",
      "  patient #247\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Srivastava, 2017_P1[PMID_36446582_Srivastava,_2017_P1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Srivastava,_2017_P1 has no variants to work with\n",
      "  patient #258\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P12[PMID_36446582_Novara,_2017_P12]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara,_2017_P12 has no variants to work with\n",
      "  patient #260\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P20[PMID_36446582_Kutkowska-Kazmierczak2021_P20]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska-Kazmierczak2021_P20 has no variants to work with\n",
      "  patient #263\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG38[PMID_36446582_KBG38]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG38 has no variants to work with\n",
      "  patient #266\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Isrie, 2012_P1[PMID_36446582_Isrie,_2012_P1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Isrie,_2012_P1 has no variants to work with\n",
      "  patient #267\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Khalifa, 2013_P1A[PMID_36446582_Khalifa,_2013_P1A]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Khalifa,_2013_P1A has no variants to work with\n",
      "  patient #270\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Willemsen2010_P4[PMID_36446582_Willemsen2010_P4]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Willemsen2010_P4 has no variants to work with\n",
      "  patient #271\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Gnazzo, 2020_P29[PMID_36446582_Gnazzo,_2020_P29]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Gnazzo,_2020_P29 has no variants to work with\n",
      "  patient #274\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P16[PMID_36446582_Kutkowska-Kazmierczak2021_P16]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska-Kazmierczak2021_P16 has no variants to work with\n",
      "  patient #280\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P3[PMID_36446582_Goldenberg2016_P3]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P3 has no variants to work with\n",
      "  patient #283\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG22[PMID_36446582_KBG22]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG22 has no variants to work with\n",
      "  patient #287\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG2[PMID_36446582_KBG2]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG2 has no variants to work with\n",
      "  patient #291\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Lim2014[PMID_36446582_Lim2014]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Lim2014 has no variants to work with\n",
      "  patient #295\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG23[PMID_36446582_KBG23]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG23 has no variants to work with\n",
      "  patient #296\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P2[PMID_36446582_Goldenberg2016_P2]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P2 has no variants to work with\n",
      "  patient #298\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P26[PMID_36446582_Goldenberg2016_P26]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P26 has no variants to work with\n",
      "  patient #299\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient KBG58[PMID_36446582_KBG58]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_KBG58 has no variants to work with\n",
      "  patient #301\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P17[PMID_36446582_Kutkowska-Kazmierczak2021_P17]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska-Kazmierczak2021_P17 has no variants to work with\n",
      "  patient #308\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Novara, 2017_P1[PMID_36446582_Novara,_2017_P1]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Novara,_2017_P1 has no variants to work with\n",
      "  patient #309\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Goldenberg2016_P10[PMID_36446582_Goldenberg2016_P10]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Goldenberg2016_P10 has no variants to work with\n",
      "  patient #315\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Kutkowska-Kazmierczak2021_P21[PMID_36446582_Kutkowska-Kazmierczak2021_P21]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Kutkowska-Kazmierczak2021_P21 has no variants to work with\n",
      "  patient #318\n",
      "    variants\n",
      "     warnings:\n",
      "     ·Expected a VCF record, a VRS CNV, or an expression with `hgvs.c` but had an error retrieving any from patient Behnert, 2018[PMID_36446582_Behnert,_2018]. Remove variant from testing\n",
      "     ·Patient PMID_36446582_Behnert,_2018 has no variants to work with\n"
     ]
    }
   ],
   "source": [
    "from genophenocorr.preprocessing import configure_caching_cohort_creator, load_phenopacket_folder\n",
    "\n",
    "fpath_phenopackets = 'phenopackets'\n",
    "cohort_creator = configure_caching_cohort_creator(hpo, cache_dir='temp_cache')\n",
    "cohort = load_phenopacket_folder(fpath_phenopackets, cohort_creator)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "768f3cc0",
   "metadata": {},
   "source": [
    "### Pick transcript\n",
    "\n",
    "We choose the [MANE Select](https://www.ncbi.nlm.nih.gov/nuccore/NM_013275.6) transcript for *ANKRD11*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6a233bd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "tx_id = 'NM_013275.6'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e73a492f",
   "metadata": {},
   "source": [
    "## Explore cohort\n",
    "\n",
    "Explore the cohort to guide selection of the genotype-phenotype analysis.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3b7a08b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import HTML, display\n",
    "from genophenocorr.view import CohortViewer\n",
    "\n",
    "viewer = CohortViewer(hpo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c4960518",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       "\n",
       "table {\n",
       "  border-collapse: collapse; \n",
       "  margin: 25px 0;\n",
       "    font-size: 0.9em;\n",
       "    font-family: sans-serif;\n",
       "    min-width: 400px;\n",
       "    box-shadow: 0 0 20px rgba(0, 0, 0, 0.15);\n",
       "}\n",
       "\n",
       "\n",
       "th {\n",
       "  background-color: #f2f2f2; \n",
       "  border: 1px solid #dddddd; \n",
       "  text-align: left;\n",
       "  padding: 2px;\n",
       "}\n",
       "\n",
       "tr {\n",
       "  border: 1px solid #dddddd; \n",
       "}\n",
       "\n",
       "td {\n",
       "  padding: 2px; \n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       "tr:nth-child(even) {\n",
       "  background-color: #f2f2f2;\n",
       "}\n",
       "</style>\n",
       "\n",
       "<table>\n",
       "<tr><td>Item</td><td>Description</td></tr>\n",
       "<caption>Description of the cohort. 77 individuals were removed from the cohort because they had no HPO terms.</caption>\n",
       "<tr><td>Total Individuals</td><td>328</td></tr>\n",
       "<tr><td>Excluded Individuals</td><td>77: Kutkowska-Kazmierczak2021_P22[PMID_36446582_Kutkowska-Kazmierczak2021_P22];Goldenberg2016_P3[PMID_36446582_Goldenberg2016_P3];KBG22[PMID_36446582_KBG22];Scarano, 2013_P11[PMID_36446582_Scarano,_2013_P11];Gnazzo, 2020_P30[PMID_36446582_Gnazzo,_2020_P30];Sacharow, 2012_P2[PMID_36446582_Sacharow,_2012_P2];Willemsen2010_P1[PMID_36446582_Willemsen2010_P1];Sacharow, 2012_P1[PMID_36446582_Sacharow,_2012_P1];Isrie, 2012_P1[PMID_36446582_Isrie,_2012_P1];Goldenberg2016_P36[PMID_36446582_Goldenberg2016_P36];Scarano, 2013_P12[PMID_36446582_Scarano,_2013_P12];Goldenberg2016_P13[PMID_36446582_Goldenberg2016_P13];Gnazzo, 2020_P31[PMID_36446582_Gnazzo,_2020_P31];Goldenberg2016_P19[PMID_36446582_Goldenberg2016_P19];KBG58[PMID_36446582_KBG58];Khalifa, 2013_P1A[PMID_36446582_Khalifa,_2013_P1A];Willemsen2010_P2[PMID_36446582_Willemsen2010_P2];Goldenberg2016_P24[PMID_36446582_Goldenberg2016_P24];Goldenberg2016_P33[PMID_36446582_Goldenberg2016_P33];Novara, 2017_P3[PMID_36446582_Novara,_2017_P3];Bucerzan2020[PMID_36446582_Bucerzan2020];Crippa2015_P3[PMID_36446582_Crippa2015_P3];Kutkowska-Kazmierczak2021_P18[PMID_36446582_Kutkowska-Kazmierczak2021_P18];Kutkowska-Kazmierczak2021_P16[PMID_36446582_Kutkowska-Kazmierczak2021_P16];Goldenberg2016_P18[PMID_36446582_Goldenberg2016_P18];Goldenberg2016_P26[PMID_36446582_Goldenberg2016_P26];KBG2[PMID_36446582_KBG2];Crippa2015_P2[PMID_36446582_Crippa2015_P2];Crippa2015_P1[PMID_36446582_Crippa2015_P1];Youngs2011[PMID_36446582_Youngs2011];Kutkowska-Kazmierczak2021_P14[PMID_36446582_Kutkowska-Kazmierczak2021_P14];Goldenberg2016_P32[PMID_36446582_Goldenberg2016_P32];Goldenberg2016_P1[PMID_36446582_Goldenberg2016_P1];Palumbo 2016[PMID_36446582_Palumbo_2016];Isrie, 2012_P2[PMID_36446582_Isrie,_2012_P2];Kutkowska-Kazmierczak2021_P19[PMID_36446582_Kutkowska-Kazmierczak2021_P19];KBG23[PMID_36446582_KBG23];Kutkowska-Kazmierczak2021_P21[PMID_36446582_Kutkowska-Kazmierczak2021_P21];Gnazzo, 2020_P29[PMID_36446582_Gnazzo,_2020_P29];Khalifa, 2013_P1B[PMID_36446582_Khalifa,_2013_P1B];Novara, 2017_P7[PMID_36446582_Novara,_2017_P7];Novara, 2017_P11[PMID_36446582_Novara,_2017_P11];Lim2014[PMID_36446582_Lim2014];Willemsen2010_P3[PMID_36446582_Willemsen2010_P3];Novara, 2017_P9[PMID_36446582_Novara,_2017_P9];Parenti2021_P23[PMID_36446582_Parenti2021_P23];KBG9[PMID_36446582_KBG9];Kutkowska-Kazmierczak2021_P15[PMID_36446582_Kutkowska-Kazmierczak2021_P15];Goldenberg2016_P22[PMID_36446582_Goldenberg2016_P22];Novara, 2017_P8[PMID_36446582_Novara,_2017_P8];Novara, 2017_P4[PMID_36446582_Novara,_2017_P4];Spengler, 2013[PMID_36446582_Spengler,_2013];Goldenberg2016_P20[PMID_36446582_Goldenberg2016_P20];Miyatake, 2013[PMID_36446582_Miyatake,_2013];Novara, 2017_P5[PMID_36446582_Novara,_2017_P5];Kutkowska-Kazmierczak2021_P20[PMID_36446582_Kutkowska-Kazmierczak2021_P20];Scarano, 2013_P10[PMID_36446582_Scarano,_2013_P10];Novara, 2017_P12[PMID_36446582_Novara,_2017_P12];Goldenberg2016_P21[PMID_36446582_Goldenberg2016_P21];Ockeloen2015_P20[PMID_36446582_Ockeloen2015_P20];Willemsen2010_P4[PMID_36446582_Willemsen2010_P4];Goldenberg2016_P12[PMID_36446582_Goldenberg2016_P12];Behnert, 2018[PMID_36446582_Behnert,_2018];Kutkowska-Kazmierczak2021_P17[PMID_36446582_Kutkowska-Kazmierczak2021_P17];KBG38[PMID_36446582_KBG38];KBG25[PMID_36446582_KBG25];KBG26[PMID_36446582_KBG26];Goldenberg2016_P4[PMID_36446582_Goldenberg2016_P4];Novara, 2017_P1[PMID_36446582_Novara,_2017_P1];Srivastava, 2017_P1[PMID_36446582_Srivastava,_2017_P1];Goldenberg2016_P28[PMID_36446582_Goldenberg2016_P28];Goldenberg2016_P29[PMID_36446582_Goldenberg2016_P29];KBG1[PMID_36446582_KBG1];Novara, 2017_P2[PMID_36446582_Novara,_2017_P2];Goldenberg2016_P2[PMID_36446582_Goldenberg2016_P2];Goldenberg2016_P10[PMID_36446582_Goldenberg2016_P10];Kutkowska-Kazmierczak2021_P23[PMID_36446582_Kutkowska-Kazmierczak2021_P23]</td></tr>\n",
       "<tr><td>Total Unique HPO Terms</td><td>27</td></tr>\n",
       "<tr><td>Total Unique Variants</td><td>251</td></tr>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(HTML(viewer.cohort_summary_table(cohort)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c39889b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       "\n",
       "table {\n",
       "  border-collapse: collapse; \n",
       "  margin: 25px 0;\n",
       "    font-size: 0.9em;\n",
       "    font-family: sans-serif;\n",
       "    min-width: 400px;\n",
       "    box-shadow: 0 0 20px rgba(0, 0, 0, 0.15);\n",
       "}\n",
       "\n",
       "\n",
       "th {\n",
       "  background-color: #f2f2f2; \n",
       "  border: 1px solid #dddddd; \n",
       "  text-align: left;\n",
       "  padding: 2px;\n",
       "}\n",
       "\n",
       "tr {\n",
       "  border: 1px solid #dddddd; \n",
       "}\n",
       "\n",
       "td {\n",
       "  padding: 2px; \n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       "tr:nth-child(even) {\n",
       "  background-color: #f2f2f2;\n",
       "}\n",
       "</style>\n",
       "\n",
       "<table>\n",
       "<tr><td>HPO Term</td><td>Count</td></tr>\n",
       "<caption>Counts of annotations to HPO terms for the 328 in the cohort</caption>\n",
       "<tr><td><a href=\"https://hpo.jax.org/app/browse/term/HP:0001572\">Macrodontia (HP:0001572)</a></td><td>182</td></tr>\n",
       "<tr><td><a href=\"https://hpo.jax.org/app/browse/term/HP:0001249\">Intellectual disability (HP:0001249)</a></td><td>159</td></tr>\n",
       "<tr><td><a href=\"https://hpo.jax.org/app/browse/term/HP:0001155\">Abnormality of the hand (HP:0001155)</a></td><td>156</td></tr>\n",
       "<tr><td><a href=\"https://hpo.jax.org/app/browse/term/HP:0001263\">Global developmental delay (HP:0001263)</a></td><td>133</td></tr>\n",
       "<tr><td><a href=\"https://hpo.jax.org/app/browse/term/HP:0004322\">Short stature (HP:0004322)</a></td><td>115</td></tr>\n",
       "<tr><td><a href=\"https://hpo.jax.org/app/browse/term/HP:0010938\">Abnormal external nose morphology (HP:0010938)</a></td><td>112</td></tr>\n",
       "<tr><td><a href=\"https://hpo.jax.org/app/browse/term/HP:0000574\">Thick eyebrow (HP:0000574)</a></td><td>105</td></tr>\n",
       "<tr><td><a href=\"https://hpo.jax.org/app/browse/term/HP:0000343\">Long philtrum (HP:0000343)</a></td><td>103</td></tr>\n",
       "<tr><td><a href=\"https://hpo.jax.org/app/browse/term/HP:0000365\">Hearing impairment (HP:0000365)</a></td><td>74</td></tr>\n",
       "<tr><td><a href=\"https://hpo.jax.org/app/browse/term/HP:0000325\">Triangular face (HP:0000325)</a></td><td>68</td></tr>\n",
       "<tr><td><a href=\"https://hpo.jax.org/app/browse/term/HP:0000356\">Abnormality of the outer ear (HP:0000356)</a></td><td>58</td></tr>\n",
       "<tr><td><a href=\"https://hpo.jax.org/app/browse/term/HP:0007018\">Attention deficit hyperactivity disorder (HP:0007018)</a></td><td>51</td></tr>\n",
       "<tr><td><a href=\"https://hpo.jax.org/app/browse/term/HP:0000729\">Autistic behavior (HP:0000729)</a></td><td>42</td></tr>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(HTML(viewer.hpo_term_counts_table(cohort))) ## Add Labels to output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f294ca99",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       "\n",
       "table {\n",
       "  border-collapse: collapse; \n",
       "  margin: 25px 0;\n",
       "    font-size: 0.9em;\n",
       "    font-family: sans-serif;\n",
       "    min-width: 400px;\n",
       "    box-shadow: 0 0 20px rgba(0, 0, 0, 0.15);\n",
       "}\n",
       "\n",
       "\n",
       "th {\n",
       "  background-color: #f2f2f2; \n",
       "  border: 1px solid #dddddd; \n",
       "  text-align: left;\n",
       "  padding: 2px;\n",
       "}\n",
       "\n",
       "tr {\n",
       "  border: 1px solid #dddddd; \n",
       "}\n",
       "\n",
       "td {\n",
       "  padding: 2px; \n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       "tr:nth-child(even) {\n",
       "  background-color: #f2f2f2;\n",
       "}\n",
       "</style>\n",
       "\n",
       "<table>\n",
       "<tr><td>Variant</td><td>Effect</td><td>Count</td><td>Key</td></tr>\n",
       "<tr><td>c.1903_1907del</td><td>FRAMESHIFT_VARIANT</td><td>33</td><td>16_89284634_89284639_GTGTTT_G</td></tr>\n",
       "<tr><td>c.2408_2412del</td><td>FRAMESHIFT_VARIANT</td><td>10</td><td>16_89284129_89284134_CTTTTT_C</td></tr>\n",
       "<tr><td>c.1381_1384del</td><td>FRAMESHIFT_VARIANT</td><td>8</td><td>16_89285157_89285161_GTTTC_G</td></tr>\n",
       "<tr><td>c.2398_2401del</td><td>FRAMESHIFT_VARIANT</td><td>8</td><td>16_89284140_89284144_TTTTC_T</td></tr>\n",
       "<tr><td>c.6792_6793insC</td><td>FRAMESHIFT_VARIANT</td><td>5</td><td>16_89279749_89279749_C_CG</td></tr>\n",
       "<tr><td>c.7481_7482insC</td><td>FRAMESHIFT_VARIANT</td><td>5</td><td>16_89275180_89275180_A_AG</td></tr>\n",
       "<tr><td>c.2182_2183del</td><td>FRAMESHIFT_VARIANT</td><td>3</td><td>16_89284358_89284360_GAT_G</td></tr>\n",
       "<tr><td>c.7570-1G>C</td><td>SPLICE_ACCEPTOR_VARIANT</td><td>3</td><td>16_89274958_89274958_C_G</td></tr>\n",
       "<tr><td>c.2175_2178del</td><td>FRAMESHIFT_VARIANT</td><td>3</td><td>16_89284363_89284367_CTTTG_C</td></tr>\n",
       "<tr><td>c.4406G>A</td><td>STOP_GAINED</td><td>3</td><td>16_89282136_89282136_C_T</td></tr>\n",
       "<tr><td>c.3224_3227del</td><td>FRAMESHIFT_VARIANT</td><td>3</td><td>16_89283314_89283318_CCTTT_C</td></tr>\n",
       "<tr><td>c.2197C>T</td><td>STOP_GAINED</td><td>3</td><td>16_89284345_89284345_G_A</td></tr>\n",
       "<tr><td>c.1977C>G</td><td>STOP_GAINED</td><td>3</td><td>16_89284565_89284565_G_C</td></tr>\n",
       "<tr><td>c.3832A>T</td><td>STOP_GAINED</td><td>3</td><td>16_89282710_89282710_T_A</td></tr>\n",
       "<tr><td>c.4384_4385insA</td><td>FRAMESHIFT_VARIANT</td><td>2</td><td>16_89282157_89282157_C_CT</td></tr>\n",
       "<tr><td>c.3590_3594del</td><td>FRAMESHIFT_VARIANT</td><td>2</td><td>16_89282947_89282952_CTTTTT_C</td></tr>\n",
       "<tr><td>c.3704_3707del</td><td>FRAMESHIFT_VARIANT</td><td>2</td><td>16_89282834_89282838_CTGTT_C</td></tr>\n",
       "<tr><td>c.6513_6514insC</td><td>FRAMESHIFT_VARIANT</td><td>2</td><td>16_89280028_89280028_C_CG</td></tr>\n",
       "<tr><td>c.2329_2332del</td><td>FRAMESHIFT_VARIANT</td><td>2</td><td>16_89284209_89284213_TTCTC_T</td></tr>\n",
       "<tr><td>c.1318C>T</td><td>STOP_GAINED</td><td>2</td><td>16_89285224_89285224_G_A</td></tr>\n",
       "<tr><td>c.7534C>T</td><td>MISSENSE_VARIANT</td><td>2</td><td>16_89275128_89275128_G_A</td></tr>\n",
       "<tr><td>c.4087C>T</td><td>STOP_GAINED</td><td>2</td><td>16_89282455_89282455_G_A</td></tr>\n",
       "<tr><td>c.1385_1388del</td><td>FRAMESHIFT_VARIANT</td><td>2</td><td>16_89285153_89285157_TTTTG_T</td></tr>\n",
       "<tr><td>c.7216C>T</td><td>STOP_GAINED</td><td>2</td><td>16_89279326_89279326_G_A</td></tr>\n",
       "<tr><td>c.3309_3310insA</td><td>FRAMESHIFT_VARIANT</td><td>2</td><td>16_89283232_89283232_C_CT</td></tr>\n",
       "<tr><td>c.5790C>A</td><td>STOP_GAINED</td><td>2</td><td>16_89280752_89280752_G_T</td></tr>\n",
       "<tr><td>c.5488G>T</td><td>STOP_GAINED</td><td>2</td><td>16_89281054_89281054_C_A</td></tr>\n",
       "<tr><td>c.3045del</td><td>FRAMESHIFT_VARIANT</td><td>2</td><td>16_89283496_89283497_CG_C</td></tr>\n",
       "<tr><td>c.1367_1370del</td><td>FRAMESHIFT_VARIANT</td><td>2</td><td>16_89285171_89285175_CTTCT_C</td></tr>\n",
       "</table>\n",
       "<p>Additionally, the following variants were observed 1 or fewer times: \n",
       "c.4964_4965del; c.4206C>G; c.1389_1390insA; c.2512C>T; c.3019C>T; c.7607G>A; c.4283_4286del; c.5145C>G; c.2751_2752insT; c.2395A>T; c.7363del; c.7834G>T; c.5146G>T; c.3180_3181insA; c.3774_3775del; c.1763C>A; c.7180C>T; c.4558del; c.4391_4392del; c.6472G>T; c.4374_4375del; c.7192C>T; c.3310_3311insG; c.1711_1723del; c.6340C>T; c.6968_6975del; c.3591_3594del; c.2297_2300del; c.831del; c.548_551del; c.1731_1732insT; c.6409_6410del; c.2412del; c.3974del; c.1940_1941delinsT; c.4107_4108del; c.4528_4529del; c.915del; c.3153del; c.4389_4390del; c.1285_1286del; c.2828_2829del; c.6015_6016insA; c.6766C>T; c.5199_5227del; c.7411_7422del; c.5123C>A; c.6817_6833del; c.3931C>T; c.5205del; c.1457C>G; c.1120G>T; c.3437_3461del; c.3382_3383del; c.4786G>T; c.3222_3223insA; c.6071_6084del; c.1785_1786delinsTT; c.4498C>T; c.6628G>T; c.3046del; c.867C>G; c.3339G>A; c.5889del; c.4103_4104del; c.3582del; c.2866G>T; c.7356_7357insC; c.6364_6367del; c.1173C>G; c.7570_7572del; c.2647G>T; c.1801C>T; c.505G>T; c.3770_3771del; c.3295_3296del; c.1893_1894insA; c.6682del; c.4218C>A; c.2305del; c.6184del; c.3771_3772insA; c.6691_6692insG; c.3448C>T; c.4177_4189del; c.211_226+1del; c.7407C>G; c.3905_3906del; c.5426_5430del; c.6701del; c.5483C>A; c.5274_5275insC; c.1460_1463del; c.3888_3889insC; c.2650del; c.7083del; c.3334del; c.3193A>T; c.5953_5954del; c.7552C>T; c.7000C>T; c.3460G>T; c.2593_2594insT; c.7189C>T; c.4408A>T; c.6053_6057del; c.4171C>T; c.7471A>C; c.3208_3209del; c.4529_4530insC; c.2692C>T; c.6187G>T; c.520C>T; c.2130del; c.3198_3199del; c.7470+2T>C; c.3123_3126del; c.7535G>A; c.7416C>G; c.866_867insA; c.2367del; c.5957_5958del; c.5712_5713insT; c.1846G>T; c.3221_3222del; c.7471-1G>C; c.2765_2766del; c.7753C>T.</p>\n",
       "<p>Use the entry in the \"Key\" column to investigate whether specific variants display genotype-phenotype correlations</p>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(HTML(viewer.variants_table(cohort, tx_id))) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbbc2cc1",
   "metadata": {},
   "source": [
    "## Configure the analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ec6c204f",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Unknown protein fallback annotator type <genophenocorr.analysis._config.CohortAnalysisConfiguration object at 0x11fe91990>",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 9\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mgenophenocorr\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01manalysis\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mpredicate\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m BooleanPredicate, GroupingPredicate\n\u001b[1;32m      4\u001b[0m analysis_config \u001b[38;5;241m=\u001b[39m CohortAnalysisConfiguration\u001b[38;5;241m.\u001b[39mbuilder()\\\n\u001b[1;32m      5\u001b[0m     \u001b[38;5;241m.\u001b[39mmissing_implies_excluded(\u001b[38;5;28;01mTrue\u001b[39;00m)\\\n\u001b[1;32m      6\u001b[0m     \u001b[38;5;241m.\u001b[39mpval_correction(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfdr_bh\u001b[39m\u001b[38;5;124m'\u001b[39m)\\\n\u001b[1;32m      7\u001b[0m     \u001b[38;5;241m.\u001b[39mmin_perc_patients_w_hpo(\u001b[38;5;241m0.1\u001b[39m)\\\n\u001b[1;32m      8\u001b[0m     \u001b[38;5;241m.\u001b[39mbuild()\n\u001b[0;32m----> 9\u001b[0m analysis \u001b[38;5;241m=\u001b[39m \u001b[43mconfigure_cohort_analysis\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcohort\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mhpo\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43manalysis_config\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/Programs/genophenocorr/src/genophenocorr/analysis/_config.py:188\u001b[0m, in \u001b[0;36mconfigure_cohort_analysis\u001b[0;34m(cohort, hpo, protein_source, cache_dir, config)\u001b[0m\n\u001b[1;32m    186\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m cache_dir \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    187\u001b[0m     cache_dir \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(os\u001b[38;5;241m.\u001b[39mgetcwd(), \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.genophenocorr_cache\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m--> 188\u001b[0m protein_metadata_service \u001b[38;5;241m=\u001b[39m \u001b[43m_configure_protein_service\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprotein_source\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcache_dir\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    191\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m CommunistCohortAnalysis(cohort, hpo, protein_metadata_service,\n\u001b[1;32m    192\u001b[0m                                missing_implies_excluded\u001b[38;5;241m=\u001b[39mconfig\u001b[38;5;241m.\u001b[39mmissing_implies_excluded,\n\u001b[1;32m    193\u001b[0m                                include_sv\u001b[38;5;241m=\u001b[39mconfig\u001b[38;5;241m.\u001b[39minclude_sv,\n\u001b[1;32m    194\u001b[0m                                p_val_correction\u001b[38;5;241m=\u001b[39mconfig\u001b[38;5;241m.\u001b[39mpval_correction,\n\u001b[1;32m    195\u001b[0m                                min_perc_patients_w_hpo\u001b[38;5;241m=\u001b[39mconfig\u001b[38;5;241m.\u001b[39mmin_perc_patients_w_hpo)\n",
      "File \u001b[0;32m~/Documents/Programs/genophenocorr/src/genophenocorr/analysis/_config.py:200\u001b[0m, in \u001b[0;36m_configure_protein_service\u001b[0;34m(protein_fallback, cache_dir)\u001b[0m\n\u001b[1;32m    197\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_configure_protein_service\u001b[39m(protein_fallback: \u001b[38;5;28mstr\u001b[39m, cache_dir) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m ProteinMetadataService:\n\u001b[1;32m    198\u001b[0m     \u001b[38;5;66;03m# (1) ProteinMetadataService\u001b[39;00m\n\u001b[1;32m    199\u001b[0m     \u001b[38;5;66;03m# Setup fallback\u001b[39;00m\n\u001b[0;32m--> 200\u001b[0m     protein_fallback \u001b[38;5;241m=\u001b[39m \u001b[43m_configure_fallback_protein_service\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprotein_fallback\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    201\u001b[0m     \u001b[38;5;66;03m# Setup protein metadata cache\u001b[39;00m\n\u001b[1;32m    202\u001b[0m     prot_cache_dir \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(cache_dir, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mprotein_cache\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m~/Documents/Programs/genophenocorr/src/genophenocorr/analysis/_config.py:213\u001b[0m, in \u001b[0;36m_configure_fallback_protein_service\u001b[0;34m(protein_fallback)\u001b[0m\n\u001b[1;32m    211\u001b[0m     fallback1 \u001b[38;5;241m=\u001b[39m UniprotProteinMetadataService()\n\u001b[1;32m    212\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 213\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mUnknown protein fallback annotator type \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mprotein_fallback\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m    214\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m fallback1\n",
      "\u001b[0;31mValueError\u001b[0m: Unknown protein fallback annotator type <genophenocorr.analysis._config.CohortAnalysisConfiguration object at 0x11fe91990>"
     ]
    }
   ],
   "source": [
    "from genophenocorr.analysis import configure_cohort_analysis, CohortAnalysisConfiguration\n",
    "from genophenocorr.analysis.predicate import BooleanPredicate, GroupingPredicate\n",
    "\n",
    "analysis_config = CohortAnalysisConfiguration.builder()\\\n",
    "    .missing_implies_excluded(True)\\\n",
    "    .pval_correction('fdr_bh')\\\n",
    "    .min_perc_patients_w_hpo(0.1)\\\n",
    "    .build()\n",
    "analysis = configure_cohort_analysis(cohort, hpo, analysis_config)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ff62a78",
   "metadata": {},
   "source": [
    "Test for presence of genotype-phenotype correlations between frameshift variants vs. others."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91809a38",
   "metadata": {},
   "outputs": [],
   "source": [
    "from genophenocorr.model import VariantEffect\n",
    "\n",
    "frameshift = analysis.compare_by_variant_effect(VariantEffect.FRAMESHIFT_VARIANT, tx_id)\n",
    "frameshift.summarize(hpo, BooleanPredicate.YES)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a83a5ca",
   "metadata": {},
   "source": [
    "Test for presence of genotype-phenotype correlations between subjects with >=1 allele of a variant vs. the other subjects:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57dd0e23",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_single = analysis.compare_by_variant_key('16_89284634_89284639_GTGTTT_G')\n",
    "var_single.summarize(hpo, BooleanPredicate.YES)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74eaa536",
   "metadata": {},
   "source": [
    "Or between subjects with one variant vs. the other variant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dfa0a50",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_double = analysis.compare_by_variant_keys('16_89284129_89284134_CTTTTT_C', '16_89284634_89284639_GTGTTT_G')\n",
    "var_double.summarize(hpo, BooleanPredicate.YES)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a12b8117",
   "metadata": {},
   "source": [
    "TODO - finalize!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
