Search.setIndex({"docnames": ["apidocs/genophenocorr", "apidocs/genophenocorr.analysis", "apidocs/genophenocorr.analysis.predicate", "apidocs/genophenocorr.constants", "apidocs/genophenocorr.constants.variant_effects", "apidocs/genophenocorr.data", "apidocs/genophenocorr.model", "apidocs/genophenocorr.preprocessing", "apidocs/genophenocorr.view", "apidocs/modules", "autosomal_recessive", "index", "installation", "tutorial", "user-guide/exploratory", "user-guide/index", "user-guide/input-data"], "filenames": ["apidocs/genophenocorr.rst", "apidocs/genophenocorr.analysis.rst", "apidocs/genophenocorr.analysis.predicate.rst", "apidocs/genophenocorr.constants.rst", "apidocs/genophenocorr.constants.variant_effects.rst", "apidocs/genophenocorr.data.rst", "apidocs/genophenocorr.model.rst", "apidocs/genophenocorr.preprocessing.rst", "apidocs/genophenocorr.view.rst", "apidocs/modules.rst", "autosomal_recessive.rst", "index.rst", "installation.rst", "tutorial.rst", "user-guide/exploratory.rst", "user-guide/index.rst", "user-guide/input-data.rst"], "titles": ["genophenocorr package", "genophenocorr.analysis package", "genophenocorr.analysis.predicate package", "genophenocorr.constants package", "genophenocorr.constants.variant_effects module", "genophenocorr.data package", "genophenocorr.model package", "genophenocorr.preprocessing package", "genophenocorr.view package", "API reference", "Genotype-Phenotype Correlations in Autosomal Recessive Diseases", "genophenocorr", "Installation", "Tutorial", "Cohort exploratory analysis", "User guide", "Input data"], "terms": {"i": [0, 1, 2, 6, 7, 10, 11, 12, 13, 16], "librari": [0, 12, 16], "analyz": [0, 1, 8], "genotyp": [0, 6, 8, 11, 13, 16], "phenotyp": [0, 1, 2, 6, 7, 8, 11, 13, 14, 16], "correl": [0, 1, 8, 11], "cohort": [0, 1, 5, 6, 7, 8, 11, 13], "rare": 0, "diseas": [0, 11], "patient": [0, 1, 2, 6, 7, 8, 11, 13, 16], "analysi": [0, 10, 11, 16], "cohortanalysi": [0, 1, 13], "constant": [0, 13], "submodul": 0, "data": [0, 6, 7, 11, 13, 15], "get_toy_cohort": [0, 5, 13], "model": [0, 16], "variant": [0, 1, 2, 6, 7, 11, 13, 14, 16], "transcriptannot": [0, 6], "variantcoordin": [0, 6, 7], "proteinmetadata": [0, 6, 7], "proteinfeatur": [0, 6], "featureinfo": [0, 6], "featuretyp": [0, 1, 2, 6], "preprocess": [0, 16], "variantcoordinatefind": [0, 7], "functionalannot": [0, 7], "proteinmetadataservic": [0, 7], "patientcr": [0, 7], "phenopacketvariantcoordinatefind": [0, 7], "phenopacketpatientcr": [0, 7, 16], "load_phenopacket_fold": [0, 7], "phenotypecr": [0, 7], "phenotypevalidationexcept": [0, 7], "proteinannotationcach": [0, 7], "protcachingfunctionalannot": [0, 7], "uniprotproteinmetadataservic": [0, 7], "vepfunctionalannot": [0, 7], "variantannotationcach": [0, 7], "varcachingfunctionalannot": [0, 7], "configure_caching_patient_cr": [0, 7, 16], "view": [0, 11, 13], "cohortview": [0, 8], "class": [1, 2, 4, 6, 7, 8, 16], "transcript": [1, 2, 6], "hpo": [1, 2, 6, 7, 8, 11, 13, 16], "recess": [1, 6], "fals": [1, 6, 13], "include_unmeasur": [1, 13], "true": [1, 6, 7], "p_val_correct": 1, "bonferroni": 1, "include_large_sv": 1, "min_perc_patients_w_hpo": 1, "10": [1, 6, 13], "sourc": [1, 2, 4, 5, 6, 7, 8, 12], "base": [1, 2, 4, 6, 7, 8, 13], "object": [1, 2, 6, 7, 8], "thi": [1, 2, 6, 7, 10, 11, 13, 16], "can": [1, 2, 6, 7, 12, 13, 14, 16], "us": [1, 6, 7, 8, 11, 13, 16], "run": [1, 7, 8, 12, 13], "differ": 1, "analys": [1, 13, 16], "given": [1, 6, 7], "analysis_cohort": 1, "The": [1, 2, 6, 7, 11, 12, 13, 16], "type": [1, 4, 6, 7, 11], "analysis_pati": 1, "A": [1, 6, 7, 8, 11, 13], "subset": 1, "structur": [1, 6, 7], "thei": [1, 11], "ar": [1, 2, 6, 10, 11, 13], "being": 1, "remov": 1, "from": [1, 6, 7, 11, 12, 13], "otherwis": [1, 6], "all": [1, 2, 6, 16], "list": [1, 6, 7, 13], "analysis_transcript": 1, "id": [1, 6, 7], "string": [1, 2, 6, 7], "is_recess": 1, "default": [1, 16], "boolean": [1, 6, 7], "we": [1, 2, 6, 7, 11, 12, 13, 14, 16], "want": [1, 13], "assum": 1, "without": [1, 13], "specif": [1, 6, 7, 11], "doe": [1, 7], "have": [1, 2], "onli": [1, 6, 7, 10], "includ": [1, 6, 7, 11], "those": 1, "explicit": 1, "sai": 1, "wa": [1, 2, 6, 7], "observ": [1, 2, 6, 7, 11], "larg": 1, "Will": 1, "test": [1, 2, 6, 8, 10, 11], "least": [1, 7], "percent": [1, 13], "integ": [1, 6], "compare_by_variant_typ": [1, 13], "var_type1": 1, "varianteffect": [1, 2, 3, 4, 13], "var_type2": 1, "option": [1, 6, 16], "fisher": [1, 11], "exact": [1, 2, 10, 11], "find": [1, 6, 7], "ani": [1, 2, 7], "between": [1, 6, 11], "effect": [1, 6, 7, 14, 16], "variant1": 1, "variant2": 1, "exon1": 1, "exon2": 1, "affect": [1, 6, 10], "exon": [1, 6], "feature1": 1, "feature2": 1, "protein": [1, 2, 6, 7, 11, 16], "featur": [1, 2, 6, 11, 12, 14], "properti": [1, 2, 6, 7], "return": [1, 2, 6, 7, 8], "none": [1, 2, 4, 6, 7], "across": 1, "paramet": [1, 6, 7, 8], "If": [1, 6], "compar": [1, 10], "panda": 1, "datafram": 1, "show": [1, 8], "result": [1, 6, 7, 11], "compare_by_vari": 1, "str": [1, 2, 6, 7, 8], "compare_by_exon": 1, "int": [1, 2, 6, 8], "within": [1, 6], "outsid": 1, "compare_by_protein_feature_typ": 1, "compare_by_protein_featur": 1, "predic": 1, "simplepred": [1, 2], "patientcategori": [1, 2], "polypred": [1, 2], "hpopresentpred": [1, 2], "varianteffectpred": [1, 2], "variantpred": [1, 2], "exonpred": [1, 2], "protfeaturepred": [1, 2], "protfeaturetypepred": [1, 2], "belong": 2, "group": [2, 10, 13], "abstract": [2, 6, 7], "bool": [2, 6, 7], "cat_id": 2, "name": [2, 4, 6, 7], "descript": [2, 6], "gener": [2, 7, 8, 11], "t": [2, 6, 7, 13], "categor": 2, "one": [2, 7, 10], "discret": 2, "categori": 2, "sequenc": [2, 6, 7], "get": [2, 7, 13], "which": [2, 13, 16], "produc": 2, "queri": 2, "assign": 2, "cannot": [2, 6, 7], "meaning": 2, "minimalontologi": [2, 7], "termid": [2, 6], "0": [2, 6, 13], "sampl": 2, "annot": [2, 6, 7, 8, 13, 16], "q": [2, 13, 16], "either": 2, "becaus": [2, 10], "match": 2, "": [2, 6, 7, 13, 16], "descend": 2, "propag": 2, "For": [2, 10, 13, 16], "instanc": [2, 7, 10, 13, 16], "seizur": [2, 13], "had": 2, "clonic": 2, "not_observ": 2, "1": [2, 4, 6, 8, 13], "Not": 2, "particular": 2, "about": [2, 6, 13, 14], "In": [2, 7, 10, 11], "other": [2, 6, 10, 11], "word": [2, 11], "exclud": [2, 6], "an": [2, 6, 7, 8, 10, 11, 13, 16], "ancestor": 2, "did": 2, "impli": 2, "not_measur": 2, "2": [2, 6, 8, 13], "measur": 2, "do": [2, 13], "know": 2, "ha": [2, 6, 7], "overlap": [2, 6], "over": 2, "repres": [2, 6, 11], "egf": 2, "like": [2, 11], "fbn1": 2, "variant_effect": [3, 6], "modul": [3, 6], "valu": [4, 6, 13], "qualnam": [4, 6], "start": [4, 6, 16], "boundari": [4, 6], "enum": [4, 6], "transcript_abl": 4, "so": 4, "0001893": 4, "splice_acceptor_vari": 4, "0001574": 4, "splice_donor_vari": 4, "0001575": 4, "stop_gain": 4, "0001587": 4, "frameshift_vari": [4, 13], "0001589": 4, "stop_lost": 4, "0001578": 4, "start_lost": 4, "0002012": 4, "transcript_amplif": 4, "0001889": 4, "inframe_insert": 4, "0001821": 4, "inframe_delet": 4, "0001822": 4, "missense_vari": [4, 13], "0001583": 4, "protein_altering_vari": 4, "0001818": 4, "splice_region_vari": 4, "0001630": 4, "splice_donor_5th_base_vari": 4, "0001787": 4, "splice_donor_region_vari": 4, "0002170": 4, "splice_polypyrimidine_tract_vari": 4, "0002169": 4, "incomplete_terminal_codon_vari": 4, "0001626": 4, "start_retained_vari": 4, "0002019": 4, "stop_retained_vari": 4, "0001567": 4, "synonymous_vari": 4, "0001819": 4, "coding_sequence_vari": 4, "0001580": 4, "mature_mirna_vari": 4, "0001620": 4, "five_prime_utr_vari": 4, "0001623": 4, "three_prime_utr_vari": 4, "0001624": 4, "non_coding_transcript_exon_vari": 4, "0001792": 4, "intron_vari": 4, "0001627": 4, "nmd_transcript_vari": 4, "0001621": 4, "non_coding_transcript_vari": 4, "0001619": 4, "upstream_gene_vari": 4, "0001631": 4, "downstream_gene_vari": 4, "0001632": 4, "tfbs_ablat": 4, "0001895": 4, "tfbs_amplif": 4, "0001892": 4, "tf_binding_site_vari": 4, "0001782": 4, "regulatory_region_abl": 4, "0001894": 4, "regulatory_region_amplif": 4, "0001891": 4, "feature_elong": 4, "0001907": 4, "regulatory_region_vari": 4, "0001566": 4, "feature_trunc": 4, "0001906": 4, "intergenic_vari": 4, "0001628": 4, "sequence_vari": 4, "0001060": 4, "defin": 6, "top": 6, "level": [6, 15], "element": 6, "follow": [6, 12], "info": [6, 7, 16], "patient_set": 6, "set": 6, "phenotype_set": 6, "variant_set": 6, "protein_set": 6, "counts_dict": 6, "size": 6, "static": [6, 8], "from_pati": [6, 16], "member": [6, 13, 14, 16], "creat": [6, 7, 11], "all_pati": 6, "frozenset": 6, "frozen": 6, "all_phenotyp": 6, "all_vari": 6, "all_protein": 6, "all_transcript": 6, "total_patient_count": 6, "count": [6, 13, 14], "list_all_pati": [6, 13], "list_all_phenotyp": [6, 13], "highest": 6, "tupl": [6, 7], "format": [6, 7], "number": [6, 8], "list_all_vari": [6, 13], "list_all_protein": [6, 13], "list_data_by_tx": [6, 13], "each": [6, 7, 10, 11, 13], "refer": [6, 11], "counter": 6, "kei": [6, 11], "total": 6, "dictionari": 6, "patient_id": 6, "iter": [6, 7], "individu": [6, 10, 11, 15], "uniqu": 6, "associ": [6, 7, 10, 11], "term_id": [6, 7], "identifi": 6, "verifi": 6, "hpotk": [6, 13, 16], "offici": 6, "respect": 6, "from_term": 6, "term": [6, 7, 8, 11, 13, 16], "minimalterm": 6, "you": [6, 13], "more": [6, 7, 16], "inform": [6, 11], "search": [6, 7], "http": [6, 7, 10, 12, 13, 16], "jax": 6, "org": [6, 7, 13, 16], "app": 6, "describ": [6, 12, 14, 16], "whether": [6, 11], "var_id": 6, "var_class": 6, "var_coordin": 6, "tx_annot": 6, "function": [6, 7, 16], "variant_coordin": [6, 7], "coordin": [6, 7, 16], "variant_str": 6, "readabl": 6, "represent": [6, 7], "e": [6, 7, 11, 13], "g": [6, 7, 11, 13], "homozyg": 6, "heterozyg": 6, "variant_class": 6, "duplic": 6, "snv": 6, "etc": 6, "create_variant_from_scratch": 6, "variant_id": 6, "gene_nam": 6, "trans_id": 6, "hgvsc_id": 6, "consequ": 6, "exons_effect": 6, "protein_effect_start": 6, "protein_effect_end": 6, "symbol": 6, "chromosome_start_refer": 6, "altern": 6, "chromosome_start_structuraltyp": 6, "requir": [6, 7, 16], "possibl": [6, 10], "hemizyg": 6, "collect": [6, 11], "singl": 6, "gene": [6, 10], "delet": 6, "gene_id": 6, "tx_id": 6, "hgvsc": 6, "affected_exon": 6, "affected_protein": 6, "transcript_id": 6, "hgv": 6, "code": [6, 8, 12, 13], "dna": 6, "avail": [6, 7, 11], "els": 6, "predict": 6, "vep": [6, 7], "overlapping_exon": 6, "protein_affect": 6, "protein_effect_loc": 6, "end": 6, "surf1": 6, "refseq": [6, 7], "nm_123456": 6, "7": 6, "c": [6, 11, 13, 16], "9876g": 6, "definit": 6, "found": [6, 7], "www": [6, 7], "sequenceontologi": 6, "alter": 6, "posit": 6, "1234": 6, "1235": 6, "chrom": 6, "ref": 6, "alt": 6, "change_length": 6, "breakend": 6, "support": [6, 10, 11], "chromosom": [6, 10], "locat": [6, 11, 16], "allel": [6, 10, 11], "chang": 6, "due": 6, "presenc": [6, 13], "is_structur": 6, "notat": 6, "as_str": 6, "label": [6, 8], "contig": 6, "where": [6, 7], "n": [6, 13], "mai": 6, "empti": [6, 7], "gg": 6, "del": 6, "vcf": 6, "charact": 6, "INS": 6, "me": 6, "sine": 6, "lead": 6, "length": 6, "zero": 6, "insert": 6, "neg": 6, "check": [6, 7], "chr5": 6, "101": 6, "svtype": 6, "120": 6, "svlen": 6, "oppos": 6, "liter": 6, "nacgtacgtac": 6, "protein_id": [6, 7], "protein_featur": 6, "full": 6, "domain": 6, "subgroup": 6, "equal": 6, "repeat": 6, "region": 6, "motif": [6, 11], "feature_typ": 6, "ank": 6, "ankyrin": 6, "contain": 6, "11": [6, 13], "short": 6, "usual": 6, "than": 6, "20": [6, 13], "amino": 6, "acid": 6, "conserv": 6, "biolog": 6, "signific": 6, "combin": 6, "secondari": 6, "organ": 6, "characterist": 6, "three": [6, 10], "dimension": 6, "fold": 6, "interest": 6, "subsect": 6, "3": [6, 13], "4": [6, 7, 13], "find_coordin": 7, "item": [7, 8, 13], "determin": 7, "some": [7, 11, 13], "sort": [7, 13], "metaclass": 7, "establish": 7, "metadata": 7, "create_pati": [7, 16], "subject": 7, "genomicinterpret": 7, "phenopacket": [7, 11, 13], "phenotype_cr": 7, "var_func_ann": 7, "pp_directori": 7, "patient_cr": [7, 16], "json": [7, 13, 16], "file": [7, 16], "directori": [7, 16], "path": [7, 16], "folder": [7, 16], "error": 7, "rais": 7, "point": 7, "creator": [7, 16], "turn": 7, "made": 7, "valid": 7, "validationrunn": 7, "create_phenotyp": 7, "satisfi": 7, "issu": [7, 11], "except": 7, "baseexcept": 7, "datadir": 7, "store": [7, 16], "retriev": 7, "pickl": 7, "get_annot": 7, "store_annot": 7, "desir": [7, 10], "load": [7, 13, 16], "storag": 7, "under": 7, "cach": [7, 16], "fallback": 7, "exist": 7, "fuction": 7, "uniprot": [7, 16], "rest": 7, "api": [7, 11, 16], "here": [7, 13, 14, 16], "help": 7, "programmatic_access": 7, "work": [7, 16], "databas": 7, "np_037407": 7, "protein_annot": 7, "predictor": [7, 16], "validation_runn": 7, "cache_dir": 7, "variant_fallback": 7, "protein_fallback": 7, "conveni": [7, 16], "configur": [7, 16], "To": [7, 13], "need": [7, 13, 16], "toolkit": [7, 16], "runner": 7, "fetch": 7, "remot": 7, "should": 7, "current": [7, 12, 16], "case": 7, "non": 7, "parent": 7, "local": 7, "choos": [7, 13], "just": 7, "implement": [7, 10], "moment": 7, "html_row": 8, "anchor": 8, "url": 8, "payload": 8, "hpo_anchor": 8, "hpo_id": 8, "hpo_term_counts_t": 8, "min_count": 8, "html": 8, "design": [8, 11], "displai": 8, "jupyt": [8, 11], "notebook": [8, 11, 13], "ipython": 8, "per": 8, "provid": [8, 11, 16], "explan": 8, "minimum": 8, "tabl": [8, 13], "variants_t": 8, "preferred_transcript": 8, "could": 8, "genophenocorr": [9, 10, 12, 13, 16], "packag": [9, 11, 12, 13], "subpackag": 9, "domin": 10, "simpl": [10, 13], "carri": 10, "while": 10, "would": [10, 11], "principl": 10, "restrict": 10, "homozygot": 10, "who": 10, "two": [10, 13], "ident": 10, "copi": 10, "matern": 10, "patern": 10, "practic": 10, "compound": 10, "heterozygot": 10, "common": 10, "perform": [10, 11, 13, 16], "comparison": 10, "might": 10, "missens": [10, 11, 13], "against": 10, "nons": 10, "nonsens": 10, "see": [10, 13, 16], "exampl": [10, 13, 16], "li": 10, "jt": 10, "et": [10, 11], "al": [10, 11], "2022": 10, "pubm": 10, "ncbi": 10, "nlm": 10, "nih": 10, "gov": 10, "36303223": 10, "suox": 10, "6": [10, 13], "field": 10, "fischer": 10, "kind": 10, "concept": 11, "denot": 11, "attribut": 11, "medic": 11, "context": 11, "deviat": 11, "normal": 11, "morphologi": 11, "physiologi": 11, "behavior": 11, "f": [11, 13, 16], "deep": 11, "precis": 11, "medicin": 11, "question": 11, "biologi": 11, "human": [11, 16], "genet": 11, "concern": 11, "relationship": 11, "abnorm": 11, "mendelian": 11, "focu": 11, "place": 11, "studi": 11, "caus": 11, "manifest": 11, "python": [11, 12], "input": [11, 13, 15], "global": 11, "allianc": 11, "genom": 11, "health": 11, "ga4gh": 11, "ingest": 11, "v": [11, 13], "prematur": 11, "termin": 11, "codon": 11, "ontologi": [11, 13, 16], "statist": 11, "report": [11, 13], "recommend": 11, "user": 11, "document": [11, 12], "instal": [11, 16], "instruct": 11, "brief": 11, "tutori": 11, "comprehens": 11, "guid": 11, "technic": 11, "read": [11, 16], "background": 11, "orgogozo": 11, "2015": 11, "differenti": 11, "best": 11, "leav": 11, "ask": 11, "bug": 11, "tracker": 11, "how": [12, 13, 14, 16], "your": [12, 13], "environ": 12, "easi": 12, "publish": 12, "index": 12, "pypi": 12, "python3": 12, "m": [12, 13], "pip": 12, "clone": 12, "github": 12, "repositori": [12, 16], "git": 12, "com": 12, "monarch": 12, "initi": 12, "cd": 12, "switch": 12, "develop": 12, "branch": 12, "access": 12, "checkout": 12, "abov": 12, "virtual": 12, "demonstr": 13, "let": [13, 16], "import": [13, 16], "load_minimal_ontologi": [13, 16], "hp": [13, 16], "toi": 13, "latest": 13, "purl": [13, 16], "obolibrari": [13, 16], "obo": [13, 16], "todo": [13, 14, 15, 16], "move": 13, "workflow": 13, "now": [13, 16], "keep": 13, "thing": 13, "ship": 13, "section": [13, 15, 16], "learn": [13, 14], "command": 13, "b": 13, "d": 13, "h": 13, "j": 13, "k": 13, "l": 13, "o": [13, 16], "p": 13, "r": 13, "u": 13, "w": 13, "x": 13, "y": 13, "z": 13, "0001166": 13, "14": 13, "0001250": 13, "0001257": 13, "17": 13, "hetvar1": 13, "13": 13, "hetvar2": 13, "homvar1": 13, "homvar2": 13, "np_09876": 13, "5": [13, 16], "26": 13, "tx_dict": 13, "nm_1234": 13, "what": [13, 14], "partit": 13, "absenc": 13, "frameshift": 13, "cohort_analysi": 13, "With": 13, "arachnodactyli": 13, "30": 13, "77": 13, "76": 13, "92": 13, "04718": 13, "84": 13, "62": 13, "9": 13, "69": 13, "23": 13, "64472": 13, "spastic": 13, "8": 13, "61": 13, "54": 13, "00000": 13, "Or": 13, "similar": 13, "81": 13, "25": 13, "00": 13, "000781": 13, "68": 13, "75": 13, "60": 13, "692449": 13, "12": 13, "80": 13, "000000": 13, "present": 13, "row": 13, "ascend": 13, "order": 13, "write": [14, 15], "mani": 14, "distribut": 14, "plot": 14, "high": 15, "overview": 15, "bridg": 15, "standard": 16, "form": 16, "consist": 16, "hold": 16, "dataset": 16, "explor": 16, "exploratori": 16, "first": 16, "step": 16, "involv": 16, "prepar": 16, "downstream": 16, "easiest": 16, "wai": 16, "schema": 16, "out": 16, "box": 16, "requisit": 16, "ll": 16, "amaz": 16, "along": 16, "next": 16, "method": 16, "By": 16, "call": 16, "respons": 16, "save": 16, "bandwidth": 16, "purpos": 16, "simple_cohort": 16, "simple_cohort_path": 16, "walk": 16, "system": 16, "transform": 16, "googl": 16, "protobuf": 16, "json_format": 16, "pars": 16, "dirpath": 16, "_": 16, "filenam": 16, "endswith": 16, "pp_path": 16, "join": 16, "open": 16, "fh": 16, "pp": 16, "append": 16, "len": 16, "construct": 16, "raw": 16}, "objects": {"": [[0, 0, 0, "-", "genophenocorr"]], "genophenocorr": [[1, 0, 0, "-", "analysis"], [3, 0, 0, "-", "constants"], [5, 0, 0, "-", "data"], [6, 0, 0, "-", "model"], [7, 0, 0, "-", "preprocessing"], [8, 0, 0, "-", "view"]], "genophenocorr.analysis": [[1, 1, 1, "", "CohortAnalysis"], [2, 0, 0, "-", "predicate"]], "genophenocorr.analysis.CohortAnalysis": [[1, 2, 1, "id0", "analysis_cohort"], [1, 2, 1, "id1", "analysis_patients"], [1, 2, 1, "id2", "analysis_transcript"], [1, 3, 1, "", "compare_by_exon"], [1, 3, 1, "", "compare_by_protein_feature"], [1, 3, 1, "", "compare_by_protein_feature_type"], [1, 3, 1, "", "compare_by_variant"], [1, 3, 1, "", "compare_by_variant_type"], [1, 2, 1, "id5", "include_large_SV"], [1, 2, 1, "id4", "include_unmeasured"], [1, 2, 1, "id3", "is_recessive"], [1, 2, 1, "id6", "min_perc_patients_w_hpo"]], "genophenocorr.analysis.predicate": [[2, 1, 1, "", "ExonPredicate"], [2, 1, 1, "", "HPOPresentPredicate"], [2, 1, 1, "", "PatientCategory"], [2, 1, 1, "", "PolyPredicate"], [2, 1, 1, "", "ProtFeaturePredicate"], [2, 1, 1, "", "ProtFeatureTypePredicate"], [2, 1, 1, "", "SimplePredicate"], [2, 1, 1, "", "VariantEffectPredicate"], [2, 1, 1, "", "VariantPredicate"]], "genophenocorr.analysis.predicate.ExonPredicate": [[2, 3, 1, "", "categories"], [2, 3, 1, "", "test"]], "genophenocorr.analysis.predicate.HPOPresentPredicate": [[2, 4, 1, "", "NOT_MEASURED"], [2, 4, 1, "", "NOT_OBSERVED"], [2, 4, 1, "", "OBSERVED"], [2, 3, 1, "", "categories"], [2, 3, 1, "", "test"]], "genophenocorr.analysis.predicate.PatientCategory": [[2, 2, 1, "", "cat_id"], [2, 2, 1, "", "description"], [2, 2, 1, "", "name"]], "genophenocorr.analysis.predicate.PolyPredicate": [[2, 2, 1, "", "categories"], [2, 3, 1, "", "test"]], "genophenocorr.analysis.predicate.ProtFeaturePredicate": [[2, 3, 1, "", "categories"], [2, 3, 1, "", "test"]], "genophenocorr.analysis.predicate.ProtFeatureTypePredicate": [[2, 3, 1, "", "categories"], [2, 3, 1, "", "test"]], "genophenocorr.analysis.predicate.SimplePredicate": [[2, 3, 1, "", "test"]], "genophenocorr.analysis.predicate.VariantEffectPredicate": [[2, 3, 1, "", "categories"], [2, 3, 1, "", "test"]], "genophenocorr.analysis.predicate.VariantPredicate": [[2, 3, 1, "", "categories"], [2, 3, 1, "", "test"]], "genophenocorr.constants": [[4, 0, 0, "-", "variant_effects"]], "genophenocorr.constants.variant_effects": [[4, 1, 1, "", "VariantEffect"]], "genophenocorr.constants.variant_effects.VariantEffect": [[4, 4, 1, "", "CODING_SEQUENCE_VARIANT"], [4, 4, 1, "", "DOWNSTREAM_GENE_VARIANT"], [4, 4, 1, "", "FEATURE_ELONGATION"], [4, 4, 1, "", "FEATURE_TRUNCATION"], [4, 4, 1, "", "FIVE_PRIME_UTR_VARIANT"], [4, 4, 1, "", "FRAMESHIFT_VARIANT"], [4, 4, 1, "", "INCOMPLETE_TERMINAL_CODON_VARIANT"], [4, 4, 1, "", "INFRAME_DELETION"], [4, 4, 1, "", "INFRAME_INSERTION"], [4, 4, 1, "", "INTERGENIC_VARIANT"], [4, 4, 1, "", "INTRON_VARIANT"], [4, 4, 1, "", "MATURE_MIRNA_VARIANT"], [4, 4, 1, "", "MISSENSE_VARIANT"], [4, 4, 1, "", "NMD_TRANSCRIPT_VARIANT"], [4, 4, 1, "", "NON_CODING_TRANSCRIPT_EXON_VARIANT"], [4, 4, 1, "", "NON_CODING_TRANSCRIPT_VARIANT"], [4, 4, 1, "", "PROTEIN_ALTERING_VARIANT"], [4, 4, 1, "", "REGULATORY_REGION_ABLATION"], [4, 4, 1, "", "REGULATORY_REGION_AMPLIFICATION"], [4, 4, 1, "", "REGULATORY_REGION_VARIANT"], [4, 4, 1, "", "SEQUENCE_VARIANT"], [4, 4, 1, "", "SPLICE_ACCEPTOR_VARIANT"], [4, 4, 1, "", "SPLICE_DONOR_5TH_BASE_VARIANT"], [4, 4, 1, "", "SPLICE_DONOR_REGION_VARIANT"], [4, 4, 1, "", "SPLICE_DONOR_VARIANT"], [4, 4, 1, "", "SPLICE_POLYPYRIMIDINE_TRACT_VARIANT"], [4, 4, 1, "", "SPLICE_REGION_VARIANT"], [4, 4, 1, "", "START_LOST"], [4, 4, 1, "", "START_RETAINED_VARIANT"], [4, 4, 1, "", "STOP_GAINED"], [4, 4, 1, "", "STOP_LOST"], [4, 4, 1, "", "STOP_RETAINED_VARIANT"], [4, 4, 1, "", "SYNONYMOUS_VARIANT"], [4, 4, 1, "", "TFBS_ABLATION"], [4, 4, 1, "", "TFBS_AMPLIFICATION"], [4, 4, 1, "", "TF_BINDING_SITE_VARIANT"], [4, 4, 1, "", "THREE_PRIME_UTR_VARIANT"], [4, 4, 1, "", "TRANSCRIPT_ABLATION"], [4, 4, 1, "", "TRANSCRIPT_AMPLIFICATION"], [4, 4, 1, "", "UPSTREAM_GENE_VARIANT"]], "genophenocorr.data": [[5, 5, 1, "", "get_toy_cohort"]], "genophenocorr.model": [[6, 1, 1, "", "Cohort"], [6, 1, 1, "", "FeatureInfo"], [6, 1, 1, "", "FeatureType"], [6, 1, 1, "", "Patient"], [6, 1, 1, "", "Phenotype"], [6, 1, 1, "", "ProteinFeature"], [6, 1, 1, "", "ProteinMetadata"], [6, 1, 1, "", "TranscriptAnnotation"], [6, 1, 1, "", "Variant"], [6, 1, 1, "", "VariantCoordinates"]], "genophenocorr.model.Cohort": [[6, 2, 1, "", "all_patients"], [6, 2, 1, "", "all_phenotypes"], [6, 2, 1, "", "all_proteins"], [6, 2, 1, "", "all_transcripts"], [6, 2, 1, "", "all_variants"], [6, 3, 1, "", "from_patients"], [6, 3, 1, "", "list_all_patients"], [6, 3, 1, "", "list_all_phenotypes"], [6, 3, 1, "", "list_all_proteins"], [6, 3, 1, "", "list_all_variants"], [6, 3, 1, "", "list_data_by_tx"], [6, 2, 1, "", "total_patient_count"]], "genophenocorr.model.FeatureInfo": [[6, 2, 1, "id37", "end"], [6, 2, 1, "id35", "name"], [6, 2, 1, "id36", "start"]], "genophenocorr.model.FeatureType": [[6, 4, 1, "id40", "DOMAIN"], [6, 4, 1, "id39", "MOTIF"], [6, 4, 1, "id41", "REGION"], [6, 4, 1, "id38", "REPEAT"]], "genophenocorr.model.Patient": [[6, 2, 1, "id0", "patient_id"], [6, 2, 1, "id1", "phenotypes"], [6, 2, 1, "id3", "proteins"], [6, 2, 1, "id2", "variants"]], "genophenocorr.model.Phenotype": [[6, 3, 1, "", "from_term"], [6, 2, 1, "id4", "identifier"], [6, 2, 1, "id5", "name"], [6, 2, 1, "id6", "observed"]], "genophenocorr.model.ProteinFeature": [[6, 3, 1, "", "create"], [6, 2, 1, "", "feature_type"], [6, 2, 1, "", "info"]], "genophenocorr.model.ProteinMetadata": [[6, 3, 1, "id31", "domains"], [6, 2, 1, "id29", "label"], [6, 3, 1, "id34", "motifs"], [6, 2, 1, "id30", "protein_features"], [6, 2, 1, "id28", "protein_id"], [6, 3, 1, "id33", "regions"], [6, 3, 1, "id32", "repeats"]], "genophenocorr.model.TranscriptAnnotation": [[6, 2, 1, "id12", "gene_id"], [6, 2, 1, "id14", "hgvsc_id"], [6, 2, 1, "id16", "overlapping_exons"], [6, 2, 1, "id17", "protein_affected"], [6, 2, 1, "id18", "protein_effect_location"], [6, 2, 1, "id13", "transcript_id"], [6, 2, 1, "id15", "variant_effects"]], "genophenocorr.model.Variant": [[6, 3, 1, "", "create_variant_from_scratch"], [6, 2, 1, "id9", "genotype"], [6, 2, 1, "id10", "tx_annotations"], [6, 2, 1, "id11", "variant_class"], [6, 2, 1, "id7", "variant_coordinates"], [6, 2, 1, "id8", "variant_string"]], "genophenocorr.model.VariantCoordinates": [[6, 2, 1, "id23", "alt"], [6, 3, 1, "id27", "as_string"], [6, 2, 1, "id24", "change_length"], [6, 2, 1, "id19", "chrom"], [6, 2, 1, "id21", "end"], [6, 2, 1, "id25", "genotype"], [6, 3, 1, "id26", "is_structural"], [6, 2, 1, "id22", "ref"], [6, 2, 1, "id20", "start"]], "genophenocorr.preprocessing": [[7, 1, 1, "", "FunctionalAnnotator"], [7, 1, 1, "", "PatientCreator"], [7, 1, 1, "", "PhenopacketPatientCreator"], [7, 1, 1, "", "PhenopacketVariantCoordinateFinder"], [7, 1, 1, "", "PhenotypeCreator"], [7, 6, 1, "", "PhenotypeValidationException"], [7, 1, 1, "", "ProtCachingFunctionalAnnotator"], [7, 1, 1, "", "ProteinAnnotationCache"], [7, 1, 1, "", "ProteinMetadataService"], [7, 1, 1, "", "UniprotProteinMetadataService"], [7, 1, 1, "", "VarCachingFunctionalAnnotator"], [7, 1, 1, "", "VariantAnnotationCache"], [7, 1, 1, "", "VariantCoordinateFinder"], [7, 1, 1, "", "VepFunctionalAnnotator"], [7, 5, 1, "", "configure_caching_patient_creator"], [7, 5, 1, "", "load_phenopacket_folder"]], "genophenocorr.preprocessing.FunctionalAnnotator": [[7, 3, 1, "", "annotate"]], "genophenocorr.preprocessing.PatientCreator": [[7, 3, 1, "", "create_patient"]], "genophenocorr.preprocessing.PhenopacketPatientCreator": [[7, 3, 1, "", "create_patient"]], "genophenocorr.preprocessing.PhenopacketVariantCoordinateFinder": [[7, 3, 1, "", "find_coordinates"]], "genophenocorr.preprocessing.PhenotypeCreator": [[7, 3, 1, "", "create_phenotype"]], "genophenocorr.preprocessing.PhenotypeValidationException": [[7, 2, 1, "", "issues"]], "genophenocorr.preprocessing.ProtCachingFunctionalAnnotator": [[7, 3, 1, "", "annotate"]], "genophenocorr.preprocessing.ProteinAnnotationCache": [[7, 3, 1, "", "get_annotations"], [7, 3, 1, "", "store_annotations"]], "genophenocorr.preprocessing.ProteinMetadataService": [[7, 3, 1, "", "annotate"]], "genophenocorr.preprocessing.UniprotProteinMetadataService": [[7, 3, 1, "", "annotate"]], "genophenocorr.preprocessing.VarCachingFunctionalAnnotator": [[7, 3, 1, "", "annotate"]], "genophenocorr.preprocessing.VariantAnnotationCache": [[7, 3, 1, "", "get_annotations"], [7, 3, 1, "", "store_annotations"]], "genophenocorr.preprocessing.VariantCoordinateFinder": [[7, 3, 1, "", "find_coordinates"]], "genophenocorr.preprocessing.VepFunctionalAnnotator": [[7, 3, 1, "", "annotate"]], "genophenocorr.view": [[8, 1, 1, "", "CohortViewer"]], "genophenocorr.view.CohortViewer": [[8, 3, 1, "", "anchor"], [8, 3, 1, "", "hpo_anchor"], [8, 3, 1, "", "hpo_term_counts_table"], [8, 3, 1, "", "html_row"], [8, 3, 1, "", "variants_table"]]}, "objtypes": {"0": "py:module", "1": "py:class", "2": "py:property", "3": "py:method", "4": "py:attribute", "5": "py:function", "6": "py:exception"}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "class", "Python class"], "2": ["py", "property", "Python property"], "3": ["py", "method", "Python method"], "4": ["py", "attribute", "Python attribute"], "5": ["py", "function", "Python function"], "6": ["py", "exception", "Python exception"]}, "titleterms": {"genophenocorr": [0, 1, 2, 3, 4, 5, 6, 7, 8, 11], "packag": [0, 1, 2, 3, 5, 6, 7, 8], "subpackag": [0, 1], "analysi": [1, 2, 13, 14], "predic": 2, "constant": [3, 4], "submodul": 3, "variant_effect": 4, "modul": 4, "data": [5, 16], "model": 6, "preprocess": 7, "view": 8, "api": 9, "refer": 9, "genotyp": 10, "phenotyp": 10, "correl": 10, "autosom": 10, "recess": 10, "diseas": 10, "tutori": [10, 13], "literatur": 11, "feedback": 11, "instal": 12, "stabl": 12, "releas": 12, "latest": 12, "set": 13, "up": 13, "prepar": 13, "sampl": 13, "cohort": [14, 16], "exploratori": 14, "user": 15, "guid": 15, "content": 15, "input": 16, "creat": 16, "from": 16, "ga4gh": 16, "phenopacket": 16, "other": 16}, "envversion": {"sphinx.domains.c": 3, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 9, "sphinx.domains.index": 1, "sphinx.domains.javascript": 3, "sphinx.domains.math": 2, "sphinx.domains.python": 4, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.viewcode": 1, "sphinx.ext.intersphinx": 1, "sphinx": 60}, "alltitles": {"genophenocorr package": [[0, "module-genophenocorr"]], "Subpackages": [[0, "subpackages"], [1, "subpackages"]], "genophenocorr.analysis package": [[1, "module-genophenocorr.analysis"]], "genophenocorr.analysis.predicate package": [[2, "module-genophenocorr.analysis.predicate"]], "genophenocorr.constants package": [[3, "module-genophenocorr.constants"]], "Submodules": [[3, "submodules"]], "genophenocorr.constants.variant_effects module": [[4, "module-genophenocorr.constants.variant_effects"]], "genophenocorr.data package": [[5, "module-genophenocorr.data"]], "genophenocorr.model package": [[6, "module-genophenocorr.model"]], "genophenocorr.preprocessing package": [[7, "module-genophenocorr.preprocessing"]], "genophenocorr.view package": [[8, "module-genophenocorr.view"]], "API reference": [[9, "api-reference"]], "Genotype-Phenotype Correlations in Autosomal Recessive Diseases": [[10, "genotype-phenotype-correlations-in-autosomal-recessive-diseases"]], "Tutorial": [[10, "tutorial"], [13, "tutorial"]], "genophenocorr": [[11, "genophenocorr"]], "Literature": [[11, "literature"]], "Feedback": [[11, "feedback"]], "Installation": [[12, "installation"]], "Stable release": [[12, "stable-release"]], "Latest release": [[12, "latest-release"]], "Set up analysis": [[13, "set-up-analysis"]], "Prepare samples": [[13, "prepare-samples"]], "Cohort exploratory analysis": [[14, "cohort-exploratory-analysis"]], "User guide": [[15, "user-guide"]], "Contents:": [[15, null]], "Input data": [[16, "input-data"]], "Create a cohort from GA4GH phenopackets": [[16, "create-a-cohort-from-ga4gh-phenopackets"]], "Create a cohort from other data": [[16, "create-a-cohort-from-other-data"]]}, "indexentries": {"genophenocorr": [[0, "module-genophenocorr"]], "module": [[0, "module-genophenocorr"], [1, "module-genophenocorr.analysis"], [2, "module-genophenocorr.analysis.predicate"], [3, "module-genophenocorr.constants"], [4, "module-genophenocorr.constants.variant_effects"], [5, "module-genophenocorr.data"], [6, "module-genophenocorr.model"], [7, "module-genophenocorr.preprocessing"], [8, "module-genophenocorr.view"]], "cohortanalysis (class in genophenocorr.analysis)": [[1, "genophenocorr.analysis.CohortAnalysis"]], "analysis_cohort (genophenocorr.analysis.cohortanalysis attribute)": [[1, "genophenocorr.analysis.CohortAnalysis.analysis_cohort"]], "analysis_cohort (genophenocorr.analysis.cohortanalysis property)": [[1, "id0"]], "analysis_patients (genophenocorr.analysis.cohortanalysis attribute)": [[1, "genophenocorr.analysis.CohortAnalysis.analysis_patients"]], "analysis_patients (genophenocorr.analysis.cohortanalysis property)": [[1, "id1"]], "analysis_transcript (genophenocorr.analysis.cohortanalysis attribute)": [[1, "genophenocorr.analysis.CohortAnalysis.analysis_transcript"]], "analysis_transcript (genophenocorr.analysis.cohortanalysis property)": [[1, "id2"]], "compare_by_exon() (genophenocorr.analysis.cohortanalysis method)": [[1, "genophenocorr.analysis.CohortAnalysis.compare_by_exon"]], "compare_by_protein_feature() (genophenocorr.analysis.cohortanalysis method)": [[1, "genophenocorr.analysis.CohortAnalysis.compare_by_protein_feature"]], "compare_by_protein_feature_type() (genophenocorr.analysis.cohortanalysis method)": [[1, "genophenocorr.analysis.CohortAnalysis.compare_by_protein_feature_type"]], "compare_by_variant() (genophenocorr.analysis.cohortanalysis method)": [[1, "genophenocorr.analysis.CohortAnalysis.compare_by_variant"]], "compare_by_variant_type() (genophenocorr.analysis.cohortanalysis method)": [[1, "genophenocorr.analysis.CohortAnalysis.compare_by_variant_type"]], "genophenocorr.analysis": [[1, "module-genophenocorr.analysis"]], "include_large_sv (genophenocorr.analysis.cohortanalysis attribute)": [[1, "genophenocorr.analysis.CohortAnalysis.include_large_SV"]], "include_large_sv (genophenocorr.analysis.cohortanalysis property)": [[1, "id5"]], "include_unmeasured (genophenocorr.analysis.cohortanalysis attribute)": [[1, "genophenocorr.analysis.CohortAnalysis.include_unmeasured"]], "include_unmeasured (genophenocorr.analysis.cohortanalysis property)": [[1, "id4"]], "is_recessive (genophenocorr.analysis.cohortanalysis attribute)": [[1, "genophenocorr.analysis.CohortAnalysis.is_recessive"]], "is_recessive (genophenocorr.analysis.cohortanalysis property)": [[1, "id3"]], "min_perc_patients_w_hpo (genophenocorr.analysis.cohortanalysis attribute)": [[1, "genophenocorr.analysis.CohortAnalysis.min_perc_patients_w_hpo"]], "min_perc_patients_w_hpo (genophenocorr.analysis.cohortanalysis property)": [[1, "id6"]], "exonpredicate (class in genophenocorr.analysis.predicate)": [[2, "genophenocorr.analysis.predicate.ExonPredicate"]], "hpopresentpredicate (class in genophenocorr.analysis.predicate)": [[2, "genophenocorr.analysis.predicate.HPOPresentPredicate"]], "not_measured (genophenocorr.analysis.predicate.hpopresentpredicate attribute)": [[2, "genophenocorr.analysis.predicate.HPOPresentPredicate.NOT_MEASURED"]], "not_observed (genophenocorr.analysis.predicate.hpopresentpredicate attribute)": [[2, "genophenocorr.analysis.predicate.HPOPresentPredicate.NOT_OBSERVED"]], "observed (genophenocorr.analysis.predicate.hpopresentpredicate attribute)": [[2, "genophenocorr.analysis.predicate.HPOPresentPredicate.OBSERVED"]], "patientcategory (class in genophenocorr.analysis.predicate)": [[2, "genophenocorr.analysis.predicate.PatientCategory"]], "polypredicate (class in genophenocorr.analysis.predicate)": [[2, "genophenocorr.analysis.predicate.PolyPredicate"]], "protfeaturepredicate (class in genophenocorr.analysis.predicate)": [[2, "genophenocorr.analysis.predicate.ProtFeaturePredicate"]], "protfeaturetypepredicate (class in genophenocorr.analysis.predicate)": [[2, "genophenocorr.analysis.predicate.ProtFeatureTypePredicate"]], "simplepredicate (class in genophenocorr.analysis.predicate)": [[2, "genophenocorr.analysis.predicate.SimplePredicate"]], "varianteffectpredicate (class in genophenocorr.analysis.predicate)": [[2, "genophenocorr.analysis.predicate.VariantEffectPredicate"]], "variantpredicate (class in genophenocorr.analysis.predicate)": [[2, "genophenocorr.analysis.predicate.VariantPredicate"]], "cat_id (genophenocorr.analysis.predicate.patientcategory property)": [[2, "genophenocorr.analysis.predicate.PatientCategory.cat_id"]], "categories (genophenocorr.analysis.predicate.polypredicate property)": [[2, "genophenocorr.analysis.predicate.PolyPredicate.categories"]], "categories() (genophenocorr.analysis.predicate.exonpredicate method)": [[2, "genophenocorr.analysis.predicate.ExonPredicate.categories"]], "categories() (genophenocorr.analysis.predicate.hpopresentpredicate method)": [[2, "genophenocorr.analysis.predicate.HPOPresentPredicate.categories"]], "categories() (genophenocorr.analysis.predicate.protfeaturepredicate method)": [[2, "genophenocorr.analysis.predicate.ProtFeaturePredicate.categories"]], "categories() (genophenocorr.analysis.predicate.protfeaturetypepredicate method)": [[2, "genophenocorr.analysis.predicate.ProtFeatureTypePredicate.categories"]], "categories() (genophenocorr.analysis.predicate.varianteffectpredicate method)": [[2, "genophenocorr.analysis.predicate.VariantEffectPredicate.categories"]], "categories() (genophenocorr.analysis.predicate.variantpredicate method)": [[2, "genophenocorr.analysis.predicate.VariantPredicate.categories"]], "description (genophenocorr.analysis.predicate.patientcategory property)": [[2, "genophenocorr.analysis.predicate.PatientCategory.description"]], "genophenocorr.analysis.predicate": [[2, "module-genophenocorr.analysis.predicate"]], "name (genophenocorr.analysis.predicate.patientcategory property)": [[2, "genophenocorr.analysis.predicate.PatientCategory.name"]], "test() (genophenocorr.analysis.predicate.exonpredicate method)": [[2, "genophenocorr.analysis.predicate.ExonPredicate.test"]], "test() (genophenocorr.analysis.predicate.hpopresentpredicate method)": [[2, "genophenocorr.analysis.predicate.HPOPresentPredicate.test"]], "test() (genophenocorr.analysis.predicate.polypredicate method)": [[2, "genophenocorr.analysis.predicate.PolyPredicate.test"]], "test() (genophenocorr.analysis.predicate.protfeaturepredicate method)": [[2, "genophenocorr.analysis.predicate.ProtFeaturePredicate.test"]], "test() (genophenocorr.analysis.predicate.protfeaturetypepredicate method)": [[2, "genophenocorr.analysis.predicate.ProtFeatureTypePredicate.test"]], "test() (genophenocorr.analysis.predicate.simplepredicate method)": [[2, "genophenocorr.analysis.predicate.SimplePredicate.test"]], "test() (genophenocorr.analysis.predicate.varianteffectpredicate method)": [[2, "genophenocorr.analysis.predicate.VariantEffectPredicate.test"]], "test() (genophenocorr.analysis.predicate.variantpredicate method)": [[2, "genophenocorr.analysis.predicate.VariantPredicate.test"]], "genophenocorr.constants": [[3, "module-genophenocorr.constants"]], "coding_sequence_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.CODING_SEQUENCE_VARIANT"]], "downstream_gene_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.DOWNSTREAM_GENE_VARIANT"]], "feature_elongation (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.FEATURE_ELONGATION"]], "feature_truncation (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.FEATURE_TRUNCATION"]], "five_prime_utr_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.FIVE_PRIME_UTR_VARIANT"]], "frameshift_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.FRAMESHIFT_VARIANT"]], "incomplete_terminal_codon_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.INCOMPLETE_TERMINAL_CODON_VARIANT"]], "inframe_deletion (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.INFRAME_DELETION"]], "inframe_insertion (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.INFRAME_INSERTION"]], "intergenic_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.INTERGENIC_VARIANT"]], "intron_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.INTRON_VARIANT"]], "mature_mirna_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.MATURE_MIRNA_VARIANT"]], "missense_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.MISSENSE_VARIANT"]], "nmd_transcript_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.NMD_TRANSCRIPT_VARIANT"]], "non_coding_transcript_exon_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.NON_CODING_TRANSCRIPT_EXON_VARIANT"]], "non_coding_transcript_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.NON_CODING_TRANSCRIPT_VARIANT"]], "protein_altering_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.PROTEIN_ALTERING_VARIANT"]], "regulatory_region_ablation (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.REGULATORY_REGION_ABLATION"]], "regulatory_region_amplification (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.REGULATORY_REGION_AMPLIFICATION"]], "regulatory_region_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.REGULATORY_REGION_VARIANT"]], "sequence_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.SEQUENCE_VARIANT"]], "splice_acceptor_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.SPLICE_ACCEPTOR_VARIANT"]], "splice_donor_5th_base_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.SPLICE_DONOR_5TH_BASE_VARIANT"]], "splice_donor_region_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.SPLICE_DONOR_REGION_VARIANT"]], "splice_donor_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.SPLICE_DONOR_VARIANT"]], "splice_polypyrimidine_tract_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.SPLICE_POLYPYRIMIDINE_TRACT_VARIANT"]], "splice_region_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.SPLICE_REGION_VARIANT"]], "start_lost (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.START_LOST"]], "start_retained_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.START_RETAINED_VARIANT"]], "stop_gained (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.STOP_GAINED"]], "stop_lost (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.STOP_LOST"]], "stop_retained_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.STOP_RETAINED_VARIANT"]], "synonymous_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.SYNONYMOUS_VARIANT"]], "tfbs_ablation (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.TFBS_ABLATION"]], "tfbs_amplification (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.TFBS_AMPLIFICATION"]], "tf_binding_site_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.TF_BINDING_SITE_VARIANT"]], "three_prime_utr_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.THREE_PRIME_UTR_VARIANT"]], "transcript_ablation (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.TRANSCRIPT_ABLATION"]], "transcript_amplification (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.TRANSCRIPT_AMPLIFICATION"]], "upstream_gene_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[4, "genophenocorr.constants.variant_effects.VariantEffect.UPSTREAM_GENE_VARIANT"]], "varianteffect (class in genophenocorr.constants.variant_effects)": [[4, "genophenocorr.constants.variant_effects.VariantEffect"]], "genophenocorr.constants.variant_effects": [[4, "module-genophenocorr.constants.variant_effects"]], "genophenocorr.data": [[5, "module-genophenocorr.data"]], "get_toy_cohort() (in module genophenocorr.data)": [[5, "genophenocorr.data.get_toy_cohort"]], "cohort (class in genophenocorr.model)": [[6, "genophenocorr.model.Cohort"]], "domain (genophenocorr.model.featuretype attribute)": [[6, "genophenocorr.model.FeatureType.DOMAIN"], [6, "id40"]], "featureinfo (class in genophenocorr.model)": [[6, "genophenocorr.model.FeatureInfo"]], "featuretype (class in genophenocorr.model)": [[6, "genophenocorr.model.FeatureType"]], "motif (genophenocorr.model.featuretype attribute)": [[6, "genophenocorr.model.FeatureType.MOTIF"], [6, "id39"]], "patient (class in genophenocorr.model)": [[6, "genophenocorr.model.Patient"]], "phenotype (class in genophenocorr.model)": [[6, "genophenocorr.model.Phenotype"]], "proteinfeature (class in genophenocorr.model)": [[6, "genophenocorr.model.ProteinFeature"]], "proteinmetadata (class in genophenocorr.model)": [[6, "genophenocorr.model.ProteinMetadata"]], "region (genophenocorr.model.featuretype attribute)": [[6, "genophenocorr.model.FeatureType.REGION"], [6, "id41"]], "repeat (genophenocorr.model.featuretype attribute)": [[6, "genophenocorr.model.FeatureType.REPEAT"], [6, "id38"]], "transcriptannotation (class in genophenocorr.model)": [[6, "genophenocorr.model.TranscriptAnnotation"]], "variant (class in genophenocorr.model)": [[6, "genophenocorr.model.Variant"]], "variantcoordinates (class in genophenocorr.model)": [[6, "genophenocorr.model.VariantCoordinates"]], "all_patients (genophenocorr.model.cohort property)": [[6, "genophenocorr.model.Cohort.all_patients"]], "all_phenotypes (genophenocorr.model.cohort property)": [[6, "genophenocorr.model.Cohort.all_phenotypes"]], "all_proteins (genophenocorr.model.cohort property)": [[6, "genophenocorr.model.Cohort.all_proteins"]], "all_transcripts (genophenocorr.model.cohort property)": [[6, "genophenocorr.model.Cohort.all_transcripts"]], "all_variants (genophenocorr.model.cohort property)": [[6, "genophenocorr.model.Cohort.all_variants"]], "alt (genophenocorr.model.variantcoordinates attribute)": [[6, "genophenocorr.model.VariantCoordinates.alt"]], "alt (genophenocorr.model.variantcoordinates property)": [[6, "id23"]], "as_string() (genophenocorr.model.variantcoordinates method)": [[6, "genophenocorr.model.VariantCoordinates.as_string"], [6, "id27"]], "change_length (genophenocorr.model.variantcoordinates attribute)": [[6, "genophenocorr.model.VariantCoordinates.change_length"]], "change_length (genophenocorr.model.variantcoordinates property)": [[6, "id24"]], "chrom (genophenocorr.model.variantcoordinates attribute)": [[6, "genophenocorr.model.VariantCoordinates.chrom"]], "chrom (genophenocorr.model.variantcoordinates property)": [[6, "id19"]], "create() (genophenocorr.model.proteinfeature static method)": [[6, "genophenocorr.model.ProteinFeature.create"]], "create_variant_from_scratch() (genophenocorr.model.variant static method)": [[6, "genophenocorr.model.Variant.create_variant_from_scratch"]], "domains() (genophenocorr.model.proteinmetadata method)": [[6, "genophenocorr.model.ProteinMetadata.domains"], [6, "id31"]], "end (genophenocorr.model.featureinfo attribute)": [[6, "genophenocorr.model.FeatureInfo.end"]], "end (genophenocorr.model.featureinfo property)": [[6, "id37"]], "end (genophenocorr.model.variantcoordinates attribute)": [[6, "genophenocorr.model.VariantCoordinates.end"]], "end (genophenocorr.model.variantcoordinates property)": [[6, "id21"]], "feature_type (genophenocorr.model.proteinfeature property)": [[6, "genophenocorr.model.ProteinFeature.feature_type"]], "from_patients() (genophenocorr.model.cohort static method)": [[6, "genophenocorr.model.Cohort.from_patients"]], "from_term() (genophenocorr.model.phenotype static method)": [[6, "genophenocorr.model.Phenotype.from_term"]], "gene_id (genophenocorr.model.transcriptannotation attribute)": [[6, "genophenocorr.model.TranscriptAnnotation.gene_id"]], "gene_id (genophenocorr.model.transcriptannotation property)": [[6, "id12"]], "genophenocorr.model": [[6, "module-genophenocorr.model"]], "genotype (genophenocorr.model.variant attribute)": [[6, "genophenocorr.model.Variant.genotype"]], "genotype (genophenocorr.model.variant property)": [[6, "id9"]], "genotype (genophenocorr.model.variantcoordinates attribute)": [[6, "genophenocorr.model.VariantCoordinates.genotype"]], "genotype (genophenocorr.model.variantcoordinates property)": [[6, "id25"]], "hgvsc_id (genophenocorr.model.transcriptannotation attribute)": [[6, "genophenocorr.model.TranscriptAnnotation.hgvsc_id"]], "hgvsc_id (genophenocorr.model.transcriptannotation property)": [[6, "id14"]], "identifier (genophenocorr.model.phenotype attribute)": [[6, "genophenocorr.model.Phenotype.identifier"]], "identifier (genophenocorr.model.phenotype property)": [[6, "id4"]], "info (genophenocorr.model.proteinfeature property)": [[6, "genophenocorr.model.ProteinFeature.info"]], "is_structural() (genophenocorr.model.variantcoordinates method)": [[6, "genophenocorr.model.VariantCoordinates.is_structural"], [6, "id26"]], "label (genophenocorr.model.proteinmetadata attribute)": [[6, "genophenocorr.model.ProteinMetadata.label"]], "label (genophenocorr.model.proteinmetadata property)": [[6, "id29"]], "list_all_patients() (genophenocorr.model.cohort method)": [[6, "genophenocorr.model.Cohort.list_all_patients"]], "list_all_phenotypes() (genophenocorr.model.cohort method)": [[6, "genophenocorr.model.Cohort.list_all_phenotypes"]], "list_all_proteins() (genophenocorr.model.cohort method)": [[6, "genophenocorr.model.Cohort.list_all_proteins"]], "list_all_variants() (genophenocorr.model.cohort method)": [[6, "genophenocorr.model.Cohort.list_all_variants"]], "list_data_by_tx() (genophenocorr.model.cohort method)": [[6, "genophenocorr.model.Cohort.list_data_by_tx"]], "motifs() (genophenocorr.model.proteinmetadata method)": [[6, "genophenocorr.model.ProteinMetadata.motifs"], [6, "id34"]], "name (genophenocorr.model.featureinfo attribute)": [[6, "genophenocorr.model.FeatureInfo.name"]], "name (genophenocorr.model.featureinfo property)": [[6, "id35"]], "name (genophenocorr.model.phenotype attribute)": [[6, "genophenocorr.model.Phenotype.name"]], "name (genophenocorr.model.phenotype property)": [[6, "id5"]], "observed (genophenocorr.model.phenotype attribute)": [[6, "genophenocorr.model.Phenotype.observed"]], "observed (genophenocorr.model.phenotype property)": [[6, "id6"]], "overlapping_exons (genophenocorr.model.transcriptannotation attribute)": [[6, "genophenocorr.model.TranscriptAnnotation.overlapping_exons"]], "overlapping_exons (genophenocorr.model.transcriptannotation property)": [[6, "id16"]], "patient_id (genophenocorr.model.patient attribute)": [[6, "genophenocorr.model.Patient.patient_id"]], "patient_id (genophenocorr.model.patient property)": [[6, "id0"]], "phenotypes (genophenocorr.model.patient attribute)": [[6, "genophenocorr.model.Patient.phenotypes"]], "phenotypes (genophenocorr.model.patient property)": [[6, "id1"]], "protein_affected (genophenocorr.model.transcriptannotation attribute)": [[6, "genophenocorr.model.TranscriptAnnotation.protein_affected"]], "protein_affected (genophenocorr.model.transcriptannotation property)": [[6, "id17"]], "protein_effect_location (genophenocorr.model.transcriptannotation attribute)": [[6, "genophenocorr.model.TranscriptAnnotation.protein_effect_location"]], "protein_effect_location (genophenocorr.model.transcriptannotation property)": [[6, "id18"]], "protein_features (genophenocorr.model.proteinmetadata attribute)": [[6, "genophenocorr.model.ProteinMetadata.protein_features"]], "protein_features (genophenocorr.model.proteinmetadata property)": [[6, "id30"]], "protein_id (genophenocorr.model.proteinmetadata attribute)": [[6, "genophenocorr.model.ProteinMetadata.protein_id"]], "protein_id (genophenocorr.model.proteinmetadata property)": [[6, "id28"]], "proteins (genophenocorr.model.patient attribute)": [[6, "genophenocorr.model.Patient.proteins"]], "proteins (genophenocorr.model.patient property)": [[6, "id3"]], "ref (genophenocorr.model.variantcoordinates attribute)": [[6, "genophenocorr.model.VariantCoordinates.ref"]], "ref (genophenocorr.model.variantcoordinates property)": [[6, "id22"]], "regions() (genophenocorr.model.proteinmetadata method)": [[6, "genophenocorr.model.ProteinMetadata.regions"], [6, "id33"]], "repeats() (genophenocorr.model.proteinmetadata method)": [[6, "genophenocorr.model.ProteinMetadata.repeats"], [6, "id32"]], "start (genophenocorr.model.featureinfo attribute)": [[6, "genophenocorr.model.FeatureInfo.start"]], "start (genophenocorr.model.featureinfo property)": [[6, "id36"]], "start (genophenocorr.model.variantcoordinates attribute)": [[6, "genophenocorr.model.VariantCoordinates.start"]], "start (genophenocorr.model.variantcoordinates property)": [[6, "id20"]], "total_patient_count (genophenocorr.model.cohort property)": [[6, "genophenocorr.model.Cohort.total_patient_count"]], "transcript_id (genophenocorr.model.transcriptannotation attribute)": [[6, "genophenocorr.model.TranscriptAnnotation.transcript_id"]], "transcript_id (genophenocorr.model.transcriptannotation property)": [[6, "id13"]], "tx_annotations (genophenocorr.model.variant attribute)": [[6, "genophenocorr.model.Variant.tx_annotations"]], "tx_annotations (genophenocorr.model.variant property)": [[6, "id10"]], "variant_class (genophenocorr.model.variant attribute)": [[6, "genophenocorr.model.Variant.variant_class"]], "variant_class (genophenocorr.model.variant property)": [[6, "id11"]], "variant_coordinates (genophenocorr.model.variant attribute)": [[6, "genophenocorr.model.Variant.variant_coordinates"]], "variant_coordinates (genophenocorr.model.variant property)": [[6, "id7"]], "variant_effects (genophenocorr.model.transcriptannotation attribute)": [[6, "genophenocorr.model.TranscriptAnnotation.variant_effects"]], "variant_effects (genophenocorr.model.transcriptannotation property)": [[6, "id15"]], "variant_string (genophenocorr.model.variant attribute)": [[6, "genophenocorr.model.Variant.variant_string"]], "variant_string (genophenocorr.model.variant property)": [[6, "id8"]], "variants (genophenocorr.model.patient attribute)": [[6, "genophenocorr.model.Patient.variants"]], "variants (genophenocorr.model.patient property)": [[6, "id2"]], "functionalannotator (class in genophenocorr.preprocessing)": [[7, "genophenocorr.preprocessing.FunctionalAnnotator"]], "patientcreator (class in genophenocorr.preprocessing)": [[7, "genophenocorr.preprocessing.PatientCreator"]], "phenopacketpatientcreator (class in genophenocorr.preprocessing)": [[7, "genophenocorr.preprocessing.PhenopacketPatientCreator"]], "phenopacketvariantcoordinatefinder (class in genophenocorr.preprocessing)": [[7, "genophenocorr.preprocessing.PhenopacketVariantCoordinateFinder"]], "phenotypecreator (class in genophenocorr.preprocessing)": [[7, "genophenocorr.preprocessing.PhenotypeCreator"]], "phenotypevalidationexception": [[7, "genophenocorr.preprocessing.PhenotypeValidationException"]], "protcachingfunctionalannotator (class in genophenocorr.preprocessing)": [[7, "genophenocorr.preprocessing.ProtCachingFunctionalAnnotator"]], "proteinannotationcache (class in genophenocorr.preprocessing)": [[7, "genophenocorr.preprocessing.ProteinAnnotationCache"]], "proteinmetadataservice (class in genophenocorr.preprocessing)": [[7, "genophenocorr.preprocessing.ProteinMetadataService"]], "uniprotproteinmetadataservice (class in genophenocorr.preprocessing)": [[7, "genophenocorr.preprocessing.UniprotProteinMetadataService"]], "varcachingfunctionalannotator (class in genophenocorr.preprocessing)": [[7, "genophenocorr.preprocessing.VarCachingFunctionalAnnotator"]], "variantannotationcache (class in genophenocorr.preprocessing)": [[7, "genophenocorr.preprocessing.VariantAnnotationCache"]], "variantcoordinatefinder (class in genophenocorr.preprocessing)": [[7, "genophenocorr.preprocessing.VariantCoordinateFinder"]], "vepfunctionalannotator (class in genophenocorr.preprocessing)": [[7, "genophenocorr.preprocessing.VepFunctionalAnnotator"]], "annotate() (genophenocorr.preprocessing.functionalannotator method)": [[7, "genophenocorr.preprocessing.FunctionalAnnotator.annotate"]], "annotate() (genophenocorr.preprocessing.protcachingfunctionalannotator method)": [[7, "genophenocorr.preprocessing.ProtCachingFunctionalAnnotator.annotate"]], "annotate() (genophenocorr.preprocessing.proteinmetadataservice method)": [[7, "genophenocorr.preprocessing.ProteinMetadataService.annotate"]], "annotate() (genophenocorr.preprocessing.uniprotproteinmetadataservice method)": [[7, "genophenocorr.preprocessing.UniprotProteinMetadataService.annotate"]], "annotate() (genophenocorr.preprocessing.varcachingfunctionalannotator method)": [[7, "genophenocorr.preprocessing.VarCachingFunctionalAnnotator.annotate"]], "annotate() (genophenocorr.preprocessing.vepfunctionalannotator method)": [[7, "genophenocorr.preprocessing.VepFunctionalAnnotator.annotate"]], "configure_caching_patient_creator() (in module genophenocorr.preprocessing)": [[7, "genophenocorr.preprocessing.configure_caching_patient_creator"]], "create_patient() (genophenocorr.preprocessing.patientcreator method)": [[7, "genophenocorr.preprocessing.PatientCreator.create_patient"]], "create_patient() (genophenocorr.preprocessing.phenopacketpatientcreator method)": [[7, "genophenocorr.preprocessing.PhenopacketPatientCreator.create_patient"]], "create_phenotype() (genophenocorr.preprocessing.phenotypecreator method)": [[7, "genophenocorr.preprocessing.PhenotypeCreator.create_phenotype"]], "find_coordinates() (genophenocorr.preprocessing.phenopacketvariantcoordinatefinder method)": [[7, "genophenocorr.preprocessing.PhenopacketVariantCoordinateFinder.find_coordinates"]], "find_coordinates() (genophenocorr.preprocessing.variantcoordinatefinder method)": [[7, "genophenocorr.preprocessing.VariantCoordinateFinder.find_coordinates"]], "genophenocorr.preprocessing": [[7, "module-genophenocorr.preprocessing"]], "get_annotations() (genophenocorr.preprocessing.proteinannotationcache method)": [[7, "genophenocorr.preprocessing.ProteinAnnotationCache.get_annotations"]], "get_annotations() (genophenocorr.preprocessing.variantannotationcache method)": [[7, "genophenocorr.preprocessing.VariantAnnotationCache.get_annotations"]], "issues (genophenocorr.preprocessing.phenotypevalidationexception property)": [[7, "genophenocorr.preprocessing.PhenotypeValidationException.issues"]], "load_phenopacket_folder() (in module genophenocorr.preprocessing)": [[7, "genophenocorr.preprocessing.load_phenopacket_folder"]], "store_annotations() (genophenocorr.preprocessing.proteinannotationcache method)": [[7, "genophenocorr.preprocessing.ProteinAnnotationCache.store_annotations"]], "store_annotations() (genophenocorr.preprocessing.variantannotationcache method)": [[7, "genophenocorr.preprocessing.VariantAnnotationCache.store_annotations"]], "cohortviewer (class in genophenocorr.view)": [[8, "genophenocorr.view.CohortViewer"]], "anchor() (genophenocorr.view.cohortviewer static method)": [[8, "genophenocorr.view.CohortViewer.anchor"]], "genophenocorr.view": [[8, "module-genophenocorr.view"]], "hpo_anchor() (genophenocorr.view.cohortviewer static method)": [[8, "genophenocorr.view.CohortViewer.hpo_anchor"]], "hpo_term_counts_table() (genophenocorr.view.cohortviewer method)": [[8, "genophenocorr.view.CohortViewer.hpo_term_counts_table"]], "html_row() (genophenocorr.view.cohortviewer static method)": [[8, "genophenocorr.view.CohortViewer.html_row"]], "variants_table() (genophenocorr.view.cohortviewer method)": [[8, "genophenocorr.view.CohortViewer.variants_table"]]}})