Search.setIndex({"docnames": ["apidocs/genophenocorr", "apidocs/genophenocorr.analysis", "apidocs/genophenocorr.analysis.predicate", "apidocs/genophenocorr.analysis.test_fisherExact", "apidocs/genophenocorr.constants", "apidocs/genophenocorr.constants.variant_effects", "apidocs/genophenocorr.data", "apidocs/genophenocorr.model", "apidocs/genophenocorr.preprocessing", "apidocs/genophenocorr.preprocessing.test_variant", "apidocs/genophenocorr.view", "apidocs/modules", "autosomal_recessive", "index", "installation", "tutorial", "user-guide/exploratory", "user-guide/index", "user-guide/input-data"], "filenames": ["apidocs/genophenocorr.rst", "apidocs/genophenocorr.analysis.rst", "apidocs/genophenocorr.analysis.predicate.rst", "apidocs/genophenocorr.analysis.test_fisherExact.rst", "apidocs/genophenocorr.constants.rst", "apidocs/genophenocorr.constants.variant_effects.rst", "apidocs/genophenocorr.data.rst", "apidocs/genophenocorr.model.rst", "apidocs/genophenocorr.preprocessing.rst", "apidocs/genophenocorr.preprocessing.test_variant.rst", "apidocs/genophenocorr.view.rst", "apidocs/modules.rst", "autosomal_recessive.rst", "index.rst", "installation.rst", "tutorial.rst", "user-guide/exploratory.rst", "user-guide/index.rst", "user-guide/input-data.rst"], "titles": ["genophenocorr package", "genophenocorr.analysis package", "genophenocorr.analysis.predicate package", "genophenocorr.analysis.test_fisherExact module", "genophenocorr.constants package", "genophenocorr.constants.variant_effects module", "genophenocorr.data package", "genophenocorr.model package", "genophenocorr.preprocessing package", "genophenocorr.preprocessing.test_variant module", "genophenocorr.view package", "API reference", "Genotype-Phenotype Correlations in Autosomal Recessive Diseases", "genophenocorr", "Installation", "Tutorial", "Cohort exploratory analysis", "User guide", "Input data"], "terms": {"analysi": [0, 12, 13, 18], "cohortanalysi": [0, 1, 15], "submodul": 0, "constant": [0, 15], "data": [0, 8, 13, 15, 17], "get_toy_cohort": [0, 6, 15], "model": [0, 18], "phenotyp": [0, 1, 2, 7, 8, 10, 13, 16, 18], "featureinfo": [0, 7], "featuretyp": [0, 1, 2, 7], "proteinfeatur": [0, 7], "proteinmetadata": [0, 7, 8], "variantcoordin": [0, 7, 8], "transcriptannot": [0, 7], "variant": [0, 1, 2, 7, 8, 13, 15, 16, 18], "patient": [0, 1, 2, 7, 8, 10, 13, 18], "cohort": [0, 1, 6, 7, 8, 10, 13, 15], "preprocess": [0, 18], "variantcoordinatefind": [0, 8], "functionalannot": [0, 8], "proteinmetadataservic": [0, 8], "patientcr": [0, 8], "phenopacketvariantcoordinatefind": [0, 8], "phenopacketpatientcr": [0, 8, 18], "load_phenopacket_fold": [0, 8], "phenotypecr": [0, 8], "phenotypevalidationexcept": [0, 8], "proteinannotationcach": [0, 8], "protcachingfunctionalannot": [0, 8], "uniprotproteinmetadataservic": [0, 8], "vepfunctionalannot": [0, 8], "variantannotationcach": [0, 8], "varcachingfunctionalannot": [0, 8], "configure_caching_patient_cr": [0, 8, 18], "view": [0, 13, 15], "cohortview": [0, 10], "class": [1, 2, 5, 7, 8, 10, 18], "transcript": [1, 2, 7], "hpo": [1, 2, 7, 8, 10, 13, 15, 18], "recess": [1, 7], "fals": [1, 7, 15], "include_unmeasur": [1, 15], "true": [1, 7, 8], "p_val_correct": 1, "bonferroni": 1, "include_large_sv": 1, "min_perc_patients_w_hpo": 1, "10": [1, 7, 15], "sourc": [1, 2, 5, 6, 7, 8, 10], "base": [1, 2, 5, 7, 8, 10], "object": [1, 2, 7, 8, 10], "thi": [1, 2, 7, 8, 12, 13, 15, 18], "can": [1, 2, 7, 8, 15, 16, 18], "us": [1, 7, 8, 10, 13, 15, 18], "run": [1, 8, 10, 15], "differ": 1, "analys": [1, 15, 18], "given": [1, 7, 8], "analysis_cohort": 1, "The": [1, 2, 7, 8, 13, 18], "type": [1, 7, 8, 13], "analysis_pati": 1, "A": [1, 7, 8, 10, 13, 15], "subset": 1, "structur": [1, 7, 8], "thei": [1, 13], "ar": [1, 2, 7, 12, 13], "being": 1, "remov": 1, "from": [1, 7, 8, 13, 15], "otherwis": [1, 7], "all": [1, 2, 7, 18], "list": [1, 7, 8, 15], "analysis_transcript": 1, "id": [1, 7, 8], "string": [1, 2, 7, 8], "is_recess": 1, "i": [1, 2, 7, 8, 12, 13, 15, 18], "default": [1, 18], "boolean": [1, 7, 8], "we": [1, 2, 8, 13, 15, 16, 18], "want": [1, 15], "assum": 1, "without": [1, 15], "specif": [1, 7, 8, 13], "doe": [1, 8], "have": [1, 2], "onli": [1, 7, 8, 12], "includ": [1, 7, 8], "those": 1, "explicit": 1, "sai": 1, "wa": [1, 2, 7, 8], "observ": [1, 2, 7, 8, 13], "larg": 1, "Will": 1, "test": [1, 2, 7, 10, 12, 13], "least": [1, 8], "percent": [1, 15], "integ": [1, 7], "compare_by_variant_typ": [1, 15], "var_type1": 1, "varianteffect": [1, 2, 4, 5, 15], "var_type2": 1, "option": [1, 7, 18], "fisher": [1, 13], "exact": [1, 2, 12, 13], "find": [1, 7, 8], "ani": [1, 2, 8], "correl": [1, 10, 13], "between": [1, 7, 13], "effect": [1, 7, 8, 16, 18], "variant1": 1, "variant2": 1, "exon1": 1, "exon2": 1, "affect": [1, 7, 12], "exon": [1, 7], "feature1": 1, "feature2": 1, "protein": [1, 2, 7, 8, 13, 18], "featur": [1, 2, 7, 13, 16], "properti": [1, 2, 7, 8], "return": [1, 2, 7, 8, 10], "analyz": [1, 10], "none": [1, 2, 7, 8], "across": 1, "paramet": [1, 7, 8, 10], "If": [1, 7], "compar": [1, 12, 15], "panda": 1, "datafram": 1, "show": [1, 10], "result": [1, 7, 8, 13], "compare_by_vari": 1, "str": [1, 2, 7, 8, 10], "compare_by_exon": 1, "int": [1, 2, 7, 10], "within": [1, 7], "outsid": 1, "compare_by_protein_feature_typ": 1, "compare_by_protein_featur": 1, "predic": 1, "simplepred": [1, 2], "patientcategori": [1, 2], "polypred": [1, 2], "hpopresentpred": [1, 2], "varianteffectpred": [1, 2], "variantpred": [1, 2], "exonpred": [1, 2], "protfeaturepred": [1, 2], "protfeaturetypepred": [1, 2], "test_fisherexact": 1, "modul": [1, 4, 8], "belong": 2, "group": [2, 12], "abstract": [2, 7, 8], "bool": [2, 7, 8], "cat_id": 2, "name": [2, 7, 8], "descript": [2, 7], "gener": [2, 8, 10, 13], "t": [2, 7, 8, 15], "categor": 2, "one": [2, 8, 12], "discret": 2, "categori": 2, "sequenc": [2, 7, 8], "get": [2, 8, 15], "which": [2, 15, 18], "produc": 2, "queri": 2, "assign": 2, "cannot": [2, 7, 8], "meaning": 2, "minimalontologi": [2, 8], "termid": [2, 7], "0": [2, 7, 15], "sampl": 2, "annot": [2, 7, 8, 10, 18], "q": [2, 15, 18], "either": 2, "becaus": [2, 12], "match": 2, "": [2, 7, 8, 15, 18], "descend": 2, "propag": 2, "For": [2, 12, 15, 18], "instanc": [2, 8, 12, 15, 18], "seizur": [2, 15], "had": 2, "clonic": 2, "not_observ": 2, "1": [2, 7, 10, 15], "Not": 2, "particular": 2, "about": [2, 7, 15, 16], "In": [2, 8, 12, 13], "other": [2, 7, 12, 13], "word": [2, 13], "exclud": [2, 7], "an": [2, 5, 7, 8, 10, 12, 13, 18], "ancestor": 2, "did": 2, "impli": 2, "not_measur": 2, "2": [2, 7, 10, 15], "measur": 2, "do": [2, 15], "know": 2, "ha": [2, 7, 8], "overlap": [2, 7], "over": 2, "repres": [2, 7, 13], "egf": 2, "like": [2, 13], "fbn1": 2, "variant_effect": [4, 7], "valu": [5, 7, 15], "enum": [5, 7], "enumer": 5, "transcript_abl": 5, "so": 5, "0001893": 5, "splice_acceptor_vari": 5, "0001574": 5, "splice_donor_vari": 5, "0001575": 5, "stop_gain": 5, "0001587": 5, "frameshift_vari": [5, 15], "0001589": 5, "stop_lost": 5, "0001578": 5, "start_lost": 5, "0002012": 5, "transcript_amplif": 5, "0001889": 5, "inframe_insert": 5, "0001821": 5, "inframe_delet": 5, "0001822": 5, "missense_vari": [5, 15], "0001583": 5, "protein_altering_vari": 5, "0001818": 5, "splice_region_vari": 5, "0001630": 5, "splice_donor_5th_base_vari": 5, "0001787": 5, "splice_donor_region_vari": 5, "0002170": 5, "splice_polypyrimidine_tract_vari": 5, "0002169": 5, "incomplete_terminal_codon_vari": 5, "0001626": 5, "start_retained_vari": 5, "0002019": 5, "stop_retained_vari": 5, "0001567": 5, "synonymous_vari": 5, "0001819": 5, "coding_sequence_vari": 5, "0001580": 5, "mature_mirna_vari": 5, "0001620": 5, "five_prime_utr_vari": 5, "0001623": 5, "three_prime_utr_vari": 5, "0001624": 5, "non_coding_transcript_exon_vari": 5, "0001792": 5, "intron_vari": 5, "0001627": 5, "nmd_transcript_vari": 5, "0001621": 5, "non_coding_transcript_vari": 5, "0001619": 5, "upstream_gene_vari": 5, "0001631": 5, "downstream_gene_vari": 5, "0001632": 5, "tfbs_ablat": 5, "0001895": 5, "tfbs_amplif": 5, "0001892": 5, "tf_binding_site_vari": 5, "0001782": 5, "regulatory_region_abl": 5, "0001894": 5, "regulatory_region_amplif": 5, "0001891": 5, "feature_elong": 5, "0001907": 5, "regulatory_region_vari": 5, "0001566": 5, "feature_trunc": 5, "0001906": 5, "intergenic_vari": 5, "0001628": 5, "sequence_vari": 5, "0001060": 5, "term_id": [7, 8], "identifi": 7, "verifi": 7, "associ": [7, 8, 12, 13], "hpotk": [7, 15, 18], "offici": 7, "respect": 7, "static": [7, 10], "from_term": 7, "term": [7, 8, 10, 13, 18], "minimalterm": 7, "uniqu": 7, "you": [7, 15], "more": [7, 8, 18], "inform": 7, "search": [7, 8], "http": [7, 8, 12, 15, 18], "jax": 7, "org": [7, 8, 15, 18], "app": 7, "describ": [7, 16, 18], "whether": [7, 13], "start": [7, 18], "end": 7, "e": [7, 8, 13, 15], "g": [7, 8, 13, 15], "repeat": 7, "ank": 7, "ankyrin": 7, "domain": 7, "contain": 7, "11": [7, 15], "posit": 7, "coordin": [7, 8, 18], "defin": 7, "avail": [7, 8], "found": [7, 8], "motif": [7, 13], "short": 7, "usual": 7, "than": 7, "20": [7, 15], "amino": 7, "acid": 7, "conserv": 7, "biolog": 7, "signific": 7, "combin": 7, "secondari": 7, "organ": 7, "characterist": 7, "three": [7, 12], "dimension": 7, "fold": 7, "region": 7, "interest": 7, "subsect": 7, "3": [7, 15], "4": [7, 8, 15], "creat": [7, 8, 13], "info": [7, 8, 18], "feature_typ": 7, "protein_id": [7, 8], "label": [7, 10], "protein_featur": 7, "full": 7, "iter": [7, 8], "subgroup": 7, "where": [7, 8], "equal": 7, "chrom": 7, "ref": 7, "alt": 7, "change_length": 7, "genotyp": [7, 10, 13, 18], "represent": [7, 8], "symbol": 7, "breakend": 7, "support": [7, 12, 13], "chromosom": [7, 12], "locat": [7, 13, 18], "allel": [7, 12, 13], "refer": [7, 13], "altern": 7, "chang": 7, "due": 7, "presenc": 7, "heterozyg": 7, "homozyg": 7, "is_structur": 7, "notat": 7, "as_str": 7, "readabl": 7, "contig": 7, "n": [7, 15], "mai": 7, "empti": [7, 8], "gg": 7, "del": 7, "follow": 7, "vcf": 7, "charact": 7, "INS": 7, "me": 7, "sine": 7, "snv": 7, "lead": 7, "length": 7, "zero": 7, "delet": 7, "insert": 7, "duplic": 7, "neg": 7, "check": [7, 8], "chr5": 7, "101": 7, "svtype": 7, "120": 7, "svlen": 7, "oppos": 7, "liter": 7, "nacgtacgtac": 7, "gene_id": 7, "tx_id": 7, "hgvsc": 7, "affected_exon": 7, "affected_protein": 7, "protein_effect_start": 7, "protein_effect_end": 7, "function": [7, 8, 18], "singl": 7, "gene": [7, 12], "transcript_id": 7, "hgvsc_id": 7, "hgv": 7, "code": [7, 10, 13, 15], "dna": 7, "els": 7, "predict": 7, "vep": [7, 8], "overlapping_exon": 7, "protein_affect": 7, "protein_effect_loc": 7, "tupl": [7, 8], "surf1": 7, "refseq": [7, 8], "nm_123456": 7, "7": 7, "c": [7, 13, 15, 18], "9876g": 7, "definit": 7, "www": [7, 8], "sequenceontologi": 7, "alter": 7, "1234": 7, "1235": 7, "var_id": 7, "var_class": 7, "var_coordin": 7, "tx_annot": 7, "variant_coordin": [7, 8], "variant_str": 7, "variant_class": 7, "etc": 7, "create_variant_from_scratch": 7, "variant_id": 7, "gene_nam": 7, "trans_id": 7, "consequ": 7, "exons_effect": 7, "format": [7, 8], "chromosome_start_refer": 7, "chromosome_start_structuraltyp": 7, "requir": [7, 8, 18], "possibl": [7, 12, 13], "hemizyg": 7, "collect": [7, 13], "each": [7, 8, 12, 13], "patient_id": 7, "individu": [7, 12, 13, 17], "patient_set": 7, "set": [7, 13], "phenotype_set": 7, "variant_set": 7, "protein_set": 7, "counts_dict": 7, "size": 7, "from_pati": [7, 18], "member": [7, 16, 18], "all_pati": 7, "frozenset": 7, "frozen": 7, "all_phenotyp": 7, "all_vari": 7, "all_protein": 7, "all_transcript": 7, "total_patient_count": 7, "count": [7, 15, 16], "list_all_pati": [7, 15], "list_all_phenotyp": [7, 15], "top": 7, "highest": 7, "number": [7, 10], "list_all_vari": [7, 15], "list_all_protein": [7, 15], "list_data_by_tx": [7, 15], "counter": 7, "kei": [7, 13], "total": 7, "dictionari": 7, "find_coordin": 8, "item": [8, 10, 15], "determin": 8, "some": [8, 13, 15], "sort": [8, 15], "metaclass": 8, "establish": 8, "metadata": 8, "create_pati": [8, 18], "subject": 8, "genomicinterpret": 8, "phenopacket": [8, 13], "phenotype_cr": 8, "var_func_ann": 8, "pp_directori": 8, "patient_cr": [8, 18], "json": [8, 15, 18], "file": [8, 18], "directori": [8, 18], "path": [8, 18], "folder": [8, 18], "error": 8, "rais": 8, "point": 8, "creator": [8, 18], "turn": 8, "made": 8, "valid": 8, "validationrunn": 8, "create_phenotyp": 8, "satisfi": 8, "issu": [8, 13], "except": 8, "baseexcept": 8, "datadir": 8, "store": [8, 18], "retriev": 8, "pickl": 8, "get_annot": 8, "store_annot": 8, "desir": [8, 12], "load": [8, 15, 18], "storag": 8, "under": 8, "cach": [8, 18], "fallback": 8, "exist": 8, "fuction": 8, "uniprot": [8, 18], "rest": 8, "api": [8, 18], "here": [8, 15, 16, 18], "help": 8, "programmatic_access": 8, "work": [8, 18], "databas": 8, "np_037407": 8, "protein_annot": 8, "predictor": [8, 18], "validation_runn": 8, "cache_dir": 8, "variant_fallback": 8, "protein_fallback": 8, "conveni": [8, 18], "configur": [8, 18], "To": [8, 15], "need": [8, 15, 18], "toolkit": [8, 18], "runner": 8, "fetch": 8, "remot": 8, "should": 8, "current": [8, 18], "case": 8, "non": 8, "parent": 8, "local": 8, "choos": [8, 15], "just": 8, "implement": [8, 12], "moment": 8, "test_vari": 8, "html_row": 10, "anchor": 10, "url": 10, "payload": 10, "hpo_anchor": 10, "hpo_id": 10, "hpo_term_counts_t": 10, "min_count": 10, "html": 10, "design": [10, 13], "displai": 10, "jupyt": [10, 13], "notebook": [10, 13, 15], "ipython": 10, "per": 10, "provid": [10, 13, 18], "explan": 10, "minimum": 10, "tabl": 10, "variants_t": 10, "preferred_transcript": 10, "could": 10, "genophenocorr": [11, 12, 15, 18], "packag": [11, 13, 15], "subpackag": 11, "domin": 12, "simpl": [12, 15], "carri": 12, "while": 12, "would": [12, 13], "principl": 12, "restrict": 12, "homozygot": 12, "who": 12, "two": 12, "ident": 12, "copi": 12, "matern": 12, "patern": 12, "practic": 12, "compound": 12, "heterozygot": 12, "common": 12, "perform": [12, 13, 18], "comparison": 12, "might": 12, "missens": [12, 13, 15], "against": 12, "nons": 12, "nonsens": 12, "see": [12, 13, 15, 18], "exampl": [12, 18], "li": 12, "jt": 12, "et": [12, 13], "al": [12, 13], "2022": 12, "pubm": 12, "ncbi": 12, "nlm": 12, "nih": 12, "gov": 12, "36303223": 12, "suox": 12, "6": [12, 15], "field": 12, "fischer": 12, "kind": 12, "concept": 13, "denot": 13, "attribut": 13, "medic": 13, "context": 13, "deviat": 13, "normal": 13, "morphologi": 13, "physiologi": 13, "behavior": 13, "f": [13, 15, 18], "deep": 13, "precis": 13, "medicin": 13, "question": 13, "biologi": 13, "human": [13, 18], "genet": 13, "concern": 13, "relationship": 13, "abnorm": 13, "mendelian": 13, "focu": 13, "place": 13, "studi": 13, "diseas": 13, "caus": 13, "manifest": 13, "python": 13, "input": [13, 15, 17], "global": 13, "allianc": 13, "genom": 13, "health": 13, "ga4gh": 13, "ingest": 13, "v": [13, 15], "prematur": 13, "termin": 13, "codon": 13, "ontologi": [13, 15, 18], "statist": 13, "report": 13, "recommend": 13, "user": 13, "document": 13, "detail": 13, "tutori": 13, "It": 13, "also": 13, "librari": [13, 18], "todo": [13, 14, 15, 16, 17, 18], "pypi": 13, "read": [13, 18], "background": 13, "orgogozo": 13, "2015": 13, "differenti": 13, "instruct": 13, "how": [13, 16, 18], "up": 13, "best": 13, "leav": 13, "ask": 13, "bug": 13, "tracker": 13, "pip": 14, "let": [15, 18], "import": [15, 18], "obograph": 15, "load_minimal_ontologi": [15, 18], "hp": [15, 18], "toi": 15, "latest": 15, "purl": [15, 18], "obolibrari": [15, 18], "obo": [15, 18], "move": 15, "workflow": 15, "now": [15, 18], "keep": 15, "thing": 15, "ship": 15, "section": [15, 17, 18], "learn": [15, 16], "your": 15, "command": 15, "b": 15, "d": 15, "h": 15, "j": 15, "k": 15, "l": 15, "m": 15, "o": [15, 18], "p": 15, "r": 15, "u": 15, "w": 15, "x": 15, "y": 15, "z": 15, "0001166": 15, "14": 15, "0001250": 15, "0001257": 15, "17": 15, "hetvar1": 15, "13": 15, "hetvar2": 15, "homvar1": 15, "homvar2": 15, "np_09876": 15, "5": [15, 18], "26": 15, "tx_dict": 15, "nm_1234": 15, "what": [15, 16], "frameshift": 15, "pprint": 15, "cohort_analysi": 15, "With": 15, "arachnodactyli": 15, "30": 15, "77": 15, "76": 15, "92": 15, "04718": 15, "84": 15, "62": 15, "9": 15, "69": 15, "23": 15, "64472": 15, "spastic": 15, "8": 15, "61": 15, "54": 15, "00000": 15, "81": 15, "25": 15, "00": 15, "000781": 15, "68": 15, "75": 15, "60": 15, "692449": 15, "12": 15, "80": 15, "000000": 15, "write": [16, 17], "mani": 16, "distribut": 16, "plot": 16, "high": 17, "level": 17, "overview": 17, "bridg": 17, "standard": 18, "form": 18, "consist": 18, "hold": 18, "dataset": 18, "explor": 18, "exploratori": 18, "first": 18, "step": 18, "involv": 18, "prepar": 18, "downstream": 18, "easiest": 18, "wai": 18, "schema": 18, "out": 18, "box": 18, "requisit": 18, "ll": 18, "amaz": 18, "instal": 18, "along": 18, "next": 18, "method": 18, "By": 18, "call": 18, "respons": 18, "save": 18, "bandwidth": 18, "purpos": 18, "simple_cohort": 18, "repositori": 18, "simple_cohort_path": 18, "walk": 18, "system": 18, "transform": 18, "googl": 18, "protobuf": 18, "json_format": 18, "pars": 18, "dirpath": 18, "_": 18, "filenam": 18, "endswith": 18, "pp_path": 18, "join": 18, "open": 18, "fh": 18, "pp": 18, "append": 18, "len": 18, "construct": 18, "raw": 18}, "objects": {"": [[0, 0, 0, "-", "genophenocorr"]], "genophenocorr": [[1, 0, 0, "-", "analysis"], [4, 0, 0, "-", "constants"], [6, 0, 0, "-", "data"], [7, 0, 0, "-", "model"], [8, 0, 0, "-", "preprocessing"], [10, 0, 0, "-", "view"]], "genophenocorr.analysis": [[1, 1, 1, "", "CohortAnalysis"], [2, 0, 0, "-", "predicate"]], "genophenocorr.analysis.CohortAnalysis": [[1, 2, 1, "id0", "analysis_cohort"], [1, 2, 1, "id1", "analysis_patients"], [1, 2, 1, "id2", "analysis_transcript"], [1, 3, 1, "", "compare_by_exon"], [1, 3, 1, "", "compare_by_protein_feature"], [1, 3, 1, "", "compare_by_protein_feature_type"], [1, 3, 1, "", "compare_by_variant"], [1, 3, 1, "", "compare_by_variant_type"], [1, 2, 1, "id5", "include_large_SV"], [1, 2, 1, "id4", "include_unmeasured"], [1, 2, 1, "id3", "is_recessive"], [1, 2, 1, "id6", "min_perc_patients_w_hpo"]], "genophenocorr.analysis.predicate": [[2, 1, 1, "", "ExonPredicate"], [2, 1, 1, "", "HPOPresentPredicate"], [2, 1, 1, "", "PatientCategory"], [2, 1, 1, "", "PolyPredicate"], [2, 1, 1, "", "ProtFeaturePredicate"], [2, 1, 1, "", "ProtFeatureTypePredicate"], [2, 1, 1, "", "SimplePredicate"], [2, 1, 1, "", "VariantEffectPredicate"], [2, 1, 1, "", "VariantPredicate"]], "genophenocorr.analysis.predicate.ExonPredicate": [[2, 3, 1, "", "categories"], [2, 3, 1, "", "test"]], "genophenocorr.analysis.predicate.HPOPresentPredicate": [[2, 4, 1, "", "NOT_MEASURED"], [2, 4, 1, "", "NOT_OBSERVED"], [2, 4, 1, "", "OBSERVED"], [2, 3, 1, "", "categories"], [2, 3, 1, "", "test"]], "genophenocorr.analysis.predicate.PatientCategory": [[2, 2, 1, "", "cat_id"], [2, 2, 1, "", "description"], [2, 2, 1, "", "name"]], "genophenocorr.analysis.predicate.PolyPredicate": [[2, 2, 1, "", "categories"], [2, 3, 1, "", "test"]], "genophenocorr.analysis.predicate.ProtFeaturePredicate": [[2, 3, 1, "", "categories"], [2, 3, 1, "", "test"]], "genophenocorr.analysis.predicate.ProtFeatureTypePredicate": [[2, 3, 1, "", "categories"], [2, 3, 1, "", "test"]], "genophenocorr.analysis.predicate.SimplePredicate": [[2, 3, 1, "", "test"]], "genophenocorr.analysis.predicate.VariantEffectPredicate": [[2, 3, 1, "", "categories"], [2, 3, 1, "", "test"]], "genophenocorr.analysis.predicate.VariantPredicate": [[2, 3, 1, "", "categories"], [2, 3, 1, "", "test"]], "genophenocorr.constants": [[5, 0, 0, "-", "variant_effects"]], "genophenocorr.constants.variant_effects": [[5, 1, 1, "", "VariantEffect"]], "genophenocorr.constants.variant_effects.VariantEffect": [[5, 4, 1, "", "CODING_SEQUENCE_VARIANT"], [5, 4, 1, "", "DOWNSTREAM_GENE_VARIANT"], [5, 4, 1, "", "FEATURE_ELONGATION"], [5, 4, 1, "", "FEATURE_TRUNCATION"], [5, 4, 1, "", "FIVE_PRIME_UTR_VARIANT"], [5, 4, 1, "", "FRAMESHIFT_VARIANT"], [5, 4, 1, "", "INCOMPLETE_TERMINAL_CODON_VARIANT"], [5, 4, 1, "", "INFRAME_DELETION"], [5, 4, 1, "", "INFRAME_INSERTION"], [5, 4, 1, "", "INTERGENIC_VARIANT"], [5, 4, 1, "", "INTRON_VARIANT"], [5, 4, 1, "", "MATURE_MIRNA_VARIANT"], [5, 4, 1, "", "MISSENSE_VARIANT"], [5, 4, 1, "", "NMD_TRANSCRIPT_VARIANT"], [5, 4, 1, "", "NON_CODING_TRANSCRIPT_EXON_VARIANT"], [5, 4, 1, "", "NON_CODING_TRANSCRIPT_VARIANT"], [5, 4, 1, "", "PROTEIN_ALTERING_VARIANT"], [5, 4, 1, "", "REGULATORY_REGION_ABLATION"], [5, 4, 1, "", "REGULATORY_REGION_AMPLIFICATION"], [5, 4, 1, "", "REGULATORY_REGION_VARIANT"], [5, 4, 1, "", "SEQUENCE_VARIANT"], [5, 4, 1, "", "SPLICE_ACCEPTOR_VARIANT"], [5, 4, 1, "", "SPLICE_DONOR_5TH_BASE_VARIANT"], [5, 4, 1, "", "SPLICE_DONOR_REGION_VARIANT"], [5, 4, 1, "", "SPLICE_DONOR_VARIANT"], [5, 4, 1, "", "SPLICE_POLYPYRIMIDINE_TRACT_VARIANT"], [5, 4, 1, "", "SPLICE_REGION_VARIANT"], [5, 4, 1, "", "START_LOST"], [5, 4, 1, "", "START_RETAINED_VARIANT"], [5, 4, 1, "", "STOP_GAINED"], [5, 4, 1, "", "STOP_LOST"], [5, 4, 1, "", "STOP_RETAINED_VARIANT"], [5, 4, 1, "", "SYNONYMOUS_VARIANT"], [5, 4, 1, "", "TFBS_ABLATION"], [5, 4, 1, "", "TFBS_AMPLIFICATION"], [5, 4, 1, "", "TF_BINDING_SITE_VARIANT"], [5, 4, 1, "", "THREE_PRIME_UTR_VARIANT"], [5, 4, 1, "", "TRANSCRIPT_ABLATION"], [5, 4, 1, "", "TRANSCRIPT_AMPLIFICATION"], [5, 4, 1, "", "UPSTREAM_GENE_VARIANT"]], "genophenocorr.data": [[6, 5, 1, "", "get_toy_cohort"]], "genophenocorr.model": [[7, 1, 1, "", "Cohort"], [7, 1, 1, "", "FeatureInfo"], [7, 1, 1, "", "FeatureType"], [7, 1, 1, "", "Patient"], [7, 1, 1, "", "Phenotype"], [7, 1, 1, "", "ProteinFeature"], [7, 1, 1, "", "ProteinMetadata"], [7, 1, 1, "", "TranscriptAnnotation"], [7, 1, 1, "", "Variant"], [7, 1, 1, "", "VariantCoordinates"]], "genophenocorr.model.Cohort": [[7, 2, 1, "", "all_patients"], [7, 2, 1, "", "all_phenotypes"], [7, 2, 1, "", "all_proteins"], [7, 2, 1, "", "all_transcripts"], [7, 2, 1, "", "all_variants"], [7, 3, 1, "", "from_patients"], [7, 3, 1, "", "list_all_patients"], [7, 3, 1, "", "list_all_phenotypes"], [7, 3, 1, "", "list_all_proteins"], [7, 3, 1, "", "list_all_variants"], [7, 3, 1, "", "list_data_by_tx"], [7, 2, 1, "", "total_patient_count"]], "genophenocorr.model.FeatureInfo": [[7, 2, 1, "id5", "end"], [7, 2, 1, "id3", "name"], [7, 2, 1, "id4", "start"]], "genophenocorr.model.FeatureType": [[7, 4, 1, "id8", "DOMAIN"], [7, 4, 1, "id7", "MOTIF"], [7, 4, 1, "id9", "REGION"], [7, 4, 1, "id6", "REPEAT"]], "genophenocorr.model.Patient": [[7, 2, 1, "id38", "patient_id"], [7, 2, 1, "id39", "phenotypes"], [7, 2, 1, "id41", "proteins"], [7, 2, 1, "id40", "variants"]], "genophenocorr.model.Phenotype": [[7, 3, 1, "", "from_term"], [7, 2, 1, "id0", "identifier"], [7, 2, 1, "id1", "name"], [7, 2, 1, "id2", "observed"]], "genophenocorr.model.ProteinFeature": [[7, 3, 1, "", "create"], [7, 2, 1, "", "feature_type"], [7, 2, 1, "", "info"]], "genophenocorr.model.ProteinMetadata": [[7, 3, 1, "id13", "domains"], [7, 2, 1, "id11", "label"], [7, 3, 1, "id16", "motifs"], [7, 2, 1, "id12", "protein_features"], [7, 2, 1, "id10", "protein_id"], [7, 3, 1, "id15", "regions"], [7, 3, 1, "id14", "repeats"]], "genophenocorr.model.TranscriptAnnotation": [[7, 2, 1, "id26", "gene_id"], [7, 2, 1, "id28", "hgvsc_id"], [7, 2, 1, "id30", "overlapping_exons"], [7, 2, 1, "id31", "protein_affected"], [7, 2, 1, "id32", "protein_effect_location"], [7, 2, 1, "id27", "transcript_id"], [7, 2, 1, "id29", "variant_effects"]], "genophenocorr.model.Variant": [[7, 3, 1, "", "create_variant_from_scratch"], [7, 2, 1, "id35", "genotype"], [7, 2, 1, "id36", "tx_annotations"], [7, 2, 1, "id37", "variant_class"], [7, 2, 1, "id33", "variant_coordinates"], [7, 2, 1, "id34", "variant_string"]], "genophenocorr.model.VariantCoordinates": [[7, 2, 1, "id21", "alt"], [7, 3, 1, "id25", "as_string"], [7, 2, 1, "id22", "change_length"], [7, 2, 1, "id17", "chrom"], [7, 2, 1, "id19", "end"], [7, 2, 1, "id23", "genotype"], [7, 3, 1, "id24", "is_structural"], [7, 2, 1, "id20", "ref"], [7, 2, 1, "id18", "start"]], "genophenocorr.preprocessing": [[8, 1, 1, "", "FunctionalAnnotator"], [8, 1, 1, "", "PatientCreator"], [8, 1, 1, "", "PhenopacketPatientCreator"], [8, 1, 1, "", "PhenopacketVariantCoordinateFinder"], [8, 1, 1, "", "PhenotypeCreator"], [8, 6, 1, "", "PhenotypeValidationException"], [8, 1, 1, "", "ProtCachingFunctionalAnnotator"], [8, 1, 1, "", "ProteinAnnotationCache"], [8, 1, 1, "", "ProteinMetadataService"], [8, 1, 1, "", "UniprotProteinMetadataService"], [8, 1, 1, "", "VarCachingFunctionalAnnotator"], [8, 1, 1, "", "VariantAnnotationCache"], [8, 1, 1, "", "VariantCoordinateFinder"], [8, 1, 1, "", "VepFunctionalAnnotator"], [8, 5, 1, "", "configure_caching_patient_creator"], [8, 5, 1, "", "load_phenopacket_folder"]], "genophenocorr.preprocessing.FunctionalAnnotator": [[8, 3, 1, "", "annotate"]], "genophenocorr.preprocessing.PatientCreator": [[8, 3, 1, "", "create_patient"]], "genophenocorr.preprocessing.PhenopacketPatientCreator": [[8, 3, 1, "", "create_patient"]], "genophenocorr.preprocessing.PhenopacketVariantCoordinateFinder": [[8, 3, 1, "", "find_coordinates"]], "genophenocorr.preprocessing.PhenotypeCreator": [[8, 3, 1, "", "create_phenotype"]], "genophenocorr.preprocessing.PhenotypeValidationException": [[8, 2, 1, "", "issues"]], "genophenocorr.preprocessing.ProtCachingFunctionalAnnotator": [[8, 3, 1, "", "annotate"]], "genophenocorr.preprocessing.ProteinAnnotationCache": [[8, 3, 1, "", "get_annotations"], [8, 3, 1, "", "store_annotations"]], "genophenocorr.preprocessing.ProteinMetadataService": [[8, 3, 1, "", "annotate"]], "genophenocorr.preprocessing.UniprotProteinMetadataService": [[8, 3, 1, "", "annotate"]], "genophenocorr.preprocessing.VarCachingFunctionalAnnotator": [[8, 3, 1, "", "annotate"]], "genophenocorr.preprocessing.VariantAnnotationCache": [[8, 3, 1, "", "get_annotations"], [8, 3, 1, "", "store_annotations"]], "genophenocorr.preprocessing.VariantCoordinateFinder": [[8, 3, 1, "", "find_coordinates"]], "genophenocorr.preprocessing.VepFunctionalAnnotator": [[8, 3, 1, "", "annotate"]], "genophenocorr.view": [[10, 1, 1, "", "CohortViewer"]], "genophenocorr.view.CohortViewer": [[10, 3, 1, "", "anchor"], [10, 3, 1, "", "hpo_anchor"], [10, 3, 1, "", "hpo_term_counts_table"], [10, 3, 1, "", "html_row"], [10, 3, 1, "", "variants_table"]]}, "objtypes": {"0": "py:module", "1": "py:class", "2": "py:property", "3": "py:method", "4": "py:attribute", "5": "py:function", "6": "py:exception"}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "class", "Python class"], "2": ["py", "property", "Python property"], "3": ["py", "method", "Python method"], "4": ["py", "attribute", "Python attribute"], "5": ["py", "function", "Python function"], "6": ["py", "exception", "Python exception"]}, "titleterms": {"genophenocorr": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 13, 14], "packag": [0, 1, 2, 4, 6, 7, 8, 10, 14], "subpackag": [0, 1], "analysi": [1, 2, 3, 15, 16], "submodul": [1, 4, 8], "predic": 2, "test_fisherexact": 3, "modul": [3, 5, 9], "constant": [4, 5], "variant_effect": 5, "data": [6, 18], "model": 7, "preprocess": [8, 9], "test_vari": 9, "view": 10, "api": 11, "refer": 11, "genotyp": 12, "phenotyp": 12, "correl": 12, "autosom": 12, "recess": 12, "diseas": 12, "tutori": [12, 15], "literatur": 13, "instal": [13, 14], "feedback": 13, "python": 14, "set": 15, "up": 15, "prepar": 15, "sampl": 15, "cohort": [16, 18], "exploratori": 16, "user": 17, "guid": 17, "content": 17, "input": 18, "creat": 18, "from": 18, "ga4gh": 18, "phenopacket": 18, "other": 18}, "envversion": {"sphinx.domains.c": 3, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 9, "sphinx.domains.index": 1, "sphinx.domains.javascript": 3, "sphinx.domains.math": 2, "sphinx.domains.python": 4, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.viewcode": 1, "sphinx.ext.intersphinx": 1, "sphinx": 60}, "alltitles": {"genophenocorr package": [[0, "module-genophenocorr"]], "Subpackages": [[0, "subpackages"], [1, "subpackages"]], "genophenocorr.analysis package": [[1, "module-genophenocorr.analysis"]], "Submodules": [[1, "submodules"], [4, "submodules"], [8, "submodules"]], "genophenocorr.analysis.predicate package": [[2, "module-genophenocorr.analysis.predicate"]], "genophenocorr.analysis.test_fisherExact module": [[3, "genophenocorr-analysis-test-fisherexact-module"]], "genophenocorr.constants package": [[4, "module-genophenocorr.constants"]], "genophenocorr.constants.variant_effects module": [[5, "module-genophenocorr.constants.variant_effects"]], "genophenocorr.data package": [[6, "module-genophenocorr.data"]], "genophenocorr.model package": [[7, "module-genophenocorr.model"]], "genophenocorr.preprocessing package": [[8, "module-genophenocorr.preprocessing"]], "genophenocorr.preprocessing.test_variant module": [[9, "genophenocorr-preprocessing-test-variant-module"]], "genophenocorr.view package": [[10, "module-genophenocorr.view"]], "API reference": [[11, "api-reference"]], "Genotype-Phenotype Correlations in Autosomal Recessive Diseases": [[12, "genotype-phenotype-correlations-in-autosomal-recessive-diseases"]], "Tutorial": [[12, "tutorial"], [15, "tutorial"]], "genophenocorr": [[13, "genophenocorr"]], "Literature": [[13, "literature"]], "Installation": [[13, "installation"], [14, "installation"]], "Feedback": [[13, "feedback"]], "Installation of genophenocorr as a Python package": [[14, "installation-of-genophenocorr-as-a-python-package"]], "Set up analysis": [[15, "set-up-analysis"]], "Prepare samples": [[15, "prepare-samples"]], "Cohort exploratory analysis": [[16, "cohort-exploratory-analysis"]], "User guide": [[17, "user-guide"]], "Contents:": [[17, null]], "Input data": [[18, "input-data"]], "Create a cohort from GA4GH phenopackets": [[18, "create-a-cohort-from-ga4gh-phenopackets"]], "Create a cohort from other data": [[18, "create-a-cohort-from-other-data"]]}, "indexentries": {"genophenocorr": [[0, "module-genophenocorr"]], "module": [[0, "module-genophenocorr"], [1, "module-genophenocorr.analysis"], [2, "module-genophenocorr.analysis.predicate"], [4, "module-genophenocorr.constants"], [5, "module-genophenocorr.constants.variant_effects"], [6, "module-genophenocorr.data"], [7, "module-genophenocorr.model"], [8, "module-genophenocorr.preprocessing"], [10, "module-genophenocorr.view"]], "cohortanalysis (class in genophenocorr.analysis)": [[1, "genophenocorr.analysis.CohortAnalysis"]], "analysis_cohort (genophenocorr.analysis.cohortanalysis attribute)": [[1, "genophenocorr.analysis.CohortAnalysis.analysis_cohort"]], "analysis_cohort (genophenocorr.analysis.cohortanalysis property)": [[1, "id0"]], "analysis_patients (genophenocorr.analysis.cohortanalysis attribute)": [[1, "genophenocorr.analysis.CohortAnalysis.analysis_patients"]], "analysis_patients (genophenocorr.analysis.cohortanalysis property)": [[1, "id1"]], "analysis_transcript (genophenocorr.analysis.cohortanalysis attribute)": [[1, "genophenocorr.analysis.CohortAnalysis.analysis_transcript"]], "analysis_transcript (genophenocorr.analysis.cohortanalysis property)": [[1, "id2"]], "compare_by_exon() (genophenocorr.analysis.cohortanalysis method)": [[1, "genophenocorr.analysis.CohortAnalysis.compare_by_exon"]], "compare_by_protein_feature() (genophenocorr.analysis.cohortanalysis method)": [[1, "genophenocorr.analysis.CohortAnalysis.compare_by_protein_feature"]], "compare_by_protein_feature_type() (genophenocorr.analysis.cohortanalysis method)": [[1, "genophenocorr.analysis.CohortAnalysis.compare_by_protein_feature_type"]], "compare_by_variant() (genophenocorr.analysis.cohortanalysis method)": [[1, "genophenocorr.analysis.CohortAnalysis.compare_by_variant"]], "compare_by_variant_type() (genophenocorr.analysis.cohortanalysis method)": [[1, "genophenocorr.analysis.CohortAnalysis.compare_by_variant_type"]], "genophenocorr.analysis": [[1, "module-genophenocorr.analysis"]], "include_large_sv (genophenocorr.analysis.cohortanalysis attribute)": [[1, "genophenocorr.analysis.CohortAnalysis.include_large_SV"]], "include_large_sv (genophenocorr.analysis.cohortanalysis property)": [[1, "id5"]], "include_unmeasured (genophenocorr.analysis.cohortanalysis attribute)": [[1, "genophenocorr.analysis.CohortAnalysis.include_unmeasured"]], "include_unmeasured (genophenocorr.analysis.cohortanalysis property)": [[1, "id4"]], "is_recessive (genophenocorr.analysis.cohortanalysis attribute)": [[1, "genophenocorr.analysis.CohortAnalysis.is_recessive"]], "is_recessive (genophenocorr.analysis.cohortanalysis property)": [[1, "id3"]], "min_perc_patients_w_hpo (genophenocorr.analysis.cohortanalysis attribute)": [[1, "genophenocorr.analysis.CohortAnalysis.min_perc_patients_w_hpo"]], "min_perc_patients_w_hpo (genophenocorr.analysis.cohortanalysis property)": [[1, "id6"]], "exonpredicate (class in genophenocorr.analysis.predicate)": [[2, "genophenocorr.analysis.predicate.ExonPredicate"]], "hpopresentpredicate (class in genophenocorr.analysis.predicate)": [[2, "genophenocorr.analysis.predicate.HPOPresentPredicate"]], "not_measured (genophenocorr.analysis.predicate.hpopresentpredicate attribute)": [[2, "genophenocorr.analysis.predicate.HPOPresentPredicate.NOT_MEASURED"]], "not_observed (genophenocorr.analysis.predicate.hpopresentpredicate attribute)": [[2, "genophenocorr.analysis.predicate.HPOPresentPredicate.NOT_OBSERVED"]], "observed (genophenocorr.analysis.predicate.hpopresentpredicate attribute)": [[2, "genophenocorr.analysis.predicate.HPOPresentPredicate.OBSERVED"]], "patientcategory (class in genophenocorr.analysis.predicate)": [[2, "genophenocorr.analysis.predicate.PatientCategory"]], "polypredicate (class in genophenocorr.analysis.predicate)": [[2, "genophenocorr.analysis.predicate.PolyPredicate"]], "protfeaturepredicate (class in genophenocorr.analysis.predicate)": [[2, "genophenocorr.analysis.predicate.ProtFeaturePredicate"]], "protfeaturetypepredicate (class in genophenocorr.analysis.predicate)": [[2, "genophenocorr.analysis.predicate.ProtFeatureTypePredicate"]], "simplepredicate (class in genophenocorr.analysis.predicate)": [[2, "genophenocorr.analysis.predicate.SimplePredicate"]], "varianteffectpredicate (class in genophenocorr.analysis.predicate)": [[2, "genophenocorr.analysis.predicate.VariantEffectPredicate"]], "variantpredicate (class in genophenocorr.analysis.predicate)": [[2, "genophenocorr.analysis.predicate.VariantPredicate"]], "cat_id (genophenocorr.analysis.predicate.patientcategory property)": [[2, "genophenocorr.analysis.predicate.PatientCategory.cat_id"]], "categories (genophenocorr.analysis.predicate.polypredicate property)": [[2, "genophenocorr.analysis.predicate.PolyPredicate.categories"]], "categories() (genophenocorr.analysis.predicate.exonpredicate method)": [[2, "genophenocorr.analysis.predicate.ExonPredicate.categories"]], "categories() (genophenocorr.analysis.predicate.hpopresentpredicate method)": [[2, "genophenocorr.analysis.predicate.HPOPresentPredicate.categories"]], "categories() (genophenocorr.analysis.predicate.protfeaturepredicate method)": [[2, "genophenocorr.analysis.predicate.ProtFeaturePredicate.categories"]], "categories() (genophenocorr.analysis.predicate.protfeaturetypepredicate method)": [[2, "genophenocorr.analysis.predicate.ProtFeatureTypePredicate.categories"]], "categories() (genophenocorr.analysis.predicate.varianteffectpredicate method)": [[2, "genophenocorr.analysis.predicate.VariantEffectPredicate.categories"]], "categories() (genophenocorr.analysis.predicate.variantpredicate method)": [[2, "genophenocorr.analysis.predicate.VariantPredicate.categories"]], "description (genophenocorr.analysis.predicate.patientcategory property)": [[2, "genophenocorr.analysis.predicate.PatientCategory.description"]], "genophenocorr.analysis.predicate": [[2, "module-genophenocorr.analysis.predicate"]], "name (genophenocorr.analysis.predicate.patientcategory property)": [[2, "genophenocorr.analysis.predicate.PatientCategory.name"]], "test() (genophenocorr.analysis.predicate.exonpredicate method)": [[2, "genophenocorr.analysis.predicate.ExonPredicate.test"]], "test() (genophenocorr.analysis.predicate.hpopresentpredicate method)": [[2, "genophenocorr.analysis.predicate.HPOPresentPredicate.test"]], "test() (genophenocorr.analysis.predicate.polypredicate method)": [[2, "genophenocorr.analysis.predicate.PolyPredicate.test"]], "test() (genophenocorr.analysis.predicate.protfeaturepredicate method)": [[2, "genophenocorr.analysis.predicate.ProtFeaturePredicate.test"]], "test() (genophenocorr.analysis.predicate.protfeaturetypepredicate method)": [[2, "genophenocorr.analysis.predicate.ProtFeatureTypePredicate.test"]], "test() (genophenocorr.analysis.predicate.simplepredicate method)": [[2, "genophenocorr.analysis.predicate.SimplePredicate.test"]], "test() (genophenocorr.analysis.predicate.varianteffectpredicate method)": [[2, "genophenocorr.analysis.predicate.VariantEffectPredicate.test"]], "test() (genophenocorr.analysis.predicate.variantpredicate method)": [[2, "genophenocorr.analysis.predicate.VariantPredicate.test"]], "genophenocorr.constants": [[4, "module-genophenocorr.constants"]], "coding_sequence_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.CODING_SEQUENCE_VARIANT"]], "downstream_gene_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.DOWNSTREAM_GENE_VARIANT"]], "feature_elongation (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.FEATURE_ELONGATION"]], "feature_truncation (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.FEATURE_TRUNCATION"]], "five_prime_utr_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.FIVE_PRIME_UTR_VARIANT"]], "frameshift_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.FRAMESHIFT_VARIANT"]], "incomplete_terminal_codon_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.INCOMPLETE_TERMINAL_CODON_VARIANT"]], "inframe_deletion (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.INFRAME_DELETION"]], "inframe_insertion (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.INFRAME_INSERTION"]], "intergenic_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.INTERGENIC_VARIANT"]], "intron_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.INTRON_VARIANT"]], "mature_mirna_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.MATURE_MIRNA_VARIANT"]], "missense_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.MISSENSE_VARIANT"]], "nmd_transcript_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.NMD_TRANSCRIPT_VARIANT"]], "non_coding_transcript_exon_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.NON_CODING_TRANSCRIPT_EXON_VARIANT"]], "non_coding_transcript_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.NON_CODING_TRANSCRIPT_VARIANT"]], "protein_altering_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.PROTEIN_ALTERING_VARIANT"]], "regulatory_region_ablation (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.REGULATORY_REGION_ABLATION"]], "regulatory_region_amplification (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.REGULATORY_REGION_AMPLIFICATION"]], "regulatory_region_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.REGULATORY_REGION_VARIANT"]], "sequence_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.SEQUENCE_VARIANT"]], "splice_acceptor_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.SPLICE_ACCEPTOR_VARIANT"]], "splice_donor_5th_base_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.SPLICE_DONOR_5TH_BASE_VARIANT"]], "splice_donor_region_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.SPLICE_DONOR_REGION_VARIANT"]], "splice_donor_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.SPLICE_DONOR_VARIANT"]], "splice_polypyrimidine_tract_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.SPLICE_POLYPYRIMIDINE_TRACT_VARIANT"]], "splice_region_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.SPLICE_REGION_VARIANT"]], "start_lost (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.START_LOST"]], "start_retained_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.START_RETAINED_VARIANT"]], "stop_gained (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.STOP_GAINED"]], "stop_lost (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.STOP_LOST"]], "stop_retained_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.STOP_RETAINED_VARIANT"]], "synonymous_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.SYNONYMOUS_VARIANT"]], "tfbs_ablation (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.TFBS_ABLATION"]], "tfbs_amplification (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.TFBS_AMPLIFICATION"]], "tf_binding_site_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.TF_BINDING_SITE_VARIANT"]], "three_prime_utr_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.THREE_PRIME_UTR_VARIANT"]], "transcript_ablation (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.TRANSCRIPT_ABLATION"]], "transcript_amplification (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.TRANSCRIPT_AMPLIFICATION"]], "upstream_gene_variant (genophenocorr.constants.variant_effects.varianteffect attribute)": [[5, "genophenocorr.constants.variant_effects.VariantEffect.UPSTREAM_GENE_VARIANT"]], "varianteffect (class in genophenocorr.constants.variant_effects)": [[5, "genophenocorr.constants.variant_effects.VariantEffect"]], "genophenocorr.constants.variant_effects": [[5, "module-genophenocorr.constants.variant_effects"]], "genophenocorr.data": [[6, "module-genophenocorr.data"]], "get_toy_cohort() (in module genophenocorr.data)": [[6, "genophenocorr.data.get_toy_cohort"]], "cohort (class in genophenocorr.model)": [[7, "genophenocorr.model.Cohort"]], "domain (genophenocorr.model.featuretype attribute)": [[7, "genophenocorr.model.FeatureType.DOMAIN"], [7, "id8"]], "featureinfo (class in genophenocorr.model)": [[7, "genophenocorr.model.FeatureInfo"]], "featuretype (class in genophenocorr.model)": [[7, "genophenocorr.model.FeatureType"]], "motif (genophenocorr.model.featuretype attribute)": [[7, "genophenocorr.model.FeatureType.MOTIF"], [7, "id7"]], "patient (class in genophenocorr.model)": [[7, "genophenocorr.model.Patient"]], "phenotype (class in genophenocorr.model)": [[7, "genophenocorr.model.Phenotype"]], "proteinfeature (class in genophenocorr.model)": [[7, "genophenocorr.model.ProteinFeature"]], "proteinmetadata (class in genophenocorr.model)": [[7, "genophenocorr.model.ProteinMetadata"]], "region (genophenocorr.model.featuretype attribute)": [[7, "genophenocorr.model.FeatureType.REGION"], [7, "id9"]], "repeat (genophenocorr.model.featuretype attribute)": [[7, "genophenocorr.model.FeatureType.REPEAT"], [7, "id6"]], "transcriptannotation (class in genophenocorr.model)": [[7, "genophenocorr.model.TranscriptAnnotation"]], "variant (class in genophenocorr.model)": [[7, "genophenocorr.model.Variant"]], "variantcoordinates (class in genophenocorr.model)": [[7, "genophenocorr.model.VariantCoordinates"]], "all_patients (genophenocorr.model.cohort property)": [[7, "genophenocorr.model.Cohort.all_patients"]], "all_phenotypes (genophenocorr.model.cohort property)": [[7, "genophenocorr.model.Cohort.all_phenotypes"]], "all_proteins (genophenocorr.model.cohort property)": [[7, "genophenocorr.model.Cohort.all_proteins"]], "all_transcripts (genophenocorr.model.cohort property)": [[7, "genophenocorr.model.Cohort.all_transcripts"]], "all_variants (genophenocorr.model.cohort property)": [[7, "genophenocorr.model.Cohort.all_variants"]], "alt (genophenocorr.model.variantcoordinates attribute)": [[7, "genophenocorr.model.VariantCoordinates.alt"]], "alt (genophenocorr.model.variantcoordinates property)": [[7, "id21"]], "as_string() (genophenocorr.model.variantcoordinates method)": [[7, "genophenocorr.model.VariantCoordinates.as_string"], [7, "id25"]], "change_length (genophenocorr.model.variantcoordinates attribute)": [[7, "genophenocorr.model.VariantCoordinates.change_length"]], "change_length (genophenocorr.model.variantcoordinates property)": [[7, "id22"]], "chrom (genophenocorr.model.variantcoordinates attribute)": [[7, "genophenocorr.model.VariantCoordinates.chrom"]], "chrom (genophenocorr.model.variantcoordinates property)": [[7, "id17"]], "create() (genophenocorr.model.proteinfeature static method)": [[7, "genophenocorr.model.ProteinFeature.create"]], "create_variant_from_scratch() (genophenocorr.model.variant static method)": [[7, "genophenocorr.model.Variant.create_variant_from_scratch"]], "domains() (genophenocorr.model.proteinmetadata method)": [[7, "genophenocorr.model.ProteinMetadata.domains"], [7, "id13"]], "end (genophenocorr.model.featureinfo attribute)": [[7, "genophenocorr.model.FeatureInfo.end"]], "end (genophenocorr.model.featureinfo property)": [[7, "id5"]], "end (genophenocorr.model.variantcoordinates attribute)": [[7, "genophenocorr.model.VariantCoordinates.end"]], "end (genophenocorr.model.variantcoordinates property)": [[7, "id19"]], "feature_type (genophenocorr.model.proteinfeature property)": [[7, "genophenocorr.model.ProteinFeature.feature_type"]], "from_patients() (genophenocorr.model.cohort static method)": [[7, "genophenocorr.model.Cohort.from_patients"]], "from_term() (genophenocorr.model.phenotype static method)": [[7, "genophenocorr.model.Phenotype.from_term"]], "gene_id (genophenocorr.model.transcriptannotation attribute)": [[7, "genophenocorr.model.TranscriptAnnotation.gene_id"]], "gene_id (genophenocorr.model.transcriptannotation property)": [[7, "id26"]], "genophenocorr.model": [[7, "module-genophenocorr.model"]], "genotype (genophenocorr.model.variant attribute)": [[7, "genophenocorr.model.Variant.genotype"]], "genotype (genophenocorr.model.variant property)": [[7, "id35"]], "genotype (genophenocorr.model.variantcoordinates attribute)": [[7, "genophenocorr.model.VariantCoordinates.genotype"]], "genotype (genophenocorr.model.variantcoordinates property)": [[7, "id23"]], "hgvsc_id (genophenocorr.model.transcriptannotation attribute)": [[7, "genophenocorr.model.TranscriptAnnotation.hgvsc_id"]], "hgvsc_id (genophenocorr.model.transcriptannotation property)": [[7, "id28"]], "identifier (genophenocorr.model.phenotype attribute)": [[7, "genophenocorr.model.Phenotype.identifier"]], "identifier (genophenocorr.model.phenotype property)": [[7, "id0"]], "info (genophenocorr.model.proteinfeature property)": [[7, "genophenocorr.model.ProteinFeature.info"]], "is_structural() (genophenocorr.model.variantcoordinates method)": [[7, "genophenocorr.model.VariantCoordinates.is_structural"], [7, "id24"]], "label (genophenocorr.model.proteinmetadata attribute)": [[7, "genophenocorr.model.ProteinMetadata.label"]], "label (genophenocorr.model.proteinmetadata property)": [[7, "id11"]], "list_all_patients() (genophenocorr.model.cohort method)": [[7, "genophenocorr.model.Cohort.list_all_patients"]], "list_all_phenotypes() (genophenocorr.model.cohort method)": [[7, "genophenocorr.model.Cohort.list_all_phenotypes"]], "list_all_proteins() (genophenocorr.model.cohort method)": [[7, "genophenocorr.model.Cohort.list_all_proteins"]], "list_all_variants() (genophenocorr.model.cohort method)": [[7, "genophenocorr.model.Cohort.list_all_variants"]], "list_data_by_tx() (genophenocorr.model.cohort method)": [[7, "genophenocorr.model.Cohort.list_data_by_tx"]], "motifs() (genophenocorr.model.proteinmetadata method)": [[7, "genophenocorr.model.ProteinMetadata.motifs"], [7, "id16"]], "name (genophenocorr.model.featureinfo attribute)": [[7, "genophenocorr.model.FeatureInfo.name"]], "name (genophenocorr.model.featureinfo property)": [[7, "id3"]], "name (genophenocorr.model.phenotype attribute)": [[7, "genophenocorr.model.Phenotype.name"]], "name (genophenocorr.model.phenotype property)": [[7, "id1"]], "observed (genophenocorr.model.phenotype attribute)": [[7, "genophenocorr.model.Phenotype.observed"]], "observed (genophenocorr.model.phenotype property)": [[7, "id2"]], "overlapping_exons (genophenocorr.model.transcriptannotation attribute)": [[7, "genophenocorr.model.TranscriptAnnotation.overlapping_exons"]], "overlapping_exons (genophenocorr.model.transcriptannotation property)": [[7, "id30"]], "patient_id (genophenocorr.model.patient attribute)": [[7, "genophenocorr.model.Patient.patient_id"]], "patient_id (genophenocorr.model.patient property)": [[7, "id38"]], "phenotypes (genophenocorr.model.patient attribute)": [[7, "genophenocorr.model.Patient.phenotypes"]], "phenotypes (genophenocorr.model.patient property)": [[7, "id39"]], "protein_affected (genophenocorr.model.transcriptannotation attribute)": [[7, "genophenocorr.model.TranscriptAnnotation.protein_affected"]], "protein_affected (genophenocorr.model.transcriptannotation property)": [[7, "id31"]], "protein_effect_location (genophenocorr.model.transcriptannotation attribute)": [[7, "genophenocorr.model.TranscriptAnnotation.protein_effect_location"]], "protein_effect_location (genophenocorr.model.transcriptannotation property)": [[7, "id32"]], "protein_features (genophenocorr.model.proteinmetadata attribute)": [[7, "genophenocorr.model.ProteinMetadata.protein_features"]], "protein_features (genophenocorr.model.proteinmetadata property)": [[7, "id12"]], "protein_id (genophenocorr.model.proteinmetadata attribute)": [[7, "genophenocorr.model.ProteinMetadata.protein_id"]], "protein_id (genophenocorr.model.proteinmetadata property)": [[7, "id10"]], "proteins (genophenocorr.model.patient attribute)": [[7, "genophenocorr.model.Patient.proteins"]], "proteins (genophenocorr.model.patient property)": [[7, "id41"]], "ref (genophenocorr.model.variantcoordinates attribute)": [[7, "genophenocorr.model.VariantCoordinates.ref"]], "ref (genophenocorr.model.variantcoordinates property)": [[7, "id20"]], "regions() (genophenocorr.model.proteinmetadata method)": [[7, "genophenocorr.model.ProteinMetadata.regions"], [7, "id15"]], "repeats() (genophenocorr.model.proteinmetadata method)": [[7, "genophenocorr.model.ProteinMetadata.repeats"], [7, "id14"]], "start (genophenocorr.model.featureinfo attribute)": [[7, "genophenocorr.model.FeatureInfo.start"]], "start (genophenocorr.model.featureinfo property)": [[7, "id4"]], "start (genophenocorr.model.variantcoordinates attribute)": [[7, "genophenocorr.model.VariantCoordinates.start"]], "start (genophenocorr.model.variantcoordinates property)": [[7, "id18"]], "total_patient_count (genophenocorr.model.cohort property)": [[7, "genophenocorr.model.Cohort.total_patient_count"]], "transcript_id (genophenocorr.model.transcriptannotation attribute)": [[7, "genophenocorr.model.TranscriptAnnotation.transcript_id"]], "transcript_id (genophenocorr.model.transcriptannotation property)": [[7, "id27"]], "tx_annotations (genophenocorr.model.variant attribute)": [[7, "genophenocorr.model.Variant.tx_annotations"]], "tx_annotations (genophenocorr.model.variant property)": [[7, "id36"]], "variant_class (genophenocorr.model.variant attribute)": [[7, "genophenocorr.model.Variant.variant_class"]], "variant_class (genophenocorr.model.variant property)": [[7, "id37"]], "variant_coordinates (genophenocorr.model.variant attribute)": [[7, "genophenocorr.model.Variant.variant_coordinates"]], "variant_coordinates (genophenocorr.model.variant property)": [[7, "id33"]], "variant_effects (genophenocorr.model.transcriptannotation attribute)": [[7, "genophenocorr.model.TranscriptAnnotation.variant_effects"]], "variant_effects (genophenocorr.model.transcriptannotation property)": [[7, "id29"]], "variant_string (genophenocorr.model.variant attribute)": [[7, "genophenocorr.model.Variant.variant_string"]], "variant_string (genophenocorr.model.variant property)": [[7, "id34"]], "variants (genophenocorr.model.patient attribute)": [[7, "genophenocorr.model.Patient.variants"]], "variants (genophenocorr.model.patient property)": [[7, "id40"]], "functionalannotator (class in genophenocorr.preprocessing)": [[8, "genophenocorr.preprocessing.FunctionalAnnotator"]], "patientcreator (class in genophenocorr.preprocessing)": [[8, "genophenocorr.preprocessing.PatientCreator"]], "phenopacketpatientcreator (class in genophenocorr.preprocessing)": [[8, "genophenocorr.preprocessing.PhenopacketPatientCreator"]], "phenopacketvariantcoordinatefinder (class in genophenocorr.preprocessing)": [[8, "genophenocorr.preprocessing.PhenopacketVariantCoordinateFinder"]], "phenotypecreator (class in genophenocorr.preprocessing)": [[8, "genophenocorr.preprocessing.PhenotypeCreator"]], "phenotypevalidationexception": [[8, "genophenocorr.preprocessing.PhenotypeValidationException"]], "protcachingfunctionalannotator (class in genophenocorr.preprocessing)": [[8, "genophenocorr.preprocessing.ProtCachingFunctionalAnnotator"]], "proteinannotationcache (class in genophenocorr.preprocessing)": [[8, "genophenocorr.preprocessing.ProteinAnnotationCache"]], "proteinmetadataservice (class in genophenocorr.preprocessing)": [[8, "genophenocorr.preprocessing.ProteinMetadataService"]], "uniprotproteinmetadataservice (class in genophenocorr.preprocessing)": [[8, "genophenocorr.preprocessing.UniprotProteinMetadataService"]], "varcachingfunctionalannotator (class in genophenocorr.preprocessing)": [[8, "genophenocorr.preprocessing.VarCachingFunctionalAnnotator"]], "variantannotationcache (class in genophenocorr.preprocessing)": [[8, "genophenocorr.preprocessing.VariantAnnotationCache"]], "variantcoordinatefinder (class in genophenocorr.preprocessing)": [[8, "genophenocorr.preprocessing.VariantCoordinateFinder"]], "vepfunctionalannotator (class in genophenocorr.preprocessing)": [[8, "genophenocorr.preprocessing.VepFunctionalAnnotator"]], "annotate() (genophenocorr.preprocessing.functionalannotator method)": [[8, "genophenocorr.preprocessing.FunctionalAnnotator.annotate"]], "annotate() (genophenocorr.preprocessing.protcachingfunctionalannotator method)": [[8, "genophenocorr.preprocessing.ProtCachingFunctionalAnnotator.annotate"]], "annotate() (genophenocorr.preprocessing.proteinmetadataservice method)": [[8, "genophenocorr.preprocessing.ProteinMetadataService.annotate"]], "annotate() (genophenocorr.preprocessing.uniprotproteinmetadataservice method)": [[8, "genophenocorr.preprocessing.UniprotProteinMetadataService.annotate"]], "annotate() (genophenocorr.preprocessing.varcachingfunctionalannotator method)": [[8, "genophenocorr.preprocessing.VarCachingFunctionalAnnotator.annotate"]], "annotate() (genophenocorr.preprocessing.vepfunctionalannotator method)": [[8, "genophenocorr.preprocessing.VepFunctionalAnnotator.annotate"]], "configure_caching_patient_creator() (in module genophenocorr.preprocessing)": [[8, "genophenocorr.preprocessing.configure_caching_patient_creator"]], "create_patient() (genophenocorr.preprocessing.patientcreator method)": [[8, "genophenocorr.preprocessing.PatientCreator.create_patient"]], "create_patient() (genophenocorr.preprocessing.phenopacketpatientcreator method)": [[8, "genophenocorr.preprocessing.PhenopacketPatientCreator.create_patient"]], "create_phenotype() (genophenocorr.preprocessing.phenotypecreator method)": [[8, "genophenocorr.preprocessing.PhenotypeCreator.create_phenotype"]], "find_coordinates() (genophenocorr.preprocessing.phenopacketvariantcoordinatefinder method)": [[8, "genophenocorr.preprocessing.PhenopacketVariantCoordinateFinder.find_coordinates"]], "find_coordinates() (genophenocorr.preprocessing.variantcoordinatefinder method)": [[8, "genophenocorr.preprocessing.VariantCoordinateFinder.find_coordinates"]], "genophenocorr.preprocessing": [[8, "module-genophenocorr.preprocessing"]], "get_annotations() (genophenocorr.preprocessing.proteinannotationcache method)": [[8, "genophenocorr.preprocessing.ProteinAnnotationCache.get_annotations"]], "get_annotations() (genophenocorr.preprocessing.variantannotationcache method)": [[8, "genophenocorr.preprocessing.VariantAnnotationCache.get_annotations"]], "issues (genophenocorr.preprocessing.phenotypevalidationexception property)": [[8, "genophenocorr.preprocessing.PhenotypeValidationException.issues"]], "load_phenopacket_folder() (in module genophenocorr.preprocessing)": [[8, "genophenocorr.preprocessing.load_phenopacket_folder"]], "store_annotations() (genophenocorr.preprocessing.proteinannotationcache method)": [[8, "genophenocorr.preprocessing.ProteinAnnotationCache.store_annotations"]], "store_annotations() (genophenocorr.preprocessing.variantannotationcache method)": [[8, "genophenocorr.preprocessing.VariantAnnotationCache.store_annotations"]], "cohortviewer (class in genophenocorr.view)": [[10, "genophenocorr.view.CohortViewer"]], "anchor() (genophenocorr.view.cohortviewer static method)": [[10, "genophenocorr.view.CohortViewer.anchor"]], "genophenocorr.view": [[10, "module-genophenocorr.view"]], "hpo_anchor() (genophenocorr.view.cohortviewer static method)": [[10, "genophenocorr.view.CohortViewer.hpo_anchor"]], "hpo_term_counts_table() (genophenocorr.view.cohortviewer method)": [[10, "genophenocorr.view.CohortViewer.hpo_term_counts_table"]], "html_row() (genophenocorr.view.cohortviewer static method)": [[10, "genophenocorr.view.CohortViewer.html_row"]], "variants_table() (genophenocorr.view.cohortviewer method)": [[10, "genophenocorr.view.CohortViewer.variants_table"]]}})