{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9807414e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import phenopackets\n",
    "import os\n",
    "from os.path import isfile\n",
    "from google.protobuf.json_format import MessageToDict, MessageToJson\n",
    "from google.protobuf.json_format import Parse, ParseDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3179a36a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from phenopackets import Phenopacket\n",
    "import varcode as vc\n",
    "#import myvariant as mv\n",
    "import numpy as np\n",
    "import pyensembl\n",
    "import glob\n",
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "import scipy.stats as stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0e95c0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "retinoblastoma = '../phenopackets/retinoblastoma.json'\n",
    "nemalineMyopathy = '../phenopackets/nemalineMyopathy.json'\n",
    "if not isfile(retinoblastoma):\n",
    "    raise FileNotFoundError(\"Could not find phenopacket\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba43f5f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "with open(nemalineMyopathy) as f:\n",
    "    data = f.read()\n",
    "jsondata = json.loads(data)\n",
    "\n",
    "NMphenopacket = Parse(json.dumps(jsondata), Phenopacket())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "271d8aae",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(retinoblastoma) as f:\n",
    "    data = f.read()\n",
    "jsondata = json.loads(data)\n",
    "\n",
    "RBphenopacket = Parse(json.dumps(jsondata), Phenopacket())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e9051d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "NMphenopacket.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24011177",
   "metadata": {},
   "outputs": [],
   "source": [
    "RBphenopacket"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "befffefb",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Patient:\n",
    "    def __init__(self, phenopackJson):\n",
    "        if not isfile(phenopackJson):\n",
    "            raise FileNotFoundError(\"Could not find phenopacket\")\n",
    "            \n",
    "        with open(phenopackJson) as f:\n",
    "            data = f.read()\n",
    "        jsondata = json.loads(data)\n",
    "        phenopack = Parse(json.dumps(jsondata), Phenopacket())\n",
    "        \n",
    "        self._id = phenopack.id\n",
    "        self._phenopack = phenopack\n",
    "        self._phenotype = self.__get_hpids()\n",
    "        self._diseases = Disease(phenopack.diseases[0])\n",
    "        if len(phenopack.interpretations) != 0:\n",
    "            self._genotype = self.__get_variants()\n",
    "        else:\n",
    "            #print('No interpretations found')\n",
    "            self._genotype = None\n",
    "                \n",
    "        \n",
    "    def __get_hpids(self):\n",
    "        hp_ids = []\n",
    "        for x in self._phenopack.phenotypic_features:\n",
    "             if not x.excluded:\n",
    "                hp_ids.append(Phenotype(x))\n",
    "        return hp_ids\n",
    "    \n",
    "    def __get_variants(self):\n",
    "        interp = self._phenopack.interpretations[0]\n",
    "        genotypes = []\n",
    "        for geno in interp.diagnosis.genomic_interpretations:\n",
    "            genotypes.append(Genotype(geno))\n",
    "        return genotypes\n",
    "            \n",
    "    \n",
    "    @property\n",
    "    def disease_id(self):\n",
    "        return self._diseases.id\n",
    "    \n",
    "    @property\n",
    "    def disease_label(self):\n",
    "        return self._diseases.label\n",
    "    \n",
    "    @property\n",
    "    def phenopacket(self):\n",
    "        return self._phenopack\n",
    "    \n",
    "    @property\n",
    "    def phenotype_ids(self):\n",
    "        allPhenotypes = [phenotype.id for phenotype in self._phenotype]\n",
    "        return allPhenotypes\n",
    "    \n",
    "    @property\n",
    "    def phenotype_labels(self):\n",
    "        allPhenotypes = [phenotype.label for phenotype in self._phenotype]\n",
    "        return allPhenotypes\n",
    "    \n",
    "    @property\n",
    "    def all_phenotypes(self):\n",
    "        return self._phenotype\n",
    "    \n",
    "    @property\n",
    "    def variant(self):\n",
    "        return [g.variant for g in self._genotype]\n",
    "    \n",
    "    @property\n",
    "    def var_effect(self):\n",
    "        return [g.top_var_effect for g in self._genotype]\n",
    "    \n",
    "    @property\n",
    "    def var_is_missense(self):\n",
    "        return [g.is_missense for g in self._genotype]\n",
    "    \n",
    "    @property\n",
    "    def var_is_nonsense(self):\n",
    "        return [g.is_nonsense for g in self._genotype]\n",
    "    \n",
    "    def describe(self):\n",
    "        stats = {\n",
    "            \"ID\": self._phenopack.id,\n",
    "            \"Disease\": self.disease_label,\n",
    "            \"Phenotypic Features\": self.phenotype_labels,\n",
    "            \"Variant\": self.variant,\n",
    "            \"Primary Effect of Variant\": self.var_effect,\n",
    "            \"Is Missense Mutation?\": self.var_is_missense,\n",
    "            \"Is Nonsense Mutation?\": self.var_is_nonsense\n",
    "        }\n",
    "        return stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79b84c4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "TestPat = Patient('../phenopackets/cohort/Bee-2015-BBS2-II_2.json')\n",
    "TestPat.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69c3fe8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Genotype:\n",
    "    def __init__(self, genInterp):\n",
    "        contig = genInterp.variant_interpretation.variation_descriptor.vcf_record.chrom\n",
    "        start = genInterp.variant_interpretation.variation_descriptor.vcf_record.pos\n",
    "        ref = genInterp.variant_interpretation.variation_descriptor.vcf_record.ref\n",
    "        alt = genInterp.variant_interpretation.variation_descriptor.vcf_record.alt\n",
    "        self._myVar = vc.Variant(str(contig), start, ref, alt, ensembl = pyensembl.ensembl_grch37)\n",
    "        self._variant = self._myVar.short_description\n",
    "        self._topVar = self._myVar.effects().top_priority_effect()\n",
    "        \n",
    "    @property\n",
    "    def variant(self):\n",
    "        return self._variant\n",
    "    \n",
    "    @property\n",
    "    def top_var_effect(self):\n",
    "        return self._topVar.short_description\n",
    "    \n",
    "    @property\n",
    "    def is_missense(self):\n",
    "        if self._topVar.short_description.endswith(\"*\") or not self._myVar.is_snv:\n",
    "            return False\n",
    "        else:\n",
    "            return True\n",
    "    \n",
    "    @property\n",
    "    def is_nonsense(self):\n",
    "        if self._topVar.short_description.endswith(\"*\") and self._myVar.is_snv:\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08a1954c",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Disease:\n",
    "    def __init__(self, phenopacket):\n",
    "        \n",
    "        self._diseaseID = phenopacket.term.id\n",
    "        self._diseaseLabel = phenopacket.term.label\n",
    "        \n",
    "    @property\n",
    "    def id(self):\n",
    "        return self._diseaseID\n",
    "    \n",
    "    @property\n",
    "    def label(self):\n",
    "        return self._diseaseLabel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e4e3d9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Phenotype:\n",
    "    def __init__(self, phenopacket):\n",
    "        \n",
    "        self._hpid = phenopacket.type.id\n",
    "        self._label = phenopacket.type.label\n",
    "        \n",
    "    @property\n",
    "    def id(self):\n",
    "        return self._hpid\n",
    "    \n",
    "    @property\n",
    "    def label(self):\n",
    "        return self._label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "299cdcce",
   "metadata": {},
   "outputs": [],
   "source": [
    "allPatients = defaultdict(Patient)\n",
    "allDiseases = []\n",
    "allDiseaseNames = set()\n",
    "\n",
    "for file in glob.glob('../phenopackets/cohort/*.json'):\n",
    "    fileName = os.path.basename(file)\n",
    "    current = Patient(file)\n",
    "    \n",
    "    if current.variant is not None and len(current.variant) != 0:\n",
    "        allPatients[fileName] = current\n",
    "\n",
    "for p in allPatients: \n",
    "    print(allPatients[p].describe())\n",
    "    allDiseaseNames.add(allPatients[p].disease_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb1f9e91",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Number of patients {len(allPatients)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b351452f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy\n",
    "\n",
    "\n",
    "def has_hpo(pat, hpo):\n",
    "    for h in pat.phenotype_ids:\n",
    "        if h == hpo:\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "for d in allDiseaseNames:\n",
    "    #print(f\"d: {d}\")\n",
    "    patients_with_d = [p for p in allPatients.values() if p.disease_id == d]\n",
    "    n_dis = len(patients_with_d)\n",
    "    if n_dis < 15: continue\n",
    "    print(f\"Patients with disease {d}: n={n_dis}\")\n",
    "    # fisher exact test is what we actually need\n",
    "    all_hpo = set()\n",
    "    total_pat = len(patients_with_d)\n",
    "    for pat in patients_with_d:\n",
    "        for hpo in pat.phenotype_ids:\n",
    "            all_hpo.add(hpo)\n",
    "    print(f\"Total hpo terms: {len(all_hpo)}\")\n",
    "    for hpo_id in all_hpo:\n",
    "        miss_with_hpo = len([ pat for pat in patients_with_d if has_hpo(pat, hpo_id) and pat.var_is_missense[0]])\n",
    "        miss_without_hpo = len([ pat for pat in  patients_with_d if not has_hpo(pat, hpo_id) and pat.var_is_missense[0]])\n",
    "        nons_with_hpo = len([ pat  for pat in patients_with_d if has_hpo(pat, hpo_id) and pat.var_is_nonsense[0]])\n",
    "        nons_without_hpo = len([ pat for pat in patients_with_d if not has_hpo(pat, hpo_id) and pat.var_is_nonsense[0]])\n",
    "        print(f\"HPO {hpo_id}: miss_with_hpo {miss_with_hpo} miss_without_hpo {miss_without_hpo} nons_with_hpo {nons_with_hpo} nons_without_hpo {nons_without_hpo}\")\n",
    "        table = np.array([[miss_with_hpo, miss_without_hpo], [nons_with_hpo, nons_without_hpo]])\n",
    "        oddsr, p =  scipy.stats.fisher_exact(table, alternative='two-sided')\n",
    "        print(f\"p: {p}\")\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aca1a060",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38ef85ce",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee5fd460",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Counts:\n",
    "    def __init__(self, disease):\n",
    "        self._disease = disease.term.id\n",
    "        self._diseaseName = disease.term.label\n",
    "        self._totals = pd.DataFrame({'key':[\"Missense\",\"Nonsense\"]}).set_index('key')\n",
    "    \n",
    "    \n",
    "    def add_patient(self, patient):\n",
    "        if not isinstance(patient, Patient):\n",
    "            raise NotAPatientError(\"ERROR: Input must be of class Patient\")\n",
    "        hasDisease = False\n",
    "        for d in patient.describe()[\"Disease\"]:\n",
    "            if d.term.id.upper() == self._disease:\n",
    "                hasDisease = True\n",
    "        if not hasDisease:\n",
    "            #print(\"Patient \" + patient._id +\" does not have disease \" + self._disease)\n",
    "            return \n",
    "        \n",
    "        miss = sum(patient.is_missense())\n",
    "        nons = sum(patient.is_nonsense())\n",
    "        \n",
    "        if miss == 0 and nons == 0:\n",
    "            #print(\"Patient \" + patient._id + \" has no missense or nonsense mutations\")\n",
    "            return \n",
    "        for p in patient.get_phenotypes:\n",
    "            if p.id not in self._totals.columns:\n",
    "                df1 = pd.DataFrame({'key':[\"Missense\", \"Nonsense\"],\n",
    "                    p.id: [miss, nons]})\n",
    "                df1 = df1.set_index('key')\n",
    "                self._totals = self._totals.join(df1, how='outer')\n",
    "            else:\n",
    "                self._totals.at[\"Missense\", p.id] += miss\n",
    "                self._totals.at[\"Nonsense\", p.id] += nons\n",
    "        return self._totals\n",
    "    \n",
    "    def run_chi2(self):\n",
    "        \"\"\"\n",
    "        self._totals is a 2x2 array with [0,0] = category 1, has HPO, \n",
    "        [0,1] category 1 does not have hpo\n",
    "        \n",
    "        \"\"\"\n",
    "        if len(self._totals.columns) != 0:\n",
    "            if self._totals.sum(axis=1).at['Nonsense'] !=0 and self._totals.sum(axis=1).at['Missense'] != 0:\n",
    "                chi2, pval, dof, exp = stats.chi2_contingency(self._totals)\n",
    "            else:\n",
    "                #print(\"There are either no Nonsense or no Missense Mutations\")\n",
    "                return\n",
    "        else:\n",
    "            #print(\"Add patients to run a Chi2 Test\")\n",
    "            return\n",
    "        print(\"The p-value is: \" + str(pval) )\n",
    "        if pval <= 0.05:\n",
    "            print(\"We can conclude the Missense vs Nonsense mutations are NOT independent of the Phenotypes for disease \" + self._disease)\n",
    "            return 'False'\n",
    "        else:\n",
    "            print(\"We can conclude the Missense vs Nonsense mutations are independent of the Phenotypes for disease \" + self._disease)\n",
    "            return 'True'\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "enviro",
   "language": "python",
   "name": "enviro"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
