{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9807414e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import phenopackets\n",
    "import os\n",
    "from os.path import isfile\n",
    "from google.protobuf.json_format import MessageToDict, MessageToJson\n",
    "from google.protobuf.json_format import Parse, ParseDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3179a36a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from phenopackets import Phenopacket\n",
    "import varcode as vc\n",
    "#import myvariant as mv\n",
    "import numpy as np\n",
    "import pyensembl\n",
    "import glob\n",
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "import scipy.stats as stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b0e95c0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "retinoblastoma = '../phenopackets/retinoblastoma.json'\n",
    "nemalineMyopathy = '../phenopackets/nemalineMyopathy.json'\n",
    "if not isfile(retinoblastoma):\n",
    "    raise FileNotFoundError(\"Could not find phenopacket\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ba43f5f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "with open(nemalineMyopathy) as f:\n",
    "    data = f.read()\n",
    "jsondata = json.loads(data)\n",
    "\n",
    "NMphenopacket = Parse(json.dumps(jsondata), Phenopacket())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "271d8aae",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(retinoblastoma) as f:\n",
    "    data = f.read()\n",
    "jsondata = json.loads(data)\n",
    "\n",
    "RBphenopacket = Parse(json.dumps(jsondata), Phenopacket())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5e9051d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'arbitrary.id'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NMphenopacket.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "24011177",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "id: \"arbitrary.id\"\n",
       "subject {\n",
       "  id: \"proband A\"\n",
       "  time_at_last_encounter {\n",
       "    age {\n",
       "      iso8601duration: \"P6M\"\n",
       "    }\n",
       "  }\n",
       "  sex: FEMALE\n",
       "  karyotypic_sex: XX\n",
       "}\n",
       "phenotypic_features {\n",
       "  type {\n",
       "    id: \"HP:0030084\"\n",
       "    label: \"Clinodactyly\"\n",
       "  }\n",
       "  modifiers {\n",
       "    id: \"HP:0012834\"\n",
       "    label: \"Right\"\n",
       "  }\n",
       "  onset {\n",
       "    age {\n",
       "      iso8601duration: \"P3M\"\n",
       "    }\n",
       "  }\n",
       "}\n",
       "phenotypic_features {\n",
       "  type {\n",
       "    id: \"HP:0000555\"\n",
       "    label: \"Leukocoria\"\n",
       "  }\n",
       "  modifiers {\n",
       "    id: \"HP:0012835\"\n",
       "    label: \"Left\"\n",
       "  }\n",
       "  onset {\n",
       "    age {\n",
       "      iso8601duration: \"P4M\"\n",
       "    }\n",
       "  }\n",
       "}\n",
       "phenotypic_features {\n",
       "  type {\n",
       "    id: \"HP:0000486\"\n",
       "    label: \"Strabismus\"\n",
       "  }\n",
       "  modifiers {\n",
       "    id: \"HP:0012835\"\n",
       "    label: \"Left\"\n",
       "  }\n",
       "  onset {\n",
       "    age {\n",
       "      iso8601duration: \"P5M15D\"\n",
       "    }\n",
       "  }\n",
       "}\n",
       "phenotypic_features {\n",
       "  type {\n",
       "    id: \"HP:0000541\"\n",
       "    label: \"Retinal detachment\"\n",
       "  }\n",
       "  modifiers {\n",
       "    id: \"HP:0012835\"\n",
       "    label: \"Left\"\n",
       "  }\n",
       "  onset {\n",
       "    age {\n",
       "      iso8601duration: \"P6M\"\n",
       "    }\n",
       "  }\n",
       "}\n",
       "measurements {\n",
       "  assay {\n",
       "    id: \"LOINC:79893-4\"\n",
       "    label: \"Left eye Intraocular pressure\"\n",
       "  }\n",
       "  value {\n",
       "    quantity {\n",
       "      unit {\n",
       "        id: \"UCUM:mm[Hg]\"\n",
       "        label: \"millimetres of mercury\"\n",
       "      }\n",
       "      value: 25.0\n",
       "      reference_range {\n",
       "        unit {\n",
       "          id: \"56844-4\"\n",
       "          label: \"Intraocular pressure of Eye\"\n",
       "        }\n",
       "        low: 10.0\n",
       "        high: 21.0\n",
       "      }\n",
       "    }\n",
       "  }\n",
       "  time_observed {\n",
       "    age {\n",
       "      iso8601duration: \"P6M\"\n",
       "    }\n",
       "  }\n",
       "}\n",
       "measurements {\n",
       "  assay {\n",
       "    id: \"LOINC:79892-6\"\n",
       "    label: \"Right eye Intraocular pressure\"\n",
       "  }\n",
       "  value {\n",
       "    quantity {\n",
       "      unit {\n",
       "        id: \"UCUM:mm[Hg]\"\n",
       "        label: \"millimetres of mercury\"\n",
       "      }\n",
       "      value: 15.0\n",
       "      reference_range {\n",
       "        unit {\n",
       "          id: \"56844-4\"\n",
       "          label: \"Intraocular pressure of Eye\"\n",
       "        }\n",
       "        low: 10.0\n",
       "        high: 21.0\n",
       "      }\n",
       "    }\n",
       "  }\n",
       "  time_observed {\n",
       "    age {\n",
       "      iso8601duration: \"P6M\"\n",
       "    }\n",
       "  }\n",
       "}\n",
       "biosamples {\n",
       "  id: \"biosample.1\"\n",
       "  sampled_tissue {\n",
       "    id: \"UBERON:0000970\"\n",
       "    label: \"eye\"\n",
       "  }\n",
       "  phenotypic_features {\n",
       "    type {\n",
       "      id: \"NCIT:C35941\"\n",
       "      label: \"Flexner-Wintersteiner Rosette Formation\"\n",
       "    }\n",
       "  }\n",
       "  phenotypic_features {\n",
       "    type {\n",
       "      id: \"NCIT:C132485\"\n",
       "      label: \"Apoptosis and Necrosis\"\n",
       "    }\n",
       "  }\n",
       "  measurements {\n",
       "    assay {\n",
       "      id: \"LOINC:33728-7\"\n",
       "      label: \"Size.maximum dimension in Tumor\"\n",
       "    }\n",
       "    value {\n",
       "      quantity {\n",
       "        unit {\n",
       "          id: \"UCUM:mm\"\n",
       "          label: \"millimeter\"\n",
       "        }\n",
       "        value: 15.0\n",
       "      }\n",
       "    }\n",
       "    time_observed {\n",
       "      age {\n",
       "        iso8601duration: \"P8M2W\"\n",
       "      }\n",
       "    }\n",
       "  }\n",
       "  tumor_progression {\n",
       "    id: \"NCIT:C8509\"\n",
       "    label: \"Primary Neoplasm\"\n",
       "  }\n",
       "  pathological_tnm_finding {\n",
       "    id: \"NCIT:C140720\"\n",
       "    label: \"Retinoblastoma pT3 TNM Finding v8\"\n",
       "  }\n",
       "  pathological_tnm_finding {\n",
       "    id: \"NCIT:C140711\"\n",
       "    label: \"Retinoblastoma pN0 TNM Finding v8\"\n",
       "  }\n",
       "  procedure {\n",
       "    code {\n",
       "      id: \"NCIT:C48601\"\n",
       "      label: \"Enucleation\"\n",
       "    }\n",
       "    body_site {\n",
       "      id: \"UBERON:0004548\"\n",
       "      label: \"left eye\"\n",
       "    }\n",
       "    performed {\n",
       "      age {\n",
       "        iso8601duration: \"P8M2W\"\n",
       "      }\n",
       "    }\n",
       "  }\n",
       "  files {\n",
       "    uri: \"file://data/fileSomaticWgs.vcf.gz\"\n",
       "    individual_to_file_identifiers {\n",
       "      key: \"biosample.1\"\n",
       "      value: \"specimen.1\"\n",
       "    }\n",
       "    file_attributes {\n",
       "      key: \"fileFormat\"\n",
       "      value: \"VCF\"\n",
       "    }\n",
       "    file_attributes {\n",
       "      key: \"genomeAssembly\"\n",
       "      value: \"GRCh38\"\n",
       "    }\n",
       "  }\n",
       "}\n",
       "interpretations {\n",
       "  id: \"interpretation.id\"\n",
       "  progress_status: SOLVED\n",
       "  diagnosis {\n",
       "    disease {\n",
       "      id: \"NCIT:C7541\"\n",
       "      label: \"Retinoblastoma\"\n",
       "    }\n",
       "    genomic_interpretations {\n",
       "      subject_or_biosample_id: \"proband A\"\n",
       "      interpretation_status: CAUSATIVE\n",
       "      variant_interpretation {\n",
       "        acmg_pathogenicity_classification: PATHOGENIC\n",
       "        therapeutic_actionability: ACTIONABLE\n",
       "        variation_descriptor {\n",
       "          variation {\n",
       "            copy_number {\n",
       "              allele {\n",
       "                sequence_location {\n",
       "                  sequence_id: \"refseq:NC_000013.14\"\n",
       "                  sequence_interval {\n",
       "                    start_number {\n",
       "                      value: 25981249\n",
       "                    }\n",
       "                    end_number {\n",
       "                      value: 61706822\n",
       "                    }\n",
       "                  }\n",
       "                }\n",
       "              }\n",
       "              number {\n",
       "                value: 1\n",
       "              }\n",
       "            }\n",
       "          }\n",
       "          extensions {\n",
       "            name: \"mosaicism\"\n",
       "            value: \"40.0%\"\n",
       "          }\n",
       "        }\n",
       "      }\n",
       "    }\n",
       "    genomic_interpretations {\n",
       "      subject_or_biosample_id: \"biosample.1\"\n",
       "      interpretation_status: CAUSATIVE\n",
       "      variant_interpretation {\n",
       "        acmg_pathogenicity_classification: PATHOGENIC\n",
       "        therapeutic_actionability: ACTIONABLE\n",
       "        variation_descriptor {\n",
       "          id: \"rs121913300\"\n",
       "          variation {\n",
       "            allele {\n",
       "              sequence_location {\n",
       "                sequence_id: \"refseq:NC_000013.11\"\n",
       "                sequence_interval {\n",
       "                  start_number {\n",
       "                    value: 48367511\n",
       "                  }\n",
       "                  end_number {\n",
       "                    value: 48367512\n",
       "                  }\n",
       "                }\n",
       "              }\n",
       "              literal_sequence_expression {\n",
       "                sequence: \"T\"\n",
       "              }\n",
       "            }\n",
       "          }\n",
       "          label: \"RB1 c.958C>T (p.Arg320Ter)\"\n",
       "          gene_context {\n",
       "            value_id: \"HGNC:9884\"\n",
       "            symbol: \"RB1\"\n",
       "          }\n",
       "          expressions {\n",
       "            syntax: \"hgvs.c\"\n",
       "            value: \"NM_000321.2:c.958C>T\"\n",
       "          }\n",
       "          expressions {\n",
       "            syntax: \"transcript_reference\"\n",
       "            value: \"NM_000321.2\"\n",
       "          }\n",
       "          vcf_record {\n",
       "            genome_assembly: \"GRCh38\"\n",
       "            chrom: \"NC_000013.11\"\n",
       "            pos: 48367512\n",
       "            ref: \"C\"\n",
       "            alt: \"T\"\n",
       "          }\n",
       "          extensions {\n",
       "            name: \"allele-frequency\"\n",
       "            value: \"25.0%\"\n",
       "          }\n",
       "          molecule_context: genomic\n",
       "          allelic_state {\n",
       "            id: \"GENO:0000135\"\n",
       "            label: \"heterozygous\"\n",
       "          }\n",
       "        }\n",
       "      }\n",
       "    }\n",
       "  }\n",
       "}\n",
       "diseases {\n",
       "  term {\n",
       "    id: \"NCIT:C7541\"\n",
       "    label: \"Retinoblastoma\"\n",
       "  }\n",
       "  onset {\n",
       "    age {\n",
       "      iso8601duration: \"P4M\"\n",
       "    }\n",
       "  }\n",
       "  disease_stage {\n",
       "    id: \"LOINC:LA24739-7\"\n",
       "    label: \"Group E\"\n",
       "  }\n",
       "  clinical_tnm_finding {\n",
       "    id: \"NCIT:C140678\"\n",
       "    label: \"Retinoblastoma cM0 TNM Finding v8\"\n",
       "  }\n",
       "  primary_site {\n",
       "    id: \"UBERON:0004548\"\n",
       "    label: \"left eye\"\n",
       "  }\n",
       "}\n",
       "medical_actions {\n",
       "  treatment {\n",
       "    agent {\n",
       "      id: \"DrugCentral:1678\"\n",
       "      label: \"melphalan\"\n",
       "    }\n",
       "    route_of_administration {\n",
       "      id: \"NCIT:C38222\"\n",
       "      label: \"Intraarterial Route of Administration\"\n",
       "    }\n",
       "    dose_intervals {\n",
       "      quantity {\n",
       "        unit {\n",
       "          id: \"UCUM:mg.kg-1\"\n",
       "          label: \"milligram per kilogram\"\n",
       "        }\n",
       "        value: 0.4\n",
       "      }\n",
       "      schedule_frequency {\n",
       "        id: \"NCIT:C64576\"\n",
       "        label: \"Once\"\n",
       "      }\n",
       "      interval {\n",
       "        start {\n",
       "          seconds: 1599004800\n",
       "        }\n",
       "        end {\n",
       "          seconds: 1599004800\n",
       "        }\n",
       "      }\n",
       "    }\n",
       "  }\n",
       "  treatment_target {\n",
       "    id: \"NCIT:C7541\"\n",
       "    label: \"Retinoblastoma\"\n",
       "  }\n",
       "  treatment_intent {\n",
       "    id: \"NCIT:C62220\"\n",
       "    label: \"Cure\"\n",
       "  }\n",
       "  adverse_events {\n",
       "    id: \"HP:0025637\"\n",
       "    label: \"Vasospasm\"\n",
       "  }\n",
       "  treatment_termination_reason {\n",
       "    id: \"NCIT:C41331\"\n",
       "    label: \"Adverse Event\"\n",
       "  }\n",
       "}\n",
       "medical_actions {\n",
       "  therapeutic_regimen {\n",
       "    ontology_class {\n",
       "      id: \"NCIT:C10894\"\n",
       "      label: \"Carboplatin/Etoposide/Vincristine\"\n",
       "    }\n",
       "    start_time {\n",
       "      age {\n",
       "        iso8601duration: \"P7M\"\n",
       "      }\n",
       "    }\n",
       "    end_time {\n",
       "      age {\n",
       "        iso8601duration: \"P8M\"\n",
       "      }\n",
       "    }\n",
       "    regimen_status: COMPLETED\n",
       "  }\n",
       "  treatment_target {\n",
       "    id: \"NCIT:C7541\"\n",
       "    label: \"Retinoblastoma\"\n",
       "  }\n",
       "  treatment_intent {\n",
       "    id: \"NCIT:C62220\"\n",
       "    label: \"Cure\"\n",
       "  }\n",
       "}\n",
       "medical_actions {\n",
       "  procedure {\n",
       "    code {\n",
       "      id: \"NCIT:C48601\"\n",
       "      label: \"Enucleation\"\n",
       "    }\n",
       "    body_site {\n",
       "      id: \"UBERON:0004548\"\n",
       "      label: \"left eye\"\n",
       "    }\n",
       "    performed {\n",
       "      age {\n",
       "        iso8601duration: \"P8M2W\"\n",
       "      }\n",
       "    }\n",
       "  }\n",
       "  treatment_target {\n",
       "    id: \"NCIT:C7541\"\n",
       "    label: \"Retinoblastoma\"\n",
       "  }\n",
       "  treatment_intent {\n",
       "    id: \"NCIT:C62220\"\n",
       "    label: \"Cure\"\n",
       "  }\n",
       "}\n",
       "files {\n",
       "  uri: \"file://data/germlineWgs.vcf.gz\"\n",
       "  individual_to_file_identifiers {\n",
       "    key: \"proband A\"\n",
       "    value: \"sample1\"\n",
       "  }\n",
       "  file_attributes {\n",
       "    key: \"fileFormat\"\n",
       "    value: \"VCF\"\n",
       "  }\n",
       "  file_attributes {\n",
       "    key: \"genomeAssembly\"\n",
       "    value: \"GRCh38\"\n",
       "  }\n",
       "}\n",
       "meta_data {\n",
       "  created {\n",
       "    seconds: 1620988500\n",
       "  }\n",
       "  created_by: \"anonymous biocurator\"\n",
       "  resources {\n",
       "    id: \"ncit\"\n",
       "    name: \"NCI Thesaurus\"\n",
       "    url: \"http://purl.obolibrary.org/obo/ncit.owl\"\n",
       "    version: \"21.05d\"\n",
       "    namespace_prefix: \"NCIT\"\n",
       "    iri_prefix: \"http://purl.obolibrary.org/obo/NCIT_\"\n",
       "  }\n",
       "  resources {\n",
       "    id: \"efo\"\n",
       "    name: \"Experimental Factor Ontology\"\n",
       "    url: \"http://www.ebi.ac.uk/efo/efo.owl\"\n",
       "    version: \"3.34.0\"\n",
       "    namespace_prefix: \"EFO\"\n",
       "    iri_prefix: \"http://purl.obolibrary.org/obo/EFO_\"\n",
       "  }\n",
       "  resources {\n",
       "    id: \"uberon\"\n",
       "    name: \"Uber-anatomy ontology\"\n",
       "    url: \"http://purl.obolibrary.org/obo/uberon.owl\"\n",
       "    version: \"2021-07-27\"\n",
       "    namespace_prefix: \"UBERON\"\n",
       "    iri_prefix: \"http://purl.obolibrary.org/obo/UBERON_\"\n",
       "  }\n",
       "  resources {\n",
       "    id: \"ncbitaxon\"\n",
       "    name: \"NCBI organismal classification\"\n",
       "    url: \"http://purl.obolibrary.org/obo/ncbitaxon.owl\"\n",
       "    version: \"2021-06-10\"\n",
       "    namespace_prefix: \"NCBITaxon\"\n",
       "    iri_prefix: \"http://purl.obolibrary.org/obo/NCBITaxon_\"\n",
       "  }\n",
       "  phenopacket_schema_version: \"2.0.0\"\n",
       "}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RBphenopacket"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e92ff783",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Genotype:\n",
    "    def __init__(self, genInterp):\n",
    "        contig = genInterp.variant_interpretation.variation_descriptor.vcf_record.chrom\n",
    "        start = genInterp.variant_interpretation.variation_descriptor.vcf_record.pos\n",
    "        ref = genInterp.variant_interpretation.variation_descriptor.vcf_record.ref\n",
    "        alt = genInterp.variant_interpretation.variation_descriptor.vcf_record.alt\n",
    "        self._myVar = vc.Variant(str(contig), start, ref, alt, ensembl = pyensembl.ensembl_grch37)\n",
    "        self._variant = self._myVar.short_description\n",
    "        self._topVar = self._myVar.effects().top_priority_effect()\n",
    "        \n",
    "    @property\n",
    "    def variant(self):\n",
    "        return self._variant\n",
    "    \n",
    "    @property\n",
    "    def top_var_effect(self):\n",
    "        return self._topVar.short_description\n",
    "    \n",
    "    @property\n",
    "    def full_var(self):\n",
    "        return self._myVar\n",
    "    \n",
    "    @property\n",
    "    def is_missense(self):\n",
    "        if self._topVar.short_description.endswith(\"*\") or not self._myVar.is_snv:\n",
    "            return False\n",
    "        else:\n",
    "            return True\n",
    "    \n",
    "    @property\n",
    "    def is_nonsense(self):\n",
    "        if self._topVar.short_description.endswith(\"*\") and self._myVar.is_snv:\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "        \n",
    "    @property \n",
    "    def is_deletion(self):\n",
    "        return self._myVar.is_deletion\n",
    "    \n",
    "    @property\n",
    "    def is_insertion(self):\n",
    "        return self._myVar.is_insertion\n",
    "    \n",
    "    @property\n",
    "    def is_transition(self):\n",
    "        return self._myVar.is_transition\n",
    "    \n",
    "    @property\n",
    "    def is_transversion(self):\n",
    "        return self._myVar.is_transversion\n",
    "    \n",
    "    @property\n",
    "    def is_indel(self):\n",
    "        return self._myVar.is_indel\n",
    "    \n",
    "    @property\n",
    "    def is_duplication(self):\n",
    "        if 'dup' in self._myVar.short_description:\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "    \n",
    "    \n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "65d42516",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Patient:\n",
    "    def __init__(self, phenopackJson):\n",
    "        if not isfile(phenopackJson):\n",
    "            raise FileNotFoundError(\"Could not find phenopacket\")\n",
    "            \n",
    "        with open(phenopackJson) as f:\n",
    "            data = f.read()\n",
    "        jsondata = json.loads(data)\n",
    "        phenopack = Parse(json.dumps(jsondata), Phenopacket())\n",
    "        \n",
    "        self._id = phenopack.id\n",
    "        self._phenopack = phenopack\n",
    "        self._phenotype = self.__get_hpids()\n",
    "        if len(phenopack.diseases) != 0:\n",
    "            self._diseases = Disease(phenopack.diseases[0])\n",
    "        else:\n",
    "            self._diseases = None\n",
    "        if len(phenopack.interpretations) != 0:\n",
    "            self._genotype = self.__get_variants()\n",
    "        else:\n",
    "            #print('No interpretations found')\n",
    "            self._genotype = None\n",
    "                \n",
    "        \n",
    "    def __get_hpids(self):\n",
    "        hp_ids = []\n",
    "        for x in self._phenopack.phenotypic_features:\n",
    "             if not x.excluded:\n",
    "                hp_ids.append(Phenotype(x))\n",
    "        return hp_ids\n",
    "    \n",
    "    def __get_variants(self):\n",
    "        interp = self._phenopack.interpretations[0]\n",
    "        genotypes = []\n",
    "        for geno in interp.diagnosis.genomic_interpretations:\n",
    "            genotypes.append(Genotype(geno))\n",
    "        return genotypes\n",
    "            \n",
    "    \n",
    "    @property\n",
    "    def disease_id(self):\n",
    "        if self._diseases is not None:\n",
    "            return self._diseases.id\n",
    "        else:\n",
    "            return None\n",
    "    \n",
    "    @property\n",
    "    def disease_label(self):\n",
    "        if self._diseases is not None:\n",
    "            return self._diseases.label\n",
    "        else:\n",
    "            return None\n",
    "    \n",
    "    @property\n",
    "    def phenopacket(self):\n",
    "        return self._phenopack\n",
    "    \n",
    "    @property\n",
    "    def phenotype_ids(self):\n",
    "        if self._phenotype is not None:\n",
    "            return [phenotype.id for phenotype in self._phenotype]\n",
    "        else:\n",
    "            return None\n",
    "    \n",
    "    @property\n",
    "    def phenotype_labels(self):\n",
    "        if self._phenotype is not None:\n",
    "            return [phenotype.label for phenotype in self._phenotype]\n",
    "        else:\n",
    "            return None\n",
    "    \n",
    "    @property\n",
    "    def all_phenotypes(self):\n",
    "        return self._phenotype\n",
    "    \n",
    "    @property\n",
    "    def variant(self):\n",
    "        if self._genotype is not None:\n",
    "            return [g.variant for g in self._genotype]\n",
    "        else:\n",
    "            return None\n",
    "    \n",
    "    @property\n",
    "    def var_effect(self):\n",
    "        if self._genotype is not None:\n",
    "            return [g.top_var_effect for g in self._genotype]\n",
    "        else:\n",
    "            return None\n",
    "    \n",
    "    @property\n",
    "    def var_is_missense(self):\n",
    "        if self._genotype is not None:\n",
    "            return [g.is_missense for g in self._genotype]\n",
    "        else:\n",
    "            return None\n",
    "    \n",
    "    @property\n",
    "    def var_is_nonsense(self):\n",
    "        if self._genotype is not None:\n",
    "            return [g.is_nonsense for g in self._genotype]\n",
    "        else:\n",
    "            return None\n",
    "        \n",
    "    @property\n",
    "    def var_is_deletion(self):\n",
    "        if self._genotype is not None:\n",
    "            return [g.is_deletion for g in self._genotype]\n",
    "        elif self._diseases is not None:\n",
    "            return [self._diseases.is_deletion]\n",
    "        else:\n",
    "            return None\n",
    "        \n",
    "    @property\n",
    "    def var_is_duplication(self):\n",
    "        if self._genotype is not None:\n",
    "            return [g.is_duplication for g in self._genotype]\n",
    "        elif self._diseases is not None:\n",
    "            return [self._diseases.is_duplication]\n",
    "        else:\n",
    "            return None\n",
    "        \n",
    "    \n",
    "    def describe(self):\n",
    "        stats = {\n",
    "            \"ID\": self._phenopack.id,\n",
    "            \"Disease\": self.disease_label,\n",
    "            \"Phenotypic Features\": self.phenotype_labels,\n",
    "            \"Variant\": self.variant,\n",
    "            \"Primary Effect of Variant\": self.var_effect\n",
    "        }\n",
    "        return stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "56a32bb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Disease:\n",
    "    def __init__(self, phenopacket):\n",
    "        self._diseaseID = phenopacket.term.id\n",
    "        self._diseaseLabel = phenopacket.term.label\n",
    "        \n",
    "    @property\n",
    "    def id(self):\n",
    "        return self._diseaseID\n",
    "    \n",
    "    @property\n",
    "    def label(self):\n",
    "        return self._diseaseLabel\n",
    "    \n",
    "    @property\n",
    "    def is_duplication(self):\n",
    "        if 'duplication' in self._diseaseLabel:\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "    \n",
    "    @property\n",
    "    def is_deletion(self):\n",
    "        if 'deletion' in self._diseaseLabel:\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "dac0e948",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Phenotype:\n",
    "    def __init__(self, phenopacket):\n",
    "        \n",
    "        self._hpid = phenopacket.type.id\n",
    "        self._label = phenopacket.type.label\n",
    "        \n",
    "    @property\n",
    "    def id(self):\n",
    "        return self._hpid\n",
    "    \n",
    "    @property\n",
    "    def label(self):\n",
    "        return self._label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2dd182d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /Users/rekerl/Library/Caches/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.cdna.all.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /Users/rekerl/Library/Caches/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.ncrna.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /Users/rekerl/Library/Caches/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.pep.all.fa.gz.pickle\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr11 g.45827853_45827855delCTT\n"
     ]
    }
   ],
   "source": [
    "TestPat = Patient('../phenopackets/cohort/Dauber-2014-SLC35C1-Proband_1.json')\n",
    "eff = TestPat._genotype\n",
    "eff = eff.pop()\n",
    "stringvar = eff.full_var.short_description\n",
    "print(stringvar)\n",
    "parts = stringvar.split(' ')\n",
    "chro = parts[0].replace('chr', '')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "299cdcce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['chr1 g.247587794C>T']\n",
      "['chr19 g.13007113G>A']\n",
      "['chr17 g.40485715C>T']\n",
      "['chr20 g.44522679T>A', 'chr20 g.44523343_44523344delTA']\n",
      "['chr21 g.38862599C>T']\n",
      "['chr1 g.1737951C>A']\n",
      "['chr1 g.120458723G>A']\n",
      "['chr19 g.7623928C>A']\n",
      "['chr8 g.28574958C>T']\n",
      "['chr17 g.26684713C>A']\n",
      "['chr18 g.10797515G>A']\n",
      "['chr2 g.189951460C>T']\n",
      "['chr1 g.1167851A>G', 'chr1 g.1167659A>G']\n",
      "['chr20 g.3214218C>G']\n",
      "['chr16 g.67645347_67645350delAAAG']\n",
      "['chr6 g.86224294_86224297delTCTC']\n",
      "['chr6 g.158613022A>T', 'chr6 g.158591564T>A']\n",
      "['chr22 g.20049061G>A']\n",
      "['chr15 g.48776128C>T']\n",
      "['chr19 g.41848135G>A']\n",
      "['chr19 g.3586872G>A']\n",
      "['chr4 g.107115874C>T']\n",
      "['chr4 g.107169430_107169431insA']\n",
      "['chr4 g.80905083_80905084delCA']\n",
      "['chr7 g.44151537T>C']\n",
      "['chr4 g.107092429T>C', 'chr4 g.107168421_107168424delTTCA']\n",
      "['chr11 g.130029877G>A']\n",
      "['chr19 g.7622129G>T', 'chr19 g.7623842G>C']\n",
      "['chr5 g.151266323T>A']\n",
      "['chr8 g.61728948C>T']\n",
      "['chr14 g.29236938_29236939insG']\n",
      "['chr1 g.8045005C>A']\n",
      "['chr17 g.65340898T>C']\n",
      "['chr11 g.88068220A>C']\n",
      "['chr11 g.64972164G>A']\n",
      "['chr15 g.62207830A>C']\n",
      "['chr21 g.44478972C>T']\n",
      "['chr22 g.21213483C>T']\n",
      "['chr5 g.16477845G>C']\n",
      "['chrX g.136649818C>T']\n",
      "['chr16 g.68712731G>A', 'chr16 g.68713839_68713839delG']\n",
      "['chr1 g.197059410_197059413delGTTT', 'chr1 g.197070598_197070599delTC']\n",
      "['chr9 g.137534120G>A']\n",
      "['chr16 g.11643549C>T']\n",
      "['chr1 g.8419976C>T']\n",
      "['chr10 g.72358357A>C']\n",
      "['chr22 g.41551019C>T']\n",
      "['chr15 g.62250807C>G', 'chr15 g.62239491_62239491delG']\n",
      "['chr4 g.65175614C>T']\n",
      "['chr10 g.104852895C>A']\n",
      "['chr21 g.47409540G>A']\n",
      "['chr4 g.107168421_107168424delTTCA']\n",
      "['chr11 g.64950393G>A', 'chr11 g.64955493C>T']\n",
      "['chr1 g.43232523_43232538delGGCGAAGAGCAGATCA', 'chr1 g.43212415G>A']\n",
      "['chr22 g.20049061G>A']\n",
      "['chr9 g.26905643G>A']\n",
      "['chr4 g.1803571C>G']\n",
      "['chr3 g.113513742C>T']\n",
      "['chr14 g.90870832A>G']\n",
      "['chr9 g.113563161A>G']\n",
      "['chr11 g.111706902C>T']\n",
      "['chr13 g.100635387C>G']\n",
      "['chr5 g.64181325_64181326insA']\n",
      "['chr10 g.134599257_134599257delG']\n",
      "['chr19 g.18897435_18897437delGCA']\n",
      "['chr3 g.8775642G>A']\n",
      "['chr1 g.17322554G>A']\n",
      "['chr9 g.35802590C>A']\n",
      "['chr19 g.5694473G>A', 'chr19 g.5694418C>T']\n",
      "['chr14 g.103389060_103389060delA', 'chr14 g.103390315T>A']\n",
      "['chr11 g.63978208C>T']\n",
      "['chr1 g.103427814G>T']\n",
      "['chr11 g.64574571C>T']\n",
      "['chr14 g.57088415C>T']\n",
      "['chr11 g.64953394C>T', 'chr11 g.64956194C>T']\n",
      "['chr11 g.64953810G>A']\n",
      "['chr15 g.101718105C>T']\n",
      "['chr15 g.67473653G>C']\n",
      "['chr22 g.18082845A>G']\n",
      "['chr19 g.51297041C>T']\n",
      "['chr10 g.50690894G>A']\n",
      "['chr9 g.79980446C>G', 'chr9 g.80020779G>A']\n",
      "['chr15 g.48752455C>T']\n",
      "['chr1 g.43804317C>T']\n",
      "['chr9 g.108484867G>A']\n",
      "['chr14 g.29237179A>T']\n",
      "['chrX g.110005963_110005963delG']\n",
      "['chr9 g.34648167A>G', 'chr9 g.34648167A>C']\n",
      "['chr14 g.51382160C>A']\n",
      "['chr22 g.17690429C>G', 'chr22 g.17663510C>T']\n",
      "['chr18 g.42531913G>A']\n",
      "['chr2 g.223161750A>G']\n",
      "['chr9 g.94519790C>T']\n",
      "['chr11 g.74204660A>G', 'chr11 g.74204372A>C']\n",
      "['chr2 g.158630626C>T']\n",
      "['chr19 g.10935787G>A']\n",
      "['chr4 g.107183260G>A']\n",
      "['chr18 g.48604664C>T']\n",
      "['chr15 g.89817462T>C', 'chr15 g.89820030_89820030delG']\n",
      "['chrX g.54492285C>T']\n",
      "['chr15 g.48752457G>A']\n",
      "['chr11 g.64974264C>T']\n",
      "['chr5 g.60194102C>G']\n",
      "['chr1 g.183155566G>A', 'chr1 g.183184700_183184701insT']\n",
      "['chr22 g.20049061G>A']\n",
      "['chr14 g.69392358C>T']\n",
      "['chr15 g.48719763C>T']\n",
      "['chr19 g.54629943G>A']\n",
      "['chr15 g.28267700G>A', 'chr15 g.28228568T>C']\n",
      "['chr4 g.107156505_107156505delT']\n",
      "['chr10 g.134598876G>C']\n",
      "['chr8 g.28574591C>T']\n",
      "['chr8 g.28575584T>G']\n",
      "['chr4 g.107152924A>G']\n",
      "['chr5 g.137222576G>A']\n",
      "['chr10 g.112349671_112349672insGCA']\n",
      "['chr17 g.10544634C>T']\n",
      "['chr5 g.36985761_36985761delA']\n",
      "['chr10 g.43609950C>G']\n",
      "['chr12 g.102158291G>A', 'chr12 g.102157979C>T']\n",
      "['chr19 g.42754086T>C']\n",
      "['chr17 g.65337056A>C']\n",
      "['chr19 g.39998096G>A', 'chr19 g.39998025_39998025delG']\n",
      "['chr5 g.94803683G>A']\n",
      "['chr17 g.4837519G>A']\n",
      "['chr6 g.45405735_45405738delCTAC']\n",
      "['chr8 g.28575584T>G']\n",
      "['chr1 g.172628600_172628600delT']\n",
      "['chr8 g.30977891C>T']\n",
      "['chr9 g.75309494C>T']\n",
      "['chr1 g.229567587T>A']\n",
      "['chr17 g.26684733G>C']\n",
      "['chr15 g.72987563A>T']\n",
      "['chr14 g.57099860_57099863delTATC']\n",
      "['chr2 g.179650717G>A']\n",
      "['chr20 g.10653644_10653645insC']\n",
      "['chr2 g.232079571G>A', 'chr2 g.232141350C>T']\n",
      "['chr1 g.8425908C>T']\n",
      "['chr22 g.20785386C>T']\n",
      "['chr8 g.90983442_90983446delTTTGT']\n",
      "['chr20 g.45317860T>A']\n",
      "['chr9 g.100616413C>A']\n",
      "['chr2 g.69298917A>T']\n",
      "['chr7 g.92762329G>A']\n",
      "['chr12 g.54338976A>C']\n",
      "['chr1 g.109445849C>A']\n",
      "['chr3 g.30713759C>G']\n",
      "['chr16 g.20359598T>G']\n",
      "['chr6 g.45399753C>T']\n",
      "['chr1 g.8418292G>A', 'chr1 g.8420275G>C']\n",
      "['chr9 g.35689793_35689795delTCT']\n",
      "['chr16 g.89620340G>C']\n",
      "['chr11 g.64950669G>A']\n",
      "['chr8 g.61654460C>T']\n",
      "['chr11 g.64974114C>T']\n",
      "['chr22 g.38565431C>A']\n",
      "['chr3 g.33114129A>G']\n",
      "['chr15 g.67524181_67524182insTT']\n",
      "['chr17 g.29665753_29665754insTT']\n",
      "['chr17 g.26684705_26684706delTT', 'chr17 g.26684785G>C']\n",
      "['chr19 g.7509101A>G', 'chr19 g.7527145C>T']\n",
      "['chr10 g.134598665G>A', 'chr10 g.134598655C>T']\n",
      "['chr7 g.92762185C>A']\n",
      "['chr8 g.17933050G>A']\n",
      "['chr8 g.28574958C>T']\n",
      "['chr21 g.36259163T>C']\n",
      "['chr1 g.156106776G>A']\n",
      "['chr1 g.8421289G>A']\n",
      "['chr22 g.20049061G>A']\n",
      "['chr15 g.40764208_40764209insA']\n",
      "['chr11 g.64951013_64951013delC', 'chr11 g.64974295G>A']\n",
      "['chr12 g.57961297C>T']\n",
      "['chr19 g.36342212C>T', 'chr19 g.36339962C>T', 'chr19 g.36322580_36322581insC']\n",
      "['chr5 g.64181274G>T']\n",
      "['chr4 g.107154202C>T']\n",
      "['chr1 g.115256532C>T']\n",
      "['chr21 g.44486397A>G', 'chr21 g.44485784G>A']\n",
      "['chr8 g.28574529C>T']\n",
      "['chr1 g.12059066G>C']\n",
      "['chr13 g.28239968C>G']\n",
      "['chr18 g.3456468G>T']\n",
      "['chr14 g.29237456_29237457insT']\n",
      "['chr3 g.9775928_9775929insA']\n",
      "['chr14 g.92361364G>C']\n",
      "['chr9 g.35065355C>G']\n",
      "['chrX g.152860129_152860130insA']\n",
      "['chr3 g.113503074G>A']\n",
      "['chr4 g.107183260G>A']\n",
      "['chr18 g.43435648G>A']\n",
      "['chr7 g.143029967G>A']\n",
      "['chr1 g.158655001T>G', 'chr1 g.158597507G>C']\n",
      "['chr17 g.26684705_26684706delTT', 'chr17 g.26684785G>C']\n",
      "['chr21 g.33039650_33039651insT']\n",
      "['chr3 g.101284167G>T', 'chr3 g.101284439A>G']\n",
      "['chr9 g.71853706G>C', 'chr9 g.71855043_71855043delC']\n",
      "['chr15 g.43508487_43508487delC']\n",
      "['chr1 g.8418276_8418277insGGTGGA']\n",
      "['chr4 g.107156512T>A']\n",
      "['chr17 g.39977197A>G']\n",
      "['chr5 g.64181274G>T']\n",
      "['chr11 g.45827443G>T', 'chr11 g.45827853_45827855delCTT']\n",
      "['chr9 g.133342110A>G']\n",
      "['chr4 g.107168421_107168424delTTCA']\n",
      "['chr1 g.8420445_8420445delG']\n",
      "['chr22 g.20024325_20024325delT']\n",
      "['chr7 g.92762329G>A']\n",
      "['chr22 g.32894440C>T']\n",
      "['chr16 g.58043896_58043897insC']\n",
      "['chr1 g.11190804C>T']\n",
      "['chr19 g.42373179_42373179delG']\n",
      "['chr21 g.45706513A>C']\n",
      "['chr10 g.123279677G>C']\n",
      "['chr17 g.36099541A>T']\n",
      "['chr14 g.35182752C>T']\n",
      "['chr12 g.977212C>T', 'chr12 g.936408_936409insT']\n",
      "['chr21 g.46313400_46313400delG']\n",
      "['chr17 g.65343511G>A']\n",
      "['chr9 g.26946976C>A']\n",
      "['chr15 g.38591587C>T']\n",
      "['chr9 g.37428529_37428530insA']\n",
      "['chr1 g.8418292G>A']\n",
      "['chr9 g.129377827A>G']\n",
      "['chr9 g.124073097G>T']\n",
      "['chr9 g.26946975_26946976insC']\n",
      "['chr11 g.66278178G>T']\n",
      "['chr15 g.48782093C>T']\n",
      "['chr16 g.69364999C>T']\n",
      "['chr17 g.17699359_17699359delC']\n",
      "['chr17 g.4836699C>G']\n",
      "['chr1 g.155204848C>T']\n",
      "['chr1 g.8418302G>T']\n",
      "['chr9 g.123298784_123298788delTGCCT', 'chr9 g.123182239C>T']\n",
      "['chr16 g.78142343G>A']\n",
      "['chr17 g.79892546C>T']\n",
      "['chr9 g.136435570G>T', 'chr9 g.136412195G>A']\n",
      "['chr11 g.64955466G>C']\n",
      "['chr22 g.18075487G>A']\n",
      "['chr10 g.123279674G>C']\n",
      "['chr8 g.28575113C>T']\n",
      "['chr12 g.25398279C>T']\n",
      "['chr3 g.113503074G>A']\n",
      "['chr1 g.8418276_8418277insGGTGGA']\n",
      "['chr19 g.1221215C>G']\n",
      "['chr17 g.39727775A>C']\n",
      "['chr9 g.130438127_130438127delC']\n",
      "['chr2 g.47388889C>G']\n",
      "['chr6 g.10404742T>A']\n",
      "['chr13 g.25463509_25463512delCTGA']\n",
      "['chr12 g.53722280G>A']\n",
      "['chr17 g.15162409A>G']\n",
      "['chr3 g.70001033C>G']\n",
      "['chr1 g.8421296_8421297insCCTGGAGGAGCTGAGGAGGGAG']\n",
      "['chr3 g.176750838T>C']\n",
      "['chr1 g.209788677_209788683delCGCTTCT']\n",
      "['chr16 g.2098650A>T']\n",
      "['chr6 g.45459711_45459714delGTTT']\n",
      "['chr19 g.48342860_48342861insG']\n",
      "['chr3 g.101284167G>T']\n",
      "['chr9 g.138651532G>A']\n",
      "['chr15 g.73622060C>G']\n",
      "['chr1 g.235926113_235926114delAC']\n",
      "['chr11 g.64955949G>A']\n",
      "['chr15 g.64455161T>C', 'chr15 g.64448943C>T']\n",
      "['chr6 g.158571548G>A', 'chr6 g.158571622C>G']\n",
      "['chr1 g.200635684_200635684delA']\n",
      "['chr17 g.26684705_26684706delTT', 'chr17 g.26684785G>C']\n",
      "['chr17 g.17129638T>C']\n",
      "['chr1 g.8555123_8555123delT']\n",
      "['chr19 g.7532387_7532410delCGAGCGGCTGGAGCAGGAGCGGGC', 'chr19 g.7532286G>T']\n",
      "['chr4 g.107183260G>A', 'chr4 g.107168395_107168396insTA']\n",
      "['chr16 g.79633624G>C']\n",
      "['chr16 g.89760580G>A']\n",
      "['chr6 g.131939657G>C']\n",
      "['chr1 g.957693A>T', 'chr1 g.977517_977518insC']\n",
      "['chr20 g.3897587_3897590delATGA']\n",
      "['chr1 g.8416255T>C']\n",
      "['chr8 g.116616998G>A', 'chr8 g.116616110G>A']\n",
      "['chr17 g.65346383G>A']\n",
      "['chr10 g.134598648C>TA']\n",
      "['chr9 g.135796816A>C', 'chr9 g.135772014C>T']\n",
      "['chr11 g.2606506G>A']\n",
      "['chrX g.76776310G>A']\n",
      "['chr22 g.41572424_41572425insATGT']\n",
      "['chr7 g.92762613A>G']\n",
      "['chr17 g.26687551G>A', 'chr17 g.26684733G>C']\n",
      "['chr7 g.92762645G>T']\n",
      "['chr15 g.28510853G>A']\n",
      "['chrX g.68059596C>T']\n",
      "['chr1 g.8418302G>C']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['chr1 g.224592155C>A']\n",
      "['chr7 g.140501302T>C']\n",
      "['chr18 g.67834203T>C']\n",
      "['chr16 g.57693480T>C']\n",
      "['chr7 g.92760808T>C']\n",
      "['chr14 g.24551909_24551911delGAA']\n",
      "['chr5 g.64082455G>A']\n",
      "['chr15 g.62305256_62305257insTCTG', 'chr15 g.62174851C>A']\n",
      "['chr1 g.227075798C>A']\n",
      "['chr1 g.218609311G>A']\n",
      "['chr7 g.92761698C>G']\n",
      "['chr4 g.107156512T>A']\n",
      "['chr1 g.8418291T>C']\n",
      "['chr10 g.50680431ATCTT>TGCACACCA']\n",
      "['chr14 g.29237342_29237343insC']\n",
      "['chr9 g.94499750C>T']\n",
      "['chr11 g.71148914C>T', 'chr11 g.71146524T>C']\n",
      "['chr2 g.170343646T>C']\n",
      "['chr2 g.44573407_44573407delT']\n",
      "['chr22 g.18075487G>A']\n",
      "['chr20 g.43251680C>T']\n",
      "['chr12 g.48381473C>T']\n",
      "['chr9 g.139265796C>A']\n",
      "['chr5 g.127668649A>C']\n",
      "['chr16 g.56530925G>A', 'chr16 g.56519631A>T']\n",
      "['chr15 g.48808561T>A']\n",
      "['chr5 g.483385C>T', 'chr5 g.477462_477462delG']\n",
      "['chr19 g.7184444C>T']\n",
      "['chr22 g.20049207G>C']\n",
      "['chrX g.69255235_69255235delG']\n",
      "['chr7 g.92760637A>G']\n",
      "['chr2 g.48050286T>C']\n",
      "['chr4 g.107156505_107156505delT']\n",
      "['chr3 g.38639391C>T']\n",
      "['chr11 g.118373871G>T']\n",
      "['chr8 g.28574591C>T']\n",
      "['chrX g.128723933G>A']\n",
      "['chr9 g.101907166A>G']\n",
      "['chr4 g.1805540C>T']\n",
      "['chr15 g.62274656A>C', 'chr15 g.62256151C>A']\n",
      "['chr5 g.37044760G>T']\n",
      "['chr19 g.38780855A>G', 'chr19 g.38774325_38774326insTA']\n",
      "['chr9 g.12695754G>TT', 'chr9 g.12695772C>T']\n",
      "['chr18 g.21487603C>T']\n",
      "['chr7 g.92762645G>T']\n",
      "['chr1 g.150315847C>G']\n",
      "['chr1 g.8716108_8716109insT']\n",
      "['chr12 g.49372436_49372436delG']\n",
      "['chr4 g.107183260G>A']\n",
      "['chr11 g.64974114C>T']\n",
      "['chrX g.17750140C>G']\n",
      "['chr3 g.58131722G>A']\n",
      "['chr1 g.8418810G>C']\n",
      "['chr19 g.8670555A>T']\n",
      "['chr16 g.51175309A>C']\n",
      "['chrX g.44833910G>T']\n",
      "['chr10 g.134598646C>T']\n",
      "['chr8 g.28575546A>G']\n",
      "['chr17 g.42334914T>A']\n",
      "['chr14 g.24709842G>A']\n",
      "['chrX g.152959399C>A']\n",
      "['chr1 g.154145610G>T']\n",
      "['chr11 g.64950354_64950355insC']\n",
      "['chr1 g.8418276_8418277insGGTGGA']\n",
      "['chr9 g.135866303C>T']\n",
      "['chr18 g.46860138G>A']\n",
      "['chr3 g.58062988G>C']\n",
      "['chr8 g.100479778_100479778delT', 'chr8 g.100712001_100712002delAT']\n",
      "['chr19 g.5719847_5719848insC', 'chr19 g.5693745G>A']\n",
      "['chr9 g.135787681G>A']\n",
      "['chr8 g.92083518G>T', 'chr8 g.92086140G>A']\n",
      "['chr1 g.8420421G>A']\n",
      "['chr17 g.17699358_17699359insC']\n",
      "['chr6 g.3225601C>T']\n",
      "['chr19 g.18896640C>T']\n",
      "['chr7 g.92762329G>A']\n",
      "['chr1 g.237947532T>A']\n",
      "['chr14 g.88434727C>T', 'chr14 g.88411945C>T']\n",
      "['chr9 g.32984702T>A']\n",
      "['chr1 g.55464998G>A']\n",
      "['chr8 g.28574958C>T']\n",
      "['chr5 g.39138767_39138768delAT']\n",
      "['chr10 g.134598798C>A']\n",
      "['chr18 g.21115443T>C', 'chr18 g.21119382C>T']\n",
      "['chr5 g.60186791G>T']\n",
      "['chr4 g.107115874C>T']\n",
      "['chr4 g.107183260G>A']\n",
      "['chr3 g.119205740T>G']\n",
      "['chr1 g.55340807C>T']\n",
      "['chr9 g.98011507_98011507delC']\n",
      "['chr4 g.107092429T>C', 'chr4 g.107168421_107168424delTTCA']\n",
      "['chr4 g.107169430_107169431insA']\n",
      "['chr10 g.134599332T>A']\n",
      "['chr6 g.45399750G>A']\n",
      "['chr19 g.7592840_7592841insC']\n"
     ]
    }
   ],
   "source": [
    "allPatients = defaultdict(Patient)\n",
    "allDiseases = []\n",
    "allDiseaseNames = set()\n",
    "\n",
    "for file in glob.glob('../phenopackets/cohort/*.json'):\n",
    "    fileName = os.path.basename(file)\n",
    "    current = Patient(file)\n",
    "    print(current.variant)\n",
    "    \n",
    "    \n",
    "    #if type(current.variant) != \"NoneType\":\n",
    "    #if current.variant is not None and len(current.variant) != 0:\n",
    "    allPatients[fileName] = current\n",
    "\n",
    "#for p in allPatients: \n",
    "    #print(allPatients[p]._diseases)\n",
    "    #allDiseaseNames.add(allPatients[p]._diseases.label)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "bb1f9e91",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of patients 384\n"
     ]
    }
   ],
   "source": [
    "print(f\"Number of patients {len(allPatients)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7777e303",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "pop expected at least 1 argument, got 0",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [18], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m test \u001b[38;5;241m=\u001b[39m allPatients\u001b[38;5;241m.\u001b[39mpop()[\u001b[38;5;241m1\u001b[39m]\n\u001b[1;32m      2\u001b[0m test\u001b[38;5;241m.\u001b[39mvariant\u001b[38;5;241m.\u001b[39mvariant\n",
      "\u001b[0;31mTypeError\u001b[0m: pop expected at least 1 argument, got 0"
     ]
    }
   ],
   "source": [
    "test = allPatients.get\n",
    "test.variant.variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8922b401",
   "metadata": {},
   "outputs": [],
   "source": [
    "import regex as re\n",
    "import requests\n",
    "\n",
    "from requests.adapters import HTTPAdapter, Retry\n",
    "\n",
    "url = 'https://rest.uniprot.org/uniprotkb/A1A519.json'\n",
    "all_files = requests.get(url).json()\n",
    "\n",
    "all_files\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e6e5e89",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_files.get('features')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75e00b95",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "874fcf7c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3da5b2e3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e58c636c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "994517de",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b351452f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy\n",
    "\n",
    "\n",
    "def has_hpo(pat, hpo):\n",
    "    for h in pat.phenotype_ids:\n",
    "        if h == hpo:\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "for d in allDiseaseNames:\n",
    "    #print(f\"d: {d}\")\n",
    "    patients_with_d = [p for p in allPatients.values() if p.disease_id == d.id]\n",
    "    n_dis = len(patients_with_d)\n",
    "    if n_dis < 15: continue\n",
    "    print(f\"Patients with disease {d.label}: n={n_dis}\")\n",
    "    # fisher exact test is what we actually need\n",
    "    all_hpo = set()\n",
    "    total_pat = len(patients_with_d)\n",
    "    for pat in patients_with_d:\n",
    "        for hpo in pat.phenotype_ids:\n",
    "            all_hpo.add(hpo)\n",
    "    print(f\"Total hpo terms: {len(all_hpo)}\")\n",
    "    #for hpo_id in all_hpo:\n",
    "        #miss_with_hpo = len([ pat for pat in patients_with_d if has_hpo(pat, hpo_id) and pat.var_is_missense[0]])\n",
    "        #miss_without_hpo = len([ pat for pat in  patients_with_d if not has_hpo(pat, hpo_id) and pat.var_is_missense[0]])\n",
    "        #nons_with_hpo = len([ pat  for pat in patients_with_d if has_hpo(pat, hpo_id) and pat.var_is_nonsense[0]])\n",
    "        #nons_without_hpo = len([ pat for pat in patients_with_d if not has_hpo(pat, hpo_id) and pat.var_is_nonsense[0]])\n",
    "        #print(f\"HPO {hpo_id}: miss_with_hpo {miss_with_hpo} miss_without_hpo {miss_without_hpo} nons_with_hpo {nons_with_hpo} nons_without_hpo {nons_without_hpo}\")\n",
    "        #table = np.array([[miss_with_hpo, miss_without_hpo], [nons_with_hpo, nons_without_hpo]])\n",
    "        #oddsr, p =  scipy.stats.fisher_exact(table, alternative='two-sided') ##Add option for chi2\n",
    "        #print(f\"p: {p}\")\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aca1a060",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38ef85ce",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee5fd460",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Counts:\n",
    "    def __init__(self, disease):\n",
    "        self._disease = disease.term.id\n",
    "        self._diseaseName = disease.term.label\n",
    "        self._totals = pd.DataFrame({'key':[\"Missense\",\"Nonsense\"]}).set_index('key')\n",
    "    \n",
    "    \n",
    "    def add_patient(self, patient):\n",
    "        if not isinstance(patient, Patient):\n",
    "            raise NotAPatientError(\"ERROR: Input must be of class Patient\")\n",
    "        hasDisease = False\n",
    "        for d in patient.describe()[\"Disease\"]:\n",
    "            if d.term.id.upper() == self._disease:\n",
    "                hasDisease = True\n",
    "        if not hasDisease:\n",
    "            #print(\"Patient \" + patient._id +\" does not have disease \" + self._disease)\n",
    "            return \n",
    "        \n",
    "        miss = sum(patient.is_missense())\n",
    "        nons = sum(patient.is_nonsense())\n",
    "        \n",
    "        if miss == 0 and nons == 0:\n",
    "            #print(\"Patient \" + patient._id + \" has no missense or nonsense mutations\")\n",
    "            return \n",
    "        for p in patient.get_phenotypes:\n",
    "            if p.id not in self._totals.columns:\n",
    "                df1 = pd.DataFrame({'key':[\"Missense\", \"Nonsense\"],\n",
    "                    p.id: [miss, nons]})\n",
    "                df1 = df1.set_index('key')\n",
    "                self._totals = self._totals.join(df1, how='outer')\n",
    "            else:\n",
    "                self._totals.at[\"Missense\", p.id] += miss\n",
    "                self._totals.at[\"Nonsense\", p.id] += nons\n",
    "        return self._totals\n",
    "    \n",
    "    def run_chi2(self):\n",
    "        \"\"\"\n",
    "        self._totals is a 2x2 array with [0,0] = category 1, has HPO, \n",
    "        [0,1] category 1 does not have hpo\n",
    "        \n",
    "        \"\"\"\n",
    "        if len(self._totals.columns) != 0:\n",
    "            if self._totals.sum(axis=1).at['Nonsense'] !=0 and self._totals.sum(axis=1).at['Missense'] != 0:\n",
    "                chi2, pval, dof, exp = stats.chi2_contingency(self._totals)\n",
    "            else:\n",
    "                #print(\"There are either no Nonsense or no Missense Mutations\")\n",
    "                return\n",
    "        else:\n",
    "            #print(\"Add patients to run a Chi2 Test\")\n",
    "            return\n",
    "        print(\"The p-value is: \" + str(pval) )\n",
    "        if pval <= 0.05:\n",
    "            print(\"We can conclude the Missense vs Nonsense mutations are NOT independent of the Phenotypes for disease \" + self._disease)\n",
    "            return 'False'\n",
    "        else:\n",
    "            print(\"We can conclude the Missense vs Nonsense mutations are independent of the Phenotypes for disease \" + self._disease)\n",
    "            return 'True'\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "enviro",
   "language": "python",
   "name": "enviro"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
