{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f806c19f",
   "metadata": {},
   "source": [
    "<H1>Comparing different variant types</H1>\n",
    "<p>Todo explain missense vs nonsense as example etc. </p>\n",
    "<p>This notebook demonstrates how to compare missense vs nonsenes ... </p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2696d536",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "#sys.path.insert(0, os.path.abspath('../genophenocorr'))\n",
    "from genophenocorr import *\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10d231b2",
   "metadata": {},
   "source": [
    "<H2>Data import</H2>\n",
    "<p>genophenocorr takes a collection of phenopackets as its input. In general, the phenopackets should represent individuals of a cohort to be investigated. For instance, we might use genophenocorr to investigate a cohort of 100 individuals with Marfan syndrome as to whether missense or nonsense variants are associated with different clinical manifestations of this disease. Therefore, the first step to use genophenocorr is to obtain or generate the phenopackets. This step is outside of the scope of the genophenocorr library, but can be done with tools such as ...(Peter to complete).</p>\n",
    "<p>For our example, we ingest phenopackets from the </p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c773d31f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /Users/rekerl/Library/Caches/pyensembl/GRCh38/ensembl107/Homo_sapiens.GRCh38.pep.all.fa.gz.pickle\n"
     ]
    }
   ],
   "source": [
    "allPatients = Cohort('../phenopackets/PlatzerData/*.json')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb15234b",
   "metadata": {},
   "source": [
    "<h2>Preview</h2>\n",
    "<p>The following optional step allows users to inspect the contents of the ingested phenopackets.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1f3d5ee3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient ID</th>\n",
       "      <th>Disease</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variant</th>\n",
       "      <th>Protein</th>\n",
       "      <th>HPO Terms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Individual 5</td>\n",
       "      <td></td>\n",
       "      <td>[MAPK8IP3]</td>\n",
       "      <td>chr16 g.1767834C&gt;T</td>\n",
       "      <td>ENSP00000501096</td>\n",
       "      <td>[HP:0001270, HP:0032988, HP:0025336, HP:003193...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12</td>\n",
       "      <td>618443</td>\n",
       "      <td>[MAPK8IP3]</td>\n",
       "      <td>chr16 g.1767834C&gt;T</td>\n",
       "      <td>ENSP00000501096</td>\n",
       "      <td>[HP:0001257, HP:0100021, HP:0010864]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>618443</td>\n",
       "      <td>[MAPK8IP3]</td>\n",
       "      <td>chr16 g.1706418G&gt;T</td>\n",
       "      <td>ENSP00000501096</td>\n",
       "      <td>[HP:0001251, HP:0010864]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>618443</td>\n",
       "      <td>[MAPK8IP3]</td>\n",
       "      <td>chr16 g.1706450C&gt;G</td>\n",
       "      <td>ENSP00000501096</td>\n",
       "      <td>[HP:0002342]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>618443</td>\n",
       "      <td>[MAPK8IP3]</td>\n",
       "      <td>chr16 g.1767834C&gt;T</td>\n",
       "      <td>ENSP00000501096</td>\n",
       "      <td>[HP:0002317, HP:0007301, HP:0002370, HP:0002342]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Individual 4</td>\n",
       "      <td></td>\n",
       "      <td>[MAPK8IP3]</td>\n",
       "      <td>chr16 g.1767834C&gt;T</td>\n",
       "      <td>ENSP00000501096</td>\n",
       "      <td>[HP:0001270, HP:0032988, HP:0032989, HP:002533...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4</td>\n",
       "      <td>618443</td>\n",
       "      <td>[MAPK8IP3]</td>\n",
       "      <td>chr16 g.1748705G&gt;A</td>\n",
       "      <td>ENSP00000501096</td>\n",
       "      <td>[HP:0001256]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>618443</td>\n",
       "      <td>[MAPK8IP3]</td>\n",
       "      <td>chr16 g.1762843C&gt;T</td>\n",
       "      <td>ENSP00000501096</td>\n",
       "      <td>[HP:0001258, HP:0010864]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Individual 3</td>\n",
       "      <td></td>\n",
       "      <td>[MAPK8IP3]</td>\n",
       "      <td>chr16 g.1762843C&gt;T</td>\n",
       "      <td>ENSP00000501096</td>\n",
       "      <td>[HP:0001270, HP:0025336, HP:0031936, HP:000218...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Individual 2</td>\n",
       "      <td></td>\n",
       "      <td>[MAPK8IP3]</td>\n",
       "      <td>chr16 g.1762843C&gt;T</td>\n",
       "      <td>ENSP00000501096</td>\n",
       "      <td>[HP:0001270, HP:0032989, HP:0031936, HP:001086...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>9</td>\n",
       "      <td>618443</td>\n",
       "      <td>[MAPK8IP3]</td>\n",
       "      <td>chr16 g.1762843C&gt;T</td>\n",
       "      <td>ENSP00000501096</td>\n",
       "      <td>[HP:0007340, HP:0002493, HP:0001257, HP:0002342]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>5</td>\n",
       "      <td>618443</td>\n",
       "      <td>[MAPK8IP3]</td>\n",
       "      <td>chr16 g.1760409T&gt;C</td>\n",
       "      <td>ENSP00000501096</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>6</td>\n",
       "      <td>618443</td>\n",
       "      <td>[MAPK8IP3]</td>\n",
       "      <td>chr16 g.1760409T&gt;C</td>\n",
       "      <td>ENSP00000501096</td>\n",
       "      <td>[HP:0001256]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Individual 1</td>\n",
       "      <td></td>\n",
       "      <td>[MAPK8IP3]</td>\n",
       "      <td>chr16 g.1762843C&gt;T</td>\n",
       "      <td>ENSP00000501096</td>\n",
       "      <td>[HP:0001270, HP:0031936, HP:0010864, HP:000075...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>7</td>\n",
       "      <td>618443</td>\n",
       "      <td>[MAPK8IP3]</td>\n",
       "      <td>chr16 g.1762388G&gt;A</td>\n",
       "      <td>ENSP00000501096</td>\n",
       "      <td>[HP:0001256]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>10</td>\n",
       "      <td>618443</td>\n",
       "      <td>[MAPK8IP3]</td>\n",
       "      <td>chr16 g.1766768C&gt;G</td>\n",
       "      <td>ENSP00000501096</td>\n",
       "      <td>[HP:0002342]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>11</td>\n",
       "      <td>618443</td>\n",
       "      <td>[MAPK8IP3]</td>\n",
       "      <td>chr16 g.1767834C&gt;T</td>\n",
       "      <td>ENSP00000501096</td>\n",
       "      <td>[HP:0001257, HP:0002342]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>618443</td>\n",
       "      <td>[MAPK8IP3]</td>\n",
       "      <td>chr16 g.1706403_1706403delG</td>\n",
       "      <td>ENSP00000501096</td>\n",
       "      <td>[HP:0001251, HP:0002342]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Patient ID Disease        Gene                      Variant  \\\n",
       "0   Individual 5          [MAPK8IP3]           chr16 g.1767834C>T   \n",
       "1             12  618443  [MAPK8IP3]           chr16 g.1767834C>T   \n",
       "2              2  618443  [MAPK8IP3]           chr16 g.1706418G>T   \n",
       "3              3  618443  [MAPK8IP3]           chr16 g.1706450C>G   \n",
       "4             13  618443  [MAPK8IP3]           chr16 g.1767834C>T   \n",
       "5   Individual 4          [MAPK8IP3]           chr16 g.1767834C>T   \n",
       "6              4  618443  [MAPK8IP3]           chr16 g.1748705G>A   \n",
       "7              8  618443  [MAPK8IP3]           chr16 g.1762843C>T   \n",
       "8   Individual 3          [MAPK8IP3]           chr16 g.1762843C>T   \n",
       "9   Individual 2          [MAPK8IP3]           chr16 g.1762843C>T   \n",
       "10             9  618443  [MAPK8IP3]           chr16 g.1762843C>T   \n",
       "11             5  618443  [MAPK8IP3]           chr16 g.1760409T>C   \n",
       "12             6  618443  [MAPK8IP3]           chr16 g.1760409T>C   \n",
       "13  Individual 1          [MAPK8IP3]           chr16 g.1762843C>T   \n",
       "14             7  618443  [MAPK8IP3]           chr16 g.1762388G>A   \n",
       "15            10  618443  [MAPK8IP3]           chr16 g.1766768C>G   \n",
       "16            11  618443  [MAPK8IP3]           chr16 g.1767834C>T   \n",
       "17             1  618443  [MAPK8IP3]  chr16 g.1706403_1706403delG   \n",
       "\n",
       "            Protein                                          HPO Terms  \n",
       "0   ENSP00000501096  [HP:0001270, HP:0032988, HP:0025336, HP:003193...  \n",
       "1   ENSP00000501096               [HP:0001257, HP:0100021, HP:0010864]  \n",
       "2   ENSP00000501096                           [HP:0001251, HP:0010864]  \n",
       "3   ENSP00000501096                                       [HP:0002342]  \n",
       "4   ENSP00000501096   [HP:0002317, HP:0007301, HP:0002370, HP:0002342]  \n",
       "5   ENSP00000501096  [HP:0001270, HP:0032988, HP:0032989, HP:002533...  \n",
       "6   ENSP00000501096                                       [HP:0001256]  \n",
       "7   ENSP00000501096                           [HP:0001258, HP:0010864]  \n",
       "8   ENSP00000501096  [HP:0001270, HP:0025336, HP:0031936, HP:000218...  \n",
       "9   ENSP00000501096  [HP:0001270, HP:0032989, HP:0031936, HP:001086...  \n",
       "10  ENSP00000501096   [HP:0007340, HP:0002493, HP:0001257, HP:0002342]  \n",
       "11  ENSP00000501096                                                 []  \n",
       "12  ENSP00000501096                                       [HP:0001256]  \n",
       "13  ENSP00000501096  [HP:0001270, HP:0031936, HP:0010864, HP:000075...  \n",
       "14  ENSP00000501096                                       [HP:0001256]  \n",
       "15  ENSP00000501096                                       [HP:0002342]  \n",
       "16  ENSP00000501096                           [HP:0001257, HP:0002342]  \n",
       "17  ENSP00000501096                           [HP:0001251, HP:0002342]  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allPatients.describe_all()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6ab9b58",
   "metadata": {},
   "source": [
    "This next step is able to look at more details of an individual phenopacket, also optional. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "df0de899",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ID                                                                   8\n",
       "Disease ID                                                      618443\n",
       "Disease Label        Neurodevelopmental disorder with or without va...\n",
       "HPO IDs                                   ['HP:0001258', 'HP:0010864']\n",
       "HPO Terms            ['Spastic paraplegia', 'Intellectual disabilit...\n",
       "Variant                                             chr16 g.1762843C>T\n",
       "Variant Type                                ['missense', 'transition']\n",
       "Gene Affected                                                 MAPK8IP3\n",
       "Gene ID                                                ENSG00000138834\n",
       "Effect of Variant                                              p.R586C\n",
       "Transcript ID                                          ENST00000673691\n",
       "Protein Affected     C-Jun-amino-terminal kinase-interacting protein 3\n",
       "Protein ID                                             ENSP00000501096\n",
       "dtype: object"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samplePatient = allPatients.all_patients.get('8')\n",
    "samplePatient.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c409ba1",
   "metadata": {},
   "source": [
    "# Running a Fisher-Exact or Chi<sup>2</sup> Test\n",
    "\n",
    "Running the Correlation test is done with the command \"run_stats\". In this example, we will be testing if there is a significant difference between missense variants and nonsense variants. <br>\n",
    "\n",
    "```\n",
    "run_stats(cohort, Fun1, Fun2, extraVar_1, extra_Var2, chi2 = False)\n",
    "```\n",
    "cohort = a Cohort class variable with patients to be analyzed. <br><br>\n",
    "In this test, we want to know whether a patient who has a specific HPO also has either a variant of type \"missense\" or of type \"nonsense\" and compare the two against each other. This is done with the function \"is_var_type\" (See below for further description). In the first test, we want to test for \"missense\" mutations, and the second test for \"nonsense\" mutations. This is why the command is set up with the same tests and different variables. <br><br>\n",
    "chi2 is defaulted to False, meaning a Fisher-Exact test will run unless you set chi2 = True, which it will then run a Chi<sup>2</sup> test. <br><br>\n",
    "The return of run_stats is a pandas DataFrame of the results. \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8d924952",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total hpo terms: 36\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1 w/ hpo</th>\n",
       "      <th>1 w/o hpo</th>\n",
       "      <th>2 w/ hpo</th>\n",
       "      <th>2 w/o hpo</th>\n",
       "      <th>pval</th>\n",
       "      <th>bonferr pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HP:0001251-Ataxia</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.117647</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0002342-Intellectual disability, moderate</th>\n",
       "      <td>4.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.514706</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0001270-Motor delay</th>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0002317-Unsteady gait</th>\n",
       "      <td>1.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0007301-Oromotor apraxia</th>\n",
       "      <td>1.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              1 w/ hpo  1 w/o hpo  2 w/ hpo  \\\n",
       "HP:0001251-Ataxia                                  0.0       15.0       1.0   \n",
       "HP:0002342-Intellectual disability, moderate       4.0       11.0       1.0   \n",
       "HP:0001270-Motor delay                             5.0       10.0       0.0   \n",
       "HP:0002317-Unsteady gait                           1.0       14.0       0.0   \n",
       "HP:0007301-Oromotor apraxia                        1.0       14.0       0.0   \n",
       "\n",
       "                                              2 w/o hpo      pval  \\\n",
       "HP:0001251-Ataxia                                   1.0  0.117647   \n",
       "HP:0002342-Intellectual disability, moderate        1.0  0.514706   \n",
       "HP:0001270-Motor delay                              2.0  1.000000   \n",
       "HP:0002317-Unsteady gait                            2.0  1.000000   \n",
       "HP:0007301-Oromotor apraxia                         2.0  1.000000   \n",
       "\n",
       "                                              bonferr pval  \n",
       "HP:0001251-Ataxia                                      1.0  \n",
       "HP:0002342-Intellectual disability, moderate           1.0  \n",
       "HP:0001270-Motor delay                                 1.0  \n",
       "HP:0002317-Unsteady gait                               1.0  \n",
       "HP:0007301-Oromotor apraxia                            1.0  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_FE = run_stats(allPatients, is_var_type, is_var_type, 'missense', 'nonsense')\n",
    "result_FE.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "cc546cc3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total hpo terms: 36\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1 w/ hpo</th>\n",
       "      <th>1 w/o hpo</th>\n",
       "      <th>2 w/ hpo</th>\n",
       "      <th>2 w/o hpo</th>\n",
       "      <th>pval</th>\n",
       "      <th>bonferr pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HP:0001251-Ataxia</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.221233</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0001270-Motor delay</th>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.884100</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0031936-Delayed ability to walk</th>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.884100</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0000219-Thin upper lip vermilion</th>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.884100</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0002342-Intellectual disability, moderate</th>\n",
       "      <td>4.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              1 w/ hpo  1 w/o hpo  2 w/ hpo  \\\n",
       "HP:0001251-Ataxia                                  0.0       15.0       1.0   \n",
       "HP:0001270-Motor delay                             5.0       10.0       0.0   \n",
       "HP:0031936-Delayed ability to walk                 5.0       10.0       0.0   \n",
       "HP:0000219-Thin upper lip vermilion                5.0       10.0       0.0   \n",
       "HP:0002342-Intellectual disability, moderate       4.0       11.0       1.0   \n",
       "\n",
       "                                              2 w/o hpo      pval  \\\n",
       "HP:0001251-Ataxia                                   1.0  0.221233   \n",
       "HP:0001270-Motor delay                              2.0  0.884100   \n",
       "HP:0031936-Delayed ability to walk                  2.0  0.884100   \n",
       "HP:0000219-Thin upper lip vermilion                 2.0  0.884100   \n",
       "HP:0002342-Intellectual disability, moderate        1.0  1.000000   \n",
       "\n",
       "                                              bonferr pval  \n",
       "HP:0001251-Ataxia                                      1.0  \n",
       "HP:0001270-Motor delay                                 1.0  \n",
       "HP:0031936-Delayed ability to walk                     1.0  \n",
       "HP:0000219-Thin upper lip vermilion                    1.0  \n",
       "HP:0002342-Intellectual disability, moderate           1.0  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_chi2 = run_stats(allPatients, is_var_type, is_var_type, 'missense', 'nonsense', chi2 = True)\n",
    "result_chi2.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd88fdab",
   "metadata": {},
   "source": [
    "# Variant Tests\n",
    "\n",
    "### Variant Type\n",
    "\n",
    "**is_var_type** - Along with the examples above, there are other options for types of variants that can be tested, as listed below.<br>\n",
    "**is_not_var_type** - This is the opposite of is_var_type. Has all of the same types of variants but will only count the patients who do not have that variant type but do and do not have that HPO. \n",
    "- missense\n",
    "- nonsense\n",
    "- duplication\n",
    "- deletion\n",
    "- insertion\n",
    "- transversion\n",
    "- transition\n",
    "- indel\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "00a29a93",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total hpo terms: 36\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1 w/ hpo</th>\n",
       "      <th>1 w/o hpo</th>\n",
       "      <th>2 w/ hpo</th>\n",
       "      <th>2 w/o hpo</th>\n",
       "      <th>pval</th>\n",
       "      <th>bonferr pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HP:0001251-Ataxia</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.019608</td>\n",
       "      <td>0.705882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0002342-Intellectual disability, moderate</th>\n",
       "      <td>4.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.245098</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0001270-Motor delay</th>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.522059</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0031936-Delayed ability to walk</th>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.522059</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0000219-Thin upper lip vermilion</th>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.522059</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              1 w/ hpo  1 w/o hpo  2 w/ hpo  \\\n",
       "HP:0001251-Ataxia                                  0.0       15.0       2.0   \n",
       "HP:0002342-Intellectual disability, moderate       4.0       11.0       2.0   \n",
       "HP:0001270-Motor delay                             5.0       10.0       0.0   \n",
       "HP:0031936-Delayed ability to walk                 5.0       10.0       0.0   \n",
       "HP:0000219-Thin upper lip vermilion                5.0       10.0       0.0   \n",
       "\n",
       "                                              2 w/o hpo      pval  \\\n",
       "HP:0001251-Ataxia                                   1.0  0.019608   \n",
       "HP:0002342-Intellectual disability, moderate        1.0  0.245098   \n",
       "HP:0001270-Motor delay                              3.0  0.522059   \n",
       "HP:0031936-Delayed ability to walk                  3.0  0.522059   \n",
       "HP:0000219-Thin upper lip vermilion                 3.0  0.522059   \n",
       "\n",
       "                                              bonferr pval  \n",
       "HP:0001251-Ataxia                                 0.705882  \n",
       "HP:0002342-Intellectual disability, moderate      1.000000  \n",
       "HP:0001270-Motor delay                            1.000000  \n",
       "HP:0031936-Delayed ability to walk                1.000000  \n",
       "HP:0000219-Thin upper lip vermilion               1.000000  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_not = run_stats(allPatients, is_var_type, is_not_var_type, 'missense', 'missense')\n",
    "result_not.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9783cb76",
   "metadata": {},
   "source": [
    "### Variant Match\n",
    "**is_var_match** - Rather than a variant type, this function tests whether a patient has a specific variant. The variable values can either be a Variant Class or a string with the format 'chromosome:start:reference:alternative\".<br>\n",
    "**is_not_var_match** - This is similarly the opposite of is_var_match. It tests whether a patient does not have a specific variant, and takes the same formats as input. \n",
    "<br> \n",
    "**Example input variables (all are for the same variant):**\n",
    "- '16:1762843:C:T'\n",
    "- varcode.Variant(16, 1762843, C, T, pyensembl)\n",
    "- allPatients.all_patients.get('8').variant.variant\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "9332db91",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total hpo terms: 36\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1 w/ hpo</th>\n",
       "      <th>1 w/o hpo</th>\n",
       "      <th>2 w/ hpo</th>\n",
       "      <th>2 w/o hpo</th>\n",
       "      <th>pval</th>\n",
       "      <th>bonferr pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HP:0002505-Loss of ambulation</th>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.012255</td>\n",
       "      <td>0.441176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0000750-Delayed speech and language development</th>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.012255</td>\n",
       "      <td>0.441176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0001264-Spastic diplegia</th>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.044118</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0004322-Precocious puberty</th>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.044118</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0001250-Seizure</th>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.065359</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    1 w/ hpo  1 w/o hpo  \\\n",
       "HP:0002505-Loss of ambulation                            3.0        2.0   \n",
       "HP:0000750-Delayed speech and language development       3.0        2.0   \n",
       "HP:0001264-Spastic diplegia                              3.0        2.0   \n",
       "HP:0004322-Precocious puberty                            3.0        2.0   \n",
       "HP:0001250-Seizure                                       2.0        3.0   \n",
       "\n",
       "                                                    2 w/ hpo  2 w/o hpo  \\\n",
       "HP:0002505-Loss of ambulation                            0.0       13.0   \n",
       "HP:0000750-Delayed speech and language development       0.0       13.0   \n",
       "HP:0001264-Spastic diplegia                              1.0       12.0   \n",
       "HP:0004322-Precocious puberty                            1.0       12.0   \n",
       "HP:0001250-Seizure                                       0.0       13.0   \n",
       "\n",
       "                                                        pval  bonferr pval  \n",
       "HP:0002505-Loss of ambulation                       0.012255      0.441176  \n",
       "HP:0000750-Delayed speech and language development  0.012255      0.441176  \n",
       "HP:0001264-Spastic diplegia                         0.044118      1.000000  \n",
       "HP:0004322-Precocious puberty                       0.044118      1.000000  \n",
       "HP:0001250-Seizure                                  0.065359      1.000000  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_match = run_stats(allPatients, is_var_match, is_not_var_match, '16:1762843:C:T', '16:1762843:C:T')\n",
    "result_match.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cbffc3f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "enviro",
   "language": "python",
   "name": "enviro"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
