{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e74be9ff0537de82",
   "metadata": {},
   "source": [
    "# LMNA\n",
    "The LMNA gene encodes lamin A and lamin C. Pathogenic variants in LMNA are associated with at least 11 clinically distinct diseases."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5a3af44bc3c0aef5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-20T14:57:02.243666600Z",
     "start_time": "2024-02-20T14:57:02.236667Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/robin/PycharmProjects/genophenocorr/gpc_venv/lib/python3.9/site-packages/urllib3/__init__.py:35: NotOpenSSLWarning: urllib3 v2 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded HPO v2024-04-26\n",
      "Using genophenocorr version 0.1.1dev\n"
     ]
    }
   ],
   "source": [
    "import genophenocorr\n",
    "from genophenocorr.preprocessing import configure_caching_cohort_creator, load_phenopacket_folder\n",
    "from genophenocorr.view import CohortViewable\n",
    "import hpotk\n",
    "from IPython.display import display, HTML\n",
    "\n",
    "\n",
    "store = hpotk.configure_ontology_store()\n",
    "hpo = store.load_minimal_hpo()\n",
    "print(f'Loaded HPO v{hpo.version}')\n",
    "print(f\"Using genophenocorr version {genophenocorr.__version__}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "550cb430",
   "metadata": {},
   "source": [
    "# Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1af8085f",
   "metadata": {},
   "outputs": [],
   "source": [
    "LMNA_transcript_id = 'NM_170707.4'\n",
    "LMNA_protein_id = \"NP_733821.1\"\n",
    "fpath_phenopackets = 'input/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a3b52e65351030dc",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-20T14:58:33.257730200Z",
     "start_time": "2024-02-20T14:57:06.570095700Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Patients Created: 100%|██████████| 127/127 [00:00<00:00, 667.18it/s]\n",
      "Validated under none policy\n",
      "No errors or warnings were found\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<!DOCTYPE html>\n",
       "<html lang=\"en\">\n",
       "<head>\n",
       "  <meta charset=\"utf-8\">\n",
       "  <title>Cohort</title>\n",
       "  <style>table {\n",
       "    border-collapse: collapse;\n",
       "    margin: 25px 0;\n",
       "      font-size: 0.9em;\n",
       "      font-family: sans-serif;\n",
       "      min-width: 400px;\n",
       "      box-shadow: 0 0 20px rgba(0, 0, 0, 0.15);\n",
       "  }\n",
       "\n",
       "  .table .column-1 {\n",
       "      text-align: left;\n",
       "  }\n",
       "  th {\n",
       "    background-color: LightSkyBlue;\n",
       "    border: 1px solid #dddddd;\n",
       "    text-align: left;\n",
       "    padding: 2px;\n",
       "  font-weight: bold;  \n",
       "  font-size: 120%;\n",
       "}\n",
       "\n",
       "\n",
       "  tr {\n",
       "    border: 1px solid #dddddd;\n",
       "  }\n",
       "\n",
       "  td {\n",
       "    padding: 2px;\n",
       "    font-weight: bold;\n",
       "  }\n",
       "\n",
       "  tr:nth-child(even) {\n",
       "    background-color: #f2f2f2;\n",
       "  }\n",
       "\n",
       "  .table td,tr {\n",
       "    text-align: right;\n",
       "}\n",
       "\n",
       ".lft {\n",
       "    text-align: left;\n",
       "}​\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "caption {\n",
       "  caption-side: bottom;\n",
       "  text-align: left;\n",
       "  padding-bottom: 10px;\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       "  </style>\n",
       "</head>\n",
       "\n",
       "<body>\n",
       "  <h1>genophenocorr cohort analysis</h1>\n",
       "  <p>Successfully ingested 127 phenopackets.</p>\n",
       "  \n",
       "  <p>No errors encountered.</p>\n",
       "  \n",
       " \n",
       "    <h3>Top 10 HPO Terms</h3>\n",
       "    <table>\n",
       "      <caption>A total of 214 HPO terms were used to annotated the cohort. </caption>\n",
       "        <tbody>\n",
       "            <tr class=\"strng\">\n",
       "                <th class=\"lft\">HPO Term</th>\n",
       "                <th>ID</th>\n",
       "                <th>Annotation Count</th>\n",
       "              </tr>\n",
       "              \n",
       "              <tr>\n",
       "                <td class=\"lft\">Ankle contracture</td>\n",
       "                <td>HP:0034677</td>\n",
       "                <td>39</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Elbow contracture</td>\n",
       "                <td>HP:0034391</td>\n",
       "                <td>37</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Spinal rigidity</td>\n",
       "                <td>HP:0003306</td>\n",
       "                <td>37</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Stiff neck</td>\n",
       "                <td>HP:0025258</td>\n",
       "                <td>36</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Proximal muscle weakness in upper limbs</td>\n",
       "                <td>HP:0008997</td>\n",
       "                <td>35</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Dilated cardiomyopathy</td>\n",
       "                <td>HP:0001644</td>\n",
       "                <td>27</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Atrial fibrillation</td>\n",
       "                <td>HP:0005110</td>\n",
       "                <td>25</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Arrhythmia</td>\n",
       "                <td>HP:0011675</td>\n",
       "                <td>23</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Lipodystrophy</td>\n",
       "                <td>HP:0009125</td>\n",
       "                <td>21</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">Proximal muscle weakness in lower limbs</td>\n",
       "                <td>HP:0008994</td>\n",
       "                <td>21</td>\n",
       "                </tr>\n",
       "            \n",
       "                </tbody>\n",
       "    </table>\n",
       "\n",
       " <h3>Top 10 Variants</h3>\n",
       " <table>\n",
       "  <caption>\n",
       "    Variants are shown according to NM_170707.4. A total of 127 unique variants were identified in the cohort.\n",
       "  </caption>\n",
       "        <tbody>\n",
       "            <tr class=\"strng\">\n",
       "                <th class=\"lft\">Variant</th>\n",
       "                <th>Variant name</th>\n",
       "                <th>Variant Count</th>\n",
       "              </tr>\n",
       "              \n",
       "              <tr>\n",
       "                <td class=\"lft\">1_156138613_156138613_C_T</td>\n",
       "                <td>c.1824C>T</td>\n",
       "                <td>15</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">1_156134474_156134474_C_A</td>\n",
       "                <td>c.585C>A</td>\n",
       "                <td>11</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">1_156134497_156134497_A_G</td>\n",
       "                <td>c.608A>G</td>\n",
       "                <td>8</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">1_156137756_156137756_C_A</td>\n",
       "                <td>c.1698+13C>A</td>\n",
       "                <td>8</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">1_156137204_156137204_G_C</td>\n",
       "                <td>c.1580G>C</td>\n",
       "                <td>8</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">1_156136984_156136984_C_T</td>\n",
       "                <td>c.1444C>T</td>\n",
       "                <td>7</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">1_156115096_156115096_C_G</td>\n",
       "                <td>c.178C>G</td>\n",
       "                <td>7</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">1_156136413_156136413_C_T</td>\n",
       "                <td>c.1357C>T</td>\n",
       "                <td>6</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">1_156136036_156136036_G_A</td>\n",
       "                <td>c.1072G>A</td>\n",
       "                <td>6</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">1_156135922_156135923_CT_C</td>\n",
       "                <td>c.960del</td>\n",
       "                <td>5</td>\n",
       "                </tr>\n",
       "            \n",
       "         </tbody>\n",
       "    </table>\n",
       "    <h3>Diseases</h3>\n",
       "    <table>\n",
       "        <tbody>\n",
       "            <tr class=\"strng\">\n",
       "                <th class=\"lft\">Disease ID</th>\n",
       "                <th>Disease Name</th>\n",
       "                <th>Annotation Count</th>\n",
       "              </tr>\n",
       "              \n",
       "              <tr>\n",
       "                <td class=\"lft\">OMIM:115200</td>\n",
       "                <td>Cardiomyopathy, dilated, 1A</td>\n",
       "                <td>44</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">OMIM:181350</td>\n",
       "                <td>Emery-Dreifuss muscular dystrophy 2, autosomal dominant</td>\n",
       "                <td>41</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">OMIM:613205</td>\n",
       "                <td>LMNA-related congenital muscular dystrophy</td>\n",
       "                <td>15</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">OMIM:176670</td>\n",
       "                <td>Hutchinson-Gilford progeria</td>\n",
       "                <td>15</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">OMIM:151660</td>\n",
       "                <td>Lipodystrophy, familial partial, type 2</td>\n",
       "                <td>12</td>\n",
       "                </tr>\n",
       "            \n",
       "           </tbody>\n",
       "    </table>\n",
       "    \n",
       "    <h3>Variant categories</h3>\n",
       "    <table>\n",
       "        <tbody>\n",
       "            <tr class=\"strng\">\n",
       "                <th class=\"lft\">Variant effect</th>\n",
       "                <th>Annotation Count</th>\n",
       "              </tr>\n",
       "              \n",
       "              <tr>\n",
       "                <td class=\"lft\">INTRON_VARIANT</td>\n",
       "                <td>8</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">SYNONYMOUS_VARIANT</td>\n",
       "                <td>15</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">FRAMESHIFT_VARIANT</td>\n",
       "                <td>5</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">STOP_GAINED</td>\n",
       "                <td>5</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">MISSENSE_VARIANT</td>\n",
       "                <td>90</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">INFRAME_DELETION</td>\n",
       "                <td>3</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">SPLICE_REGION_VARIANT</td>\n",
       "                <td>4</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td class=\"lft\">SPLICE_ACCEPTOR_VARIANT</td>\n",
       "                <td>1</td>\n",
       "                </tr>\n",
       "            \n",
       "           </tbody>\n",
       "    </table>\n",
       "  \n",
       "\n",
       "\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cohort_creator = configure_caching_cohort_creator(hpo)\n",
    "cohort = load_phenopacket_folder(fpath_phenopackets, cohort_creator)\n",
    "viewer = CohortViewable(hpo, transcript_id=LMNA_transcript_id)\n",
    "display(HTML(viewer.process(cohort, LMNA_transcript_id)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3fe6dac6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<!DOCTYPE html>\n",
       "<html lang=\"en\">\n",
       "<head>\n",
       "  <meta charset=\"utf-8\">\n",
       "  <title>Cohort</title>\n",
       "  <style>table {\n",
       "    border-collapse: collapse;\n",
       "    margin: 25px 0;\n",
       "      font-size: 0.9em;\n",
       "      font-family: sans-serif;\n",
       "      min-width: 400px;\n",
       "      box-shadow: 0 0 20px rgba(0, 0, 0, 0.15);\n",
       "  }\n",
       "\n",
       "\n",
       "  th {\n",
       "    background-color: #f2f2f2;\n",
       "    border: 1px solid #dddddd;\n",
       "    text-align: left;\n",
       "    padding: 2px;\n",
       "  }\n",
       "\n",
       "  tr {\n",
       "    border: 1px solid #dddddd;\n",
       "  }\n",
       "\n",
       "  td {\n",
       "    padding: 2px;\n",
       "    font-weight: bold;\n",
       "  }\n",
       "\n",
       "  tr:nth-child(even) {\n",
       "    background-color: #f2f2f2;\n",
       "  }\n",
       "\n",
       "  .table td,tr {\n",
       "    text-align: right;\n",
       "}\n",
       "\n",
       ".table td:first-child, tr:first-child {\n",
       "    text-align: left;\n",
       "}​\n",
       "\n",
       "  </style>\n",
       "</head>\n",
       "\n",
       "<body>\n",
       "  <h1>genophenocorr disease analysis</h1>\n",
       "\n",
       "    <h3>Diseases</h3>\n",
       "    <table>\n",
       "        <tbody>\n",
       "            <tr>\n",
       "                <th>Disease ID</th>\n",
       "                <th>Disease Name</th>\n",
       "                <th>Annotation Count</th>\n",
       "              </tr>\n",
       "              \n",
       "              <tr>\n",
       "                <td>OMIM:115200</td>\n",
       "                <td>Cardiomyopathy, dilated, 1A</td>\n",
       "                <td>44</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>OMIM:181350</td>\n",
       "                <td>Emery-Dreifuss muscular dystrophy 2, autosomal dominant</td>\n",
       "                <td>41</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>OMIM:613205</td>\n",
       "                <td>LMNA-related congenital muscular dystrophy</td>\n",
       "                <td>15</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>OMIM:176670</td>\n",
       "                <td>Hutchinson-Gilford progeria</td>\n",
       "                <td>15</td>\n",
       "                </tr>\n",
       "            \n",
       "              <tr>\n",
       "                <td>OMIM:151660</td>\n",
       "                <td>Lipodystrophy, familial partial, type 2</td>\n",
       "                <td>12</td>\n",
       "                </tr>\n",
       "            \n",
       "           </tbody>\n",
       "    </table>\n",
       "    <h3>Variants by Disease</h3>\n",
       "    \n",
       "    <h4>OMIM:115200</h4>\n",
       "    <table>\n",
       "        <tbody>\n",
       "            <tr>\n",
       "                <th>Variant Key</th>\n",
       "                <th>Variant Counts</th>\n",
       "            </tr>\n",
       "            \n",
       "            <tr>\n",
       "                <th>1_156134474_156134474_C_A</th>\n",
       "                <th>11</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156135922_156135923_CT_C</th>\n",
       "                <th>5</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156134497_156134497_A_G</th>\n",
       "                <th>8</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156137756_156137756_C_A</th>\n",
       "                <th>8</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156115096_156115096_C_G</th>\n",
       "                <th>7</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156115172_156115172_T_G</th>\n",
       "                <th>5</th>\n",
       "            </tr>\n",
       "        \n",
       "        </table></th>\n",
       "        \n",
       "    <h4>OMIM:181350</h4>\n",
       "    <table>\n",
       "        <tbody>\n",
       "            <tr>\n",
       "                <th>Variant Key</th>\n",
       "                <th>Variant Counts</th>\n",
       "            </tr>\n",
       "            \n",
       "            <tr>\n",
       "                <th>1_156136121_156136121_G_A</th>\n",
       "                <th>4</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156137183_156137183_G_C</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156115052_156115052_A_G</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156136413_156136413_C_T</th>\n",
       "                <th>6</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156136036_156136036_G_A</th>\n",
       "                <th>2</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156115248_156115251_TGAG_T</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156137204_156137204_G_C</th>\n",
       "                <th>8</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156115067_156115067_G_C</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156134830_156134830_A_C</th>\n",
       "                <th>3</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156114934_156114934_C_T</th>\n",
       "                <th>5</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156115106_156115106_T_G</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156137207_156137207_C_A</th>\n",
       "                <th>2</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156134860_156134860_G_A</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156135257_156135257_A_C</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156136076_156136076_T_A</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156136424_156136424_C_A</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156134942_156134945_TAAG_T</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156134911_156134911_G_A</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "        </table></th>\n",
       "        \n",
       "    <h4>OMIM:613205</h4>\n",
       "    <table>\n",
       "        <tbody>\n",
       "            <tr>\n",
       "                <th>Variant Key</th>\n",
       "                <th>Variant Counts</th>\n",
       "            </tr>\n",
       "            \n",
       "            <tr>\n",
       "                <th>1_156136420_156136420_G_C</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156136103_156136103_T_C</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156134910_156134910_C_T</th>\n",
       "                <th>2</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156136036_156136036_G_A</th>\n",
       "                <th>4</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156115067_156115067_G_C</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156136919_156136919_A_G</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156115034_156115034_A_G</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156135281_156135281_T_C</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156136414_156136414_G_C</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156136422_156136422_A_G</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156115009_156115012_GAGA_G</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "        </table></th>\n",
       "        \n",
       "    <h4>OMIM:176670</h4>\n",
       "    <table>\n",
       "        <tbody>\n",
       "            <tr>\n",
       "                <th>Variant Key</th>\n",
       "                <th>Variant Counts</th>\n",
       "            </tr>\n",
       "            \n",
       "            <tr>\n",
       "                <th>1_156138613_156138613_C_T</th>\n",
       "                <th>15</th>\n",
       "            </tr>\n",
       "        \n",
       "        </table></th>\n",
       "        \n",
       "    <h4>OMIM:151660</h4>\n",
       "    <table>\n",
       "        <tbody>\n",
       "            <tr>\n",
       "                <th>Variant Key</th>\n",
       "                <th>Variant Counts</th>\n",
       "            </tr>\n",
       "            \n",
       "            <tr>\n",
       "                <th>1_156136984_156136984_C_T</th>\n",
       "                <th>7</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156136251_156136251_C_T</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156115093_156115093_C_G</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156134853_156134853_G_A</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156138507_156138507_C_T</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>1_156114947_156114947_C_T</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "        </table></th>\n",
       "        \n",
       "        </tbody>\n",
       "    </table>\n",
       "    <h3>Variant Effects by Disease</h3>\n",
       "    \n",
       "    <h4>OMIM:115200</h4>\n",
       "    <table>\n",
       "        <tbody>\n",
       "            <tr>\n",
       "                <th>Variant Effect</th>\n",
       "                <th>Variant Effect Counts</th>\n",
       "            </tr>\n",
       "            \n",
       "            <tr>\n",
       "                <th>MISSENSE_VARIANT</th>\n",
       "                <th>31</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>FRAMESHIFT_VARIANT</th>\n",
       "                <th>5</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>INTRON_VARIANT</th>\n",
       "                <th>8</th>\n",
       "            </tr>\n",
       "        \n",
       "        </table></th>\n",
       "        \n",
       "    <h4>OMIM:181350</h4>\n",
       "    <table>\n",
       "        <tbody>\n",
       "            <tr>\n",
       "                <th>Variant Effect</th>\n",
       "                <th>Variant Effect Counts</th>\n",
       "            </tr>\n",
       "            \n",
       "            <tr>\n",
       "                <th>MISSENSE_VARIANT</th>\n",
       "                <th>34</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>SPLICE_REGION_VARIANT</th>\n",
       "                <th>4</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>INFRAME_DELETION</th>\n",
       "                <th>2</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>STOP_GAINED</th>\n",
       "                <th>5</th>\n",
       "            </tr>\n",
       "        \n",
       "        </table></th>\n",
       "        \n",
       "    <h4>OMIM:613205</h4>\n",
       "    <table>\n",
       "        <tbody>\n",
       "            <tr>\n",
       "                <th>Variant Effect</th>\n",
       "                <th>Variant Effect Counts</th>\n",
       "            </tr>\n",
       "            \n",
       "            <tr>\n",
       "                <th>MISSENSE_VARIANT</th>\n",
       "                <th>13</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>SPLICE_ACCEPTOR_VARIANT</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "            <tr>\n",
       "                <th>INFRAME_DELETION</th>\n",
       "                <th>1</th>\n",
       "            </tr>\n",
       "        \n",
       "        </table></th>\n",
       "        \n",
       "    <h4>OMIM:176670</h4>\n",
       "    <table>\n",
       "        <tbody>\n",
       "            <tr>\n",
       "                <th>Variant Effect</th>\n",
       "                <th>Variant Effect Counts</th>\n",
       "            </tr>\n",
       "            \n",
       "            <tr>\n",
       "                <th>SYNONYMOUS_VARIANT</th>\n",
       "                <th>15</th>\n",
       "            </tr>\n",
       "        \n",
       "        </table></th>\n",
       "        \n",
       "    <h4>OMIM:151660</h4>\n",
       "    <table>\n",
       "        <tbody>\n",
       "            <tr>\n",
       "                <th>Variant Effect</th>\n",
       "                <th>Variant Effect Counts</th>\n",
       "            </tr>\n",
       "            \n",
       "            <tr>\n",
       "                <th>MISSENSE_VARIANT</th>\n",
       "                <th>12</th>\n",
       "            </tr>\n",
       "        \n",
       "        </table></th>\n",
       "        \n",
       "        </tbody>\n",
       "    </table>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from genophenocorr.view import DiseaseViewable\n",
    "\n",
    "dis_view = DiseaseViewable(hpo, transcript_id=LMNA_transcript_id)\n",
    "dis_dict = dis_view._prepare_context(cohort)\n",
    "display(HTML(dis_view.process(cohort)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3cde1a84a1b62676",
   "metadata": {},
   "source": [
    "## Configure the analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d0ee56e0c371dcd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from genophenocorr.analysis import configure_cohort_analysis, CohortAnalysisConfiguration\n",
    "\n",
    "analysis_config = CohortAnalysisConfiguration()\n",
    "analysis_config.missing_implies_excluded = True\n",
    "analysis_config.pval_correction = 'fdr_bh'\n",
    "\n",
    "analysis = configure_cohort_analysis(cohort, hpo, config=analysis_config)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0be03eb2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>MISSENSE_VARIANT on NM_170707.4</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Yes</th>\n",
       "      <th colspan=\"2\" halign=\"left\">No</th>\n",
       "      <th colspan=\"2\" halign=\"left\"></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>p value</th>\n",
       "      <th>Corrected p value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Lipodystrophy [HP:0009125]</th>\n",
       "      <td>10/75</td>\n",
       "      <td>13%</td>\n",
       "      <td>15/35</td>\n",
       "      <td>43%</td>\n",
       "      <td>0.001167</td>\n",
       "      <td>0.080007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Alopecia [HP:0001596]</th>\n",
       "      <td>0/3</td>\n",
       "      <td>0%</td>\n",
       "      <td>15/15</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.001225</td>\n",
       "      <td>0.080007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Micrognathia [HP:0000347]</th>\n",
       "      <td>0/3</td>\n",
       "      <td>0%</td>\n",
       "      <td>15/15</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.001225</td>\n",
       "      <td>0.080007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Proptosis [HP:0000520]</th>\n",
       "      <td>0/3</td>\n",
       "      <td>0%</td>\n",
       "      <td>15/15</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.001225</td>\n",
       "      <td>0.080007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Failure to thrive [HP:0001508]</th>\n",
       "      <td>0/3</td>\n",
       "      <td>0%</td>\n",
       "      <td>15/15</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.001225</td>\n",
       "      <td>0.080007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Osteolysis involving bones of the upper limbs [HP:0045039]</th>\n",
       "      <td>0/0</td>\n",
       "      <td>0%</td>\n",
       "      <td>15/15</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Abnormality of the Achilles tendon [HP:0005109]</th>\n",
       "      <td>3/3</td>\n",
       "      <td>100%</td>\n",
       "      <td>2/2</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Skeletal muscle hypertrophy [HP:0003712]</th>\n",
       "      <td>3/3</td>\n",
       "      <td>100%</td>\n",
       "      <td>1/1</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Flexion contracture [HP:0001371]</th>\n",
       "      <td>35/35</td>\n",
       "      <td>100%</td>\n",
       "      <td>7/7</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Abnormality of the skeletal system [HP:0000924]</th>\n",
       "      <td>37/37</td>\n",
       "      <td>100%</td>\n",
       "      <td>23/23</td>\n",
       "      <td>100%</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>457 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "MISSENSE_VARIANT on NM_170707.4                       Yes             No  \\\n",
       "                                                    Count Percent  Count   \n",
       "Lipodystrophy [HP:0009125]                          10/75     13%  15/35   \n",
       "Alopecia [HP:0001596]                                 0/3      0%  15/15   \n",
       "Micrognathia [HP:0000347]                             0/3      0%  15/15   \n",
       "Proptosis [HP:0000520]                                0/3      0%  15/15   \n",
       "Failure to thrive [HP:0001508]                        0/3      0%  15/15   \n",
       "...                                                   ...     ...    ...   \n",
       "Osteolysis involving bones of the upper limbs [...    0/0      0%  15/15   \n",
       "Abnormality of the Achilles tendon [HP:0005109]       3/3    100%    2/2   \n",
       "Skeletal muscle hypertrophy [HP:0003712]              3/3    100%    1/1   \n",
       "Flexion contracture [HP:0001371]                    35/35    100%    7/7   \n",
       "Abnormality of the skeletal system [HP:0000924]     37/37    100%  23/23   \n",
       "\n",
       "MISSENSE_VARIANT on NM_170707.4                                       \\\n",
       "                                                   Percent   p value   \n",
       "Lipodystrophy [HP:0009125]                             43%  0.001167   \n",
       "Alopecia [HP:0001596]                                 100%  0.001225   \n",
       "Micrognathia [HP:0000347]                             100%  0.001225   \n",
       "Proptosis [HP:0000520]                                100%  0.001225   \n",
       "Failure to thrive [HP:0001508]                        100%  0.001225   \n",
       "...                                                    ...       ...   \n",
       "Osteolysis involving bones of the upper limbs [...    100%  1.000000   \n",
       "Abnormality of the Achilles tendon [HP:0005109]       100%  1.000000   \n",
       "Skeletal muscle hypertrophy [HP:0003712]              100%  1.000000   \n",
       "Flexion contracture [HP:0001371]                      100%  1.000000   \n",
       "Abnormality of the skeletal system [HP:0000924]       100%  1.000000   \n",
       "\n",
       "MISSENSE_VARIANT on NM_170707.4                                       \n",
       "                                                   Corrected p value  \n",
       "Lipodystrophy [HP:0009125]                                  0.080007  \n",
       "Alopecia [HP:0001596]                                       0.080007  \n",
       "Micrognathia [HP:0000347]                                   0.080007  \n",
       "Proptosis [HP:0000520]                                      0.080007  \n",
       "Failure to thrive [HP:0001508]                              0.080007  \n",
       "...                                                              ...  \n",
       "Osteolysis involving bones of the upper limbs [...          1.000000  \n",
       "Abnormality of the Achilles tendon [HP:0005109]             1.000000  \n",
       "Skeletal muscle hypertrophy [HP:0003712]                    1.000000  \n",
       "Flexion contracture [HP:0001371]                            1.000000  \n",
       "Abnormality of the skeletal system [HP:0000924]             1.000000  \n",
       "\n",
       "[457 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from genophenocorr.model import VariantEffect\n",
    "from genophenocorr.analysis.predicate import PatientCategories\n",
    "from genophenocorr.analysis.predicate.genotype import VariantEffectPredicate\n",
    "\n",
    "by_missense = analysis.compare_by_variant_effect(VariantEffect.MISSENSE_VARIANT, tx_id=LMNA_transcript_id)\n",
    "by_missense.summarize(hpo, PatientCategories.YES)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1b18c688",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>MISSENSE_VARIANT on NM_170707.4</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Yes</th>\n",
       "      <th colspan=\"2\" halign=\"left\">No</th>\n",
       "      <th colspan=\"2\" halign=\"left\"></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>Count</th>\n",
       "      <th>Percent</th>\n",
       "      <th>p value</th>\n",
       "      <th>Corrected p value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OMIM:176670</th>\n",
       "      <td>0/90</td>\n",
       "      <td>0%</td>\n",
       "      <td>15/37</td>\n",
       "      <td>41%</td>\n",
       "      <td>8.025624e-10</td>\n",
       "      <td>4.012812e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OMIM:151660</th>\n",
       "      <td>12/90</td>\n",
       "      <td>13%</td>\n",
       "      <td>0/37</td>\n",
       "      <td>0%</td>\n",
       "      <td>1.807470e-02</td>\n",
       "      <td>4.518674e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OMIM:181350</th>\n",
       "      <td>34/90</td>\n",
       "      <td>38%</td>\n",
       "      <td>7/37</td>\n",
       "      <td>19%</td>\n",
       "      <td>5.904112e-02</td>\n",
       "      <td>9.840187e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OMIM:613205</th>\n",
       "      <td>13/90</td>\n",
       "      <td>14%</td>\n",
       "      <td>2/37</td>\n",
       "      <td>5%</td>\n",
       "      <td>2.276691e-01</td>\n",
       "      <td>2.845863e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OMIM:115200</th>\n",
       "      <td>31/90</td>\n",
       "      <td>34%</td>\n",
       "      <td>13/37</td>\n",
       "      <td>35%</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "MISSENSE_VARIANT on NM_170707.4    Yes             No                        \\\n",
       "                                 Count Percent  Count Percent       p value   \n",
       "OMIM:176670                       0/90      0%  15/37     41%  8.025624e-10   \n",
       "OMIM:151660                      12/90     13%   0/37      0%  1.807470e-02   \n",
       "OMIM:181350                      34/90     38%   7/37     19%  5.904112e-02   \n",
       "OMIM:613205                      13/90     14%   2/37      5%  2.276691e-01   \n",
       "OMIM:115200                      31/90     34%  13/37     35%  1.000000e+00   \n",
       "\n",
       "MISSENSE_VARIANT on NM_170707.4                    \n",
       "                                Corrected p value  \n",
       "OMIM:176670                          4.012812e-09  \n",
       "OMIM:151660                          4.518674e-02  \n",
       "OMIM:181350                          9.840187e-02  \n",
       "OMIM:613205                          2.845863e-01  \n",
       "OMIM:115200                          1.000000e+00  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "by_dis = analysis.compare_disease_vs_genotype(VariantEffectPredicate(transcript_id=LMNA_transcript_id, effect=VariantEffect.MISSENSE_VARIANT))\n",
    "by_dis.summarize(hpo, PatientCategories.YES)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9285e5dc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gpc_venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
